<div class="blog-post">
  <div class="section-title">
    <p class="blog-post-title">Building a simple DNA basecaller in Python</p>
  </div>

  <p class="blog-post_paragraph">
    I went to the Oxford Nanopore Community meetup 2022 to learn more about how
    the Oxford Nanopore sequencing devices work. One of the workshops I
    participated in educated us about the history of basecalling algorithms and
    had us try to build our own basecaller using sample raw data.
  </p>
  <p class="blog-post_paragraph">
    As a brief summary, nanopore sequencing is a technology that allows you to
    determine the sequences of DNA (and potentially RNA) in your sample of
    interest. How this works is that extracted DNA are prepped with reagents to
    make the DNA molecules nanopore sequencing compatible. The prepped sample is
    then loaded onto the sequencing device that contains a membrane full of
    biological pore proteins, termemd &quot;nanopores&quot;, that can allow
    molecules to pass from one side of the membrane to the other. Samples are
    loaded alongside so-called &quot;motor proteins&quot; that bind to DNA
    molecules, tether themselves near a pore, and begin to feed the strand of
    DNA through the nanopore by unwinding the double-stranded and passing it
    into the pore as a single strand of DNA (in most cases).
  </p>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_sequencing.gif"
      alt="Nanopore Sequencing"
    />
  </div>
  <p class="blog-post_paragraph">
    Throughout this whole process, an electrical current is being applied
    through the membrane. As molecules pass through the pores, disruptions in
    the electrical current can be observed and measured.
  </p>
  <p class="blog-post_paragraph">
    This is where the concept of
    <a href="https://en.wikipedia.org/wiki/Hidden_Markov_model" target="_blank"
      >Hidden Markov Models</a
    >
    (HMMs) come in. HMMs are statistical models that make use of probabilistic
    transitions between hidden states to explain observed outputs. In the case
    of nanopore sequencing and DNA basecalling, the hidden states represent the
    unknown sequence of DNA bases, while the observed outputs correspond to the
    electrical signals measured by the sequencing device. Based on the observed
    electrical signals, we can design a program that estimates the probabilities
    of transitioning between different hidden states, and use this information
    to infer the most likely sequence of DNA bases that passed through the
    nanopore.
  </p>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_basecalling.gif"
      alt="DNA Basecaller"
    />
  </div>
  <p class="blog-post_paragraph">
    The premise of our project was simple, given a list of electrical readings
    at many different timepoints, decipher the sequence of DNA that went through
    the nanopore molecule. All we were given were three files to start with:
  </p>
  <pre class="no-code">
<code class="language-treeview special-treeview">signals.json &larr; A file containing electrical signal data obtained from a nanopore device

kmer_table.json &larr; A file containing a table each DNA dinucleotide and their corresponding electrical signature

sequences.json &larr; A file containing the correct sequences corresponding to the raw electrical data (to test the accuracy of the basecaller)
</code></pre>
  <p class="blog-post_paragraph">
    I&#39;ll elaborate on these further as they come up in the code.
  </p>
  <p class="blog-post_paragraph">
    The workshop was very independent. We were left to our own devices and given
    3 hours to come up with a potential solution. Here is the solution that I
    came up with.
  </p>
  <p class="blog-post_paragraph">
    I decided to tackle this challenge in Python. The first thing I did was
    import the modules I was going to use.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Import the required modules
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import json</code></pre>
  <p class="blog-post_paragraph">
    Then I imported all files and saved them into variables.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Opening JSON files and JSON objects as lists or dictionaries
with open('sequences.json') as f:
sequencing = json.load(f)

with open('signals.json') as f2:
signals = json.load(f2)

with open('kmer_table.json') as f3:
kmer_table = json.load(f3)</code></pre>
  <p class="blog-post_paragraph">
    The next thing I needed to do is investigate the contents of each file.
  </p>
  <p class="blog-post_paragraph">Starting off with the sequencing file.</p>
  <pre
    class="line-numbers"
  ><code class="language-python">print(type(sequencing))</code></pre>
  <pre
    class="output no-lang-display"
  ><code class="language-output">&lt;class 'list'&gt;</code></pre>
  <pre
    class="line-numbers"
  ><code class="language-python"># Prints the first 80 base pairs of each sequence
for i, seq in enumerate(sequencing):
print(f"sequence {i + 1}: {seq[0:80]}")</code></pre>
  <pre
    class="truncated-output"
  ><code class="language-truncout">sequence 1: CCTTACCAGTTTCATAATAAATAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGGGCGAATACCGCGCTGCACGTG
sequence 2: TGACCGTTTGCAGGCATTGCTTGGATCCATATAGCGCACAAATAGTCGCGAGTATGGTTGGTGATCAAAAGACCTGTAGA
sequence 3: AGCGCTCGGCTAAAACTAGCCTGGCCATCGTTGTCATAAACCTATCCCGGCGTCCGCTTAGCGTACTGAAGGTGTTCGTA
sequence 4: GCCCGTAGAAAAACAGGTTGAGGACAATAGCAGTTACGAATACAATTATCGATCCGGTGGTGAAAGACGTCATCGCACCT
sequence 5: GACACGAAGACTTCAGGCGTGCAGATGCGAAACGCGCAGTCGCACCGATATATACACGATTGCTCTCAGGCCGCATGGTT
sequence 6: CAACTATCATGCAGTATTGTCACGCGTTGCCTGCCTTGACTAGCGGCTATTATATGATGATGCGATCGTTTAGGCCAGCT
sequence 7: AGGCTGGTGAGAGTACTCGTCGCTTCCAATAATAATCGATACAAGCTATCCAGTGTATCCCATGAACCACGTACCCTTCG
sequence 8: GCTGGGCTATTTTCCAGACCACATGAATCGGTTTCTGCGGCTGGGTCTTCAGTTGGCACGTCCTTTGGGGGTCCCAGGCA
sequence 9: AGTTCTCTCCGATTGGCTGACAGTAACAAGCCGGGTGGACCCGGGAGGTTTTTTGTAACTGTTAGCTAGAATGCGCTCAA
sequence 10: GCGAAACCCGATACAAAACAGCATACACGCAGTAGATACTATTGGTCCGGAAAAAATGTCACTCCAGTAATGACTTAGTC
sequence 11: TAGATTAATTTGCTTAAGGAACACTCGGTTGGCGCCTGTAACTTGAACAGGCCTTAGAGCCACCGATATGGATGCTTGTA
sequence 12: GGCGAGAAGACGGTAATAAGCGAGTGTTCCCACGGGCTTTGGGATGCTTGTCTCAAGTCTTTCCCGACAGCATTAAGTCT
sequence 13: CTGCATTGCGGCTCGTGATACGTAATGCGTTACGCAAATACTTCAGTCAGTGTCTAGACGTCCAAGTTTTCTTTAGAGCC
sequence 14: ATGGGTGCCAGCGACCGATGAGAAGACATGAACCGGGGGGTCCTTTGGTTACCGAGCTTCAACCACGGGAAATGCTAATG
sequence 15: AAGCGGCGTAACTCTCTGGGCAGTGGTATTCTACTCGACTACTCCTTAAGCGCCCATGCTGACAATGTGCGGACGTGGCC
sequence 16: CTGAATGACGGATAAGGTGCAGAGGACACATGTAGGACTAATAGAAAAGAACTCGGCAAAGTCTCCACCAACGCCTCGGC
sequence 17: GGATTGCCACAGGAAACATCTAACGAGACCTACTGGTTACGTTAGTAGTCATGCGCACCCATCTTGAGATATGCGAACGT
sequence 18: TCACCTCCCTGAGGTGTGACTACTAATCTCCCGCCGGGAGTCCACTACTGGTAAATTATCACAGCTAGCACTGTTTGGCA
sequence 19: AACATGTGGCTGCCCACGGTTAGTCCGCGCCACCCAATTTAGAGTCAATAATTATTCCAGGTAAGAAACAACACAGCTCT
sequence 20: CGTAATAATTATACGAAAACTTCCCCGTCTGTTCGCAATGAGGGGTATCGTTGGGTACTTCCCTACTCGAACCTCTTATC
sequence 21: CCAAACCCCGTAAAGAGATCACGAGACACGGCAGGCCGCAGACCTATGTACGGCGGCATTTATAGCCATAGGACCCCGTA
sequence 22: CACGATTTTGGCACTGCCGCCGATTAACCTATCTTGCGGTTTACAATGAGTCTAGTTAACGGCTGTTGAATGAACCCGTG
sequence 23: GGCTTTCCGTCGGTCAAGATTTTGCTGTTGAAAGGATGAGGTCTTGAATCGAACCCTTAGTTGTGAATAACGAATGACTA
sequence 24: TTTACCTTATCGAATAGTTTATAGAAATTCGCTCGTTGAGTTCACGGTGGACAGTGTGCGTCCCGGGCGACGCGCTTGGT
sequence 25: GGGTGCCAAAGCCTGTTCAAGCCGGGTCGTTTATATAGGTCCGAGATTGTAGATGAATTATTAAGGTCATTTCGAAGCAG
... 
sequence 197: TCCAGACACGAACTGAAGGACTGGAGCTTCGAACTTGCGTCGGTCGAGATGTATGCTATTTAATCAGGCACACGACTATT
sequence 198: GTTGATCTCCTATACGTGGGCTCGGGATTGAACTAGACTACCTCGGTTACCCCTCCCAATAATGTATACAATAACCCTCT
sequence 199: CGTCGTGGCAGAAACTTTCTTTCGCTGAGATACACTACGGTAACTTCTCTAAATCCATCTAAAGGCCTATATTCACCTAA
sequence 200: GATTTGTTTTAACTTGCACACCCGGTGGTCCTAACTTATCACGGTTCCTGTCACGGTAAAGCTTCTACTGCGACCAGCTC</code></pre>
  <p class="blog-post_paragraph">
    The sequencing file contains a list of 200 sequences that will be used to
    benchmark the accuracy of my basecaller.
  </p>
  <p class="blog-post_paragraph">
    The next file contains the raw numerical data obtained from the sequencing
    device. Let's see what this looks like.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">print(type(signals))</code></pre>
  <pre
    class="output no-lang-display"
  ><code class="language-output">&lt;class 'list'&gt;</code></pre>
  <pre
    class="line-numbers"
  ><code class="language-python"># Prints the first 80 electrical current readings for each sequencing run
# Only print the data for the first and last 3 readings to not overclutter the output
for i, sig in enumerate(signals):
if i < 3 or i >= len(signals) - 3:
print(f"signal {i + 1}: {sig[0:80]}", end="\n");
if i == 3:
print("...")</code></pre>
  <pre class="output"><code class="language-output">
signal 1: [0.3100529743453997, 0.3158634462325644, 0.31208845112970585, 0.31708760150559795, 0.3123011896833653, 0.31436657306891536, 0.441634747471758, 0.4407353774410627, 0.4446590187328251, 0.43930513606416066, 0.43076376005652706, 0.4328593195800709, 0.4429004272747269, 0.9361748673202694, 0.9390461565805324, 0.9340941773738767, 0.9415035256947212, 0.9359606710046688, 0.943332432954438, 0.9353710639164073, 0.9419725335868708, 0.9401975844192835, 0.7471104323783434, 0.7543959853309887, 0.746393462038096, 0.746710131244383, 0.7528868451684507, 0.747498631994136, 0.7476246277309819, 0.7486243971631584, 0.7459993232405099, 0.06667997763869558, 0.0649340587939394, 0.062222685213238066, 0.06206643173130584, 0.06430062205106872, 0.06956134040207629, 0.06598387921012897, 0.0604661714628984, 0.06560967682099546, 0.31202491531158766, 0.3117040072647851, 0.3199658855006085, 0.3108242484092875, 0.31058326033949035, 0.30469157885101905, 0.24439068341818745, 0.2518044917549384, 0.24722031639525904, 0.252061655207133, 0.24554051579053626, 0.24693076069467307, 0.25269828282782403, 0.25584356952185233, 0.12736989120569608, 0.129791323366885, 0.13181837668548962, 0.12916035661261638, 0.1213262398467999, 0.6870786211499699, 0.6900207826827397, 0.6858406123948471, 0.6893077841492848, 0.6811833748584222, 0.9341854880845744, 0.9396155189823407, 0.9285516539555502, 0.9330691499908618, 0.9374291267283594, 0.9306224221015219, 0.9377284839843284, 0.934301810037233, 0.9363336216179967, 0.9364640685863979, 0.9382571727902347, 0.9352448450616708, 0.9378587456292904, 0.9350953727974025, 0.9397242940403095, 0.9442752364531524]
signal 2: [0.8718274187396214, 0.8701425065460943, 0.8736259998975986, 0.873563917571122, 0.8650604183657802, 0.873097399690971, 0.8727233417770178, 0.8694257721321107, 0.871501837917242, 0.8759874130843427, 0.8763655561883015, 0.8762994742089532, 0.8726961912153857, 0.8717003819220065, 0.8755815519018014, 0.8734274313657613, 0.8791456680423432, 0.8796572345884945, 0.4981095474949926, 0.493466052165371, 0.5007334390525425, 0.4973244180014974, 0.5032891882812756, 0.06530039807092146, 0.0582975625632877, 0.06669292223744569, 0.06806910936103525, 0.06404489553159515, 0.06494058414660187, 0.06041628781925487, 0.05906335840732782, 0.06751954748570266, 0.3134644364681886, 0.31213329920146377, 0.3122242999220129, 0.3155110458612274, 0.3116373949351636, 0.3105062883956581, 0.31128102557675413, 0.31185001119590655, 0.3105541145917271, 0.3128531292426612, 0.37303360549249376, 0.3731802886295113, 0.3829461567909086, 0.3750290419359434, 0.3818319447236752, 0.3709809415682388, 0.3780192669371244, 0.3750452720284999, 0.694264040073273, 0.6856392546114898, 0.6946692919574114, 0.6851454156060066, 0.6875305097928515, 0.682875656267178, 0.6873617953935149, 0.6866727075854471, 0.6830624567091332, 0.6801160137914215, 0.6910856277693966, 0.6894805503128675, 0.6847146459314587, 0.6870728254541224, 0.6845991846173449, 0.6869477013746157, 0.6885001707549948, 0.6889124626282804, 0.9354960925055359, 0.9366607824920513, 0.9395644675378666, 0.9348626941960132, 0.9352777715381704, 0.9378274008343893, 0.9394790489943761, 0.9358042009950395, 0.9406861231946123, 0.9372197435360158, 0.9348090298787833, 0.9378776616289848]
signal 3: [0.1274964684895317, 0.11763153197627063, 0.12233869311664558, 0.12254135138028113, 0.12329276649411786, 0.12496405394528928, 0.12305279351256104, 0.12239640235761852, 0.1246080684691115, 0.12432869653873103, 0.5560826306023586, 0.5553832011121335, 0.5602097558363068, 0.5652997841669394, 0.5617131749677041, 0.5595141922509749, 0.5662509925302252, 0.5519983950524241, 0.5654846593366483, 0.5651964400615108, 0.37905706999889854, 0.3746771430684578, 0.3741134090613114, 0.3724145635578148, 0.38049633862938553, 0.3713483578883748, 0.37691884470929016, 0.3779694670140296, 0.37592619157277657, 0.3713721450438278, 0.384927780972506, 0.37590647784363934, 0.5619483350547104, 0.5603338030303621, 0.5592816320030317, 0.566222509541765, 0.5686722608390449, 0.5680271476000597, 0.5618333249682713, 0.5558656685026228, 0.5602826033697548, 0.564526733886477, 0.5631099032475467, 0.5610104754500993, 0.5580318158535241, 0.43578874108032767, 0.4412567352358483, 0.4338171603530718, 0.4357294506306183, 0.44022261476155106, 0.43335530477482764, 0.4304955471531175, 0.4396365723016269, 0.44029933079579625, 0.4383513242791009, 0.8067275170443323, 0.8119974772736225, 0.8188054252021972, 0.8067538465592148, 0.8136093042092316, 0.8102415519807998, 0.8133409264855411, 0.807648751586652, 0.8057651962969663, 0.8146957560581631, 0.8130323584640134, 0.8134334644374099, 0.8065623716874014, 0.8142275377946145, 0.8175714067476539, 0.37283845202236987, 0.3763459626275503, 0.372998084137685, 0.37856234526935145, 0.37169400880892456, 0.37449915023108193, 0.3784276749805034, 0.37684731771537266, 0.37120261139844263, 0.374844623437569]
...
signal 198: [0.6921936583913131, 0.6929469439559184, 0.6831279102032887, 0.6887385736647821, 0.6777946896132349, 0.6849696392478317, 0.6921244754934774, 0.6905801930851725, 0.6854101468617635, 0.6896460991103326, 0.6825116574630986, 0.6824764078060804, 0.6898633248858523, 0.6937365112132673, 0.6887585074530919, 0.6890768597122101, 0.9421082551685703, 0.9400501622697477, 0.9314443889323214, 0.9390270120195359, 0.939450621016608, 0.9397182138953516, 0.9315150824279522, 0.8743698500861806, 0.8745155587037003, 0.8719907884673278, 0.8748959598514212, 0.8800718023963146, 0.8694379088762506, 0.8706182690992401, 0.8699051995162409, 0.8792076188806851, 0.8730294199990056, 0.8804206030956366, 0.5037697406367723, 0.48984054839486557, 0.4951452166933016, 0.49288082282212947, 0.497623584340674, 0.5065239477765119, 0.5066215863440017, 0.18898116216060534, 0.187459740748378, 0.18698611034484586, 0.18582029956678953, 0.18575776415235454, 0.19097358490149405, 0.1893103419723928, 0.19055735676203864, 0.19143450240450882, 0.18671323718151872, 0.18744343782208284, 0.19501530992791913, 0.813879660298781, 0.8154418116163831, 0.8062478317124162, 0.8192301836464629, 0.8092749712733941, 0.8045974896150151, 0.8090206224361796, 0.8148429176407843, 0.8155292175624317, 0.8130456302359815, 0.8018364537533587, 0.8115358885897485, 0.8174555727790886, 0.4341396165459761, 0.4426843838282123, 0.43745119696158646, 0.4367867577631802, 0.4360314265173659, 0.44028734943250203, 0.43099919960952465, 0.4390299538389471, 0.43403530527287376, 0.43713693726849395, 0.43806148458515565, 0.4397573820298558, 0.4423007887421328, 0.4329612343142859]
signal 199: [0.37511356666345974, 0.3732188172574157, 0.3766826805391929, 0.378946188667255, 0.36980891541241895, 0.37291900736031325, 0.36740332725669994, 0.3781855107568417, 0.3765533575109063, 0.3730304691556473, 0.3734847396981682, 0.36777457317812745, 0.37766467277211924, 0.6866468970868385, 0.6869252220047867, 0.6809483946213962, 0.6884042060334641, 0.6822242664994125, 0.6914621143743624, 0.8082132388272841, 0.8177249607122832, 0.8152733609209599, 0.8094833048173833, 0.8117879213139008, 0.8079605599257219, 0.8211145826397316, 0.8133131061251264, 0.818826895730872, 0.8195377801343101, 0.8103341177706156, 0.37114487135846885, 0.36897149168144283, 0.3729882864864854, 0.3659396094037252, 0.3754084402772783, 0.3707015265583655, 0.3736264577945466, 0.6916405560469775, 0.6896243152907189, 0.6818406854424618, 0.6875521546689678, 0.6924355482330501, 0.6943283057458314, 0.6874446035363492, 0.6818928881262954, 0.6871875379591865, 0.6817051917806289, 0.6814097008856999, 0.8808241360150421, 0.8754977601246159, 0.8753877219204552, 0.8736839997387661, 0.6258349679787194, 0.6268309625624313, 0.6235885396088324, 0.6244147352226377, 0.6237884828008239, 0.6279050081077552, 0.6203130253973985, 0.6257366348678228, 0.623006583912816, 0.6260556177067069, 0.6230116778691505, 0.6211624827759077, 0.5632572371750116, 0.5635742337798384, 0.5629627783637978, 0.5674693244955594, 0.558453986962408, 0.5695069358034348, 0.5643719332096544, 0.5571968236072049, 0.24686692351918185, 0.24803103883222286, 0.25473947518326495, 0.24670410154849873, 0.12487128051740992, 0.12336858821290289, 0.12886237036632436, 0.12632991434803018]
signal 200: [0.4965896260247073, 0.503147546863643, 0.5074540271373705, 0.4994575363838496, 0.5030391863427802, 0.5023133051148065, 0.497618338740552, 0.190495466063656, 0.19452453470652062, 0.18866880895420154, 0.19211215236658935, 0.18489948629042924, 0.19360314760236613, 0.18916225871773007, 0.1885724015113846, 0.1877425021836601, 0.1835116492162438, 0.18780699883694185, 0.1843148767948086, 0.19455680573867482, 0.1863005240695882, 0.19170712655405187, 0.9371328000646023, 0.9374187126678553, 0.936239194011627, 0.9280312160085781, 0.935771580873965, 0.9360535537661905, 0.9359144171135485, 0.9410242631402418, 0.9437963026167502, 0.9308311909199294, 0.9432783267221693, 0.9323361432921073, 0.9329231280729392, 0.938816407045115, 0.9374749656743449, 0.9346029317122143, 0.938464122130661, 0.8721400372206707, 0.872300786463957, 0.8807978292977977, 0.8756186512537023, 0.8749175914720416, 0.8749445859757259, 0.8712607513249992, 0.8791549988635969, 0.8776937589692215, 0.8795066449257479, 0.6921780038475386, 0.6813975224046545, 0.688040297675312, 0.6891537559103618, 0.6894304023484992, 0.6894564992010884, 0.6845481835437235, 0.6814274234761432, 0.6914817484473015, 0.6824467369833629, 0.6872203763146844, 0.6815172712427255, 0.6888938537467207, 0.6844644676698142, 0.6861460620553707, 0.9387423224092766, 0.9306285260059989, 0.9380209458916023, 0.9367948908398598, 0.9377807630316575, 0.9369310681540363, 0.9330864861693335, 0.9419121541713907, 0.9424454593140845, 0.939219939281965, 0.9408843796150103, 0.930354401821948, 0.9389099341077157, 0.9417874521707633, 0.9385513932566601, 0.9397627128626431]
</code></pre>
  <p class="blog-post_paragraph">
    The signals file contains a list of 200 signals that correspond to the 200
    sequences found in the sequences file. As the DNA sequences pass through a
    nanopore protein the electrical current is disrupted. The current is
    measured at consistent time points and outputted.
  </p>
  <p class="blog-post_paragraph">
    Lastly, the last file is the k-mer table file. k-mers are sequences of DNA
    of length k. This file is what will be used to decode the signal and turn
    the electrical data into DNA code.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">print(type(kmer_table))</code></pre>
  <pre
    class="output no-lang-display"
  ><code class="language-output">&lt;class 'dict'&gt;</code></pre>
  <pre
    class="line-numbers"
  ><code class="language-python"># Output the table with each key, value pair on a newline
print(json.dumps(kmer_table, separators = (",\n ", ": ")))</code></pre>

  <pre class="output no-lang-display"><code class="language-output">{"AA": 0.0,
"AC": 0.0625,
"AG": 0.125,
"AT": 0.1875,
"CA": 0.25,
"CC": 0.3125,
"CG": 0.375,
"CT": 0.4375,
"GA": 0.5,
"GC": 0.5625,
"GG": 0.625,
"GT": 0.6875,
"TA": 0.75,
"TC": 0.8125,
"TG": 0.875,
"TT": 0.9375}</code></pre>
  <p class="blog-post_paragraph">
    As you can see, the data provided in this k-mer table are for k-mers of
    length 2, also referred to as <em>dinucleotides</em>. The nanopores inside
    the sequencing device span the length of a couple of DNA bases and as so
    multiple bases are inside a nanopore protein at any given time as the strand
    of DNA is fed through the pore. Each permutation of DNA sequences will have
    a unique disruption pattern to the electrical current and therefore have a
    unique electrical signal that can be used to decode the sequence if these
    signatures are known beforehand (as in the table above). In real-life, many
    more bases can exist inside a nanopore at a given time than just 2 at once,
    and real life DNA bases can contain modifications that would need them to be
    treated as separate bases (termed
    <a href="https://en.wikipedia.org/wiki/Epigenetics" target="_blank"
      >Epigenetics</a
    >). However, encoding all these permututations of k-mers would be too
    exhaustive for this challenge, as the goal of this problem is to get us to
    help us develop a conceptual understanding of what is going on at the
    molecular level within these sequencing devices and how various basecalling
    algorithms work and not so much to develop the most accurate basecalling
    algorithm possible. Therefore, a simplified dinucleotide table was provided.
  </p>
  <p class="blog-post_paragraph">
    I'll also create a way to group the signal data and the sequence data in
    order to keep things organized in my code and not have to remember which
    index I need to access corresponding data between my lists. I did this by
    creating a signal class whose objects contain the corresponding signals and
    correct sequences extracted using their index in the original lists.
  </p>
  <pre class="line-numbers"><code class="language-python">class Signal:

def __init__(self, signal_index):
self.signal_index = signal_index
self.signal = np.row_stack(signals[signal_index]) # Convert the signal data into a numpy array for easier slicing
self.correct_seq = sequencing[signal_index]

def __str__(self) -> str:
nl = "\n" # Trick I used to format newlines in f-strings
smoothed_signal = self.smoothed_signal if hasattr(self, "smoothed_signal") else "n/a"
return f"Signals array: {nl}{self.signal} {nl} {nl}Smoothed Signal: {nl}{smoothed_signal} {nl} {nl}Correct sequence: {nl}{self.correct_seq}"

def __repr__(self) -> str:
nl = "\n"
smoothed_signal = self.smoothed_signal.__repr__() if hasattr(self, "smoothed_signal") else "n/a"
return f"Signal index: {nl}{self.signal_index} {nl} {nl}Signals array (numpy array): {self.signal.__repr__()} {nl} {nl}Smoothed Signal (numpy array): {nl}{smoothed_signal}{nl} {nl}Correct sequence: {nl}{self.correct_seq} "</code></pre>
  <p class="blog-post_paragraph">
    Now that all the provided data has been previewed and organized, it would be
    nice to visualize it to explore it further. Plotting out the electrical
    signals in relation to their time points can help the user visualize the
    story of what actually happened to the DNA molecule as it passed through the
    pore.
  </p>
  <p class="blog-post_paragraph">
    In order to do this, I will select one of the series of signals from the
    signals data comprised of 200 sets of signals. I will be using this set of
    signals for the rest of the project to try to get my basecaller to work and
    then apply my basecaller on the rest of the 199 signals once I believe it's
    finished to see how well it performs.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Extract signal data from first sequencing run (index 0)
signal_1 = Signal(0)

# Preview the contents of the object
print(signal_1)</code></pre>
  <pre class="truncated-output"><code class="language-truncout">Signals array: 
[[ 3.10052974e-01]
[ 3.15863446e-01]
[ 3.12088451e-01]
[ 3.17087602e-01]
[ 3.12301190e-01]
[ 3.14366573e-01]
[ 4.41634747e-01]
[ 4.40735377e-01]
[ 4.44659019e-01]
[ 4.39305136e-01]
[ 4.30763760e-01]
[ 4.32859320e-01]
[ 4.42900427e-01]
[ 9.36174867e-01]
[ 9.39046157e-01]
[ 9.34094177e-01]
[ 9.41503526e-01]
[ 9.35960671e-01]
[ 9.43332433e-01]
[ 9.35371064e-01]
[ 9.41972534e-01]
[ 9.40197584e-01]
[ 7.47110432e-01]
[ 7.54395985e-01]
[ 7.46393462e-01]
[ 7.46710131e-01]
[ 7.52886845e-01]
[ 7.47498632e-01]
[ 7.47624628e-01]
[ 7.48624397e-01]
[ 7.45999323e-01]
[ 6.66799776e-02]
[ 6.49340588e-02]
[ 6.22226852e-02]
[ 6.20664317e-02]
[ 6.43006221e-02]
[ 6.95613404e-02]
[ 6.59838792e-02]
[ 6.04661715e-02]
...  
[ 6.85390322e-01]
[ 6.90598138e-01]
[ 6.82551341e-01]
[ 6.85252495e-01]
[ 8.76283884e-01]
[ 8.83671748e-01]
[ 8.77991853e-01]
[ 8.76095308e-01]
[ 8.74319666e-01]
[ 8.69710228e-01]] 


Correct sequence: 
CCTTACCAGTTTCATAATAAATAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGGGCGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT</code></pre>
  <p class="blog-post_paragraph">
    Next I'll visualize the sequencing data using pyplot
  </p>
  <pre class="line-numbers"><code class="language-python"># Plot the signal
plt.plot(signal_1.signal);
# Add the axis labels
plt.xlabel('Time point');
plt.ylabel('Signal');</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_signal.png"
      alt="signal_graph_with_dinucleotide_labels"
    />
  </div>
  <p class="blog-post_paragraph">
    As expected, the entire sequencing data looks like a chaotic electrogram
    that is difficult to decipher. It would be helpful to zoom in and see a
    subsection on this plot in higher resolution to see what features we can
    pick up on to begin basecalling.
  </p>
  <p class="blog-post_paragraph">
    To do this, I'll plot only a subsection of the signal data (only the first
    100 time points).
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Plot the data from the first 100 timepoints 
plot_range = 100
plt.plot(signal_1.signal[0:100]);
plt.xlabel('Time point');
plt.ylabel('Signal');</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_signal_zoomed.png"
      alt="signal_graph_with_dinucleotide_labels"
    />
  </div>
  <p class="blog-post_paragraph">
    At this resolution we can see a lot more clearly what may be happening as
    the DNA is going through the nanopore. How I like to conceptualize this is
    by breaking down the plot into 2 components: vertical-like elements and
    horizontal-like elements. Remember that inside the sequencer the DNA is
    being threaded through the nanopore by a motor protein. More specifically,
    the DNA is being
    <em>ratcheted</em> through the pore in a stepwise fashion. As the motor
    protein steps through and unwinds the DNA, there are periods where the DNA
    is let go of temporarilty, moves quickly through the nanopore channel, and
    then the motor protein stops the DNA. As the DNA is static, the sequencer
    will take a reading, and then take another reading, and take another
    reading, etc. until eventually the DNA will take a step. The DNA steps are
    random and it&#39;s driven by the dynamics of the motor protein. The
    &quot;static phase&quot; is therefore represented by the horzontal-like
    elements on the plot. These will be hereby known as &quot;sequencing
    events&quot; as they are the only time the DNA is stationary enough to
    provide an accurate reading. The vertical-like elements thereby represent
    the moments of time when the DNA is moving through the pore. This happens so
    quickly that there is no extractable sequence data from these timepoints.
  </p>

  <p class="blog-post_paragraph">
    Another thing you can notice at this high resolution is that the horizontal,
    sequencing events are a little jagged and not very smooth. This is "noise"
    that can be caused by a variety of events occuring inside the pore. The DNA
    molecule is not rigid inside the pore. It is moving around and displacing
    within the pore itself. Additionally, the motor protein is constantly
    undergoing conformation changes that if picked up by the nanopore can appear
    in the final signal.
  </p>
  <p class="blog-post_paragraph">
    Now that we have a good idea of what is going on on the molecular level in
    our data, we can begin to think about what steps we need to take to build
    the basecaller.
  </p>
  <p class="blog-post_paragraph">
    There are 2 big-picture approaches that can be implemented to tackle this
    challenge:
  </p>
  <ol>
    <li>
      We can filter the data at a sequence level by converting each signal from
      every timepoint into a corresponding dinucleotide and then trying to cut
      the sequence down to it&#39;s correct length
    </li>
    <li>
      We can filter the data at the signal level by trying to warp the signal in
      a way that would reduce incorrect sequencing calls and then perform the
      conversion using the k-mer table.
    </li>
  </ol>
  <p class="blog-post_paragraph">
    None of these strategies is the correct or incorrect one per-se, but the
    steps needed to implement these strategies differ greatly. Some people at
    the workshop attempted to build their basecallers using the first one,
    whereas others chose the second one. I personally opted to go with the
    second strategy because I liked the visual aid that the signal data
    provides. I have more of an intuition for when things are or aren&#39;t
    making sense looking at the signal plot compared to pure DNA code which
    would help me know if I&#39;m on the right track.
  </p>
  <p class="blog-post_paragraph">
    Now that I have a strategy chosen, I'll begin to attempt to analyze the
    data.
  </p>
  <p class="blog-post_paragraph">
    The first challenge I have to tackle is to &quot;smooth&quot; the data. Data
    smoothing is a technique used in data analysis and data science to reduce
    the noise or variability in a dataset. It involves applying a mathematical
    algorithm to the data to remove fluctuations, outliers, or other sources of
    noise in the data. The purpose of data smoothing is to reveal underlying
    trends, patterns, or relationships in the data that may be obscured by
    random variations. There are different methods of data smoothing, but for
    this case whats important to find is any jumps in the electrical signal
    indicating the strand of DNA taking a &quot;step&quot; through the motor
    protein and introducing a new DNA base into the nanopore protein.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Create a function called smooth_data 
def smooth_data(signal):
  # Subtract the first element from the second, the second from the third, and so on. Get the absolute value of the result
  smoothed_signal = np.abs(np.array(signal.signal[0:len(signal.signal) - 1]) - np.array(signal.signal[1:len(signal.signal)]))

  # Return the smoothed signal
  return smoothed_signal</code></pre>
  <p class="blog-post_paragraph">
    Now I'll apply the smoothing function to the signal from the first
    sequencing run and check the ouptut
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Calculate the smoothed signal
smoothed_signal = smooth_data(signal_1)

# Set the smoothed_signal attribute of Signal 1 to the smoothed signal
setattr(signal_1, "smoothed_signal", smoothed_signal)

# Output the smoothed signal
signal_1.smoothed_signal</code></pre>
  <pre
    class="truncated-output"
  ><code class="language-truncout">array([[5.81047189e-03],
[3.77499510e-03],
[4.99915038e-03],
[4.78641182e-03],
[2.06538339e-03],
[1.27268174e-01],
[8.99370031e-04],
[3.92364129e-03],
[5.35388267e-03],
[8.54137601e-03],
[2.09555952e-03],
[1.00411077e-02],
[4.93274440e-01],
[2.87128926e-03],
[4.95197921e-03],
[7.40934832e-03],
[5.54285469e-03],
[7.37176195e-03],
[7.96136904e-03],
[6.60146967e-03],
[1.77494917e-03],
[1.93087152e-01],
[7.28555295e-03],
[8.00252329e-03],
[3.16669206e-04],
...
[7.38786395e-03],
[5.67989496e-03],
[1.89654495e-03],
[1.77564261e-03],
[4.60943763e-03]])</code></pre>
  <p class="blog-post_paragraph">
    Now I overlayed the smoothed signal on the signal plot to see if I could
    find and identify any jumps in the electrical signal
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Plot the signal and the smoothed signal overlayed
plt.plot(signal_1.signal[0:plot_range]);
plt.plot(signal_1.smoothed_signal[0:plot_range]);
# Add the axis labels
plt.xlabel('Time point');
plt.ylabel('Signal');</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_signal_and_smoothed_signal.png"
      alt="signal_graph_with_smooth_signal_overlayed"
    />
  </div>
  <p class="blog-post_paragraph">
    I also plotted the smoothed data on a histogram to visualize the
    distribution of the differences between subsequent time points and to see if
    theres a any differences that appear less frequently than the others. These
    would likely indicate steps of the motor protein since these events occur at
    much less frequent time points and have a larger impact on the amplitude of
    the electrical current than static DNA residing in the pore.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Plot histogram of the smoothed signal
plt.hist(signal_1.smoothed_signal, bins = 100, range=[0,0.1]);</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/histogram.png"
      alt="histogram of smooth signal"
    />
  </div>
  <p class="blog-post_paragraph">
    From this histogram, it looks like most of the differences between
    subsequent time points are less than 0.03. I decided to then try calling a
    difference of 0.03 as the determining cutoff for whether DNA is static or
    whether it has taken a step inside the pore.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">cutoff = 0.03</code></pre>

  <p class="blog-post_paragraph">
    Overlaying the cutoff value on the histogram you can see the small group of
    bars belonging to large changes in the signal between 2 successive
    timepoints.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Plot histogram with cutoff line
plt.hist(signal_1.smoothed_signal, bins = 100, range=[0,0.1]);
# Add a vertical line at the cutoff point
plt.axvline(x = cutoff, color = 'red')
# Limit the x-axis to the right of the cutoff
plt.xlim(right=0.105)
# Highlight the area to the right of the cutoff
plt.axvspan(xmin=cutoff, xmax=0.12, color='orange', alpha=0.1);
# Add text to the plot
plt.text(0.069, 60, 'Any signal change in this area \nrepresents a change in \nbasepairs in the pore', fontsize = 12, color = "red", horizontalalignment = "center");</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/histogram_w_cutoff.png"
      alt="histogram of smooth signal with cutoff line"
    />
  </div>
  <p class="blog-post_paragraph">
    Going back to the signals plot, you can better visualize the moments that
    the motor protein takes a step and the DNA inside the pore changes by
    overlay this cutoff point.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Plot the smoothed signal with the cutoff line
plt.plot(signal_1.signal[0:plot_range]);
plt.plot(signal_1.smoothed_signal[0:plot_range]);
# Add the axis labels
plt.xlabel('Time point');
plt.ylabel('Signal');
# Add the cutoff line
plt.axhline(y = cutoff, color = 'red');
# Limit the y-axis to just below 1
plt.ylim(top=0.99)
# Highlight the area above the cutoff
plt.axhspan(ymin=cutoff, ymax=1, color='orange', alpha=0.1);
# Add text to the plot
plt.text(82, 0.2, f'Signal deltas \n(orange lines) \nin this area \n (exceeding the \nthreshold of {cutoff})\nrepresents a \n change in \nbasepairs \nin the pore', fontsize = 10, color = "red", horizontalalignment="center");</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/nanopore_signal_w_cutoff.png"
      alt="signal graph with cutoff line"
    />
  </div>
  <p class="blog-post_paragraph">
    Now that I have a cutoff value for determining when the DNA in the pore
    change, I can use this value to grab the signal value of the timepoints at
    which the steps happened.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">def get_movement_timepoints(signal, cutoff):
  # Checks each timepoint and assigns it boolean value if value is greater than the cutoff 
  timepoints_bool = signal.smoothed_signal > cutoff

  # Grab the indicies of the timepoints where the value was greater than the cutoff
  dna_step_timepoints = np.array([i for i, x in enumerate(timepoints_bool) if x])

  # Insert a timepoint 0 at the beginning of the array to represent the inital DNA that was present
  dna_step_timepoints = np.insert(dna_step_timepoints, 0, 0)

  # Output the timepoints
  return dna_step_timepoints

# Output the timepoints where the DNA starts moving
dna_movement_timepoints = get_movement_timepoints(signal_1, cutoff)</code></pre>
  <pre
    class="output no-lang-display"
  ><code class="language-output">array([  0,   5,  12,  21,  30,  39,  45,  53,  58,  63,  93, 101, 106,
121, 129, 136, 146, 157, 173, 185, 197, 214, 229, 235, 244, 260,
272, 298, 306, 318, 327, 336, 351, 359, 373, 388, 394, 410, 422,
436, 446, 452, 463, 476, 499, 503, 506, 513, 524, 535, 542, 549,
556, 567, 577, 588, 599, 609, 621, 625, 636, 649, 656, 664, 676,
693, 708, 717, 728, 734, 740, 753, 759, 765, 775, 788, 798, 807,
815, 826, 839, 850, 861, 869, 883, 895, 904, 913, 924, 931, 939,
951, 961, 972, 982, 993])
  </code></pre>
  <p class="blog-post_paragraph">
    Overlaying these timepoints on the electrical current graph, you can
    visualize the moments these steps occur.
  </p>
  <pre class="line-numbers"><code class="language-python"># Plot the signal 
plt.plot(signal_1.signal[0:100]);
# Limit the x-axis to the first 100 timepoints
plt.xlim(right=100)
# Add axis labels
plt.xlabel('Time point');
plt.ylabel('Signal');

# Highlight the locations where DNA movement was detected
for i in dna_movement_timepoints:
  if i > 0:
    plt.axvspan(xmin=i, xmax=i+1, color='red', alpha=0.3);</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signals_highlighted_thresholds_only.png"
      alt="signal graph with threshold regions highlighted"
    />
  </div>
  <p>
    As expected, the step values calculated correspond perfectly to the large
    vertical-like elements in the electrical signal graph. This suggests that
    the threshold value determined through the data smoothing was a good value
    to use to determine when the DNA in the pore changes.
  </p>
  <p class="blog-post_paragraph">
    At this stage, I can to start to work towards isolating data from the
    periods where the DNA is static in the pore. The plan is for each DNA
    capture event, isolate one value that I can use to compare to the k-mer
    table and start basecalling. Practically, this can be broken down into two
    steps:
  </p>
  <ol>
    <li>
      <p class="blog-post_paragraph">
        Grabbing all the timepoints between DNA movement events to isolate all
        the data for each DNA capture event
      </p>
    </li>
    <li>
      <p class="blog-post_paragraph">
        Grab the average value of the electrical current for that DNA capture
        event
      </p>
    </li>
  </ol>
  <p class="blog-post_paragraph">
    Having a list of these these average values from the DNA capture events that
    we can use to cross-reference the k-mer table will almost be all the
    information needed to create the corresponding DNA sequence.
  </p>
  <p class="blog-post_paragraph">Let's start with the first step.</p>
  <pre
    class="line-numbers"
  ><code class="language-python">def get_capture_timepoints(signal, cutoff):
  # Create a list to store the timepoints where the DNA was captured
  dna_capture_timepoints = list()

  # Loop through each timepoint and calculate the mean of the signal values between each timepoint
  for i in range(0,len(dna_movement_timepoints)-1):
    minimum = dna_movement_timepoints[i]
    maximum = dna_movement_timepoints[(i+1)]+1 # Add 1 to capture 100% of time before the DNA moves
    dna_capture_timepoints.append(list(range(minimum,maximum)))
    
  # Return the list of timepoints
  return dna_capture_timepoints

# Save it to a variable
dna_capture_timepoints = get_capture_timepoints(signal_1, cutoff)

# Print the groups of timepoints where the DNA was captured
print(f"The DNA was captured at the following timepoints: {dna_capture_timepoints}")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The DNA was captured at the following timepoints: [[0, 1, 2, 3, 4, 5], [5, 6, 7, 8, 9, 10, 11, 12], [12, 13, 14, 15, 16, 17, 18, 19, 20, 21], [21, 22, 23, 24, 25, 26, 27, 28, 29, 30], [30, 31, 32, 33, 34, 35, 36, 37, 38, 39], [39, 40, 41, 42, 43, 44, 45], [45, 46, 47, 48, 49, 50, 51, 52, 53], [53, 54, 55, 56, 57, 58], [58, 59, 60, 61, 62, 63], [63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93], [93, 94, 95, 96, 97, 98, 99, 100, 101], [101, 102, 103, 104, 105, 106], [106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121], [121, 122, 123, 124, 125, 126, 127, 128, 129], [129, 130, 131, 132, 133, 134, 135, 136], [136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146], [146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157], [157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173], [173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185], [185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197], [197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214], [214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229], [229, 230, 231, 232, 233, 234, 235], [235, 236, 237, 238, 239, 240, 241, 242, 243, 244], [244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260], [260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272], [272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298], [298, 299, 300, 301, 302, 303, 304, 305, 306], [306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318], [318, 319, 320, 321, 322, 323, 324, 325, 326, 327], [327, 328, 329, 330, 331, 332, 333, 334, 335, 336], [336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351], [351, 352, 353, 354, 355, 356, 357, 358, 359], [359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373], [373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388], [388, 389, 390, 391, 392, 393, 394], [394, 395, 396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410], [410, 411, 412, 413, 414, 415, 416, 417, 418, 419, 420, 421, 422], [422, 423, 424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435, 436], [436, 437, 438, 439, 440, 441, 442, 443, 444, 445, 446], [446, 447, 448, 449, 450, 451, 452], [452, 453, 454, 455, 456, 457, 458, 459, 460, 461, 462, 463], [463, 464, 465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475, 476], [476, 477, 478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499], [499, 500, 501, 502, 503], [503, 504, 505, 506], [506, 507, 508, 509, 510, 511, 512, 513], [513, 514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524], [524, 525, 526, 527, 528, 529, 530, 531, 532, 533, 534, 535], [535, 536, 537, 538, 539, 540, 541, 542], [542, 543, 544, 545, 546, 547, 548, 549], [549, 550, 551, 552, 553, 554, 555, 556], [556, 557, 558, 559, 560, 561, 562, 563, 564, 565, 566, 567], [567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577], [577, 578, 579, 580, 581, 582, 583, 584, 585, 586, 587, 588], [588, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599], [599, 600, 601, 602, 603, 604, 605, 606, 607, 608, 609], [609, 610, 611, 612, 613, 614, 615, 616, 617, 618, 619, 620, 621], [621, 622, 623, 624, 625], [625, 626, 627, 628, 629, 630, 631, 632, 633, 634, 635, 636], [636, 637, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649], [649, 650, 651, 652, 653, 654, 655, 656], [656, 657, 658, 659, 660, 661, 662, 663, 664], [664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 676], [676, 677, 678, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692, 693], [693, 694, 695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 707, 708], [708, 709, 710, 711, 712, 713, 714, 715, 716, 717], [717, 718, 719, 720, 721, 722, 723, 724, 725, 726, 727, 728], [728, 729, 730, 731, 732, 733, 734], [734, 735, 736, 737, 738, 739, 740], [740, 741, 742, 743, 744, 745, 746, 747, 748, 749, 750, 751, 752, 753], [753, 754, 755, 756, 757, 758, 759], [759, 760, 761, 762, 763, 764, 765], [765, 766, 767, 768, 769, 770, 771, 772, 773, 774, 775], [775, 776, 777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787, 788], [788, 789, 790, 791, 792, 793, 794, 795, 796, 797, 798], [798, 799, 800, 801, 802, 803, 804, 805, 806, 807], [807, 808, 809, 810, 811, 812, 813, 814, 815], [815, 816, 817, 818, 819, 820, 821, 822, 823, 824, 825, 826], [826, 827, 828, 829, 830, 831, 832, 833, 834, 835, 836, 837, 838, 839], [839, 840, 841, 842, 843, 844, 845, 846, 847, 848, 849, 850], [850, 851, 852, 853, 854, 855, 856, 857, 858, 859, 860, 861], [861, 862, 863, 864, 865, 866, 867, 868, 869], [869, 870, 871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882, 883], [883, 884, 885, 886, 887, 888, 889, 890, 891, 892, 893, 894, 895], [895, 896, 897, 898, 899, 900, 901, 902, 903, 904], [904, 905, 906, 907, 908, 909, 910, 911, 912, 913], [913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923, 924], [924, 925, 926, 927, 928, 929, 930, 931], [931, 932, 933, 934, 935, 936, 937, 938, 939], [939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949, 950, 951], [951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961], [961, 962, 963, 964, 965, 966, 967, 968, 969, 970, 971, 972], [972, 973, 974, 975, 976, 977, 978, 979, 980, 981, 982], [982, 983, 984, 985, 986, 987, 988, 989, 990, 991, 992, 993]]</code></pre>
  <p class="blog-post_paragraph">
    Once again it would be helpful to visualize these ranges on the signals
    graph to ensure the ranges are correct.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">plt.plot(signal_1.signal[0:100]);
# Limit the x-axis to the first 100 timepoints
plt.xlim(right=100)
# Add axis labels
plt.xlabel('Time point');
plt.ylabel('Signal');

for i,v in enumerate(dna_capture_timepoints):
  # Highlight ranges where the DNA was captured
  plt.axvspan(xmin=v[0], xmax=v[-1], color='green', alpha=0.1);
  
  # Highlight the locations where DNA movement was detected
  if i > 0 and i in dna_movement_timepoints :
    plt.axvspan(xmin=i, xmax=i+1, color='red', alpha=0.3);</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signals_highlighted_w_overlaps.png"
      alt="signal graph with capture regions highlighted but with overlapping capture regions and threshold regions"
    />
  </div>
  <p class="blog-post_paragraph">
    Here you can see the ranges that were just calculated highlighted in green.
    Upon first glance everything looks perfect. As you'd expect, the ranges
    correspond to the flat-like parts of the signal graph, corresponding to the
    moments in time where the DNA is captured in the pore. However the problem
    is that by capturing all the timepoints between the moments the DNA moves
    determined through the threshold, there may be some brief moments in time
    where the DNA starts to move or is just about to finish moving through the
    pore that are also classified as a moment of time where the DNA is captured.
    Because the slope of the electrical current is so steep at the moments the
    DNA moves, these brief moments can have a huge effect on the average signal
    value and can cause the basecaller to make a mistake when comparing the
    average signal value to the dinucleotide table.
  </p>
  <p class="blog-post_paragraph">
    You can notice this on the graph by looking at where the green regions and
    the red regions touch. If you look closely, there are a few moments where
    the green regions may be picking up or close to picking up values where the
    DNA is moving through the pore. And because this graph is only a small
    portion of the entire signal graph, there are probably many more moments
    like this in the entire signal graph that have a more severe effect on the
    average signal value. To fix this problem I decided add a little bit of
    buffer between the minimum and maximum capture values to shorten the range
    of capture. This is what the code looks like:
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">def get_capture_timepoints(signal, cutoff):
  # Reinitialize the list to store the timepoints where the DNA was captured
  dna_capture_timepoints = list()

  # Loop through each timepoint and calculate the mean of the signal values between each timepoint
  for i in range(0,len(dna_movement_timepoints)-1):
    
    """Modify the miniminum and maximum values to make the capture range a bit shorter
    Add a little less buffer to the very start to capture more of the initial capture event""" 
    minimum_buffer = 2
    maximum_buffer = 0
    minimum = dna_movement_timepoints[i] + 1 if i == 0 else dna_movement_timepoints[i] + minimum_buffer
    maximum = dna_movement_timepoints[(i+1)] - maximum_buffer
    
    # The rest of the code is the same as before
    dna_capture_timepoints.append(list(range(minimum,maximum)))
    
  # Return the list of timepoints
  return dna_capture_timepoints

# Save it to a variable
dna_capture_timepoints = get_capture_timepoints(signal_1, cutoff)
  
# Print the groups of timepoints where the DNA was captured
print(f"The DNA was captured at the following timepoints: {dna_capture_timepoints}")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The DNA was captured at the following timepoints: [[1, 2, 3, 4], [7, 8, 9, 10, 11], [14, 15, 16, 17, 18, 19, 20], [23, 24, 25, 26, 27, 28, 29], [32, 33, 34, 35, 36, 37, 38], [41, 42, 43, 44], [47, 48, 49, 50, 51, 52], [55, 56, 57], [60, 61, 62], [65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92], [95, 96, 97, 98, 99, 100], [103, 104, 105], [108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120], [123, 124, 125, 126, 127, 128], [131, 132, 133, 134, 135], [138, 139, 140, 141, 142, 143, 144, 145], [148, 149, 150, 151, 152, 153, 154, 155, 156], [159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172], [175, 176, 177, 178, 179, 180, 181, 182, 183, 184], [187, 188, 189, 190, 191, 192, 193, 194, 195, 196], [199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213], [216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228], [231, 232, 233, 234], [237, 238, 239, 240, 241, 242, 243], [246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259], [262, 263, 264, 265, 266, 267, 268, 269, 270, 271], [274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297], [300, 301, 302, 303, 304, 305], [308, 309, 310, 311, 312, 313, 314, 315, 316, 317], [320, 321, 322, 323, 324, 325, 326], [329, 330, 331, 332, 333, 334, 335], [338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350], [353, 354, 355, 356, 357, 358], [361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372], [375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387], [390, 391, 392, 393], [396, 397, 398, 399, 400, 401, 402, 403, 404, 405, 406, 407, 408, 409], [412, 413, 414, 415, 416, 417, 418, 419, 420, 421], [424, 425, 426, 427, 428, 429, 430, 431, 432, 433, 434, 435], [438, 439, 440, 441, 442, 443, 444, 445], [448, 449, 450, 451], [454, 455, 456, 457, 458, 459, 460, 461, 462], [465, 466, 467, 468, 469, 470, 471, 472, 473, 474, 475], [478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498], [501, 502], [505], [508, 509, 510, 511, 512], [515, 516, 517, 518, 519, 520, 521, 522, 523], [526, 527, 528, 529, 530, 531, 532, 533, 534], [537, 538, 539, 540, 541], [544, 545, 546, 547, 548], [551, 552, 553, 554, 555], [558, 559, 560, 561, 562, 563, 564, 565, 566], [569, 570, 571, 572, 573, 574, 575, 576], [579, 580, 581, 582, 583, 584, 585, 586, 587], [590, 591, 592, 593, 594, 595, 596, 597, 598], [601, 602, 603, 604, 605, 606, 607, 608], [611, 612, 613, 614, 615, 616, 617, 618, 619, 620], [623, 624], [627, 628, 629, 630, 631, 632, 633, 634, 635], [638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648], [651, 652, 653, 654, 655], [658, 659, 660, 661, 662, 663], [666, 667, 668, 669, 670, 671, 672, 673, 674, 675], [678, 679, 680, 681, 682, 683, 684, 685, 686, 687, 688, 689, 690, 691, 692], [695, 696, 697, 698, 699, 700, 701, 702, 703, 704, 705, 706, 707], [710, 711, 712, 713, 714, 715, 716], [719, 720, 721, 722, 723, 724, 725, 726, 727], [730, 731, 732, 733], [736, 737, 738, 739], [742, 743, 744, 745, 746, 747, 748, 749, 750, 751, 752], [755, 756, 757, 758], [761, 762, 763, 764], [767, 768, 769, 770, 771, 772, 773, 774], [777, 778, 779, 780, 781, 782, 783, 784, 785, 786, 787], [790, 791, 792, 793, 794, 795, 796, 797], [800, 801, 802, 803, 804, 805, 806], [809, 810, 811, 812, 813, 814], [817, 818, 819, 820, 821, 822, 823, 824, 825], [828, 829, 830, 831, 832, 833, 834, 835, 836, 837, 838], [841, 842, 843, 844, 845, 846, 847, 848, 849], [852, 853, 854, 855, 856, 857, 858, 859, 860], [863, 864, 865, 866, 867, 868], [871, 872, 873, 874, 875, 876, 877, 878, 879, 880, 881, 882], [885, 886, 887, 888, 889, 890, 891, 892, 893, 894], [897, 898, 899, 900, 901, 902, 903], [906, 907, 908, 909, 910, 911, 912], [915, 916, 917, 918, 919, 920, 921, 922, 923], [926, 927, 928, 929, 930], [933, 934, 935, 936, 937, 938], [941, 942, 943, 944, 945, 946, 947, 948, 949, 950], [953, 954, 955, 956, 957, 958, 959, 960], [963, 964, 965, 966, 967, 968, 969, 970, 971], [974, 975, 976, 977, 978, 979, 980, 981], [984, 985, 986, 987, 988, 989, 990, 991, 992]]</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signals_highlighted.png"
      alt="signal graph with capture regions highlighted containing no overlaps"
    />
  </div>
  <p class="blog-post_paragraph">
    You can see that by shortening the ranges for when the signal is considered
    captured, you can ensure that all the electrical signal data present in
    those ranges corresponds to moments where the DNA is static only. You lose a
    little bit of accuracy but not capturing all the data from when the DNA is
    captured, but this is seemingly marginal as enough data is still captured in
    the ranges to calculate a summary value for each timepoint and compare to
    the dinucleotide table.
  </p>
  <p class="blog-post_paragraph">
    You can see this distinction better when you overlay the two graphs in gif
    form. Notice the lack of overlaps between the capture ranges and the ranges
    when the DNA is moving in the second graph.
  </p>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signal_highlights.gif"
      alt="comparison of full capture ranges and shortened capture ranges"
    />
  </div>
  <p>
    Now that I have a strong certainty that timepoints corresponding to the DNA
    capture events are accurate, I can begin to determine what the true signal
    value should be for each capture event. The easiest way to do this would be
    to just take the average value for each capture event.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python">def get_mean(signal, capture_ranges):    
  # Get the minimum and maximum timepoints for the capture range 
  minimum = capture_ranges[0]
  maximum = capture_ranges[-1]
  
  # If the minimum and maximum are different, take the mean of the signal between those timepoints
  if minimum != maximum:
    signal_means = np.mean(signal.signal[minimum:maximum])
    
  # If the minimum and maximum are the same, the mean is just the signal at that timepoint
  elif minimum == maximum:
    signal_means = signal.signal[minimum]
    
  # Return the mean signal value
  return signal_means


def get_capture_means (signal, capture_ranges):
  # Initialize a list to store the mean signal values
  dna_capture_means = list()

  # Loop through each capture range and calculate the mean of the signal values between each timepoint
  for i in capture_ranges:
    dna_capture_means.append(get_mean(signal, i))
  
  # Return the list of mean signal values
  return dna_capture_means

# Save it to a variable
dna_capture_means = get_capture_means(signal_1, dna_capture_timepoints)

# Print the means of the signal values between each timepoint
print(f"The mean signal values between each timepoint were: {dna_capture_means}")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The mean signal values between each timepoint were: [0.3150131662892894, 0.4388658230736439, 0.9382180045874408, 0.7492516139178393, 0.0648448362336262, 0.31416471372489374, 0.24871154796850795, 0.1308048500261873, 0.6879306975387933, 0.9370147095493105, 0.8159946817442971, 0.24863837283436427, 0.18605255407371368, 0.751014020900838, 0.004255596959964134, 0.18929806587588077, 0.7514755426453339, 0.0011823610953436306, 0.187467075517471, 0.7507884372502139, 2.5870896072124716e-05, 0.1262400931431966, 0.6240087553876384, 0.6866547933581, 0.9386703979027454, 0.8128445892628149, 0.31167593379749114, 0.2516419592575371, 0.06132959678341394, 0.24987594891890097, 0.18731848359063394, 0.8749174700728205, 0.5608392456658148, 0.4379674418878848, 0.7514539619115898, 0.06256210834128777, 0.3131973515651855, 0.4376931440899407, 0.8755278991756956, 0.6893415491698264, 0.8153524156074078, 0.311137568574452, 0.3737306713764218, 0.6253556444287833, 0.5589792437418337, array([0.37872502]), 0.5013373929581939, 0.0633991898846784, 0.31232163205244223, 0.37451955745237364, 0.6851198704100576, 0.7497175931477327, 0.12481725738905967, 0.6217719306961202, 0.5622003206562043, 0.37614806672730927, 0.5022579756624813, -0.00047589780351690376, 0.19090715855006127, 0.7499095545393254, 0.06350294507341912, 0.314077071113244, 0.37549389148237877, 0.5629573518749411, 0.3748965636638899, 0.5626400191947448, 0.4364733853164748, 0.8757241408520289, 0.5606866969884582, 0.2482026347335776, 0.06072948755345618, 0.3761473448185078, 0.6883768600160365, 0.8748179870203537, 0.5624995114267668, 0.3147353655147443, 0.43951941642678233, 0.8135594790161494, 0.37368900538425576, 0.5634998303778214, 0.4389919970670458, 0.7494307820780715, 0.06355513434040884, 0.43699445180493124, 0.9394337642485967, 0.8754225300569566, 0.6872994354595927, 0.7498742724964946, 0.06354843720884656, 0.2512742474294959, 0.06273223469201125, 0.24919664305885744, 0.19024481785117403, 0.8747400619650362, 0.6912431853106157]</code></pre>
  <p class="blog-post_paragraph">
    Now that we have the average current values, the next step is to find a way
    to determine which k-mer signature value it's closest to. One common way of
    performing this is to calculate the Euclidean distance between event and
    each kmer. The Euclidean distance measures the straight-line distance
    between two points in any-dimensional space (2-dimensions in this case). To
    compute the Euclidean distance, you calculate the squared difference between
    corresponding coordinates of the two points, sum up these squared
    differences across all dimensions, and take the square root of the result to
    obtain the final distance value. By finding the k-mer signature with the
    shortest euclidean distance to the average current value for a particular
    capture range, you can begin to basecall the DNA captured in the pore at
    that moment.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python"># Write a function to calculate the euclidian distance between two values
def euclidian_distance(x,y):
  dist = np.sqrt(np.square((x - y)))
  return dist

# Write a function to basecall the dinucleotide based on the mean signal value
def basecall(signal, capture_ranges, kmer_table):
  # Get the mean signal value for the capture range
  capture_mean = get_mean(signal, capture_ranges)
  
  # Initialize an empty list to store the euclidian distances between the capture mean and each dinucleotide signature
  euc_distances = list()
  
  # Loop through each dinucleotide signature and calculate the euclidian distance between the capture mean and the signature
  for dinuc, signature in kmer_table.items():
    euc_distances.append((dinuc, euclidian_distance(capture_mean, signature)))
    
  # Get the minimum euclidian distance
  min_dist = np.min([x[1] for x in euc_distances])
  
  # Make a basecall
  dinuccall = [x[0] for x in euc_distances if x[1] == min_dist][0]
  
  # Return the basecall
  return dinuccall

def basecall_signal(signal, capture_ranges, kmer_table):
  # Initialize a list to store the basecalls
  basecalls = []

  # Loop through each capture range and basecall each capture range
  for cap_range in capture_ranges:
    basecalls.append(basecall(signal, cap_range, kmer_table))
  
  # Return the list of basecalls
  return basecalls

# Initialize a list to store the basecalls
basecalls = basecall_signal(signal_1, dna_capture_timepoints, kmer_table)

# Print the basecalls
print(basecalls)</code></pre>
  <pre
    class="output"
  ><code class="language-output">['CC', 'CT', 'TT', 'TA', 'AC', 'CC', 'CA', 'AG', 'GT', 'TT', 'TC', 'CA', 'AT', 'TA', 'AA', 'AT', 'TA', 'AA', 'AT', 'TA', 'AA', 'AG', 'GG', 'GT', 'TT', 'TC', 'CC', 'CA', 'AC', 'CA', 'AT', 'TG', 'GC', 'CT', 'TA', 'AC', 'CC', 'CT', 'TG', 'GT', 'TC', 'CC', 'CG', 'GG', 'GC', 'CG', 'GA', 'AC', 'CC', 'CG', 'GT', 'TA', 'AG', 'GG', 'GC', 'CG', 'GA', 'AA', 'AT', 'TA', 'AC', 'CC', 'CG', 'GC', 'CG', 'GC', 'CT', 'TG', 'GC', 'CA', 'AC', 'CG', 'GT', 'TG', 'GC', 'CC', 'CT', 'TC', 'CG', 'GC', 'CT', 'TA', 'AC', 'CT', 'TT', 'TG', 'GT', 'TA', 'AC', 'CA', 'AC', 'CA', 'AT', 'TG', 'GT']</code></pre>
  <pre
    class="line-numbers"
  > <code class="language-python"># Define a function to plot the signal and basecalls
def visualize_basecalls(signal, capture_ranges, basecalls):
  
  # Define a function to help place the basecalls where they belong on the graph
  def get_label_details(timepoints: "list") -> "tuple(int, str)":
    # Get the corresponding signals array for the given time range
    signals_from_timepoints = signal_1.signal[timepoints]
    
    # Get the maximum value of the signal for the given time range (this will be used to generate the y-axis value of the label)
    max_signal = np.max(signals_from_timepoints)
    
    # Basecall the average of the signal for each timepoint
    dinuccall = basecall(signal_1, timepoints, kmer_table)
    
    # Return y-axis value (max_signal + some padding) and label (dinculeotide)
    return (max_signal + 0.0225, dinuccall)

  # Define a function to plot the colorized and decolorized subplots
  def plot_subplot(plot_num: "int (0 or 1)", with_color: "bool") -> None:
    #Plot subplots of the signal with annotations
    axs[plot_num].plot(signal_1.signal[0:100])
    axs[plot_num].set_xlim(right=100)
    axs[plot_num].set_ylim(top=1)
    axs[plot_num].set_xlabel('Time point')
    axs[plot_num].set_ylabel('Signal')
    
    # Loop through each capture range and plot the corresponding label
    for i, v in enumerate(dna_capture_timepoints):

      # Color the background of the plot if with_color is True
      if with_color:
        axs[plot_num].axvspan(xmin=v[0], xmax=v[-1], color='green', alpha=0.1)
        if i > 0 and i in dna_movement_timepoints:
          axs[0].axvspan(xmin=i, xmax=i+1, color='red', alpha=0.3)
          
      # Only label the first 11 labels because there are 11 visible DNA capture events in the first 100 time points
      if i < 11:
        label_data = get_label_details(v)
        x_center = (v[0] + v[-1]) / 2  # calculate the center of the green area
        y_center = label_data[0] 
        basecall = label_data[1] 
        axs[plot_num].text(x_center, y_center, basecall, ha='center', va='center', fontsize=10) # ha = horizontal alignment, va = vertical alignment

  # Set up the figure with two subplots
  fig, axs = plt.subplots(2, 1, figsize=(7, 10))

  # Plot the colorized graph on top
  plot_subplot(plot_num = 0, with_color = True)
  # Plot the decolorized graph on top
  plot_subplot(plot_num = 1, with_color = False) 

# Output the graphs
visualize_basecalls(signal_1, dna_capture_timepoints, basecalls)</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signals_with_dinucleotide_labels.png"
      alt="signal graph with dinucleotide labels"
    />
  </div>
  <p class="blog-post_paragraph">
    We are almost done. You can see now that each capture point has the
    nucleotides that are present in the pore resolved. So at this stage it would
    seem intuitive to just collapse all the dinucleotides down to a single
    sequence and have that be the final sequence of DNA. However remember that
    the motor protein feeds the DNA through the pore one step (or one DNA base)
    at a time. Even though the pore can hold two basepairs (in this example),
    each base within the pore at adjacent capture periods is not unique. If you
    look at the graph, you can notice that the last DNA base at any capture
    event matches the first DNA base at the next capture event. This means that
    to get the final sequence, you have to only take last DNA base into account
    (the new one that just entered the pore), in order to not "double-count"
    bases. The only exception to this is the first capture event, where we can
    clearly see in these examples that it starts with both bases already in the
    pore.
  </p>
  <pre
    class="line-numbers"
  ><code class="language-python">def collapse_seq(basecalls: "list") -> "str":
  """Collapse a list of basecalls into a single string"""
  
  # Initialize an empty string to store the final sequence
  final_seq = ""
  
  # Loop through each dinucleotide in the list of basecalls
  for i, v in enumerate(basecalls):
    
    # If it's the first basecall, add both bases to the final sequence
    if i == 0:
      final_seq += v
      continue
    
    # Otherwise, add the second base to the final sequence
    final_seq += v[-1]
    
  # Return the final sequence
  return final_seq

# Assign the collapsed sequence to a variable
basecalled_seq = collapse_seq(basecalls)

# Set the basecalled sequence as the sequence attribute of the signal
setattr(signal_1, "basecalled_seq", basecalled_seq)

# Print the collapsed sequence
print(f"The basecalled sequence for the first signal is: {signal_1.basecalled_seq}")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The basecalled sequence for the first signal is: CCTTACCAGTTCATAATAATAAGGTTCCACATGCTACCTGTCCGGCGACCGTAGGCGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT</code></pre>
  <p class="blog-post_paragraph">
    Now that we have the first version of the basecaller built, we can measure
    its accuracy by comparing it to the correct sequence for that signal.
  </p>
  <p class="blog-post_paragraph">
    edlib is a quick sequence alignment package that can be used in python. I'll
    use this to measure my alignment accuracy.
  </p>
  <pre class="line-numbers"> <code class="language-python">import edlib

def align(basecalled_seq: "str", correct_seq: "str") -> "dict":
  """Align two sequences using edlib"""
  
  # Perform the alignment
  align_result = edlib.align(basecalled_seq, correct_seq, task='path')
  
  # Format the alignment result
  nice_alignment = edlib.getNiceAlignment(align_result, basecalled_seq, correct_seq)
  
  # Return the alignment result
  # print(align_result)
  return align_result, nice_alignment

# Perform the alignment
align_result, nice_alignment = align(signal_1.basecalled_seq, signal_1.correct_seq)

# Print the alignment
print("basecalled: " + nice_alignment['query_aligned'])
print("            " + nice_alignment['matched_aligned'])
print("   correct: " + nice_alignment['target_aligned'])</code></pre>
  <pre
    class="output"
  ><code class="language-output">basecalled: CCTTACCAGTT-CATAATAA-TAAGGTTCC-ACATGCTACCTGTCCGG-CGACCGTAGG-CGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT
            |||||||||||-||||||||-|||||||||-|||||||||||||||||-||||||||||-|||||||||||||||||||||||||||||||||||||||||
   correct: CCTTACCAGTTTCATAATAAATAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGGGCGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT
  </code></pre>

  <pre
    class="line-numbers"
  > <code class="language-python"># Measure the accuracy of the basecaller
def measure_accuracy (alignment: "dict", correct_seq: "str") -> "float":
  acc = (1 - alignment['editDistance'] / len(correct_seq)) * 100  
  return acc

# Store the accuracy of the basecaller for the first signal
seq_acc = measure_accuracy(align_result, signal_1.correct_seq)

# Print the accuracy
print(f"The accuracy of the basecaller is: {seq_acc}%")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The accuracy of the basecaller is: 95.04950495049505%</code></pre>
  <p class="blog-post_paragraph">
    The first version of the basecaller was able to achieve an accuracy of 95%.
    Not a bad start. If you look closely at the sequence, you'll notice that the
    mismatches only occur at repetative sequences. This is a notorious problem
    for nanopore sequencing, known as "homopolymer indel errors". Homopolymers
    are repetitive sequences of the same nucleotide base (A, T, C, or G) within
    a DNA or RNA molecule. They can occur naturally or as a result of DNA or RNA
    synthesis or replication errors. For example, a homopolymer could consist of
    a repeated sequence of "A" nucleotides, such as "AAAAAA". The main challenge
    with homopolymers in nanopore sequencing arises from the difficulty in
    distinguishing the duration of time each nucleotide spends in the pore.
    Since each nucleotide in a homopolymer has the same base, the electrical
    signals produced by the pore can be similar or indistinguishable for
    consecutive nucleotides. As a result, the sequencing process may struggle to
    accurately determine the exact length or number of repeated nucleotides in a
    homopolymer region, and as a result insert or delete a DNA base at that
    sequence. To mitigate the challenges posed by homopolymers, various
    strategies are employed in nanopore sequencing, such as using specialized
    base-calling algorithms, employing specific sequencing chemistries or pore
    types that improve homopolymer resolution, and combining nanopore sequencing
    with other technologies for validation or error correction.
  </p>
  <p class="blog-post_paragraph">
    For this example, I'll apply a similar approach to determining homopolymers
    as I did with the signals. I'll plot out a histogram of a certain signal
    feature and I'll try to find a cutoff point that can distingusish between
    two groupings, homopolymer-containing signals and non-homopolymer-continaing
    signals. Since homopolymers typically don't of change to the electrical
    current, organizing the predicted capture regions by their size and then
    looking at the regions with longer durations may be best way at finding
    homopolymers in the signal.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python"># Initialize a list to store lengths of the capture ranges
capture_ranges_lengths = []

# Loop through each capture range and append the length of the capture range to the list
for i in dna_capture_timepoints:
  capture_ranges_lengths.append(len(i))
  
# Print the lengths of the capture ranges
print(f"The lengths (aka number of timepoints in) of the capture ranges are: {sorted(capture_ranges_lengths)}")</code></pre>
  <pre
    class="output"
  ><code class="language-output">The lengths (aka number of timepoints in) of the capture ranges are: [1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 11, 11, 11, 11, 11, 12, 12, 12, 13, 13, 13, 13, 13, 14, 14, 14, 15, 15, 21, 24, 28]</code></pre>
  <pre
    class="line-numbers"
  > <code class="language-python"># Plot a histogram of the capture range lengths
plt.hist(capture_ranges_lengths, bins=25, range=[0,30]);</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/capture_length_histogram.png"
      alt="histogram of capture range lengths"
    />
  </div>
  <p class="blog-post_paragraph">
    Here you can once again see what looks to be one grouping of ranges followed
    by a couple of outliers on the histogram. The outliers are longer signal
    reads so this indicates that these are likely homopolymers. Since they all
    seem to exist past the the 20 timepoints mark, that seems like a good place
    to try putting a threshold to find homopolymers.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python">homopolymer_cutoff = 20</code></pre>
  <pre
    class="line-numbers"
  > <code class="language-python"># Plot a histogram of the capture range lengths
plt.hist(capture_ranges_lengths, bins=25, range=[0,30]);
# add a vertical line to show the cutoff
plt.axvline(x = homopolymer_cutoff, color = 'purple');
# Limit the x-axis to the right of the cutoff
plt.xlim(right=31)
# Highlight the area to the right of the cutoff
plt.axvspan(xmin=homopolymer_cutoff, xmax=31, color='purple', alpha=0.1);
# Add text to the plot
plt.text(26, 
         10, 
         'Capture events \nthis long \nare likely \nhomopolymers', 
         fontsize = 12, 
         color = "purple", 
         horizontalalignment = "center");</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/capture_length_histogram_w_cutoff.png"
      alt="histogram of capture range lengths with homopolymer cutoff"
    />
  </div>
  <p class="blog-post_paragraph">
    Now that we have a threshold point, we can try to adjust the capture
    timepoints factor in homopolymers. If we find suspected homopolymers in the
    signal, we can split up the capture ranges bases on how many homopolymers
    are suspected to be there.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python">def adjust_capture_timepoints(capture_timepoints, homopolymer_cutoff):
  # Make a copy of the capture timepoints to not modify the original argument
  capture_ranges = capture_timepoints.copy()

  # Loop through each capture range
  for i, capture_range in enumerate(capture_ranges):
  
    # If the capture range is long enough to be a suspected homopolymer
    if len(capture_range) >= homopolymer_cutoff:
      # Split the capture range in half
      split_point = int(len(capture_range) / 2)
      capture_ranges[i] = capture_range[:split_point]

      # Add the second half of the capture range to the next index in the list
      # If this range is also a suspected homopolymer, it will be split in the next iteration
      capture_ranges.insert(i+1, capture_range[split_point:])
    
  # Remove any empty lists that may have been created by the splitting
  capture_ranges = [x for x in capture_ranges if x != []]
  
  # Return the updated capture ranges
  return capture_ranges

dna_capture_timepoints = adjust_capture_timepoints(dna_capture_timepoints, homopolymer_cutoff)</code></pre>
  <p class="blog-post_paragraph">
    In the first 100 timepoints that we graphed, there was a homopolymer (a
    "TTT" subsequence that was called as a "TT"). Therefore we can visualize the
    same range as before and check if this was correctly adjusted.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python">visualize_basecalls(signal_1, dna_capture_timepoints, basecalls)</code></pre>
  <pre
    class="output-img no-code"
  ><code class="language-output">&nbsp;</code></pre>
  <div class="workshowcase-img-container">
    <img
      class="img-fluid"
      src="assets/img/basecaller_blog_post/signals_with_dinucleotide_labels_homopolymers_fixed.png"
      alt="signal graph with dinucleotide labels and homopolymers fixed"
    />
  </div>
  <p class="blog-post_paragraph">
    It worked perfectly! You can see the longest flat portion of the signal of
    this graph get split into 2 basecalls. The "GT", "TT" and "TT" calls will
    get flattened into a "TTT" call at the very end.
  </p>
  <p class="blog-post_paragraph">
    Lets run the rest of the pipeline to see if this improved the total
    sequencing accuracy.
  </p>
  <pre class="line-numbers"> <code class="language-python"># Basecall the signal
basecalls = basecall_signal(signal_1, dna_capture_timepoints, kmer_table)

# Collapse the basecalls into a single sequence
final_sequence = collapse_seq(basecalls)

# Set the basecalled sequence as an attribute of the signal object
setattr(signal_1, "basecalled_seq", final_sequence)

# Align the basecalled sequence to the correct sequence
alignment, nice_alignment = align(signal_1.basecalled_seq, signal_1.correct_seq)

# Calculate the accuracy of the basecaller
accuracy = measure_accuracy(alignment, signal_1.correct_seq)

# Print the alignment and the accuracy
print("basecalled: " + nice_alignment['query_aligned'])
print("            " + nice_alignment['matched_aligned'])
print("   correct: " + nice_alignment['target_aligned'])
print("\n")
print(f"The accuracy of the basecaller is: {accuracy}%")</code></pre>
  <pre
    class="output"
  ><code class="language-output">basecalled: CCTTACCAGTTTCATAATAA-TAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGG-CGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT
    ||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||
correct: CCTTACCAGTTTCATAATAAATAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGGGCGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT


The accuracy of the basecaller is: 98.01980198019803%</code></pre>
  <p class="blog-post_paragraph">
    With a simple tweak, I was able to improve the accuracy of the basecaller to
    98% percent for this particular sequnece. Now I can go ahead an run it on
    all 200 signals, and determine the average accuracy of the basecaller I just
    designed.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python"># Putting it all together
def basecall_all(
  signals: "list", 
  kmer_signatures: "dict", 
  correct_seqs: "list", 
  movement_cutoff: "float" = 0.03, 
  homopolymer_cutoff: "int" = 20,
  detailed: "bool" = True) -> "list":
  
  # Initialize a list to store the accuracies of each signal
  accuracies = []
  
  # Loop through each signal in the list of signals
  for i, v in enumerate(signals):
    
    # Instantiate a new signal object
    current_signal = Signal(i)
    
    # Smooth the signal
    smoothed_signal = smooth_data(current_signal)

    # Set the smoothed signal as an attribute of the signal object
    setattr(current_signal, "smoothed_signal", smoothed_signal)
    
    # Get the timepoints where the DNA moves through the pore
    dna_movement_timepoints = get_movement_timepoints(current_signal, movement_cutoff)
    
    # Set the DNA capture timepoints as an attribute of the signal object
    setattr(current_signal, "movement_timepoints", dna_movement_timepoints)
    
    # Get the timepoints where the DNA is captured in the pore
    dna_capture_timepoints = get_capture_timepoints(current_signal, movement_cutoff)

    # Adjust the DNA movement timepoints to account for homopolymers
    dna_capture_timepoints = adjust_capture_timepoints(dna_capture_timepoints, homopolymer_cutoff)
    
    # Set the DNA capture timepoints as an attribute of the signal object
    setattr(current_signal, "capture_timepoints", dna_capture_timepoints)
    
    # Basecall the signal
    basecalls_list = basecall_signal(current_signal, current_signal.capture_timepoints, kmer_signatures)
    
    # Collapse the basecalls into a single sequence
    final_sequence = collapse_seq(basecalls_list)
    
    # Set the basecalled sequence as an attribute of the signal object
    setattr(current_signal, "basecalled_seq", final_sequence)
    
    # Align the basecalled sequence to the correct sequence
    alignment, nice_alignment = align(current_signal.basecalled_seq, current_signal.correct_seq)
    
    # Calculate the accuracy of the basecaller
    accuracy = measure_accuracy(alignment, current_signal.correct_seq)
    accuracies.append(accuracy)
    
    # Print out the alignment
    if detailed:
      print(f"Basecalling signal {i+1} out of {len(signals)}")
      print(f"The alignment for signal {i + 1} is:")
      print("basecalled: " + nice_alignment['query_aligned'])
      print("            " + nice_alignment['matched_aligned'])
      print("   correct: " + nice_alignment['target_aligned'])
      print(f"The accuracy for this sequence is: {accuracy}%")
      print("\n")
      
  # Determine the average accuracy of the basecaller
  basecaller_accuracy = sum(accuracies) / len(accuracies) if len(accuracies) > 0 else 0
  print(f"Finished basecalling {len(signals)} signals")
  print(f"The average accuracy of the basecaller is: {basecaller_accuracy}%")
  
  
basecall_all(signals, kmer_table, sequencing, movement_cutoff = 0.03, homopolymer_cutoff = 20, detailed=True)</code></pre>
  <pre
    class="output"
  ><code class="language-output">Basecalling signal 1 out of 200
The alignment for signal 1 is:
basecalled: CCTTACCAGTTTCATAATAA-TAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGG-CGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT
            ||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||
   correct: CCTTACCAGTTTCATAATAAATAAGGTTCCCACATGCTACCTGTCCGGGCGACCGTAGGGCGAATACCGCGCTGCACGTGCCTCGCTACTTGTACACATGT
The accuracy for this sequence is: 98.01980198019803%


Basecalling signal 2 out of 200
The alignment for signal 2 is:
basecalled: TGACCGTTTGCAGGCATTGCTTGGATCCATATAGCGCACAA-TAGTCGCGAGTATGGTTGGTGATCAAA-GACCTGTAGATGTTATACTGTCAGCCCATATG
            |||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||
   correct: TGACCGTTTGCAGGCATTGCTTGGATCCATATAGCGCACAAATAGTCGCGAGTATGGTTGGTGATCAAAAGACCTGTAGATGTTATACTGTCAGCCCATATG
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 3 out of 200
The alignment for signal 3 is:
basecalled: AGCGCTCGGCTAAA-CTAGCCTGGCCATCGTTGTCATAA-CCTATCCCGGCGTCCGCTTAGCGTACTGAAGGTGTTCGTAATGTCTCC-TCGGCCAGCGCGTCG
            ||||||||||||||-||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||
   correct: AGCGCTCGGCTAAAACTAGCCTGGCCATCGTTGTCATAAACCTATCCCGGCGTCCGCTTAGCGTACTGAAGGTGTTCGTAATGTCTCCCTCGGCCAGCGCGTCG
The accuracy for this sequence is: 97.11538461538461%


Basecalling signal 4 out of 200
The alignment for signal 4 is:
basecalled: GCCCGTAGAAA--CAGGTTGAGGACAATAGCAGTTACGAATACAATTATCGATCCGGTGGTGAA-GACGTCATCGCACCTCCAACACACC-ATGAGAGAACGG
            |||||||||||--|||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||-||||||||||||
   correct: GCCCGTAGAAAAACAGGTTGAGGACAATAGCAGTTACGAATACAATTATCGATCCGGTGGTGAAAGACGTCATCGCACCTCCAACACACCCATGAGAGAACGG
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 5 out of 200
The alignment for signal 5 is:
basecalled: GACACGAAGACTTCAGGCGTGCAGATGCGAA-CGCGCAGTCGCACC-ATATATACACGATTGCTCT-AGGCCGCATGGTTACCTCCAGTTTCAA-GCATA
            |||||||||||||||||||||||||||||||-||||||||||||||-|||||||||||||||||||-|||||||||||||||||||||||||||-|||||
   correct: GACACGAAGACTTCAGGCGTGCAGATGCGAAACGCGCAGTCGCACCGATATATACACGATTGCTCTCAGGCCGCATGGTTACCTCCAGTTTCAAAGCATA
The accuracy for this sequence is: 96.0%


Basecalling signal 6 out of 200
The alignment for signal 6 is:
basecalled: CAACTATCATGCAGTATTGTCACGCGTTGCCTGCCTTGACTAGCGGCTATTATATGATGATGCGATCGTT-AGGCCAGCTCCCATT-CGATACGCACTAAC
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||-||||||||||||||
   correct: CAACTATCATGCAGTATTGTCACGCGTTGCCTGCCTTGACTAGCGGCTATTATATGATGATGCGATCGTTTAGGCCAGCTCCCATTTCGATACGCACTAAC
The accuracy for this sequence is: 98.01980198019803%


Basecalling signal 7 out of 200
The alignment for signal 7 is:
basecalled: AGGCTGGTGAG-GTACTCGTCGCTTCCAATAATAATCGATACAAGCTATCCAGTGTATCCCATGAACCACGTACC-TTCGCGGCCGACAA-TCGTATCCCGA-
            |||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||-|||||||||||-
   correct: AGGCTGGTGAGAGTACTCGTCGCTTCCAATAATAATCGATACAAGCTATCCAGTGTATCCCATGAACCACGTACCCTTCGCGGCCGACAAATCGTATCCCGAC
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 8 out of 200
The alignment for signal 8 is:
basecalled: GCTGG-CTATTT-CCAGACCACATGAATCGGTT-CTGCGGCTGG-T-TTCAGTTGGCACGTCCTTTGGG--TCC-AGGCACGTGCATCGGCCACAGGCACACGGC
            |||||-||||||-||||||||||||||||||||-||||||||||-|-||||||||||||||||||||||--|||-||||||||||||||||||||||||||||||
   correct: GCTGGGCTATTTTCCAGACCACATGAATCGGTTTCTGCGGCTGGGTCTTCAGTTGGCACGTCCTTTGGGGGTCCCAGGCACGTGCATCGGCCACAGGCACACGGC
The accuracy for this sequence is: 92.38095238095238%


Basecalling signal 9 out of 200
The alignment for signal 9 is:
basecalled: AGTTCTCTCCGATTGGCTGACAGTAACAAGCCGGGTGGACCCGGGAGGTTTT--GTAACTGTTAGCTAGAATGCGCTCAATAGCACC-ACACTT-ATCTTGA
            ||||||||||||||||||||||||||||||||||||||||||||||||||||--|||||||||||||||||||||||||||||||||-||||||-|||||||
   correct: AGTTCTCTCCGATTGGCTGACAGTAACAAGCCGGGTGGACCCGGGAGGTTTTTTGTAACTGTTAGCTAGAATGCGCTCAATAGCACCCACACTTTATCTTGA
The accuracy for this sequence is: 96.07843137254902%


Basecalling signal 10 out of 200
The alignment for signal 10 is:
basecalled: GCGAA-CC-GATACAAA-CAGCATACACGCAGTAGATACTATTGGTCCGGAAAA--TGTCACTCCAGTAATGACTTAGTCCAGTCCAGGTGTGCCTGGGTCTAGAA
            |||||-||-||||||||-||||||||||||||||||||||||||||||||||||--||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GCGAAACCCGATACAAAACAGCATACACGCAGTAGATACTATTGGTCCGGAAAAAATGTCACTCCAGTAATGACTTAGTCCAGTCCAGGTGTGCCTGGGTCTAGAA
The accuracy for this sequence is: 95.28301886792453%


Basecalling signal 11 out of 200
The alignment for signal 11 is:
basecalled: TAGATTAATT-GCTTAAGGAACACTCGGTTGGCGCCTGTAACTTGAACAGGCCTTAGAGCCACCGATATGGATGCTTGTAGTAGCCACACCTGTAAGCTG
            ||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TAGATTAATTTGCTTAAGGAACACTCGGTTGGCGCCTGTAACTTGAACAGGCCTTAGAGCCACCGATATGGATGCTTGTAGTAGCCACACCTGTAAGCTG
The accuracy for this sequence is: 99.0%


Basecalling signal 12 out of 200
The alignment for signal 12 is:
basecalled: GGCGAGAAGACGGTAATAAGCGAGTGTTCCCACGGGCTT-GG-ATGCTTGTCTCAAGTCTTTCC-GACAGCATTAAGTCTAAGCGCATAAGACCGCTGATC
            |||||||||||||||||||||||||||||||||||||||-||-|||||||||||||||||||||-||||||||||||||||||||||||||||||||||||
   correct: GGCGAGAAGACGGTAATAAGCGAGTGTTCCCACGGGCTTTGGGATGCTTGTCTCAAGTCTTTCCCGACAGCATTAAGTCTAAGCGCATAAGACCGCTGATC
The accuracy for this sequence is: 97.02970297029702%


Basecalling signal 13 out of 200
The alignment for signal 13 is:
basecalled: CTGCATTGCGGCTCGTGATACGTAATGCGTTACGCAA-TACTTCAGTCAGTGTCTAGACGTCCAAGTTT-CTT-AGAGCC-GCCGTGGCCTGGCGGTAATATAAA
            |||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||-|||-||||||-||||||||||||||||||||||||
   correct: CTGCATTGCGGCTCGTGATACGTAATGCGTTACGCAAATACTTCAGTCAGTGTCTAGACGTCCAAGTTTTCTTTAGAGCCCGCCGTGGCCTGGCGGTAATATAAA
The accuracy for this sequence is: 96.19047619047619%


Basecalling signal 14 out of 200
The alignment for signal 14 is:
basecalled: ATGG-TGCCAGCGACCGATGAGAAGACATGAACCGGGG--TCCTT-GGTTACCGAGCTTCAACCACGG-AA-TGCTAATGGG-AGGCAGAGTTACAT-
            ||||-|||||||||||||||||||||||||||||||||--|||||-||||||||||||||||||||||-||-||||||||||-||||||||||||||-
   correct: ATGGGTGCCAGCGACCGATGAGAAGACATGAACCGGGGGGTCCTTTGGTTACCGAGCTTCAACCACGGGAAATGCTAATGGGGAGGCAGAGTTACATC
The accuracy for this sequence is: 91.83673469387756%


Basecalling signal 15 out of 200
The alignment for signal 15 is:
basecalled: AAGCGGCGTAACTCTCTGGGCAGTGGTATTCTACTCGACTACTCCTTAAGCGCC-ATGCTGACAATGTGCGGACGTGGCCACTTATTAA-GATGGCTGG-AG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||-|||||||||-||
   correct: AAGCGGCGTAACTCTCTGGGCAGTGGTATTCTACTCGACTACTCCTTAAGCGCCCATGCTGACAATGTGCGGACGTGGCCACTTATTAAAGATGGCTGGGAG
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 16 out of 200
The alignment for signal 16 is:
basecalled: CTGA-TGACGGATAAGGTGCAGAGGACACATGTAGGACTAATAGAAA-GAACTCGGCAA-GTCTCCACCAACGCCTCGGCGCGAATCACTAATAGACC-TTTT
            ||||-||||||||||||||||||||||||||||||||||||||||||-|||||||||||-||||||||||||||||||||||||||||||||||||||-||||
   correct: CTGAATGACGGATAAGGTGCAGAGGACACATGTAGGACTAATAGAAAAGAACTCGGCAAAGTCTCCACCAACGCCTCGGCGCGAATCACTAATAGACCCTTTT
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 17 out of 200
The alignment for signal 17 is:
basecalled: GGATTGCCACAGGAA-CATCTAACGAGACCTACTGGTTACGTTAG-AGTCATGCGCACC-ATCTTGAGATATGCGAACGTAAGCCGTTACGTCG-CTAGCCGTC
            |||||||||||||||-|||||||||||||||||||||||||||||-|||||||||||||-||||||||||||||||||||||||||||||||||-|||||||||
   correct: GGATTGCCACAGGAAACATCTAACGAGACCTACTGGTTACGTTAGTAGTCATGCGCACCCATCTTGAGATATGCGAACGTAAGCCGTTACGTCGGCTAGCCGTC
The accuracy for this sequence is: 96.15384615384616%


Basecalling signal 18 out of 200
The alignment for signal 18 is:
basecalled: TCACCTCC-TGAGGTGTGACTACTAATCTCCCGCCGG-AGTCCACTACTGGTAA-TTATCACAGCTAGCACTGTTTGGCATAGTAAGCACCGTCAATT-C
            ||||||||-||||||||||||||||||||||||||||-||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||-|
   correct: TCACCTCCCTGAGGTGTGACTACTAATCTCCCGCCGGGAGTCCACTACTGGTAAATTATCACAGCTAGCACTGTTTGGCATAGTAAGCACCGTCAATTTC
The accuracy for this sequence is: 96.0%


Basecalling signal 19 out of 200
The alignment for signal 19 is:
basecalled: AACATGTGGC-GCCCACGGTTAGTCCGCGCCACCCAATT-AGAGTCAATAATTATTCCAGGTAAAGAA-CAACACAGCTCTTT-GACTTGGCTCAA
            ||||||||||-||||||||||||||||||||||||||||-||||||||||||||||||||||||-|||-||||||||||||||-||||||||||||
   correct: AACATGTGGCTGCCCACGGTTAGTCCGCGCCACCCAATTTAGAGTCAATAATTATTCCAGGTAA-GAAACAACACAGCTCTTTTGACTTGGCTCAA
The accuracy for this sequence is: 94.73684210526316%


Basecalling signal 20 out of 200
The alignment for signal 20 is:
basecalled: CGTAATAATTATACGAAA-CTTCCC-GTCTGTTCGCAATGAGGG-TATCGTTGG-TACTTCC-TACTCGAACCTCTTATCGGGCCACC-ACGCAGTCC
            ||||||||||||||||||-||||||-||||||||||||||||||-|||||||||-|||||||-|||||||||||||||||||-|||||-|||||||||
   correct: CGTAATAATTATACGAAAACTTCCCCGTCTGTTCGCAATGAGGGGTATCGTTGGGTACTTCCCTACTCGAACCTCTTATCGG-CCACCCACGCAGTCC
The accuracy for this sequence is: 92.78350515463917%


Basecalling signal 21 out of 200
The alignment for signal 21 is:
basecalled: CCAAACCC-GTAAAGAGATCACGAGACACGGCAGGCCGCAGACCTATGTACGGCGGCATT-ATAGCCATAGGACCC-GTAGGAGATAGCACGTTAGCT
            ||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||-|||||||||||||||||||||
   correct: CCAAACCCCGTAAAGAGATCACGAGACACGGCAGGCCGCAGACCTATGTACGGCGGCATTTATAGCCATAGGACCCCGTAGGAGATAGCACGTTAGCT
The accuracy for this sequence is: 96.93877551020408%


Basecalling signal 22 out of 200
The alignment for signal 22 is:
basecalled: CACGATTT-GGCACTGCCGCCGATTAACCTATCTTG-GGTTTACAATGAGTCTAGTTAACGGCTGTTGAATGAACC-GTGTAATCAACATAA-CGGCTGTCTAA
            ||||||||-|||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||-|||||||||||||||-|||||||||||
   correct: CACGATTTTGGCACTGCCGCCGATTAACCTATCTTGCGGTTTACAATGAGTCTAGTTAACGGCTGTTGAATGAACCCGTGTAATCAACATAAACGGCTGTCTAA
The accuracy for this sequence is: 96.15384615384616%


Basecalling signal 23 out of 200
The alignment for signal 23 is:
basecalled: GGCTTTCCGTCGGTCAAGATTT-GCTGTTGAA-GGAT-AGGTCTTGAATCGAACCCTTAGTTGTGAATAACGAATGACTAG-TCCATTGCAATT-CTTGTATTGCT
            ||||||||||||||||||||||-|||||||||-||||-|||||||||||||||||||||||||||||||||||||||||||-||||||||||||-|||||||||||
   correct: GGCTTTCCGTCGGTCAAGATTTTGCTGTTGAAAGGATGAGGTCTTGAATCGAACCCTTAGTTGTGAATAACGAATGACTAGATCCATTGCAATTTCTTGTATTGCT
The accuracy for this sequence is: 95.28301886792453%


Basecalling signal 24 out of 200
The alignment for signal 24 is:
basecalled: TT-ACCTTATCGAATAGTTTATAGAA-TTCGCTCGTTGAGTTCACGGTGGACAGTGTGCGTCC-GG-CGACGCGCTTGGTAATAGTAATCCACCCC-AGA--
            ||-|||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||-||-|||||||||||||||||||||||||||||-|||--
   correct: TTTACCTTATCGAATAGTTTATAGAAATTCGCTCGTTGAGTTCACGGTGGACAGTGTGCGTCCCGGGCGACGCGCTTGGTAATAGTAATCCACCCCCAGAAA
The accuracy for this sequence is: 93.13725490196079%


Basecalling signal 25 out of 200
The alignment for signal 25 is:
basecalled: GGGTGCCAAAGCCTGTTCAAGCCGGGTCGTT-ATATAGGTCCGAGATTGTAGATGAATTATTAAGGTCATT-CGAAGCAGAGTCTAGAGG-AGCAATGCCT
            |||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||-||||||||||||||||||-||||||||||
   correct: GGGTGCCAAAGCCTGTTCAAGCCGGGTCGTTTATATAGGTCCGAGATTGTAGATGAATTATTAAGGTCATTTCGAAGCAGAGTCTAGAGGGAGCAATGCCT
The accuracy for this sequence is: 97.02970297029702%


Basecalling signal 26 out of 200
The alignment for signal 26 is:
basecalled: ATAGGCATGGAGTGAGGGACTGAAGTTGGCGAGTCTTCGGAATTAGAGTCACGGTCTCCAAGGATTACGCCGCAGTATTGTGCCGGCTGGTAGCTGTCGAT
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ATAGGCATGGAGTGAGGGACTGAAGTTGGCGAGTCTTCGGAATTAGAGTCACGGTCTCCAAGGATTACGCCGCAGTATTGTGCCGGCTGGTAGCTGTCGAT
The accuracy for this sequence is: 100.0%


Basecalling signal 27 out of 200
The alignment for signal 27 is:
basecalled: TTGGG-CGAGATGTAGAACATTCGAAGCC-TG-CAGGTCCGGTTAAGTTAAGGTCGGGTACTCAACCTTT-CAATTGAGAAGG-CC-AGATAGCTCTGAG
            |||||-|||||||||||||||||||||||-||-|||||||||||||||||||||||||||||||||||||-||||||||||||-||-|||||||||||||
   correct: TTGGGGCGAGATGTAGAACATTCGAAGCCCTGGCAGGTCCGGTTAAGTTAAGGTCGGGTACTCAACCTTTTCAATTGAGAAGGGCCCAGATAGCTCTGAG
The accuracy for this sequence is: 94.0%


Basecalling signal 28 out of 200
The alignment for signal 28 is:
basecalled: GTCTGCTTCTGTATGAAGTCCGCTCCGGG-TCGGTCCGAAACC-GCTTCAGCATGTTTGTTT-GCGGCTCATCCAAGTCGG-CCGGCTCGCTTACACTCGCACTGGCCA
            |||||||||||||||||||||||||||||-|||||||||||||-||||||||||||||||||-||||||||||||||||||-|||||||||||||||||||||||||||
   correct: GTCTGCTTCTGTATGAAGTCCGCTCCGGGGTCGGTCCGAAACCCGCTTCAGCATGTTTGTTTTGCGGCTCATCCAAGTCGGGCCGGCTCGCTTACACTCGCACTGGCCA
The accuracy for this sequence is: 96.3302752293578%


Basecalling signal 29 out of 200
The alignment for signal 29 is:
basecalled: GTGCAAGCTGTAGTAGGACGTCGTCATGTGTCTGGAGAAGATGTCTAA-CATCTGAAA-TACGCTGTAGTCGTGGAGAGCC-TTTGCACATTAATAGA
            ||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||-||||||||||||||||||||||-||||||||||||||||
   correct: GTGCAAGCTGTAGTAGGACGTCGTCATGTGTCTGGAGAAGATGTCTAAACATCTGAAAATACGCTGTAGTCGTGGAGAGCCCTTTGCACATTAATAGA
The accuracy for this sequence is: 96.93877551020408%


Basecalling signal 30 out of 200
The alignment for signal 30 is:
basecalled: ATACTCGATAG-CGCGATCACACGTGCACGTGGGTAGTTAACTATATACCCGCGTCGAGGAGTGGTCGCAATATGATACCTCGCTCTCATGTGCGTCTTCT
            |||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ATACTCGATAGACGCGATCACACGTGCACGTGGGTAGTTAACTATATACCCGCGTCGAGGAGTGGTCGCAATATGATACCTCGCTCTCATGTGCGTCTTCT
The accuracy for this sequence is: 99.00990099009901%


Basecalling signal 31 out of 200
The alignment for signal 31 is:
basecalled: TGGG-TGGACACTGG-AATT-GACCATTACCACGATACCATAGTGTCACGGACTGCCTAGCGGCCCAGACCC-AA-GTGTTAAGTGGGCATATGTCGAGC
            ||||-||||||||||-||||-|||||||||||||||||||||||||||||||||||||||||||||||||||-||-||||||||||||||||||||||||
   correct: TGGGGTGGACACTGGGAATTTGACCATTACCACGATACCATAGTGTCACGGACTGCCTAGCGGCCCAGACCCCAAAGTGTTAAGTGGGCATATGTCGAGC
The accuracy for this sequence is: 95.0%


Basecalling signal 32 out of 200
The alignment for signal 32 is:
basecalled: TACTACGAATCCCAACTGTCGAA-TAAGCTGCT--CC-GAGTGCCGATAGCAGCTCATTGTTCCTCATGAGAGGTATTACTCAGTGCTGAACCTATGAAATA-
            |||||||||||||||||||||||-|||||||||--||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-
   correct: TACTACGAATCCCAACTGTCGAAATAAGCTGCTTTCCCGAGTGCCGATAGCAGCTCATTGTTCCTCATGAGAGGTATTACTCAGTGCTGAACCTATGAAATAA
The accuracy for this sequence is: 95.14563106796116%


Basecalling signal 33 out of 200
The alignment for signal 33 is:
basecalled: AAGGTAGAAACTTCGGATAGGCTATT-CTCATTACGGCTTGCACGATGCAGATGCATACGGTCCTGCCAGTACAGCCC-ACCTTGAACTAATGTAATTGCCGCAT-
            ||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||-
   correct: AAGGTAGAAACTTCGGATAGGCTATTTCTCATTACGGCTTGCACGATGCAGATGCATACGGTCCTGCCAGTACAGCCCCACCTTGAACTAATGTAATTGCCGCATA
The accuracy for this sequence is: 97.16981132075472%


Basecalling signal 34 out of 200
The alignment for signal 34 is:
basecalled: GATTCATTCATGCACATAGTGACCACCGGCCTAACCTCGGCACTGATGTGAGGTGTCGCGAGACCAAGTTGATATACC-AGCTCACTACGCTGTCATAGTG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||
   correct: GATTCATTCATGCACATAGTGACCACCGGCCTAACCTCGGCACTGATGTGAGGTGTCGCGAGACCAAGTTGATATACCCAGCTCACTACGCTGTCATAGTG
The accuracy for this sequence is: 99.00990099009901%


Basecalling signal 35 out of 200
The alignment for signal 35 is:
basecalled: GGTTCC-GTTCCTTCGGCATCCTGCTGGCCTCGCGGCC-GTGTCACGTGAGCC-GTAGATTTT----GGCCGTAGAGTTT--CAAA-GTT-CTTCCTTCCACG
            ||||||-|||||||||||||||||||||||||||||||-||||||||||||||-|||||||||----|||||||||||||--||||-|||-||||||||||||
   correct: GGTTCCCGTTCCTTCGGCATCCTGCTGGCCTCGCGGCCCGTGTCACGTGAGCCCGTAGATTTTTTTTGGCCGTAGAGTTTTTCAAAAGTTTCTTCCTTCCACG
The accuracy for this sequence is: 89.32038834951457%


Basecalling signal 36 out of 200
The alignment for signal 36 is:
basecalled: GCAATAAGTTGGAGATGCTCAGCCGATCAGCTT-GCTCGGTGATTGAGTTTACTAGCGAATGCAGAGCATATGGCAATGCCGACTCGTGCGGTTCC
            |||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GCAATAAGTTGGAGATGCTCAGCCGATCAGCTTTGCTCGGTGATTGAGTTTACTAGCGAATGCAGAGCATATGGCAATGCCGACTCGTGCGGTTCC
The accuracy for this sequence is: 98.95833333333334%


Basecalling signal 37 out of 200
The alignment for signal 37 is:
basecalled: GGCGTGCTCGCGGGCCGAGCTGCTACATAAATG-CGACTCGGTAATCGCAACCATT-CAGTTCCTCTTCGCTGCAGTGGATGGCGTCTTGTCTCTCTCCC-
            |||||||||||||||||||||||||||||||||-||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||-
   correct: GGCGTGCTCGCGGGCCGAGCTGCTACATAAATGACGACTCGGTAATCGCAACCATTTCAGTTCCTCTTCGCTGCAGTGGATGGCGTCTTGTCTCTCTCCCC
The accuracy for this sequence is: 97.02970297029702%


Basecalling signal 38 out of 200
The alignment for signal 38 is:
basecalled: ATCAAGCCGAATCGTTGCACATCAACAAGATTCGTGGAGGAGAAGATTGACATCGTGGCGTCGTTCAACCTTAGAGTAGGCTGCGGTCACTGGACCTGAACG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ATCAAGCCGAATCGTTGCACATCAACAAGATTCGTGGAGGAGAAGATTGACATCGTGGCGTCGTTCAACCTTAGAGTAGGCTGCGGTCACTGGACCTGAACG
The accuracy for this sequence is: 100.0%


Basecalling signal 39 out of 200
The alignment for signal 39 is:
basecalled: ATCAGTCAGGTT-CAAGCTGAAGCCTGTCACCTGGCTGGTAAGGCC-GGCTTT-CTGGTCCAATAAATTAGTCCTCAGAGTGAGG-TGCGCAAA-TG
            ||||||||||||-|||||||||||||||||||||||||||||||||-||||||-|||||||||||||||||||||||||||||||-||||||||-||
   correct: ATCAGTCAGGTTTCAAGCTGAAGCCTGTCACCTGGCTGGTAAGGCCCGGCTTTTCTGGTCCAATAAATTAGTCCTCAGAGTGAGGGTGCGCAAAATG
The accuracy for this sequence is: 94.84536082474226%


Basecalling signal 40 out of 200
The alignment for signal 40 is:
basecalled: TGCATCCTCGGCGACTGAGG-CAACAAGGCACCGCAA-CGAAACGCCAAGCGGCGCTAGCGACTTCTTGATCATCTTGAACTCTGATTGAGAAGCGACCCG-
            ||||||||||||||||||||-||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-
   correct: TGCATCCTCGGCGACTGAGGGCAACAAGGCACCGCAAACGAAACGCCAAGCGGCGCTAGCGACTTCTTGATCATCTTGAACTCTGATTGAGAAGCGACCCGT
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 41 out of 200
The alignment for signal 41 is:
basecalled: TTCATCAGAATGCCTGTAAGTAGCTACAA-TGTTT-CACGGCCGCGCCAA-GCGGTGCACACTATAA-GCCAAGTTCTCCTCCGACGTATACG-AGATAA-
            |||||||||||||||||||||||||||||-|||||-||||||||||||||-||||||||||||||||-|||||||||||||||||||||||||-||||||-
   correct: TTCATCAGAATGCCTGTAAGTAGCTACAAATGTTTTCACGGCCGCGCCAAAGCGGTGCACACTATAAAGCCAAGTTCTCCTCCGACGTATACGTAGATAAA
The accuracy for this sequence is: 94.05940594059405%


Basecalling signal 42 out of 200
The alignment for signal 42 is:
basecalled: ATTCCATT-CAGTGGAACGAATAGATAAGACTTTCGTGGTTGTAAGAGTACCTGAATAGTCTTGTTAAGGGCC-GTAGCTTACTTT-CATTCGTAA
            ||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||-|||||||||
   correct: ATTCCATTTCAGTGGAACGAATAGATAAGACTTTCGTGGTTGTAAGAGTACCTGAATAGTCTTGTTAAGGGCCCGTAGCTTACTTTTCATTCGTAA
The accuracy for this sequence is: 96.875%


Basecalling signal 43 out of 200
The alignment for signal 43 is:
basecalled: ATAATGTTTGTCACTCCTTGTCGGACCC-ATAGTGAACGTTGTCTCGAGCCTT-AATCCTTAAA-GCCGTTCACTATTGTTAGACC-TT-CGGTCAGGATACA
            ||||||||||||||||||||||||||||-||||||||||||||||||||||||-||||||||||-|||||||||||||||||||||-||-|||||||||||||
   correct: ATAATGTTTGTCACTCCTTGTCGGACCCCATAGTGAACGTTGTCTCGAGCCTTTAATCCTTAAAAGCCGTTCACTATTGTTAGACCCTTTCGGTCAGGATACA
The accuracy for this sequence is: 95.14563106796116%


Basecalling signal 44 out of 200
The alignment for signal 44 is:
basecalled: CTATATTACCCGCTGATAGTGATCCTTAGCCTAATTACGGACAGTCGGTGTAACCAAGCTATTCACACCACAA-GACTCGTCCTT-GCACCCGCCGTAGTG-
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||-|||||||||||||||-
   correct: CTATATTACCCGCTGATAGTGATCCTTAGCCTAATTACGGACAGTCGGTGTAACCAAGCTATTCACACCACAAAGACTCGTCCTTTGCACCCGCCGTAGTGG
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 45 out of 200
The alignment for signal 45 is:
basecalled: ACGGCTGTGG-TATGGCCGACGGG-ATCCATTGTTACGACATGGTTAGAAAGGAGTCGCGGACACGAAGATGAGAAGTAATT-ACGCATCCGGTC
            ||||||||||-|||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||
   correct: ACGGCTGTGGGTATGGCCGACGGGGATCCATTGTTACGACATGGTTAGAAAGGAGTCGCGGACACGAAGATGAGAAGTAATTTACGCATCCGGTC
The accuracy for this sequence is: 96.84210526315789%


Basecalling signal 46 out of 200
The alignment for signal 46 is:
basecalled: GTTACGTTGGGTCGCAGCGTCACATTACGAATGAACCAATTGTTGGGTGTACTCGTATTT-ATTATCTGGTACATTGATCGTGGAGACCGAGCATTGCCGC
            ||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||
   correct: GTTACGTTGG-TCGCAGCGTCACATTACGAATGAACCAATTGTTGGGTGTACTCGTATTTTATTATCTGGTACATTGATCGTGGAGACCGAGCATTGCCGC
The accuracy for this sequence is: 98.0%


Basecalling signal 47 out of 200
The alignment for signal 47 is:
basecalled: GGCATAGTCAGGAATGCGG-CTCGTTAGATACGTTT-GTCCGGG-AGG-CTGTTACAGTAGTTCTACAGTGCGTT-GATT-GCAGCATCAATGTCGAATTCG-
            |||||||||||||||||||-||||||||||||||||-|||||||-|||-||||||||||||||||||||||||||-||||-|||||||||||||||||||||-
   correct: GGCATAGTCAGGAATGCGGGCTCGTTAGATACGTTTTGTCCGGGGAGGGCTGTTACAGTAGTTCTACAGTGCGTTTGATTTGCAGCATCAATGTCGAATTCGG
The accuracy for this sequence is: 93.20388349514563%


Basecalling signal 48 out of 200
The alignment for signal 48 is:
basecalled: GTCCACAGGTACAGCCAGCGCTGTAATCGGTGATTCGCCCTATTGGCGCCTCTCGCACATCATCCGCTTT-AGGCATGTATAATCCATT-GGTCCAG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||-|||||||
   correct: GTCCACAGGTACAGCCAGCGCTGTAATCGGTGATTCGCCCTATTGGCGCCTCTCGCACATCATCCGCTTTTAGGCATGTATAATCCATTTGGTCCAG
The accuracy for this sequence is: 97.9381443298969%


Basecalling signal 49 out of 200
The alignment for signal 49 is:
basecalled: GGTTACGTGCTGTCC-TCTGGTGCTACGAGGACCC-TACCGAGGAATAAGGCAGTATCAACCTATCTAGCTTGTGGTTATTGTT-CAGACGCGGACCTCG
            |||||||||||||||-|||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||
   correct: GGTTACGTGCTGTCCCTCTGGTGCTACGAGGACCCCTACCGAGGAATAAGGCAGTATCAACCTATCTAGCTTGTGGTTATTGTTTCAGACGCGGACCTCG
The accuracy for this sequence is: 97.0%


Basecalling signal 50 out of 200
The alignment for signal 50 is:
basecalled: CAGTCCACACGTTGGTCTTCGACCTTCCTACGCATAGATTACATTATATCTTACGATACCCC-ACTTAAGACGTACC-GTGGGG-TT-GGACTAAGACG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||-||||||-||-|||||||||||
   correct: CAGTCCACACGTTGGTCTTCGACCTTCCTACGCATAGATTACATTATATCTTACGATACCCCCACTTAAGACGTACCCGTGGGGGTTTGGACTAAGACG
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 51 out of 200
The alignment for signal 51 is:
basecalled: ACATACTTCGATAGTGCGAGCAATTAATTCGCCTTAACTACCTCCCGTGAA-GTTT-CATCCGCGGCCATCGACATAGCC-TGACTAGACCATGTTAGAC
            |||||||||||||||||||||||||||||||||||||||||||||||||||-||||-|||||||||||||||||||||||-|||||||||||||||||||
   correct: ACATACTTCGATAGTGCGAGCAATTAATTCGCCTTAACTACCTCCCGTGAAAGTTTTCATCCGCGGCCATCGACATAGCCCTGACTAGACCATGTTAGAC
The accuracy for this sequence is: 97.0%


Basecalling signal 52 out of 200
The alignment for signal 52 is:
basecalled: TGGACCTAAGCTTGTTCGCACCATACACAATTATTGATCTAGATCAGGCTCTAAGGCTAACCAGACCCC--AAA--CTCTGCTAACACGCCGTACTCTG
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||--|||--|||||||||||||||||||||||
   correct: TGGACCTAAGCTTGTTCGCACCATACACAATTATTGATCTAGATCAGGCTCTAAGGCTAACCAGACCCCCCAAAAACTCTGCTAACACGCCGTACTCTG
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 53 out of 200
The alignment for signal 53 is:
basecalled: CATGCATGACGCTGCCGGATGCTCGCGGATGATCTGTCTT-CGATTGTAATGTTGAAA--GGAGGTTAGTGCATATTCCGATTCGACAGGCGCACAT
            ||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||--|||||||||||||||||||||||||||||||||||||
   correct: CATGCATGACGCTGCCGGATGCTCGCGGATGATCTGTCTTTCGATTGTAATGTTGAAAAAGGAGGTTAGTGCATATTCCGATTCGACAGGCGCACAT
The accuracy for this sequence is: 96.90721649484536%


Basecalling signal 54 out of 200
The alignment for signal 54 is:
basecalled: CAA-GATTTAACCGATAGTGCACACAAGCTTGAGGT-GCC-AGG-TTTACGGATAACATCGGCAATCTAATTCATTCGCACCACTGCTT-AGAATGCG-
            |||-||||||||||||||||||||||||||||||||-|||-|||-||||||||||||||||||||||||||||||||||||||||||||-||||||||-
   correct: CAAAGATTTAACCGATAGTGCACACAAGCTTGAGGTCGCCCAGGGTTTACGGATAACATCGGCAATCTAATTCATTCGCACCACTGCTTTAGAATGCGT
The accuracy for this sequence is: 93.93939393939394%


Basecalling signal 55 out of 200
The alignment for signal 55 is:
basecalled: CAACCGCGTCC-TCGAACCATTAGTCTACTATCCTAGCTGATAGATCTCGATTCAGAAGGAACTAACGGCCTATTT-CCTCATCATAGTGTT-ATAGTC
            |||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||-||||||
   correct: CAACCGCGTCCCTCGAACCATTAGTCTACTATCCTAGCTGATAGATCTCGATTCAGAAGGAACTAACGGCCTATTTTCCTCATCATAGTGTTTATAGTC
The accuracy for this sequence is: 96.96969696969697%


Basecalling signal 56 out of 200
The alignment for signal 56 is:
basecalled: AGTAGGAAGG-ACGAGTATACTGTGCACTACTT-GTTGCTGATGGAACGTCGCCACGTGTCAAGTCC-TTAGCGACAAGAGGCATCAACATAACGCTTGGTCCAA
            ||||||||||-||||||||||||||||||||||-|||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||
   correct: AGTAGGAAGGGACGAGTATACTGTGCACTACTTTGTTGCTGATGGAACGTCGCCACGTGTCAAGTCCCTTAGCGACAAGAGGCATCAACATAACGCTTGGTCCAA
The accuracy for this sequence is: 97.14285714285714%


Basecalling signal 57 out of 200
The alignment for signal 57 is:
basecalled: ACGATGTAGATCTGACTTAGGGCGTTT-CC-TACCAACGTACTTCTT-GGCCTAACAAA-CTGG-ATAGTTGAGTACTACTAATCGGTACTACGCTA
            |||||||||||||||||||||||||||-||-||||||||||||||||-|||||||||||-||||-||||||||||||||||||||||||||||||||
   correct: ACGATGTAGATCTGACTTAGGGCGTTTTCCCTACCAACGTACTTCTTTGGCCTAACAAAACTGGGATAGTTGAGTACTACTAATCGGTACTACGCTA
The accuracy for this sequence is: 94.84536082474226%


Basecalling signal 58 out of 200
The alignment for signal 58 is:
basecalled: TTAAGGAGTAGACGAGGTCACTGAAGGCGCTGCGGACGTAGCGTGTTCCGAAGGTAGCGCGTAGCTCAGCGG-AGTTCACATTAAATGAAA--GAC
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||--|||
   correct: TTAAGGAGTAGACGAGGTCACTGAAGGCGCTGCGGACGTAGCGTGTTCCGAAGGTAGCGCGTAGCTCAGCGGGAGTTCACATTAAATGAAAAAGAC
The accuracy for this sequence is: 96.875%


Basecalling signal 59 out of 200
The alignment for signal 59 is:
basecalled: CGGCGACTGGCCAGAAATGCCCC--TGGCAATGTGGAATAGTTGACCTAGGGGAGTTACACCTCGTGTCCAGTCCTATGGTGGCTCATCCACACA
            |||||||||||||||||||||||--||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: CGGCGACTGGCCAGAAATGCCCCCCTGGCAATGTGGAATAGTTGACCTAGGGGAGTTACACCTCGTGTCCAGTCCTATGGTGGCTCATCCACACA
The accuracy for this sequence is: 97.89473684210527%


Basecalling signal 60 out of 200
The alignment for signal 60 is:
basecalled: AACTGACCTAGTTCTCACTCC-ATTCCTT-AA-GGCAACAGTCAGCTTGCTATTCTTACC-TCTCC-GAAGAATGGAAA-TCCGTGGCGGTTCATAATGCGACT
            |||||||||||||||||||||-|||||||-||-|||||||||||||||||||||||||||-|||||-||||||||||||-||||||||||||||||||||||||
   correct: AACTGACCTAGTTCTCACTCCCATTCCTTTAAAGGCAACAGTCAGCTTGCTATTCTTACCCTCTCCCGAAGAATGGAAAATCCGTGGCGGTTCATAATGCGACT
The accuracy for this sequence is: 94.23076923076923%


Basecalling signal 61 out of 200
The alignment for signal 61 is:
basecalled: CTCGACGTGCTGCTCTACTCCACGCATTAGTCGGG--TCGAGCACCAA-TT-CAACTCAAATGTCGATCTCGTTAATAATTCCCC--TGTGCCGGTTGACGCC
            |||||||||||||||||||||||||||||||||||--|||||||||||-||-|||||||||||||||||||||||||||||||||--||||||||||||||||
   correct: CTCGACGTGCTGCTCTACTCCACGCATTAGTCGGGGGTCGAGCACCAAATTTCAACTCAAATGTCGATCTCGTTAATAATTCCCCCCTGTGCCGGTTGACGCC
The accuracy for this sequence is: 94.1747572815534%


Basecalling signal 62 out of 200
The alignment for signal 62 is:
basecalled: TAGTACCCGCTGATATGTAAGTCAGGTATCTAA-TT-GCTTGTCTTACCGCGCTCGCATCGGAGACCC-ATTAACCTTAGCAGG-CTGACGCGGCTGTATTGG
            |||||||||||||||||||||||||||||||||-||-|||||||||||||||||||||||||||||||-|||||||||||||||-||||||||||||||||||
   correct: TAGTACCCGCTGATATGTAAGTCAGGTATCTAAATTTGCTTGTCTTACCGCGCTCGCATCGGAGACCCCATTAACCTTAGCAGGGCTGACGCGGCTGTATTGG
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 63 out of 200
The alignment for signal 63 is:
basecalled: GGATT-CGAGCGTTAAATTAGG-CAA-CAGAACCACAACAACGTTGCGTGACTAATGAGGCGACTAGGCATGGAGACGACGTTCCGAGACAACTGGACC-GG-CA
            |||||-||||||||||||||||-|||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||-||
   correct: GGATTTCGAGCGTTAAATTAGGGCAAACAGAACCACAACAACGTTGCGTGACTAATGAGGCGACTAGGCATGGAGACGACGTTCCGAGACAACTGGACCCGGGCA
The accuracy for this sequence is: 95.23809523809523%


Basecalling signal 64 out of 200
The alignment for signal 64 is:
basecalled: GGTCTCAGCAAGCACTCGACTACGGTCCC-TACGTGTGACGAGTCTAGGCCTACAAA--TTT-GTTGC-TCATAGAGCGCCGACTATAGGTTCTCGCCTCA
            |||||||||||||||||||||||||||||-|||||||||||||||||||||||||||--|||-|||||-||||||||||||||||||||||||||||||||
   correct: GGTCTCAGCAAGCACTCGACTACGGTCCCCTACGTGTGACGAGTCTAGGCCTACAAAAATTTTGTTGCCTCATAGAGCGCCGACTATAGGTTCTCGCCTCA
The accuracy for this sequence is: 95.04950495049505%


Basecalling signal 65 out of 200
The alignment for signal 65 is:
basecalled: CTGGTCCGGTAGGTT-AACTAGTTGATATCGG-ACATATAATGCCAGTACGTAATTACCTGACGCC-TTCGTCCAA-TAAGCGAGGATTT-GCGCCAGTGATCG
            |||||||||||||||-||||||||||||||||-|||||||||||||||||||||||||||||||||-|||||||||-|||||||||||||-|||||||||||||
   correct: CTGGTCCGGTAGGTTTAACTAGTTGATATCGGGACATATAATGCCAGTACGTAATTACCTGACGCCCTTCGTCCAAATAAGCGAGGATTTTGCGCCAGTGATCG
The accuracy for this sequence is: 95.1923076923077%


Basecalling signal 66 out of 200
The alignment for signal 66 is:
basecalled: TTACGCCGCC-ACAA-GGATGAGCACAGACGGTACC-ACCTCGTCCTGACTCGGAGAGAAAACACTTCCGTCTGGACTTGG-CGATATACCAGGTTACA
            ||||||||||-||||-||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||
   correct: TTACGCCGCCCACAAAGGATGAGCACAGACGGTACCCACCTCGTCCTGACTCGGAGAGAAAACACTTCCGTCTGGACTTGGGCGATATACCAGGTTACA
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 67 out of 200
The alignment for signal 67 is:
basecalled: AGGCTTGAA-GAGTGAGTTCAGATCAGCGAATGGAAAGGCAA-CTTTGATT-CTCGAATTTTAATTT-ACAATTTCAGACCATTGATGTAGCAGATG
            |||||||||-|||||||||||||||||||||||-||||||||-||||||||-|||||||||||||||-|||||||||||||||||||||||||||||
   correct: AGGCTTGAAAGAGTGAGTTCAGATCAGCGAATG-AAAGGCAAACTTTGATTTCTCGAATTTTAATTTTACAATTTCAGACCATTGATGTAGCAGATG
The accuracy for this sequence is: 94.79166666666666%


Basecalling signal 68 out of 200
The alignment for signal 68 is:
basecalled: AGAACTCGG-TACCAGCATTCAGCACGTATCAGGTCTTACCTTGAGAAAGAATTT-GACCTAATTGTGCGCAGCCATCGAAGTTGCTCGAGGCAGATGA
            |||||||||-|||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||
   correct: AGAACTCGGGTACCAGCATTCAGCACGTATCAGGTCTTACCTTGAGAAAGAATTTTGACCTAATTGTGCGCAGCCATCGAAGTTGCTCGAGGCAGATGA
The accuracy for this sequence is: 97.97979797979798%


Basecalling signal 69 out of 200
The alignment for signal 69 is:
basecalled: TAACTCATCGGG-AATCAGCCGCAAGATCGAGTGGTGCGATTCAATAGAGGTTT-AGAGG-TGTGCGCC-GTCTGAACACTACTACGGCCTGGCGTGTTCG
            ||||||||||||-|||||||||||||||||||||||||||||||||||||||||-|||||-||||||||-|||||||||||||||||||||||||||||||
   correct: TAACTCATCGGGGAATCAGCCGCAAGATCGAGTGGTGCGATTCAATAGAGGTTTTAGAGGGTGTGCGCCCGTCTGAACACTACTACGGCCTGGCGTGTTCG
The accuracy for this sequence is: 96.03960396039604%


Basecalling signal 70 out of 200
The alignment for signal 70 is:
basecalled: CTTACCTCACGTCTGAATAA-CC-GCGTTCATGCTGAATAAATT-GCAGCCC-GTGATAATAGGAAACCTCTTCGTAGAGGTCAA-GAGGCTCGTTGCGTCCACA-
            ||||||||||||||||||||-||-||||||||||||||||||||-|||||||-||||||||||||||||||||||||||||||||-|||||||||||||||||||-
   correct: CTTACCTCACGTCTGAATAAACCCGCGTTCATGCTGAATAAATTTGCAGCCCCGTGATAATAGGAAACCTCTTCGTAGAGGTCAAAGAGGCTCGTTGCGTCCACAC
The accuracy for this sequence is: 94.33962264150944%


Basecalling signal 71 out of 200
The alignment for signal 71 is:
basecalled: AATGCGGTCTACAGTCAATCTGACAGGATCCATCCAAGCC-GATGTGGCGGAAGGTATTGCATTCTGGCTACGCCATTACGCAACGGAACAAGAGTATGCTG
            ||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: AATGCGGTCTACAGTCAATCTGACAGGATCCATCCAAGCCCGATGTGGCGGAAGGTATTGCATTCTGGCTACGCCATTACGCAACGGAACAAGAGTATGCTG
The accuracy for this sequence is: 99.01960784313727%


Basecalling signal 72 out of 200
The alignment for signal 72 is:
basecalled: ACGAGG-AGGCCGCACTCTACAAGCAGTACTCAAGTGTGAATGCACAGACTGGCAAGGGAGTATACAATCGATGGTTACCTTCCTCTGCCTGCCTT-GTT
            ||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||
   correct: ACGAGGGAGGCCGCACTCTACAAGCAGTACTCAAGTGTGAATGCACAGACTGGCAAGGGAGTATACAATCGATGGTTACCTTCCTCTGCCTGCCTTTGTT
The accuracy for this sequence is: 98.0%


Basecalling signal 73 out of 200
The alignment for signal 73 is:
basecalled: TT-GGTCTACACAACCCGGCC-GCGATCGATATTGAGGCAATATGGTACTT-CGGAGATGCTCTGTGAAGAATGCGCGCTCTTCGTAGTTACCAGCTACTTT
            ||-||||||||||||||||||-|||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TTTGGTCTACACAACCCGGCCCGCGATCGATATTGAGGCAATATGGTACTTTCGGAGATGCTCTGTGAAGAATGCGCGCTCTTCGTAGTTACCAGCTACTTT
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 74 out of 200
The alignment for signal 74 is:
basecalled: TCGGGTCGAGGG-CAACGAACGCC-GATGG-ATTTAGATATT--GGTTGGATTATACC-TAAAGATATTCGGCACGAGTACAATCTACTAGTTCTGAATT
            ||||||||||||-|||||||||||-|||||-|||||||||||--||||||||||||||-|||||||||||||||||||||||||||||||||||||||||
   correct: TCGGGTCGAGGGGCAACGAACGCCCGATGGGATTTAGATATTTTGGTTGGATTATACCCTAAAGATATTCGGCACGAGTACAATCTACTAGTTCTGAATT
The accuracy for this sequence is: 94.0%


Basecalling signal 75 out of 200
The alignment for signal 75 is:
basecalled: TGTGTCGTACAAGTGACAGACTCGCCC-ATAACGCTATAATGGCAACAAGCCCGGCACACACCCAGAATAGTCGGCCGCTGCGCCGAACTGACCGCCCGAGA
            |||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TGTGTCGTACAAGTGACAGACTCGCCCCATAACGCTATAATGGCAACAAGCCCGGCACACACCCAGAATAGTCGGCCGCTGCGCCGAACTGACCGCCCGAGA
The accuracy for this sequence is: 99.01960784313727%


Basecalling signal 76 out of 200
The alignment for signal 76 is:
basecalled: CGTGAAGATTCACACGTCGCAAATTTGTCTGGGCAGCTCTACC-GTTAGTTACCGGTACAGTATCCATCAAGCGCGCAGAAGTGCCTAATATTGGAGG
            |||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: CGTGAAGATTCACACGTCGCAAATTTGTCTGGGCAGCTCTACCCGTTAGTTACCGGTACAGTATCCATCAAGCGCGCAGAAGTGCCTAATATTGGAGG
The accuracy for this sequence is: 98.9795918367347%


Basecalling signal 77 out of 200
The alignment for signal 77 is:
basecalled: ACGTGAGGCGG-CTTATAGACCCGCCATTCTCAGTGCCCGATTCCTGTTGCAGTGATCTCTTGAATCGGGAGTAA-GATTCCTTTGGAGG-CCGTTAGTG
            |||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||-|||||||||
   correct: ACGTGAGGCGGGCTTATAGACCCGCCATTCTCAGTGCCCGATTCCTGTTGCAGTGATCTCTTGAATCGGGAGTAAAGATTCCTTTGGAGGGCCGTTAGTG
The accuracy for this sequence is: 97.0%


Basecalling signal 78 out of 200
The alignment for signal 78 is:
basecalled: TACCTAACGCATTATTAAGATACAGCATCCGAATATCGATGATGCGCAGCACATGATCAGTACACGATCGACGCAA-GCGG-CCTAGTCCAACCGCGCGGTGGTG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||-|||||||||||||||||||||||
   correct: TACCTAACGCATTATTAAGATACAGCATCCGAATATCGATGATGCGCAGCACATGATCAGTACACGATCGACGCAAAGCGGGCCTAGTCCAACCGCGCGGTGGTG
The accuracy for this sequence is: 98.09523809523809%


Basecalling signal 79 out of 200
The alignment for signal 79 is:
basecalled: GTCTT-CTCGTGAGGTTCCAGGTTCTGATCGGCTCGCTTT-CGG-ATCGCTGCATCGCTCGGGAAGAGCTGTT-CC-TCGTTCTGG-TTGGGT-
            |||||-||||||||||||||||||||||||||||||||||-|||-||||||||||||||||||||||||||||-||-|||||||||-||||||-
   correct: GTCTTTCTCGTGAGGTTCCAGGTTCTGATCGGCTCGCTTTTCGGGATCGCTGCATCGCTCGGGAAGAGCTGTTTCCCTCGTTCTGGGTTGGGTT
The accuracy for this sequence is: 92.5531914893617%


Basecalling signal 80 out of 200
The alignment for signal 80 is:
basecalled: CCAGCGCAGAGCGG-CACCTT--CC-GTGAACGAGCTATTAACGACGGTTACTGGTCTCAGCTCTCTTCTGGAGGGAGTAAGCCGTTGTAGGTT-AGGATTT
            ||||||||||||||-||||||--||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||
   correct: CCAGCGCAGAGCGGGCACCTTTTCCCGTGAACGAGCTATTAACGACGGTTACTGGTCTCAGCTCTCTTCTGGAGGGAGTAAGCCGTTGTAGGTTTAGGATTT
The accuracy for this sequence is: 95.09803921568627%


Basecalling signal 81 out of 200
The alignment for signal 81 is:
basecalled: ATGAGGG-ACCAATTCAGCAGTTCTTCACTAGTTCGCTG-AAGGAAGCTTGTCTCGGCTTTT-GGCGATCCGAGACGG-CTCCTGGAGCCGCCGGTTCGCGTATGG
            |||||||-|||||||||||||||||||||||||||||||-||||||||||||||||||||||-|||||||||||||||-|||||||||||||||||||||||||||
   correct: ATGAGGGGACCAATTCAGCAGTTCTTCACTAGTTCGCTGTAAGGAAGCTTGTCTCGGCTTTTTGGCGATCCGAGACGGGCTCCTGGAGCCGCCGGTTCGCGTATGG
The accuracy for this sequence is: 96.22641509433963%


Basecalling signal 82 out of 200
The alignment for signal 82 is:
basecalled: CGAGTATTGGCAACC-GTGGCGAGTCAACTACGGTGCTGAAAA--CGG-CCAGTATATGGCTAATA-GACACATGTAGACAGTTGTCGTTCTCCGCTCT--
            |||||||||||||||-|||||||||||||||||||||||||||--|||-|||||||||||||||||-||||||||||||||||||||||||||||||||--
   correct: CGAGTATTGGCAACCCGTGGCGAGTCAACTACGGTGCTGAAAAAACGGGCCAGTATATGGCTAATATGACACATGTAGACAGTTGTCGTTCTCCGCTCTTT
The accuracy for this sequence is: 93.06930693069306%


Basecalling signal 83 out of 200
The alignment for signal 83 is:
basecalled: GCCCATGGATACC-GCTGGTATAAGATTATAAGGCCTATCAGACAGGTAATTAAGGG-TTCAGACGTCGG-TATAGGACGGGCATCGAGACG--
            |||||||||||||-|||||||||||||||||||||||||||||||||||||||||||-||||||||||||-|||||||||||||||||||||--
   correct: GCCCATGGATACCCGCTGGTATAAGATTATAAGGCCTATCAGACAGGTAATTAAGGGGTTCAGACGTCGGGTATAGGACGGGCATCGAGACGGG
The accuracy for this sequence is: 94.68085106382979%


Basecalling signal 84 out of 200
The alignment for signal 84 is:
basecalled: CTAGCCCAGTTT-ATTAGAGCAGTTT-CGTAAA--CGAACACTAGCCACGATTGCCACC-GTCGTACCAACGTACAGCTACGGTAGGTTT-ACAATAAA
            ||||||||||||-|||||||||||||-||||||--||||||||||||||||||||||||-||||||||||||||||||||||||||||||-||||||||
   correct: CTAGCCCAGTTTTATTAGAGCAGTTTTCGTAAAAACGAACACTAGCCACGATTGCCACCCGTCGTACCAACGTACAGCTACGGTAGGTTTTACAATAAA
The accuracy for this sequence is: 93.93939393939394%


Basecalling signal 85 out of 200
The alignment for signal 85 is:
basecalled: TAAGTGCTCTT-AGATGCGGTTAGGTGGCCAACAGTGTT-CTAA-CTGATACTTGGAACGTGAAGGAACTGATATGATTCGG-CGACCATTCC-GTGGTGCT
            |||||||||||-|||||||||||||||||||||||||||-||||-||||||||||||-||||||||||||||||||||||||-||||||||||-||||||||
   correct: TAAGTGCTCTTTAGATGCGGTTAGGTGGCCAACAGTGTTTCTAAACTGATACTTGGA-CGTGAAGGAACTGATATGATTCGGGCGACCATTCCCGTGGTGCT
The accuracy for this sequence is: 94.05940594059405%


Basecalling signal 86 out of 200
The alignment for signal 86 is:
basecalled: GTGTTAGACCATAACGCGCAACGGTGCTGTAA-CTATTTGGACCTCTGCG-GCAAGTACAATTGTAATGATGAAACAATTAGCGGATCCTGGCA-
            ||||||||||||||||||||||||||||||||-|||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||-
   correct: GTGTTAGACCATAACGCGCAACGGTGCTGTAAACTATTTGGACCTCTGCGCGCAAGTACAATTGTAATGATGAAACAATTAGCGGATCCTGGCAA
The accuracy for this sequence is: 96.84210526315789%


Basecalling signal 87 out of 200
The alignment for signal 87 is:
basecalled: TTGTTT--ACTATTGTGCTGATATATATAGCCAATAACACCGAGTCCTAGCGGAGAGTGGTGGATAGCCGG-CCGGAAGG-AATACATATCAGCTGGCG-
            ||||||--|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||-||||||||||||||||||-
   correct: TTGTTTTTACTATTGTGCTGATATATATAGCCAATAACACCGAGTCCTAGCGGAGAGTGGTGGATAGCCGGGCCGGAAGGGAATACATATCAGCTGGCGG
The accuracy for this sequence is: 95.0%


Basecalling signal 88 out of 200
The alignment for signal 88 is:
basecalled: GAACTGCCGGTTGAGGTCGACC-TTGGGTATGGAACACCAATCTGCAAA-GCCGGAACGCAAGATCACTTCTGGACGTTGCGTTGTT-GGTACTA-
            ||||||||||||||||||||||-||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||-|||||||-
   correct: GAACTGCCGGTTGAGGTCGACCCTTGGGTATGGAACACCAATCTGCAAAAGCCGGAACGCAAGATCACTTCTGGACGTTGCGTTGTTTGGTACTAA
The accuracy for this sequence is: 95.83333333333334%


Basecalling signal 89 out of 200
The alignment for signal 89 is:
basecalled: TTCGAA-TATGGTACATTAGTTCATGGG-ACGAGGGATCTCCTTACCCTGAAA-TTT-CCGTGGTGTGTTGCACTTGTGAGTTGTCGTCGCCGGA-
            ||||||-|||||||||||||||||||||-||||||||||||||||||||||||-|||-|||||||||||||||||||||||||||||||||||||-
   correct: TTCGAAATATGGTACATTAGTTCATGGGGACGAGGGATCTCCTTACCCTGAAAATTTTCCGTGGTGTGTTGCACTTGTGAGTTGTCGTCGCCGGAA
The accuracy for this sequence is: 94.79166666666666%


Basecalling signal 90 out of 200
The alignment for signal 90 is:
basecalled: GTTGTAATCTACAGCGAAGATCGGAGGCGATAGTGTAGGGTCTT-GCTGAAGG-CTTAGGCGTGGCTTCATAAGAAATTAGTAACGCATTCGCCGAAAA--
            ||||||||||||||||||||||||||||||||||||||||||||-||||||||-|||||||||||||||||||||||||||||||||||||||||||||--
   correct: GTTGTAATCTACAGCGAAGATCGGAGGCGATAGTGTAGGGTCTTTGCTGAAGGGCTTAGGCGTGGCTTCATAAGAAATTAGTAACGCATTCGCCGAAAAAA
The accuracy for this sequence is: 96.03960396039604%


Basecalling signal 91 out of 200
The alignment for signal 91 is:
basecalled: CGCGGTATTTGCGAGTTAGG-CCTGATTGCATATAACATCACCTGGG--TTAGAGTCATGTATGTACCTGTGCGTTAAA-TATCAACATACACC-GG
            ||||||||||||||||||||-||||||||||||||||||||||||||--||||||||||||||||||||||||||||||-||||||||||||||-||
   correct: CGCGGTATTTGCGAGTTAGGGCCTGATTGCATATAACATCACCTGGGGGTTAGAGTCATGTATGTACCTGTGCGTTAAAATATCAACATACACCCGG
The accuracy for this sequence is: 94.84536082474226%


Basecalling signal 92 out of 200
The alignment for signal 92 is:
basecalled: ACAACCGACTCC-AGTCACGGCCTGCGGACTTGGGACCTTT-CTGCATTAGGCTTCTAAA-TCCTCTGAA-CTTAGCCTCTAGCC-TCCGTGTACTT-AACG
            ||||||||||||-||||||||||||||||||||||||||||-||||||||||||||||||-|||||||||-||||||||||||||-|||||||||||-||||
   correct: ACAACCGACTCCCAGTCACGGCCTGCGGACTTGGGACCTTTTCTGCATTAGGCTTCTAAAATCCTCTGAAACTTAGCCTCTAGCCCTCCGTGTACTTTAACG
The accuracy for this sequence is: 94.11764705882352%


Basecalling signal 93 out of 200
The alignment for signal 93 is:
basecalled: GTAGTATGCGCTGTAATTCTTGTAGCC-GCGCCGCCGGTGCGG-CAGAGCC-TATATGTCTGAACCCGCACACGTACTT-AGCGGCGTCAAACTCCATGCTGACCTACTC
            |||||||||||||||||||||||||||-|||||||||||||||-|||||||-|||||||||||||||||||||||||||-||||||||||||||||||||||||||||||
   correct: GTAGTATGCGCTGTAATTCTTGTAGCCCGCGCCGCCGGTGCGGGCAGAGCCCTATATGTCTGAACCCGCACACGTACTTTAGCGGCGTCAAACTCCATGCTGACCTACTC
The accuracy for this sequence is: 96.36363636363636%


Basecalling signal 94 out of 200
The alignment for signal 94 is:
basecalled: CTCGG-CCGCC--GAATAACAACGTCCGGGTTGCGTACGCGGACCCGGCC-GCAGATGCACGCAGG-CACAGGG-AAGCATAGAGACGATTATTACGCCG
            |||||-|||||--|||||||||||||||||||||||||||||||||||||-|||||||||||||||-|||||||-|||||||||||||||||||||||||
   correct: CTCGGGCCGCCCCGAATAACAACGTCCGGGTTGCGTACGCGGACCCGGCCCGCAGATGCACGCAGGGCACAGGGGAAGCATAGAGACGATTATTACGCCG
The accuracy for this sequence is: 94.0%


Basecalling signal 95 out of 200
The alignment for signal 95 is:
basecalled: GAGATTTCCGGG-CACTTGGAGCACGAA-TATTGGACGCTCTTGAAACGGTACGTAATCGGCAATATTCGTACCGACCAGTGAGCTTGATAGTTGTATTGAAG
            ||||||||||||-|||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GAGATTTCCGGGGCACTTGGAGCACGAAATATTGGACGCTCTTGAAACGGTACGTAATCGGCAATATTCGTACCGACCAGTGAGCTTGATAGTTGTATTGAAG
The accuracy for this sequence is: 98.05825242718447%


Basecalling signal 96 out of 200
The alignment for signal 96 is:
basecalled: TATACATCAGAACAGTGTATTCTCCTCCAGACGTTCGCAGGCACCGCCTGTACTCTATCAGTGAGG-AAGACGGAGTCCAACATAGGTGCACTT-AC
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||-||
   correct: TATACATCAGAACAGTGTATTCTCCTCCAGACGTTCGCAGGCACCGCCTGTACTCTATCAGTGAGGGAAGACGGAGTCCAACATAGGTGCACTTTAC
The accuracy for this sequence is: 97.9381443298969%


Basecalling signal 97 out of 200
The alignment for signal 97 is:
basecalled: GTATGGAGTTT-GTGCATT-CTGTTT-CTTGAGCCTATCCAGGAACCCTTGAAGATGGTTCTAACGTCAGTCATT-GACATCATGAA--CTAGCACCGCC-
            |||||||||||-|||||||-||||||-||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||--|||||||||||-
   correct: GTATGGAGTTTTGTGCATTTCTGTTTTCTTGAGCCTATCCAGGAACCCTTGAAGATGGTTCTAACGTCAGTCATTTGACATCATGAAAACTAGCACCGCCA
The accuracy for this sequence is: 93.06930693069306%


Basecalling signal 98 out of 200
The alignment for signal 98 is:
basecalled: CCGG-ATTGCTCGGAGCGGTTCCCAACTCGACTGAAGCGGACAA-CTATCGCGTACGATGGG--CAGTCCGGAATCCGACC-GCGATACAGTAACCAGGAGA
            ||||-|||||||||||||||||||||||||||||||||||||||-|||||||||||||||||--|||||||||||||||||-||||||||||||||||||||
   correct: CCGGGATTGCTCGGAGCGGTTCCCAACTCGACTGAAGCGGACAAACTATCGCGTACGATGGGGGCAGTCCGGAATCCGACCCGCGATACAGTAACCAGGAGA
The accuracy for this sequence is: 95.09803921568627%


Basecalling signal 99 out of 200
The alignment for signal 99 is:
basecalled: ACACGGAACGTCCCAATGACCGAT-TCTGTTGATGTTCACTTGTTGCGACGCTAAA--TACCGACCTCACGCGGGCGCACTTTACGGTCTGGTTGGAT
            ||||||||||||||||||||||||-|||||||||||||||||||||||||||||||--||||||||||||||||||||||||||||||||||||||||
   correct: ACACGGAACGTCCCAATGACCGATATCTGTTGATGTTCACTTGTTGCGACGCTAAAAATACCGACCTCACGCGGGCGCACTTTACGGTCTGGTTGGAT
The accuracy for this sequence is: 96.93877551020408%


Basecalling signal 100 out of 200
The alignment for signal 100 is:
basecalled: ACGGAATTGCAGTGACAGAACACACGGAGACGCGATAATACATA-TAGATTG-CAA-GCTTCAGACCCTT-GTCAAGTTGCGCGTTCCAAGTTCCACCC-TCCCAG
            ||||||||||||||||||||||||||||||||||||||||||||-|||||||-|||-|||||||||||||-||||||||||||||||||||||||||||-||||||
   correct: ACGGAATTGCAGTGACAGAACACACGGAGACGCGATAATACATAATAGATTGGCAAAGCTTCAGACCCTTTGTCAAGTTGCGCGTTCCAAGTTCCACCCCTCCCAG
The accuracy for this sequence is: 95.28301886792453%


Basecalling signal 101 out of 200
The alignment for signal 101 is:
basecalled: CATCCTGTCACCGCTTGCACCAGGTAATGTAGTTGGTCATTACGCGATGTCTCAATCTAACGACTCTCGCGCGTCATCGGATCGGTACGG-CGTAA
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||
   correct: CATCCTGTCACCGCTTGCACCAGGTAATGTAGTTGGTCATTACGCGATGTCTCAATCTAACGACTCTCGCGCGTCATCGGATCGGTACGGGCGTAA
The accuracy for this sequence is: 98.95833333333334%


Basecalling signal 102 out of 200
The alignment for signal 102 is:
basecalled: CGTGCTCAA-TGCCAAGGCGTCGTGAACATGGG-CTATTGCCGCGTGTGTGAA-CACCGTTGTAGCCGTTCATACATGAA-CTACGCGG-TATTAA
            |||||||||-|||||||||||||||||||||||-|||||||||||||||||||-||||||||||||||||||||||||||-||||||||-||||||
   correct: CGTGCTCAAATGCCAAGGCGTCGTGAACATGGGGCTATTGCCGCGTGTGTGAAACACCGTTGTAGCCGTTCATACATGAAACTACGCGGGTATTAA
The accuracy for this sequence is: 94.79166666666666%


Basecalling signal 103 out of 200
The alignment for signal 103 is:
basecalled: TACCGTTCCTTTGCTAGATATCGTGAA-CCAGCGACGACCTCGGGCAGGATGGTTCACCGTACCTCC-TGAACCTGAGCCGTGTGATCATCTCCGTAA
            |||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||
   correct: TACCGTTCCTTTGCTAGATATCGTGAAACCAGCGACGACCTCGGGCAGGATGGTTCACCGTACCTCCCTGAACCTGAGCCGTGTGATCATCTCCGTAA
The accuracy for this sequence is: 97.95918367346938%


Basecalling signal 104 out of 200
The alignment for signal 104 is:
basecalled: CAAGTGACTAGGTTGTCCGTT-ACAACGATTTGTAGACGCACCGTGCCCGCAATGTGTGCAGGCCAA-CGACTTACCGTACACTAGGATCGGTGACGTCTTA
            |||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||
   correct: CAAGTGACTAGGTTGTCCGTTTACAACGATTTGTAGACGCACCGTGCCCGCAATGTGTGCAGGCCAAACGACTTACCGTACACTAGGATCGGTGACGTCTTA
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 105 out of 200
The alignment for signal 105 is:
basecalled: GCCGTCTCCATAGTT-GTAAGGTAAA-TATACGTACTATGCCCTGCTGGTACATCTACCC-ATATTCCGAGCTAGAAGGAGAGAAACCTGAA-CTGTTCAGCG
            |||||||||||||||-||||||||||-|||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||-||||||||||
   correct: GCCGTCTCCATAGTTTGTAAGGTAAAATATACGTACTATGCCCTGCTGGTACATCTACCCCATATTCCGAGCTAGAAGGAGAGAAACCTGAAACTGTTCAGCG
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 106 out of 200
The alignment for signal 106 is:
basecalled: TATATGGAAGCTATCGGACATATTATCCC-TTACAGTTAAACATAAA-CGGTGCGTTATTGTCC-TCTCTATATTGTCAGAACCTAGCAGGG-TGACC-
            |||||||||||||||||||||||||||||-|||||||||||||||||-||||||||||||||||-|||||||||||||||||||||||||||-|||||-
   correct: TATATGGAAGCTATCGGACATATTATCCCCTTACAGTTAAACATAAAACGGTGCGTTATTGTCCCTCTCTATATTGTCAGAACCTAGCAGGGGTGACCC
The accuracy for this sequence is: 94.94949494949495%


Basecalling signal 107 out of 200
The alignment for signal 107 is:
basecalled: CGGTGAGTAGCCGACTT-GTTTCTCTATAGAGGCTT-AACGCTATCAA-CACC-GATACGTGTAGGTAATCTCATACAGATGGCCATTTCTCTTAGCCGT
            |||||||||||||||||-||||||||||||||||||-|||||||||||-||||-||||||||||||||||||||||||||||||||||||||||||||||
   correct: CGGTGAGTAGCCGACTTTGTTTCTCTATAGAGGCTTTAACGCTATCAAACACCCGATACGTGTAGGTAATCTCATACAGATGGCCATTTCTCTTAGCCGT
The accuracy for this sequence is: 96.0%


Basecalling signal 108 out of 200
The alignment for signal 108 is:
basecalled: CCTGCTCC-AGTCC-TCGCC-GCATTCC-ACTGATAGCGTACTGCC--AAA-GGAGG-CAAGCGTGAGG-TGTTAA-GTCTGCC-GGAGCCATAGTTGCACAC-
            ||||||||-|||||-|||||-|||||||-|||||||||||||||||--|||-|||||-|||||||||||-||||||-|||||||-||||||||||||||||||-
   correct: CCTGCTCCCAGTCCCTCGCCCGCATTCCCACTGATAGCGTACTGCCCCAAAAGGAGGGCAAGCGTGAGGGTGTTAAAGTCTGCCCGGAGCCATAGTTGCACACT
The accuracy for this sequence is: 88.46153846153845%


Basecalling signal 109 out of 200
The alignment for signal 109 is:
basecalled: AGACTGATCAA-TAATGAACCCTTATCCC-AATACGAATGAATACTACTGTCTGCGG-CAGGTCGGTCACGAGCAGAGCCTCTACTGCGGTAA-CTCTA-
            |||||||||||-|||||||||||||||||-|||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||-|||||-
   correct: AGACTGATCAAATAATGAACCCTTATCCCCAATACGAATGAATACTACTGTCTGCGGGCAGGTCGGTCACGAGCAGAGCCTCTACTGCGGTAAACTCTAT
The accuracy for this sequence is: 95.0%


Basecalling signal 110 out of 200
The alignment for signal 110 is:
basecalled: ATCAAGGCTCCCGTGTCGGGTAGGCATACGGCATAATCGTGACATGCACGG-TGCCTTT-CGCCTTCAGCACGAAGTCCACCTTT-GACTGCTTGTAA
            |||||||||||||||||||||||||||||||||||-|||||||||||||||-|||||||-|||||||||||||||||||||||||-||||||||||||
   correct: ATCAAGGCTCCCGTGTCGGGTAGGCATACGGCATA-TCGTGACATGCACGGGTGCCTTTTCGCCTTCAGCACGAAGTCCACCTTTTGACTGCTTGTAA
The accuracy for this sequence is: 95.87628865979381%


Basecalling signal 111 out of 200
The alignment for signal 111 is:
basecalled: ATTATCAACTGTAACATCGATCACAA-TAGTTAGCTACCC-GGACCTTACCTCGCTCTAAGGGG-CACACGCGATACGCGATGCC-AATACGTACAGT
            ||||||||||||||||||||||||||-|||||||||||||-|||||||||||||||||||||||-||||||||||||||||||||-||||||||||||
   correct: ATTATCAACTGTAACATCGATCACAAATAGTTAGCTACCCCGGACCTTACCTCGCTCTAAGGGGGCACACGCGATACGCGATGCCCAATACGTACAGT
The accuracy for this sequence is: 95.91836734693877%


Basecalling signal 112 out of 200
The alignment for signal 112 is:
basecalled: GAGCCAGATTGCACGCATTGCGTCATCTTCAGGAAGGG-CTTGTGTTAACTAATCGGCACCGAGGACCGGCTACGATATAATGATAAGTAGCTTTCAT
            ||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GAGCCAGATTGCACGCATTGCGTCATCTTCAGGAAGGGGCTTGTGTTAACTAATCGGCACCGAGGACCGGCTACGATATAATGATAAGTAGCTTTCAT
The accuracy for this sequence is: 98.9795918367347%


Basecalling signal 113 out of 200
The alignment for signal 113 is:
basecalled: GGTCGGTAAA-GCACTGGGACTGTGCCAGCATTGATCGAA-TAATCGGCCGGG-CTCTTGTTT-CTCTAGGTTATT-ATTCCC-GTCC-AACAGGCC
            ||||||||||-|||||||||||||||||||||||||||||-||||||||||||-|||||||||-||||||||||||-||||||-||||-||||||||
   correct: GGTCGGTAAAAGCACTGGGACTGTGCCAGCATTGATCGAAATAATCGGCCGGGGCTCTTGTTTTCTCTAGGTTATTTATTCCCCGTCCCAACAGGCC
The accuracy for this sequence is: 92.78350515463917%


Basecalling signal 114 out of 200
The alignment for signal 114 is:
basecalled: TGGATCTAGGCCACCGCACCCAGACCTCTTCGCACCTGCTTT-CATGAGTGTGTTTGG-CTGTGAGGTATGTATTGTGTAATT-CAGGATGACAGCA-
            ||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||-||||||||||||||||||||||||-|||||||||||||-
   correct: TGGATCTAGGCCACCGCACCCAGACCTCTTCGCACCTGCTTTTCATGAGTGTGTTTGGGCTGTGAGGTATGTATTGTGTAATTTCAGGATGACAGCAA
The accuracy for this sequence is: 95.91836734693877%


Basecalling signal 115 out of 200
The alignment for signal 115 is:
basecalled: TCGTTT-CACTCC-GTATGTACGCAGTTACAAGTT-AAA-CACTTCACTTATGACGTGTGAGTAGTAGGTACTCGTTAGTAAAA----GTACGCTCCCTA-
            ||||||-||||||-|||||||||||||||||||||-|||-||||||||||||||||||||||||||||||||||||||||||||----||||||||||||-
   correct: TCGTTTTCACTCCCGTATGTACGCAGTTACAAGTTTAAAACACTTCACTTATGACGTGTGAGTAGTAGGTACTCGTTAGTAAAAAAAAGTACGCTCCCTAA
The accuracy for this sequence is: 91.0891089108911%


Basecalling signal 116 out of 200
The alignment for signal 116 is:
basecalled: CTTATTATGACTCGCACTCAGTCTTCCGACGGTTAGGTAGACAACAGGAGAATCCATCCGTGCAAGCCGTCCGATTCTGTGGACAACCTCGGCATAATGGCC-
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-
   correct: CTTATTATGACTCGCACTCAGTCTTCCGACGGTTAGGTAGACAACAGGAGAATCCATCCGTGCAAGCCGTCCGATTCTGTGGACAACCTCGGCATAATGGCCT
The accuracy for this sequence is: 99.02912621359224%


Basecalling signal 117 out of 200
The alignment for signal 117 is:
basecalled: ATCGACAACTCGGCGG-CAAA-TATTCGACTTGGATCTGAAGG-CCGTCTACGTTACATTCATCACATTCGCCGTCAGGTGAGTCTTT-CCAGCTCCTCG
            ||||||||||||||||-||||-|||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||-|||||||||||
   correct: ATCGACAACTCGGCGGGCAAAATATTCGACTTGGATCTGAAGGGCCGTCTACGTTACATTCATCACATTCGCCGTCAGGTGAGTCTTTTCCAGCTCCTCG
The accuracy for this sequence is: 96.0%


Basecalling signal 118 out of 200
The alignment for signal 118 is:
basecalled: GATTACCGGTCCTCGATATTATAATTGGTGACC-TACTCGTGTGCATGTACAAGCATTCTCTTAGTGCCC-GTACGACGTCCAGAGAGTCATTCGCTA
            |||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||
   correct: GATTACCGGTCCTCGATATTATAATTGGTGACCCTACTCGTGTGCATGTACAAGCATTCTCTTAGTGCCCCGTACGACGTCCAGAGAGTCATTCGCTA
The accuracy for this sequence is: 97.95918367346938%


Basecalling signal 119 out of 200
The alignment for signal 119 is:
basecalled: GATGTATACATGTACGCGCAGATGTACTTATGGTTCCAGTTGAGAGCTCCTGCTACCTATT-ATGGCC-AGCCGGTAA-CGGCGCAA-GCCTC
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||-|||||||||-||||||||-|||||
   correct: GATGTATACATGTACGCGCAGATGTACTTATGGTTCCAGTTGAGAGCTCCTGCTACCTATTTATGGCCCAGCCGGTAAACGGCGCAAAGCCTC
The accuracy for this sequence is: 95.6989247311828%


Basecalling signal 120 out of 200
The alignment for signal 120 is:
basecalled: CGTTAGG-TTCCGGTTTCAGTACCGAAGAGGTGCC-ATGCATTATCTACTGTCAGGATGG-AGTCGTGAGCCTTT-CAATCGTAGACCTCTTAAGTGCATC
            |||||||-|||||||||||||||||||||||||||-||||||||||||||||||||||||-||||||||||||||-|||||||||||||||||||||||||
   correct: CGTTAGGGTTCCGGTTTCAGTACCGAAGAGGTGCCCATGCATTATCTACTGTCAGGATGGGAGTCGTGAGCCTTTTCAATCGTAGACCTCTTAAGTGCATC
The accuracy for this sequence is: 96.03960396039604%


Basecalling signal 121 out of 200
The alignment for signal 121 is:
basecalled: CGTGATGG-CGATAGGTT-ACCGTTTTGTCAACTATACACTTAAATGTCAAATAATGACCAGGTTCGATCAGGCGGTGACGAACCTCGACTTGACA
            ||||||||-|||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: CGTGATGGGCGATAGGTTTACCGTTTTGTCAACTATACACTTAAATGTCAAATAATGACCAGGTTCGATCAGGCGGTGACGAACCTCGACTTGACA
The accuracy for this sequence is: 97.91666666666666%


Basecalling signal 122 out of 200
The alignment for signal 122 is:
basecalled: CAATCTAACC--AACCACACC-TTATGACGCCACTA-ATCCGATATGCC-TGTGAGACACGCTGATGAAA-TACCAATAAGCACATTCCGACG-GTTAGCCACCT
            ||||||||||--|||||||||-||||||||||||||-||||||||||||-||||||||||||||||||||-||||||||||||||||||||||-|||||||||||
   correct: CAATCTAACCCCAACCACACCCTTATGACGCCACTACATCCGATATGCCCTGTGAGACACGCTGATGAAAATACCAATAAGCACATTCCGACGAGTTAGCCACCT
The accuracy for this sequence is: 93.33333333333333%


Basecalling signal 123 out of 200
The alignment for signal 123 is:
basecalled: CGTGATACTGCACAATT--CGTGTATAAGGTCCGCTTTACTGG-AACATAATTCC-TGATTTAAGCGGACTCCGCTGAGCC-GAGCTGG-ACTTT
            |||||||||||||||||--||||||||||||||||||||||||-|||||||||||-|||||||||||||||||||||||||-|||||||-|||||
   correct: CGTGATACTGCACAATTTTCGTGTATAAGGTCCGCTTTACTGGGAACATAATTCCCTGATTTAAGCGGACTCCGCTGAGCCCGAGCTGGGACTTT
The accuracy for this sequence is: 93.6842105263158%


Basecalling signal 124 out of 200
The alignment for signal 124 is:
basecalled: GTATGCTTATTAAA-GGCAGTCGCCTGACCGTCGCGACGATAACGAGACGGGCTTTAATGTAGCAACGTCC-GGCCCC-ACAAACTCAATAA-AGGC
            ||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||-|||||||||||||-||||
   correct: GTATGCTTATTAAAAGGCAGTCGCCTGACCGTCGCGACGATAACGAGACGGGCTTTAATGTAGCAACGTCCCGGCCCCCACAAACTCAATAAGAGGC
The accuracy for this sequence is: 95.87628865979381%


Basecalling signal 125 out of 200
The alignment for signal 125 is:
basecalled: TGCAGGGTGCGATATCTCCGTACACAGGTCC-TGGATAGACCTTCTGACGAACATGTTTCGCCGCCGTAATCACGTAACTGGACTTAGTGACACCTAT
            |||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TGCAGGGTGCGATATCTCCGTACACAGGTCCCTGGATAGACCTTCTGACGAACATGTTTCGCCGCCGTAATCACGTAACTGGACTTAGTGACACCTAT
The accuracy for this sequence is: 98.9795918367347%


Basecalling signal 126 out of 200
The alignment for signal 126 is:
basecalled: TTACTTCCCTGTAGGCTTCGTTACCTAGTAGCTTT-ATTCGTGGACACTGGCGCATGGAGTGGGCTTAA-CCGTAGACTAGATTCGTGCCAAGCGTAA-TAGT
            |||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||-||||
   correct: TTACTTCCCTGTAGGCTTCGTTACCTAGTAGCTTTTATTCGTGGACACTGGCGCATGGAGTGGGCTTAAACCGTAGACTAGATTCGTGCCAAGCGTAAATAGT
The accuracy for this sequence is: 97.0873786407767%


Basecalling signal 127 out of 200
The alignment for signal 127 is:
basecalled: AACAGGCTCGCACGCATAAGTGCAACTACAATGTCATACTCATGGCCAGTTTGCAGAAAA-TGTT-ATAATTGTCAGGG-TGGTGATTGAGCAG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||-|||||||||||||-||||||||||||||
   correct: AACAGGCTCGCACGCATAAGTGCAACTACAATGTCATACTCATGGCCAGTTTGCAGAAAAATGTTTATAATTGTCAGGGGTGGTGATTGAGCAG
The accuracy for this sequence is: 96.80851063829788%


Basecalling signal 128 out of 200
The alignment for signal 128 is:
basecalled: TCTTCGGTGCGGTTTCAATCGTCGAATGTTATT-CGGCGCTCCGCGTCAGCTGAGATGATTGCGTTCACCCACGCC-GAGCTGTT-GGGCTCTGG-ACTGC
            |||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||-||||||||-|||||||||-|||||
   correct: TCTTCGGTGCGGTTTCAATCGTCGAATGTTATTTCGGCGCTCCGCGTCAGCTGAGATGATTGCGTTCACCCACGCCCGAGCTGTTTGGGCTCTGGGACTGC
The accuracy for this sequence is: 96.03960396039604%


Basecalling signal 129 out of 200
The alignment for signal 129 is:
basecalled: ACATGGTTCACAAGAAGAGGGTTCCGGAAGG-ATCCCACTATACAATTTT-GCCGGCGAAAGGGCGCCGAGTCATATAGG-AGCCCATT-AGTTGCACG
            |||||||||||||||||||||||||||||||-||||||||||||||||||-|||||||||||||||||||||||||||||-||||||||-|||||||||
   correct: ACATGGTTCACAAGAAGAGGGTTCCGGAAGGGATCCCACTATACAATTTTTGCCGGCGAAAGGGCGCCGAGTCATATAGGGAGCCCATTGAGTTGCACG
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 130 out of 200
The alignment for signal 130 is:
basecalled: GGCGGCCAACCAGATACTGTGAACGTAGTGAGTAATGAACACTTCAACCAGGATTATT-CAGACCACGCGCACTAGTGGTACACGCTCAGTGTTGAATTCGCTG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||
   correct: GGCGGCCAACCAGATACTGTGAACGTAGTGAGTAATGAACACTTCAACCAGGATTATTTCAGACCACGCGCACTAGTGGTACACGCTCAGTGTTGAATTCGCTG
The accuracy for this sequence is: 99.03846153846155%


Basecalling signal 131 out of 200
The alignment for signal 131 is:
basecalled: GGGAGGAAAGCTTAGTCCGAACCTTAACTTAATTGGCACTACTCGTGATGCAGAAGAGCCAAATCAGGCGCGTGGTGGAATAGAAATATGTCGAGGTG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GGGAGGAAAGCTTAGTCCGAACCTTAACTTAATTGGCACTACTCGTGATGCAGAAGAGCCAAATCAGGCGCGTGGTGGAATAGAAATATGTCGAGGTG
The accuracy for this sequence is: 100.0%


Basecalling signal 132 out of 200
The alignment for signal 132 is:
basecalled: CC-TGGAACCTATACACTCTATGCATGCCGGG-CCATT-CGTTGGCTGGG-TCATTACTCGCGCTTGAGTTACATTATGCTTCCGTGCCAGCACTTCTA
            ||-|||||||||||||||||||||||||||||-|||||-|||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||
   correct: CCCTGGAACCTATACACTCTATGCATGCCGGGGCCATTTCGTTGGCTGGGGTCATTACTCGCGCTTGAGTTACATTATGCTTCCGTGCCAGCACTTCTA
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 133 out of 200
The alignment for signal 133 is:
basecalled: GCTACAGTTTGCCAGTTCCAGCTTCTGGTGATCATGCTGTCTT-AATT-AAA-GAACGG-CATCTATTTCCAA-GACACCTTGAATACGAGGCCC-AATTT-ACT
            |||||||||||||||||||||||||||||||||||||||||||-||||-|||-||||||-|||||||||||||-|||||||||||||||||||||-|||||-|||
   correct: GCTACAGTTTGCCAGTTCCAGCTTCTGGTGATCATGCTGTCTTTAATTTAAAAGAACGGACATCTATTTCCAAAGACACCTTGAATACGAGGCCCCAATTTTACT
The accuracy for this sequence is: 93.33333333333333%


Basecalling signal 134 out of 200
The alignment for signal 134 is:
basecalled: GCGAA-TGG-ATGGTGCACGCGAAATTACGATTAGAATAACACACAGGAAAGTGGAACTAAGAATTGTGGCCTTAAGTATTTCGTATAGATAGGTTGCCGAAGT
            |||||-|||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GCGAAATGGGATGGTGCACGCGAAATTACGATTAGAATAACACACAGGAAAGTGGAACTAAGAATTGTGGCCTTAAGTATTTCGTATAGATAGGTTGCCGAAGT
The accuracy for this sequence is: 98.07692307692307%


Basecalling signal 135 out of 200
The alignment for signal 135 is:
basecalled: TTGCGTAACC-TCTAAGTCCACTT-GTGACGACGCAGTACTCCATGTTAGG-AA-CTGG--CACAGGCCCTT-GACGAACAACTATTTT-ACTCAACAC--
            ||||||||||-|||||||||||||-||||||||||||||||||||||||||-||-||||--|||||||||||-||||||||||||||||-|||||||||--
   correct: TTGCGTAACCCTCTAAGTCCACTTTGTGACGACGCAGTACTCCATGTTAGGGAAACTGGGGCACAGGCCCTTTGACGAACAACTATTTTTACTCAACACCC
The accuracy for this sequence is: 90.0990099009901%


Basecalling signal 136 out of 200
The alignment for signal 136 is:
basecalled: ACGTTT-CATCTTT--AGTCAGGACCATTAGTGCGG-AGTCACCGCCGAAGG-CC-GCTACCTCCTGTCCCC--GCAGGACACGCAAAGGTTAAGGCTCACT
            ||||||-|||||||--||||||||||||||||||||-|||||||||||||||-||-||||||||||||||||--||||||||||||||||||||||||||||
   correct: ACGTTTTCATCTTTTTAGTCAGGACCATTAGTGCGGGAGTCACCGCCGAAGGGCCCGCTACCTCCTGTCCCCCCGCAGGACACGCAAAGGTTAAGGCTCACT
The accuracy for this sequence is: 92.15686274509804%


Basecalling signal 137 out of 200
The alignment for signal 137 is:
basecalled: TTT-CGAGGACCGGAATGCTAAAGACAGTCGCTAAGACAGATTCTCTGTCCTCTATCCTAGGCTCTAATACTCGGTCGTAATCAACCTGCTTCCGTTCTGC
            |||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TTTTCGAGGACCGGAATGCTAAAGACAGTCGCTAAGACAGATTCTCTGTCCTCTATCCTAGGCTCTAATACTCGGTCGTAATCAACCTGCTTCCGTTCTGC
The accuracy for this sequence is: 99.00990099009901%


Basecalling signal 138 out of 200
The alignment for signal 138 is:
basecalled: TCTCACGGTTCGCATCAA-CAAA-CACCCTTTGGCGTAAGACACTACCTTATGGTGCCTTGTGAAGTGCTGGTCGTATCCTGTCGCGCTCGCTCCGCTT-GGCGT
            ||||||||||||||||||-||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||
   correct: TCTCACGGTTCGCATCAAACAAAACACCCTTTGGCGTAAGACACTACCTTATGGTGCCTTGTGAAGTGCTGGTCGTATCCTGTCGCGCTCGCTCCGCTTTGGCGT
The accuracy for this sequence is: 97.14285714285714%


Basecalling signal 139 out of 200
The alignment for signal 139 is:
basecalled: TAGGG--TACTAGTTCTCACACATCTCTCCCTT-CGCCATCTAATCGAATACGATTAACACTTATCCATGCGCGCGCTATTCCGAGATGCAGCTGCCTGCTGAT
            |||||--||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TAGGGGGTACTAGTTCTCACACATCTCTCCCTTTCGCCATCTAATCGAATACGATTAACACTTATCCATGCGCGCGCTATTCCGAGATGCAGCTGCCTGCTGAT
The accuracy for this sequence is: 97.11538461538461%


Basecalling signal 140 out of 200
The alignment for signal 140 is:
basecalled: AGACTGGCTCGTACGGG-CGTT-GCTCAACGTCAGACATCCC-AGACTATT-GAAA-CGTCTAATCAAGCATACAACC-TGTGTGCTAACAAGACACAAGAA-C
            |||||||||||||||||-||||-|||||||||||||||||||-||||||||-||||-|||||||||||||||||||||-|||||||||||||||||||||||-|
   correct: AGACTGGCTCGTACGGGGCGTTTGCTCAACGTCAGACATCCCCAGACTATTTGAAAACGTCTAATCAAGCATACAACCCTGTGTGCTAACAAGACACAAGAAAC
The accuracy for this sequence is: 93.26923076923077%


Basecalling signal 141 out of 200
The alignment for signal 141 is:
basecalled: TTATACACGGTCGGGG--TCCGTGTTATCCGAACGGTGTATTGATTCGTGCCATAAGCTGATCTACGGCGAACCCTCCCGGATAAGGGACTGACA
            ||||||||||||||||--|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TTATACACGGTCGGGGGGTCCGTGTTATCCGAACGGTGTATTGATTCGTGCCATAAGCTGATCTACGGCGAACCCTCCCGGATAAGGGACTGACA
The accuracy for this sequence is: 97.89473684210527%


Basecalling signal 142 out of 200
The alignment for signal 142 is:
basecalled: AGCGCCGCACGTGTATTTGGCCATTGTGATCATTGG-CCAACCC-AGAAGCGACGCCTATCCC-TAGACTGAACAGTAGTGTCACTACACTGAGAATCCC--G-
            ||||||||||||||||||||||||||||||||||||-|||||||-||||||||||||||||||-||||||||||||||||||||||||||||||||||||--|-
   correct: AGCGCCGCACGTGTATTTGGCCATTGTGATCATTGGGCCAACCCCAGAAGCGACGCCTATCCCCTAGACTGAACAGTAGTGTCACTACACTGAGAATCCCCCGG
The accuracy for this sequence is: 94.23076923076923%


Basecalling signal 143 out of 200
The alignment for signal 143 is:
basecalled: CATAGTCTAGTTGCTATT-CCC--AA-GCCTGCGTGTGATGATGAGAACTTCAGCCTATT-CAAGATGGAGGACGAACCTCTGGCGGAACATCTGCCGGGA
            ||||||||||||||||||-|||--||-|||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||
   correct: CATAGTCTAGTTGCTATTTCCCCCAAAGCCTGCGTGTGATGATGAGAACTTCAGCCTATTTCAAGATGGAGGACGAACCTCTGGCGGAACATCTGCCGGGA
The accuracy for this sequence is: 95.04950495049505%


Basecalling signal 144 out of 200
The alignment for signal 144 is:
basecalled: CCTTAGCCTCTTT-AATAATGCCGG-ATACAGCGCGAGAACGTGTCCTT-GACACAACACC-ATTTGCAGGAATACATAGACAGAGCATAGGAAGCCAGA
            |||||||||||||-|||||||||||-|||||||||||||||||||||||-|||||||||||-||||||||||||||||||||||||||||||||||||||
   correct: CCTTAGCCTCTTTTAATAATGCCGGGATACAGCGCGAGAACGTGTCCTTTGACACAACACCCATTTGCAGGAATACATAGACAGAGCATAGGAAGCCAGA
The accuracy for this sequence is: 96.0%


Basecalling signal 145 out of 200
The alignment for signal 145 is:
basecalled: CCC-TTAA-CCTTATGGG-AAGAGATGCCAACGACGTGATCGGG-TCCTAGAGTCATT-CCACGCACTTGCCGCTAACGCAGTTGGACATTCACG--
            |||-||||-|||||||||-|||||||||||||||||||||||||-|||||||||||||-||||||||||||||||||||||||||||||||||||--
   correct: CCCCTTAAACCTTATGGGGAAGAGATGCCAACGACGTGATCGGGGTCCTAGAGTCATTTCCACGCACTTGCCGCTAACGCAGTTGGACATTCACGGG
The accuracy for this sequence is: 92.78350515463917%


Basecalling signal 146 out of 200
The alignment for signal 146 is:
basecalled: TGACAATCTACGACTGTACATTCGTACGCAGACCGGAA-GGAACAGACTGTGGTGG-CATTACCTTTT-AGTTAACTGATACATTAGGTTCCGAGCTCC-GG
            ||||||||||||||||||||||||||||||||||||||-|||||||||||||||||-|||||||||||-||||||||||||||||||||||||||||||-||
   correct: TGACAATCTACGACTGTACATTCGTACGCAGACCGGAAAGGAACAGACTGTGGTGGGCATTACCTTTTTAGTTAACTGATACATTAGGTTCCGAGCTCCCGG
The accuracy for this sequence is: 96.07843137254902%


Basecalling signal 147 out of 200
The alignment for signal 147 is:
basecalled: GATGGGTAGGACGGGAA-TGCCGGTCATTGATGCTGTATTATAA-GTCTAGATGATTCCCGGGTTT-AGGCTAAGGCC-TTAGTGCTGGG-CAGCT--
            |||||||||||||||||-||||||||||||||||||||||||||-|||||||||||||||||||||-|||||||||||-|||||||||||-|||||--
   correct: GATGGGTAGGACGGGAAATGCCGGTCATTGATGCTGTATTATAAAGTCTAGATGATTCCCGGGTTTTAGGCTAAGGCCCTTAGTGCTGGGGCAGCTTT
The accuracy for this sequence is: 92.85714285714286%


Basecalling signal 148 out of 200
The alignment for signal 148 is:
basecalled: TATTGAA-TATCTTCACTGTAATACCTGGCCGAGAAGAATATGCTGTAAGTAAGCTTCGGCGATTCTCAGGACAA-TATAGTAACCGGCACACTTGGCGACA
            |||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||
   correct: TATTGAAATATCTTCACTGTAATACCTGGCCGAGAAGAATATGCTGTAAGTAAGCTTCGGCGATTCTCAGGACAAATATAGTAACCGGCACACTTGGCGACA
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 149 out of 200
The alignment for signal 149 is:
basecalled: ATAAGACTCATCTGTCAT-TTGGCTGGTTACGCCAAGGTCCAACTAAGTCGTATGGCTAGATACGGGCCATAAAGAAATCTGTAGATACGTCATT-ATGG-
            ||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||-
   correct: ATAAGACTCATCTGTCATATTGGCTGGTTACGCCAAGGTCCAACTAAGTCGTATGGCTAGATACGGGCCATAAAGAAATCTGTAGATACGTCATTTATGGG
The accuracy for this sequence is: 97.02970297029702%


Basecalling signal 150 out of 200
The alignment for signal 150 is:
basecalled: GGGAGCTATCACCAGGTCGTCTGTCGCTAGCCAATATTAATTTGTAGTAATATTGGATGGCGTATCGACAA-CTGATTTGTCGCTACGCTGGCAAGATCT
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||
   correct: GGGAGCTATCACCAGGTCGTCTGTCGCTAGCCAATATTAATTTGTAGTAATATTGGATGGCGTATCGACAAACTGATTTGTCGCTACGCTGGCAAGATCT
The accuracy for this sequence is: 99.0%


Basecalling signal 151 out of 200
The alignment for signal 151 is:
basecalled: GGGG--ACGATGCTTACCATAACCACTACCAGTTCATACAGTA-ATGACGAGCGACAGAGCGCAGCGCGGCGAAGCCTGTTGG-CCTCAAGCCTGTTGAGA
            ||||--|||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||-|||||||||||||||||
   correct: GGGGGGACGATGCTTACCATAACCACTACCAGTTCATACAGTAGATGACGAGCGACAGAGCGCAGCGCGGCGAAGCCTGTTGGGCCTCAAGCCTGTTGAGA
The accuracy for this sequence is: 96.03960396039604%


Basecalling signal 152 out of 200
The alignment for signal 152 is:
basecalled: ACTGAA-GTCGAACACCGCCGCAGAACAACAATCGTAGAGAGATAGCTTAGG-CCATGGACGACTTATGCCAGG-CATAACAAGTGTCGACAA-TATGA
            ||||||-|||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||-||||||||||||||||||-|||||
   correct: ACTGAAAGTCGAACACCGCCGCAGAACAACAATCGTAGAGAGATAGCTTAGGGCCATGGACGACTTATGCCAGGTCATAACAAGTGTCGACAAATATGA
The accuracy for this sequence is: 95.95959595959596%


Basecalling signal 153 out of 200
The alignment for signal 153 is:
basecalled: CTGG-TGGTATCTTAA-TGCGGAAACAACAAGATTTCGCGCCGACTTACGCTATCTAAGACCC-TACGTGGAACGCTCCTCAGAAGCTTCCGACCCGTCC-G-
            ||||-|||||||||||-||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||-|-
   correct: CTGGGTGGTATCTTAAATGCGGAAACAACAAGATTTCGCGCCGACTTACGCTATCTAAGACCCCTACGTGGAACGCTCCTCAGAAGCTTCCGACCCGTCCCGG
The accuracy for this sequence is: 95.14563106796116%


Basecalling signal 154 out of 200
The alignment for signal 154 is:
basecalled: AGGTTTCGCTCGTGACGTCGCAA-TTGATCAA-TACGG--CAGAATTGGTACGACGCTTCCGGTCAAGGCGTACCGGAACTAACTATTATTCGATATT-CGCTTT
            |||||||||||||||||||||||-||||||||-|||||--||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||
   correct: AGGTTTCGCTCGTGACGTCGCAAATTGATCAAATACGGGGCAGAATTGGTACGACGCTTCCGGTCAAGGCGTACCGGAACTAACTATTATTCGATATTTCGCTTT
The accuracy for this sequence is: 95.23809523809523%


Basecalling signal 155 out of 200
The alignment for signal 155 is:
basecalled: GTTCGCGCTGAGATGGATGCGCTAAGCTTT--AATCGCATGGACAGAA-CTGAGG-CCACTTGATGTTCTGTCTGATTT-GCGCTGG-CACTAGTCAGGA
            ||||||||||||||||||||||||||||||--||||||||||||||||-||||||-|||||||||||||||||||||||-|||||||-||||||||||||
   correct: GTTCGCGCTGAGATGGATGCGCTAAGCTTTTTAATCGCATGGACAGAAACTGAGGGCCACTTGATGTTCTGTCTGATTTTGCGCTGGGCACTAGTCAGGA
The accuracy for this sequence is: 94.0%


Basecalling signal 156 out of 200
The alignment for signal 156 is:
basecalled: ACAAGG-CCC-AACCGCTGTCAGTGACTTAGGTACATT-CAACGTCTAGCTACACCATCCCC--AGCCACGGG-TCATGATCCGTTATAACGACCGACCGCTCA-
            ||||||-|||-|||||||||||||||||||||||||||-|||||||||||||||||||||||--|||||||||-||||||||||||||||||||||||||||||-
   correct: ACAAGGGCCCCAACCGCTGTCAGTGACTTAGGTACATTTCAACGTCTAGCTACACCATCCCCCCAGCCACGGGGTCATGATCCGTTATAACGACCGACCGCTCAT
The accuracy for this sequence is: 93.33333333333333%


Basecalling signal 157 out of 200
The alignment for signal 157 is:
basecalled: TACTCTGCGAA-CTAGACCAAGCCC-AGCAAGTGGAA-GGACGCGAAGATGCTCTGACTTGTGTCGTTGCAGCAACTCTCTAAGTCGAGTATTCAGGAT-
            |||||||||||-|||||||||||||-|||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-
   correct: TACTCTGCGAAACTAGACCAAGCCCCAGCAAGTGGAAAGGACGCGAAGATGCTCTGACTTGTGTCGTTGCAGCAACTCTCTAAGTCGAGTATTCAGGATT
The accuracy for this sequence is: 96.0%


Basecalling signal 158 out of 200
The alignment for signal 158 is:
basecalled: GAACCGCCATCGTATGCAA-CC-TT-CAGTAAGATTCCTTCACGTACC-AGATGTAGCTAGAGCTCCACTTGCATCCTCAATAACGGCTAGCCACACGTCTTC
            |||||||||||||||||||-||-||-||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GAACCGCCATCGTATGCAAACCCTTTCAGTAAGATTCCTTCACGTACCCAGATGTAGCTAGAGCTCCACTTGCATCCTCAATAACGGCTAGCCACACGTCTTC
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 159 out of 200
The alignment for signal 159 is:
basecalled: GCGGCGGTGCATTCAGAATGGAGAGCATCAGAGCCGTGCGTGAA-CTAATGGCATTAGGATCCGGTCTCAATCCC-TATTCGG-CGG-TCAGAGCGACTTAAT
            ||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||-|||||||-|||-|||||||||||||||
   correct: GCGGCGGTGCATTCAGAATGGAGAGCATCAGAGCCGTGCGTGAAACTAATGGCATTAGGATCCGGTCTCAATCCCCTATTCGGGCGGGTCAGAGCGACTTAAT
The accuracy for this sequence is: 96.11650485436894%


Basecalling signal 160 out of 200
The alignment for signal 160 is:
basecalled: TAA-TACCGCACGCAACGACGTCGGATCAATTGTT-CTCATGG-ATCAAA-GAACTTGATATTGAAGGATACCTCGCACGGATGGAATCGGACTGATT-CTGCG
            |||-|||||||||||||||||||||||||||||||-|||||||-||||||-|||||||||||||||||||||||||||||||||||||||||||||||-|||||
   correct: TAAATACCGCACGCAACGACGTCGGATCAATTGTTTCTCATGGGATCAAAAGAACTTGATATTGAAGGATACCTCGCACGGATGGAATCGGACTGATTTCTGCG
The accuracy for this sequence is: 95.1923076923077%


Basecalling signal 161 out of 200
The alignment for signal 161 is:
basecalled: GAGCTACAGCTTCGCAGATATTGTAACGAGGTTGAACTT-GCTGAATGTGAAACTTCGGCCGCTCGCTCTTT-AGTCGTCACATCGGCCAAA-GGAGA--
            |||||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||-|||||||||||||||||||-|||||--
   correct: GAGCTACAGCTTCGCAGATATTGTAACGAGGTTGAACTTTGCTGAATGTGAAACTTCGGCCGCTCGCTCTTTTAGTCGTCACATCGGCCAAAAGGAGAAA
The accuracy for this sequence is: 95.0%


Basecalling signal 162 out of 200
The alignment for signal 162 is:
basecalled: AGG-CTAGCGCCTGTCAAGCTGATGATTGATGGTGACCACAGGAGCGTGGCTGTCTCAGGTTATCGGG-ATCGCGGATAGAATGTCCAGTTGAGACGAGACT
            |||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||
   correct: AGGGCTAGCGCCTGTCAAGCTGATGATTGATGGTGACCACAGGAGCGTGGCTGTCTCAGGTTATCGGGGATCGCGGATAGAATGTCCAGTTGAGACGAGACT
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 163 out of 200
The alignment for signal 163 is:
basecalled: TAGACTGATCCC--GCCGGATCTTCCACATCAC-TGGCGCTACTCGATAGGTTTCGG-TCCGTGATGTAA-CGCCGCACTGTATGAGCC-ATGAGGCGAGTAT
            ||||||||||||--|||||||||||||||||||-|||||||||||||||||||||||-||||||||||||-||||||||||||||||||-|||||||||||||
   correct: TAGACTGATCCCCCGCCGGATCTTCCACATCACGTGGCGCTACTCGATAGGTTTCGGGTCCGTGATGTAAACGCCGCACTGTATGAGCCCATGAGGCGAGTAT
The accuracy for this sequence is: 94.1747572815534%


Basecalling signal 164 out of 200
The alignment for signal 164 is:
basecalled: CACC-T-AG-CAGTTGGG--TCC-GAGTTAGTGGGTTGTAACCTAATCCTAGATATATTTACACGTAGAAA-GCTCTACCATCTCGTGTCC-AGG-AGTGGGA
            ||||-|-||-||||||||--|||-|||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||-|||-|||||||
   correct: CACCCTTAGGCAGTTGGGGGTCCCGAGTTAGTGGGTTGTAACCTAATCCTAGATATATTTACACGTAGAAAAGCTCTACCATCTCGTGTCCCAGGGAGTGGGA
The accuracy for this sequence is: 91.2621359223301%


Basecalling signal 165 out of 200
The alignment for signal 165 is:
basecalled: TCCAAA---CC-TGG-TTATCCAGGG-AATGAAACGCCGAACAGGTCGACATATTCAGATCTTCGGCGTGTCAGCGCGACACCATGACTGACTACAGGTCT
            ||||||---||-|||-||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TCCAAAAAACCCTGGGTTATCCAGGGGAATGAAACGCCGAACAGGTCGACATATTCAGATCTTCGGCGTGTCAGCGCGACACCATGACTGACTACAGGTCT
The accuracy for this sequence is: 94.05940594059405%


Basecalling signal 166 out of 200
The alignment for signal 166 is:
basecalled: GGG-TATGGCAG-TTAA-TAGAGAACGCCAACTTCATTCGCGTT-GGCCATATTATCTATGACATT-CCTAGACACGGTTCGTCGG-TAAGG-TCTTAGGC
            |||-||||||||-||||-||||||||||||||||||||||||||-|||||||||||||||||||||-|||||||||||||||||||-|||||-||||||||
   correct: GGGGTATGGCAGATTAAATAGAGAACGCCAACTTCATTCGCGTTTGGCCATATTATCTATGACATTTCCTAGACACGGTTCGTCGGGTAAGGGTCTTAGGC
The accuracy for this sequence is: 93.06930693069306%


Basecalling signal 167 out of 200
The alignment for signal 167 is:
basecalled: ATGCTGCCGGTTGAGTGGCGGACATTAACAGCCTACCCCAGGTTAA-GACAGAGTGTT-GATCTCGTGTGGTAACCTCGTTGCCAGGTTCGGATTCGG-C
            ||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||-|||||||||||||||||||||||||||||||||||||||-|
   correct: ATGCTGCCGGTTGAGTGGCGGACATTAACAGCCTACCCCAGGTTAAAGACAGAGTGTTTGATCTCGTGTGGTAACCTCGTTGCCAGGTTCGGATTCGGAC
The accuracy for this sequence is: 97.0%


Basecalling signal 168 out of 200
The alignment for signal 168 is:
basecalled: TGTACGCATTTAAACCGTATGGTCATGCTGTAGGCTCTCACAGTCACGACTAAACCTGCGGGATCACGCCATTATTAGCGTTATT-ACTCGAGGTCCTTA
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||
   correct: TGTACGCATTTAAACCGTATGGTCATGCTGTAGGCTCTCACAGTCACGACTAAACCTGCGGGATCACGCCATTATTAGCGTTATTTACTCGAGGTCCTTA
The accuracy for this sequence is: 99.0%


Basecalling signal 169 out of 200
The alignment for signal 169 is:
basecalled: AGTCGCGG-TCCACACTT-GACACTTT-CGAAGGTATTGCACCTACCGGATACACAGCATCGTGGTTGCCGAGCTCTCCACTGGACATGGCCATTGTCCTG
            ||||||||-|||||||||-||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: AGTCGCGGGTCCACACTTTGACACTTTTCGAAGGTATTGCACCTACCGGATACACAGCATCGTGGTTGCCGAGCTCTCCACTGGACATGGCCATTGTCCTG
The accuracy for this sequence is: 97.02970297029702%


Basecalling signal 170 out of 200
The alignment for signal 170 is:
basecalled: GTGTTCCGG-ATGTGAACGGCGGCGTCTCCTCTCCAGGCTAA-GACATGCCTTTCTAGTTACCTGAGACAATCTTATCAGAACACTCCGTGGCTCTCCACGA
            |||||||||-||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GTGTTCCGGGATGTGAACGGCGGCGTCTCCTCTCCAGGCTAAAGACATGCCTTTCTAGTTACCTGAGACAATCTTATCAGAACACTCCGTGGCTCTCCACGA
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 171 out of 200
The alignment for signal 171 is:
basecalled: GGCGGTTGTTGCATCGCGCC-GATAAGTCTTAGTCCTCGTGTCAGTGCTTACGTGGACTCATCGATCGTTACACGAAGCAGCTCAACAGCGTAACTGCTT
            ||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GGCGGTTGTTGCATCGCGCCCGATAAGTCTTAGTCCTCGTGTCAGTGCTTACGTGGACTCATCGATCGTTACACGAAGCAGCTCAACAGCGTAACTGCTT
The accuracy for this sequence is: 99.0%


Basecalling signal 172 out of 200
The alignment for signal 172 is:
basecalled: CTGTGTAAGGTTCAAA-GA-GTCACTCATGTTTT-GTCCGGCGGG--TTAAACTACCTCAAACCGACAACAGGG-AAAGTGG-CATTCGTTT-CTGG-CCGGTTAC
            ||||||||||||||||-||-||||||||||||||-||||||||||--|||||||||||||||||||||||||||-|||||||-|||||||||-||||-||||||||
   correct: CTGTGTAAGGTTCAAAAGAAGTCACTCATGTTTTTGTCCGGCGGGGGTTAAACTACCTCAAACCGACAACAGGGGAAAGTGGGCATTCGTTTTCTGGGCCGGTTAC
The accuracy for this sequence is: 91.50943396226415%


Basecalling signal 173 out of 200
The alignment for signal 173 is:
basecalled: GCATGCCATGTAGCGCCGCCGG-CCC-GACATTAGATGCGACGACAACTGCCTTGAACATTCGTATGTAGCCTTCTCAGTTGTTTGATATTGTAGCTTGC
            ||||||||||||||||||||||-|||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GCATGCCATGTAGCGCCGCCGGGCCCCGACATTAGATGCGACGACAACTGCCTTGAACATTCGTATGTAGCCTTCTCAGTTGTTTGATATTGTAGCTTGC
The accuracy for this sequence is: 98.0%


Basecalling signal 174 out of 200
The alignment for signal 174 is:
basecalled: TACATTT-CGGCTCTATAATTAGAAAA--GGATGGCAATT-AAA-CAGCACAGGCAATCATGCTGTCTTATACTATTAA-CATGGTCAT-AGAACAGTCGCTG-
            |||||||-|||||||||||||||||||--|||||||||||-|||-||||||||||||||||||||||||||||||||||-|||||||||-|||||||||||||-
   correct: TACATTTTCGGCTCTATAATTAGAAAAAAGGATGGCAATTTAAAACAGCACAGGCAATCATGCTGTCTTATACTATTAAACATGGTCATCAGAACAGTCGCTGG
The accuracy for this sequence is: 92.3076923076923%


Basecalling signal 175 out of 200
The alignment for signal 175 is:
basecalled: ATTCGTAGCTCTGGACCG-CTTATGCTACGTACTATGGCC-TACGCACAGCCGGTCCGCGAGTAACTCGTCGGAGCGTAAGTTCAGCGACCGGCCGGCGCCA
            ||||||||||||||||||-|||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ATTCGTAGCTCTGGACCGGCTTATGCTACGTACTATGGCCCTACGCACAGCCGGTCCGCGAGTAACTCGTCGGAGCGTAAGTTCAGCGACCGGCCGGCGCCA
The accuracy for this sequence is: 98.0392156862745%


Basecalling signal 176 out of 200
The alignment for signal 176 is:
basecalled: ACTCCACGTCACTGTAAGCTCCTCATTTT-AGGG-CCGTGACTACTTATCTCGAACAAGTGACTCATATTCAACTCCTGCGTTATTTGCCACCGTCT
            |||||||||||||||||||||||||||||-||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ACTCCACGTCACTGTAAGCTCCTCATTTTTAGGGGCCGTGACTACTTATCTCGAACAAGTGACTCATATTCAACTCCTGCGTTATTTGCCACCGTCT
The accuracy for this sequence is: 97.9381443298969%


Basecalling signal 177 out of 200
The alignment for signal 177 is:
basecalled: GGTTGTTACATATCCATGCTCAATGTGCGAAATCCG-TGATCGCCCTTCGCAGATGTATGCGAAGCATGCGGG-AACCTGGCGATGTT-CAACCT
            ||||||||||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||-||||||||||||||-||||||
   correct: GGTTGTTACATATCCATGCTCAATGTGCGAAATCCGGTGATCGCCCTTCGCAGATGTATGCGAAGCATGCGGGGAACCTGGCGATGTTTCAACCT
The accuracy for this sequence is: 96.84210526315789%


Basecalling signal 178 out of 200
The alignment for signal 178 is:
basecalled: GCC-TCATTCCTGCGAATTCACTACCGTACTGGCGGTCGAAGCC-GTGAAGACCATATATTGGCCCAA-TAGAAA-TAATCCCCATCGAGGG-TAGG-AGACCA
            |||-||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||-||||||-||||||||||||||||-||||-||||||
   correct: GCCCTCATTCCTGCGAATTCACTACCGTACTGGCGGTCGAAGCCCGTGAAGACCATATATTGGCCCAAATAGAAAATAATCCCCATCGAGGGGTAGGGAGACCA
The accuracy for this sequence is: 94.23076923076923%


Basecalling signal 179 out of 200
The alignment for signal 179 is:
basecalled: GACAAGCCATGTCGCTCTGTCCTCGACGTTAAA-CATCGCGTACTGGGTACTGATGTTT-ACCGACACTCAAGG-CTTCC-GGCTAATATTT-AGCGACTCTAAGGGC
            |||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||-||||||||||||||-|||||-|||||||||||-|||||||||||||||
   correct: GACAAGCCATGTCGCTCTGTCCTCGACGTTAAAACATCGCGTACTGGGTACTGATGTTTTACCGACACTCAAGGGCTTCCCGGCTAATATTTTAGCGACTCTAAGGGC
The accuracy for this sequence is: 95.37037037037037%


Basecalling signal 180 out of 200
The alignment for signal 180 is:
basecalled: TAGTAAGAAGCAGACGGCCTAGGATTCTATCGTGG-ACGTGTTGACTGAAGTGCATCCTGTATTT-GCATGCGCGATAATTCCCGTGAGCGCATATCC
            |||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||-||||||||||||||||||||||||||||||||
   correct: TAGTAAGAAGCAGACGGCCTAGGATTCTATCGTGGGACGTGTTGACTGAAGTGCATCCTGTATTTTGCATGCGCGATAATTCCCGTGAGCGCATATCC
The accuracy for this sequence is: 97.95918367346938%


Basecalling signal 181 out of 200
The alignment for signal 181 is:
basecalled: CTCACGG-TCGCGTGCCGCCTGGAGACGAAA-CCAACC-TTAACACAGTGAACGGAGAGTACTGAAGCATACGAAA-CTGGTACTAAGAA-CACCAGCG-
            |||||||-|||||||||||||||||||||||-||||||-|||||||||||||||||||||||||||||||||||||-|||||||||||||-||||||||-
   correct: CTCACGGGTCGCGTGCCGCCTGGAGACGAAAACCAACCCTTAACACAGTGAACGGAGAGTACTGAAGCATACGAAAACTGGTACTAAGAAACACCAGCGG
The accuracy for this sequence is: 94.0%


Basecalling signal 182 out of 200
The alignment for signal 182 is:
basecalled: GGAATTATTGAACGG-GCCGGTTCCACGAACC-TGTGTACC-GGTTTGTTAGGCC-ATCTGACACACGCTCAGAGTTCCGGAGGTGTGGAGTGCTGAACCGGGATAC
            |||||||||||||||-||||||||||||||||-||||||||-|||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GGAATTATTGAACGGTGCCGGTTCCACGAACCCTGTGTACCCGGTTTGTTAGGCCCATCTGACACACGCTCAGAGTTCCGGAGGTGTGGAGTGCTGAACCGGGATAC
The accuracy for this sequence is: 96.26168224299066%


Basecalling signal 183 out of 200
The alignment for signal 183 is:
basecalled: TTCTCAGGCTACATTGCGAGCTATACCGAAGCGCCC-GTATT-ATCAAGTATCGCTAGCACACAA-GTCGCAGCCGCC-GCATGGTTAAGGCAATT-AA
            ||||||||||||||||||||||||||||||||||||-|||||-||||||||||||||||||||||-||||||||||||-|||||||||||||||||-||
   correct: TTCTCAGGCTACATTGCGAGCTATACCGAAGCGCCCCGTATTTATCAAGTATCGCTAGCACACAAAGTCGCAGCCGCCCGCATGGTTAAGGCAATTTAA
The accuracy for this sequence is: 94.94949494949495%


Basecalling signal 184 out of 200
The alignment for signal 184 is:
basecalled: CCTATATGTCCATGTAGCATCGAAA-CGCGTGTGCGCAAGCTACTGTCTGCTCAACCTATCAAGGATGCACCAGATTCGCGGAAGGCCCC-AGACGACC-GC
            |||||||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||-||
   correct: CCTATATGTCCATGTAGCATCGAAAACGCGTGTGCGCAAGCTACTGTCTGCTCAACCTATCAAGGATGCACCAGATTCGCGGAAGGCCCCCAGACGACCCGC
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 185 out of 200
The alignment for signal 185 is:
basecalled: ATGGG-CCAGTTGAGTAATCTACGAGCACAGTAGTAGGAAGTCGATGAACGGCGTTGTCTGACGAGAGAATGGAGCTATTCTTAGATTAGCTCCAT
            |||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: ATGGGGCCAGTTGAGTAATCTACGAGCACAGTAGTAGGAAGTCGATGAACGGCGTTGTCTGACGAGAGAATGGAGCTATTCTTAGATTAGCTCCAT
The accuracy for this sequence is: 98.95833333333334%


Basecalling signal 186 out of 200
The alignment for signal 186 is:
basecalled: TT-CAATTGGAGTGAAACATTAGATTCGCCTATCTCGCAGCAAGCGACGCGCATTGTAACGCTGGCCGCC-GAGATGGATAAGTT-GGACCGTGCGTT-AAGAGAC
            ||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||-||||||||||||-|||||||
   correct: TTTCAATTGGAGTGAAACATTAGATTCGCCTATCTCGCAGCAAGCGACGCGCATTGTAACGCTGGCCGCCCGAGATGGATAAGTTTGGACCGTGCGTTTAAGAGAC
The accuracy for this sequence is: 96.22641509433963%


Basecalling signal 187 out of 200
The alignment for signal 187 is:
basecalled: AGCAA-TCGACACGTCACTTGCGCCGAAGTTAGCCTGG-TCGGTTGAGAACCAGATCAATACC-TTT-GGGACATCACATAA-GAAGACATCC--T
            |||||-||||||||||||||||||||||||||||||||-||||||||||||||||||||||||-|||-||||||||||||||-||||||||||--|
   correct: AGCAAATCGACACGTCACTTGCGCCGAAGTTAGCCTGGGTCGGTTGAGAACCAGATCAATACCCTTTTGGGACATCACATAAAGAAGACATCCCCT
The accuracy for this sequence is: 92.70833333333334%


Basecalling signal 188 out of 200
The alignment for signal 188 is:
basecalled: GAGCTGATTGCTCACACGGGATGACGCGCTGGACGCCGGCGATTCCCGTATCCGATATCGTCAGTTT-AGG-AGTGGAGTTCGCTGTCGAATCGCGATC
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||-|||||||||||||||||||||||||||
   correct: GAGCTGATTGCTCACACGGGATGACGCGCTGGACGCCGGCGATTCCCGTATCCGATATCGTCAGTTTTAGGGAGTGGAGTTCGCTGTCGAATCGCGATC
The accuracy for this sequence is: 97.97979797979798%


Basecalling signal 189 out of 200
The alignment for signal 189 is:
basecalled: GAGTGTGTTT----GGACGGTCCGCTCTTCGCCC-TCGGATACGTGTGTGTATTCCTCACGTGTCCTAGGTTGTCCTAGTATACCGGACATACGAAGGACAAG
            ||||||||||----||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GAGTGTGTTTTTTTGGACGGTCCGCTCTTCGCCCCTCGGATACGTGTGTGTATTCCTCACGTGTCCTAGGTTGTCCTAGTATACCGGACATACGAAGGACAAG
The accuracy for this sequence is: 95.14563106796116%


Basecalling signal 190 out of 200
The alignment for signal 190 is:
basecalled: GATCATCATGCAAGCATGACTTGATATAATTCTGGACTTATGTGCTGACACGCGTACCTCAGAATGATACATGGCC-GACTCAACACGTT-GCC-GCTGAG
            |||||||||||||||||||||-||||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||-|||-||||||
   correct: GATCATCATGCAAGCATGACT-GATATAATTCTGGACTTATGTGCTGACACGCGTACCTCAGAATGATACATGGCCCGACTCAACACGTTTGCCCGCTGAG
The accuracy for this sequence is: 96.0%


Basecalling signal 191 out of 200
The alignment for signal 191 is:
basecalled: AACTTCAGGATAGGAGCCAAACACACCCGAACATCCTTCTCAGCATGGCCTT-CTCTCAATCAATTGGTTGAATCTACACTAAGTTCCAGGTTATG
            ||||||||||||||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||||||||||||||||||||||
   correct: AACTTCAGGATAGGAGCCAAACACACCCGAACATCCTTCTCAGCATGGCCTTTCTCTCAATCAATTGGTTGAATCTACACTAAGTTCCAGGTTATG
The accuracy for this sequence is: 98.95833333333334%


Basecalling signal 192 out of 200
The alignment for signal 192 is:
basecalled: TCTGAGTGAACGTCTCAATAGTTCGACTCAAGGGAAATTT-GCCACTGTATAATAATGTACC-GCCAGATGGCTGCGGACCGCGCACC-ATCGAGACCAT
            ||||||||||||||||||||||||||||||||||||||||-|||||||||||||||||||||-|||||||||||||||||||||||||-|||||||||||
   correct: TCTGAGTGAACGTCTCAATAGTTCGACTCAAGGGAAATTTTGCCACTGTATAATAATGTACCCGCCAGATGGCTGCGGACCGCGCACCCATCGAGACCAT
The accuracy for this sequence is: 97.0%


Basecalling signal 193 out of 200
The alignment for signal 193 is:
basecalled: GTGTCGCTTATGACGGTGCCTTCCAGGAGATGTGCTCAACTATTGTAACGGTTCTCTGTAA-TGCGATCTAGAGGTAA-CAGCCATGG-CGTTACCTAACGT
            |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||-|||||||||-|||||||||||||
   correct: GTGTCGCTTATGACGGTGCCTTCCAGGAGATGTGCTCAACTATTGTAACGGTTCTCTGTAAATGCGATCTAGAGGTAAACAGCCATGGGCGTTACCTAACGT
The accuracy for this sequence is: 97.05882352941177%


Basecalling signal 194 out of 200
The alignment for signal 194 is:
basecalled: CC-GTAAATCACCATCTGCAGAATCTCGAGGCAGAAGGCATTCTGGGTGCCCGAAGCTAAGGTTATTCATAGCTATTGATACCGACGACACTGCGGCC-G
            ||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||-|
   correct: CCCGTAAATCACCATCTGCAGAATCTCGAGGCAGAAGGCATTCTGGGTGCCCGAAGCTAAGGTTATTCATAGCTATTGATACCGACGACACTGCGGCCCG
The accuracy for this sequence is: 98.0%


Basecalling signal 195 out of 200
The alignment for signal 195 is:
basecalled: AGCGATCGTTTGCTTCC-GAGCC-TAATAAGCGATTATGTCTCTCAGACGCATACCGCGTGCCTTT-GCATGATAGTGTACAACC-GGCAGGTCAAT
            |||||||||||||||||-|||||-||||||||||||||||||||||||||||||||||||||||||-||||||||||||||||||-|||||||||||
   correct: AGCGATCGTTTGCTTCCCGAGCCCTAATAAGCGATTATGTCTCTCAGACGCATACCGCGTGCCTTTTGCATGATAGTGTACAACCCGGCAGGTCAAT
The accuracy for this sequence is: 95.87628865979381%


Basecalling signal 196 out of 200
The alignment for signal 196 is:
basecalled: AGGGTGGAGAATAACGAATGAA-GCCGTCGATTGTT-CCTGAGGTTTAGCCAGCATATACGCTATGTACATT-CCATGGAAGATTGGTATTTCTTT--C
            ||||||||||||||||||||||-|||||||||||||-|||||||||||||||||||||||||||||||||||-|||||||||||||||||||||||--|
   correct: AGGGTGGAGAATAACGAATGAAAGCCGTCGATTGTTTCCTGAGGTTTAGCCAGCATATACGCTATGTACATTTCCATGGAAGATTGGTATTTCTTTTTC
The accuracy for this sequence is: 94.94949494949495%


Basecalling signal 197 out of 200
The alignment for signal 197 is:
basecalled: TCCAGACACGAACTGAAGGACTGGAGCTTCGAACTTGCGTCGGTCGAGATGTATGCTATTTAATCAGGCACACGACTATTAGATGAGTCCGCAATCGT
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: TCCAGACACGAACTGAAGGACTGGAGCTTCGAACTTGCGTCGGTCGAGATGTATGCTATTTAATCAGGCACACGACTATTAGATGAGTCCGCAATCGT
The accuracy for this sequence is: 100.0%


Basecalling signal 198 out of 200
The alignment for signal 198 is:
basecalled: GTTGATCTCCTATACGTGG-CTCGG-ATTGAACTAGACTACCTCGGTTACCC-TCCCAATAATGTATACAATAACC-TCTCGTCTTCACGACCAAGAGCT
            |||||||||||||||||||-|||||-||||||||||||||||||||||||||-|||||||||||||||||||||||-|||||||||||||||||||||||
   correct: GTTGATCTCCTATACGTGGGCTCGGGATTGAACTAGACTACCTCGGTTACCCCTCCCAATAATGTATACAATAACCCTCTCGTCTTCACGACCAAGAGCT
The accuracy for this sequence is: 96.0%


Basecalling signal 199 out of 200
The alignment for signal 199 is:
basecalled: CGTCGTGGCAGAAACTTTCTT-CGCTGAGATACACTACGGTAACTTCTCTAA-TCCATCTAA-G-CCTATATTCACCTAAGGTCCACATGCTAAATGGG-
            |||||||||||||||||||||-||||||||||||||||||||||||||||||-|||||||||-|-||||||||||||||||||||||||||||||||||-
   correct: CGTCGTGGCAGAAACTTTCTTTCGCTGAGATACACTACGGTAACTTCTCTAAATCCATCTAAAGGCCTATATTCACCTAAGGTCCACATGCTAAATGGGG
The accuracy for this sequence is: 95.0%


Basecalling signal 200 out of 200
The alignment for signal 200 is:
basecalled: GATT-GTTT-AACTTGCACACC-GGTGGTCCTAACTTATCACGGTTCCTGTCACGGTAAAGCTTCTACTGCGACCAGCTCGATATCCGCGACTTAGCTGT
            ||||-||||-||||||||||||-|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
   correct: GATTTGTTTTAACTTGCACACCCGGTGGTCCTAACTTATCACGGTTCCTGTCACGGTAAAGCTTCTACTGCGACCAGCTCGATATCCGCGACTTAGCTGT
The accuracy for this sequence is: 97.0%


Finished basecalling 200 signals
The average accuracy of the basecaller is: 95.98872015787889%</code></pre>
  <p class="blog-post_paragraph">
    So based on the 200 raw signals, the accuacy of this basecaller is 95.9%.
    And now that I have an easy method to basecall and measure the accuracy at
    all signals, I can quickly tweak some of the parameters to see if I can
    improve it any further.
  </p>
  <p class="blog-post_paragraph">For example:</p>
  <pre
    class="line-numbers"
  > <code class="language-python">basecall_all(signals, kmer_table, sequencing, movement_cutoff = 0.03, homopolymer_cutoff = 17, detailed=False)</code></pre>
  <pre
    class="output"
  ><code class="language-output">Finished basecalling 200 signals
The average accuracy of the basecaller is: 96.63414334298976%</code></pre>
  <p class="blog-post_paragraph">
    I found that changing the homopolymer cutoff to be 17 timepoints provided
    the most result in finding homopolymers in the signal, and I was able to
    improve the basecaller accuracy to 96.6%
  </p>
  <p class="blog-post_paragraph">
    The last thing to do is just to simply format everything into a command line
    executable, and then this basecaller is complete.
  </p>
  <pre
    class="line-numbers"
  > <code class="language-python"># Import the required modules
import numpy as np
import json
import edlib
import argparse
import sys

# Take in all the arguments and set up the help text
parser = argparse.ArgumentParser(description='Nanopore signal basecaller')
parser.add_argument("-s", "--signals", help = "File with nanopore signals")
parser.add_argument("-k", "--kmers", help = "File with k-mer signatures")
parser.add_argument("-c", "--correct", help = "File with the correct sequences")
parser.add_argument("-q", "--quiet", action = "store_false", help = "Hide the alignments from the output")

args = parser.parse_args()

# Extract the command line arguments
try:
  signals_file = args.signals
  kmer_table_file = args.kmers
  sequences_file = args.correct
  detailed_bool = args.quiet
except:
  # Exit if the arguments are not provided
  print("Please provide all arguments")
  sys.exit(1)
  
# Load the data from the files
with open(signals_file) as f:
  signals = json.load(f)

with open(kmer_table_file) as f2:
  kmer_table = json.load(f2)

with open(sequences_file) as f3:
  sequences = json.load(f3)

# Define the Signal class
class Signal:

  def __init__(self, signal_index):
    self.signal_index = signal_index
    self.signal = np.row_stack(signals[signal_index]) 
    self.correct_seq = sequences[signal_index]

  def __str__(self) -> str:
    nl = "\n" 
    return f"Signals array: {nl}{self.signal} {nl} {nl}Correct sequence: {nl}{self.correct_seq}"

  def __repr__(self) -> str:
    nl = "\n"
    return f"Signal index: {nl}{self.signal_index} {nl} {nl}Signals array (numpy array): {self.signal.__repr__()} {nl} {nl}Correct sequence: {nl}{self.correct_seq} "


def smooth_data(signal):
  # Subtract the first element from the second, the second from the third, and so on. Get the absolute value of the result
  smoothed_signal = np.abs(np.array(signal.signal[0:len(signal.signal) - 1]) - np.array(signal.signal[1:len(signal.signal)]))

  # Return the smoothed signal
  return smoothed_signal


def get_movement_timepoints(signal, cutoff):
  # Checks each timepoint and assigns it boolean value if value is greater than the cutoff 
  timepoints_bool = signal.smoothed_signal > cutoff

  # Grab the indicies of the timepoints where the value was greater than the cutoff
  dna_movement_timepoints = np.array([i for i, x in enumerate(timepoints_bool) if x])

  # Insert a timepoint 0 at the beginning of the array to represent the inital DNA that was present
  dna_movement_timepoints = np.insert(dna_movement_timepoints, 0, 0)

  # Output the timepoints
  return dna_movement_timepoints


def get_capture_timepoints(signal, cutoff):
  # Reinitialize the list to store the timepoints where the DNA was captured
  dna_capture_timepoints = list()

  # Loop through each timepoint and calculate the mean of the signal values between each timepoint
  for i in range(0,len(signal.movement_timepoints)-1):
    
    """Modify the miniminum and maximum values to make the capture range a bit shorter
    Add a little less buffer to the very start to capture more of the initial capture event""" 
    minimum_buffer = 2
    maximum_buffer = 0
    minimum = signal.movement_timepoints[i] + 1 if i == 0 else signal.movement_timepoints[i] + minimum_buffer
    maximum = signal.movement_timepoints[(i+1)] - maximum_buffer
    
    # Append the range of timepoints to the list of capture timepoints
    dna_capture_timepoints.append(list(range(minimum,maximum)))
    
  # Return the list of timepoints
  return dna_capture_timepoints


def adjust_capture_timepoints(capture_timepoints, homopolymer_cutoff):
  # Make a copy of the capture timepoints to not modify the original argument
  capture_ranges = capture_timepoints.copy()

  # Loop through each capture range
  for i, capture_range in enumerate(capture_ranges):
  
    # If the capture range is long enough to be a suspected homopolymer
    if len(capture_range) >= homopolymer_cutoff:
      # Split the capture range in half
      split_point = int(len(capture_range) / 2)
      capture_ranges[i] = capture_range[:split_point]

      # Add the second half of the capture range to the next index in the list
      # If this range is also a suspected homopolymer, it will be split in the next iteration
      capture_ranges.insert(i+1, capture_range[split_point:])
    
  # Remove any empty lists that may have been created by the splitting
  capture_ranges = [x for x in capture_ranges if x != []]
  
  # Return the updated capture ranges
  return capture_ranges


def get_mean(signal, capture_ranges):    
  # Get the minimum and maximum timepoints for the capture range 
  minimum = capture_ranges[0]
  maximum = capture_ranges[-1]
  
  # If the minimum and maximum are different, take the mean of the signal between those timepoints
  if minimum != maximum:
    signal_means = np.mean(signal.signal[minimum:maximum])
    
  # If the minimum and maximum are the same, the mean is just the signal at that timepoint
  elif minimum == maximum:
    signal_means = signal.signal[minimum]
    
  # Return the mean signal value
  return signal_means


def get_capture_means (signal, capture_ranges):
  # Initialize a list to store the mean signal values
  dna_capture_means = list()

  # Loop through each capture range and calculate the mean of the signal values between each timepoint
  for i in capture_ranges:
    dna_capture_means.append(get_mean(signal, i))
  
  # Return the list of mean signal values
  return dna_capture_means


def euclidian_distance(x,y):
  dist = np.sqrt(np.square((x - y)))
  return dist


def basecall(signal, capture_ranges, kmer_table):
  # Get the mean signal value for the capture range
  capture_mean = get_mean(signal, capture_ranges)
  
  # Initialize an empty list to store the euclidian distances between the capture mean and each dinucleotide signature
  euc_distances = list()
  
  # Loop through each dinucleotide signature and calculate the euclidian distance between the capture mean and the signature
  for dinuc, signature in kmer_table.items():
    euc_distances.append((dinuc, euclidian_distance(capture_mean, signature)))
    
  # Get the minimum euclidian distance
  min_dist = np.min([x[1] for x in euc_distances])
  
  # Make a basecall
  dinuccall = [x[0] for x in euc_distances if x[1] == min_dist][0]
  
  # Return the basecall
  return dinuccall


def basecall_signal(signal, capture_ranges, kmer_table):
  # Initialize a list to store the basecalls
  basecalls = []

  # Loop through each capture range and basecall each capture range
  for cap_range in capture_ranges:
    basecalls.append(basecall(signal, cap_range, kmer_table))
  
  # Return the list of basecalls
  return basecalls


def collapse_seq(basecalls: "list") -> "str":
  """Collapse a list of basecalls into a single string"""
  
  # Initialize an empty string to store the final sequence
  final_seq = ""
  
  # Loop through each dinucleotide in the list of basecalls
  for i, v in enumerate(basecalls):

    # If it's the first basecall, add both bases to the final sequence
    if i == 0:
      final_seq += v
      continue

    # Otherwise, add the second base to the final sequence
    final_seq += v[-1]

  # Return the final sequence
  return final_seq


def align(basecalled_seq: "str", correct_seq: "str") -> "dict":
  """Align two sequences using edlib"""
  
  # Perform the alignment
  align_result = edlib.align(basecalled_seq, correct_seq, task='path')
  
  # Format the alignment result
  nice_alignment = edlib.getNiceAlignment(align_result, basecalled_seq, correct_seq)
  
  # Return the alignment result
  # print(align_result)
  return align_result, nice_alignment


def measure_accuracy (alignment: "dict", correct_seq: "str") -> "float":
  acc = (1 - alignment['editDistance'] / len(correct_seq)) * 100  
  return acc


# Putting it all together
def basecall_all(
  signals: "list", 
  kmer_signatures: "dict", 
  correct_seqs: "list", 
  movement_cutoff: "float" = 0.03, 
  homopolymer_cutoff: "int" = 20,
  detailed: "bool" = True) -> "list":
  
  # Initialize a list to store the accuracies of each signal
  accuracies = []
  
  # Loop through each signal in the list of signals
  for i, v in enumerate(signals):
    
    # Instantiate a new signal object
    current_signal = Signal(i)
    
    # Smooth the signal
    smoothed_signal = smooth_data(current_signal)

    # Set the smoothed signal as an attribute of the signal object
    setattr(current_signal, "smoothed_signal", smoothed_signal)
    
    # Get the timepoints where the DNA moves through the pore
    dna_movement_timepoints = get_movement_timepoints(current_signal, movement_cutoff)
    
    # Set the DNA capture timepoints as an attribute of the signal object
    setattr(current_signal, "movement_timepoints", dna_movement_timepoints)
    
    # Get the timepoints where the DNA is captured in the pore
    dna_capture_timepoints = get_capture_timepoints(current_signal, movement_cutoff)

    # Adjust the DNA movement timepoints to account for homopolymers
    dna_capture_timepoints = adjust_capture_timepoints(dna_capture_timepoints, homopolymer_cutoff)
    
    # Set the DNA capture timepoints as an attribute of the signal object
    setattr(current_signal, "capture_timepoints", dna_capture_timepoints)
    
    # Basecall the signal
    basecalls_list = basecall_signal(current_signal, current_signal.capture_timepoints, kmer_signatures)
    
    # Collapse the basecalls into a single sequence
    final_sequence = collapse_seq(basecalls_list)
    
    # Set the basecalled sequence as an attribute of the signal object
    setattr(current_signal, "basecalled_seq", final_sequence)
    
    # Align the basecalled sequence to the correct sequence
    alignment, nice_alignment = align(current_signal.basecalled_seq, current_signal.correct_seq)
    
    # Calculate the accuracy of the basecaller
    accuracy = measure_accuracy(alignment, current_signal.correct_seq)
    accuracies.append(accuracy)
    
    # Print out the alignment
    if detailed:
      print(f"Basecalling signal {i+1} out of {len(signals)}")
      print(f"The alignment for signal {i + 1} is:")
      print("basecalled: " + nice_alignment['query_aligned'])
      print("            " + nice_alignment['matched_aligned'])
      print("   correct: " + nice_alignment['target_aligned'])
      print(f"The accuracy for this sequence is: {accuracy}%")
      print("\n")
      
  # Determine the average accuracy of the basecaller
  basecaller_accuracy = sum(accuracies) / len(accuracies) if len(accuracies) > 0 else 0
  print(f"Finished basecalling {len(signals)} signals")
  print(f"The average accuracy of the basecaller is: {basecaller_accuracy}%")
  

# Main function entry point
if __name__ == "__main__":
  basecall_all(signals, kmer_table, sequences, movement_cutoff = 0.03, homopolymer_cutoff = 17, detailed=detailed_bool)</code></pre>
  <p class="blog-post_paragraph">Save all this into here:</p>
  <pre
    class="no-code"
  ><code class="language-treeview special-treeview">basecaller.py</code></pre>
  <p class="blog-post_paragraph">
    Now this can be run from the command line as:
  </p>
  <pre
    class="command-line"
    data-user=""
    data-host=""
    style="padding-top: 2.2em"
  > <code class="language-commandline">python3 basecaller.py -s signals.json -k kmer_table.json -c sequences.json -q</code></pre>
  <pre
    class="output"
  ><code class="language-output">Finished basecalling 200 signals
The average accuracy of the basecaller is: 96.63414334298976%</code></pre>
  <p class="blog-post_paragraph">
    So to summarize I was able to build a nanopore sequencing basecaller that
    was 96.6% accurate on average. Obviously in the real world this basecaller
    wouldn't be good enough to be used for any purpose. Most basecallers
    nowadays are built on sophisticated deep learning models. However this
    excercise does provide a good way to gain intuition into what is going on at
    the molecular level and helps you think more critically about the results of
    your sequencing methods. Doing this excerise has provided me with a
    comprehensive understanding of nanopore sequencing at the molecular level,
    and a much better understanding of what the machine learning models are
    looking at when they're trying to improve the basecalling accuracy. Nanopore
    sequencing has the potential to revolutionize various fields of science and
    have a profound impact on the future in many ways, and I hope to see the day
    where basecalling accuracy is no longer an issue.
  </p>
</div>
