<div class="blog-post">

<div class="section-title">
    <p class="blog-post-title">
      Modern Genome Assembly Techniques
    </p>
</div>

<p class="blog-post_paragraph">In November of 2024, I took part in an online workshop run by Physalia and the Wellcome Sanger Institute in order to learn the cutting edge techniques to assemble high quality Eukaryotic genomes using PacBio and Hi-C data. This was to help bolster my capabilities as a genomics scientist. I started off my research journey in the wet lab, working hands on with solutions and organisms in the lab, but I quickly realized that the amount of biological insight you can achieve at this scale of research is very limited. Current day -omics technologies allow us to gain far better resolution and give us much more of a complete picture about the biological realities of the organisms we work with. Unfortunately performing this kind of research requires a lot of computational and statistical know-how that a lot of biologists do not possess. During my masters, I challenged myself to take on a computational project in addition to my main project in the lab. I tried to remap all the existing gene models of an organism to a newer and more improved version of it's genome assembly. This project forced me to learn a lot of important foundational programming and computer science concepts which enabled me to pursue bioinformatics work, but for the duration of my bioinformatics journey I was forced to learn all the important concepts through self-study. When I started my PhD, one of the first projects I started working on was to sequence and assemble the genomes of a library of living photosynthetic microorganisms that were collected from the region I was working on. This is a humongous task as genome assembly is still not a trivial feat. Despite more and more genome assembly pipelines being published by researchers in recent years, genome assembly is still a process that requires a lot of biological knowledge, refinement, iterating, and curation by someone with a lot of genome-scale understanding of biology. Every organism will pose it's own unique challenges and quirks and so sequencing a lot of biodiverse organisms from an understudied area of the Tree of Life felt a little daunting. Luckily, my supervisor was gratious enough to enroll me in a week-long, hands-on online workshop to get me up to speed with the history and current state of genome assembly techniques. The workshop was run by Dr. Marcela Uliano-Silva from the Darwin tree of life project, a project aimed at sequencing the genomes of thousands of eukaryotic organisms from the UK region. We learned a lot of foundational and in-depth topics of the genome assembly process and were tasked with assembling a chromosome-level genome by the end of the week and presenting our thought process behind the assembly choices that we made.</p>

<p class="blog-post_paragraph">In the workshop I worked on assembling the genome of the plant species <em>Urtica urens</em>, commonly known as a small nettle. I chose this organism because it came from a plant lineage and so I figured it would have many similar biological features to the organisms that I was sequencing in my lab, such as chloroplasts with photosynthetic machinery and their own DNA, as well as cell walls that would make DNA isolation from samples difficult.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_1.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here I will outline the commands I ran from the workshops server and my local machine and the thought process I had at every step to create a chromosome level assembly of <em>Urtrica urens</em>, as well as a couple of background notes on certain topics related to the field of genomics and assembling genomes.</p>

<p class="blog-post_paragraph">
    <u><em>Background Information on Modern Genome Sequencing</em></u>
  </p>

<p class="blog-post_paragraph">A little background on genome assembly technologies before I start with the workshop portion. The first full DNA genome to be sequenced was that of bacteriophage &phi;X174 in 1977. Since then, it took researchers until 2022 to create a fully-resolved, human genome assembly. This breakthrough was powered by innovations in genome sequencing technologies that overcame decades long challenges in genomics.</p>

<p class="blog-post_paragraph">The main hurdles to creating contiguous genome assemblies have been, and to this day still is, dealing with repetative regions in the genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_2.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">These regions confuse assembly algorithms and lead to fragmented or misassembled genomes. In order to effectively resolve repetative regions in the genome, you need sequencing reads that are longer than the span of the repetative regions, and in high-enough abundance to resolve any ambiguities in the assembly algorithms.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_3.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Modern sequencing technologies, also known as long-read technologies, capture much more of the genome per-read, and have enabled us resolve a lot of tricky genomic areas, and investigate genetic phenomena across the tree of life that previously was never possible. Some interesting patterns were noticed. Firstly, the human genome contains around 26,000 genes. This number is pretty consistent across the tree of life. It's become evident that a lot of the genetic diversity that we see comes at the regulatory, or epigenomic level. Another interesting phenomenon that emerged across biodiversity genomics studies is something known as the <strong>C-value paradox</strong>. The <strong><em>C</em></strong>-value refers to the amount of DNA in the haploid genome of an organism. Very similar organisms can show a large difference in C-value (e.g. Flowering plants and amphibians).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_4.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Assemblies using long-read technologies also improve the quality control process, reduce computational intesity for assembly algorithms, and reduce the coverage-demands per region of the genomes.</p>

<p class="blog-post_paragraph">One of the current long-read sequencing providers is Pacific Biosciences. As a high-level overview, their technology works by capturing single molecules of DNA in small pores and polymerizing a complimentary strand of DNA with flourescent DNA bases.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_5.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The main problem with this technology is that the DNA polymerase can synthesize the complementary strand at a rate of 1000 base pairs per second. This rate can actually be too fast sometimes for the machine to measure the fluorescence accurately at the same rate and so it might miss some insertions or get confused. This means that <strong> the main error type for this technology are deletions and insertions</strong> (indels).</p>

<p class="blog-post_paragraph">The first generation of this technology was called PacBio continuous long reads, or PacBio CLR reads. These reads were very errorneous, and could have a per-read error rate of around 15%. This is a problem in the assembly process because most assembly algorithms require you to have good quality reads and so you had to correct your reads before assembly. This is computationally a very costly process but also causes you to lose a lot of your data afterwards. Noisy reads also pose problems for certain hard to assemble regions, especially if you don't have even sequencing coverage across the genome.</p>

<p class="blog-post_paragraph">The next generation of reads were called PacBio Circular Consensus Reads, also known as PacBio HiFi reads. These are the current gold-standard reads for performing genome assembly.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_6.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The innovation here is that the DNA sequence is circularized with special sequencing adapters and the molecule is read over-and-over again. Because the errors in each pass of the sequencing should be randomly distributed accross the reads, this strategy will give you enough sequencing depth and every base in the sequence to create an accurate consensus sequence, which in most cases is extremely close to the reality of what the original sequence was.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_7.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The other major player in the long-read sequencing space is Oxford Nanopore Technologies sequencing. These technologies work on the native molecule of DNA, sequencing it by analysing the electrical signal that is disrupted as the DNA molecular passes through a pore in a membrane.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_8.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This technology produces the longest DNA sequencing reads out of any technology on the market, with the maximum read length only restricted by the input fragment size. Typical read length is around 10,000 base pairs, but this can very significaly depending on which strategy you choose during preparation of the sequencing library.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_9.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The biggest downside of this technology is the poor quality per-read. For genome assembly, you want the highest-accuracy reads you can find. The trade-off for higher read lengths in Nanopore sequencing is usually not worth it. These reads are instead useful for resolving long repetative regions that can't be resolved by PacBio HiFi reads.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_10.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Recently, there have been two technolgies that have been released to correct Nanopore reads after they've been basecalled, called <a href="https://www.biorxiv.org/content/10.1101/2024.05.18.594796v1" target="_blank">HERRO</a> and <a href="https://github.com/LuoGroup2023/DeChat" target="_blank">DeChat</a>. These technologies are promising, but each still have enough problems (overcorrection, loss of coverage, and potentially-introduced AI bias) where making the decision to just use pure Nanopore sequencing for a genome assembly project is not yet recommended. PacBio HiFi reads are still more accurate and still reign king.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_11.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Now on with the workshop.</p>

<p class="blog-post_paragraph">
    <u><em>Step 1: Exploring the data and performing a k-mer analysis</em></u>
  </p>


<p class="blog-post_paragraph">Inside the workshop's server, there are files.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">pwd
scp /home/ubuntu/Share</code></pre>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">ls -1</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">database.fasta
drUrtUren1_data
idEriPert2
iHemFuc1_data
ilNotDrom1_data
ilVanAtal1_data
MitoHiFi_data
scripts
softwares</code></pre>

<p class="blog-post_paragraph"><code>drUrtUren1</code> is the Sanger Tree of Life project code for <em>Urtica urens</em>. When I change into this directory, I can see the following files.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd drUrtUren1_data/
ls -1</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1.600.fasta
drUrtUren1.ccs.total.fasta.gz
drUrtUren1.hicanu.total.contigs.fasta.gz
drUrtUren1.hifiasm.total.a_ctg.fa.gz
drUrtUren1.hifiasm.total.p_ctg.fa.gz
drUrtUren1.pacbio.k21.meryl
drUrtUren1.total.histo
HiC
merqury_hifiasm_eval
purged
results
run_drUrtUren1.hifiasm.a.busco_odb10
run_drUrtUren1.hifiasm.p.busco_odb10
run_drUrtUren1.purged.busco_odb10
run_drUrtUren1.purged_htigs.busco_odb10</code></pre>

<p class="blog-post_paragraph"><code>drUrtUren1.ccs.total.fasta.gz</code> contains all the sequencing reads collected for this organism. This reads are PacBio Circular Consensus (Hifi) Reads, processessed with with PacBio <code>css</code> program</p>

<p class="blog-post_paragraph">The first thing I wanted to do was to perform a <u><em>k</em>-mer analysis</u> on the sequencing reads for this organism.</p>

<p class="blog-post_paragraph">In biology, <em>k</em>-mers are unique subsequences of a sequence of length k.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_12.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Performing a <em>k</em>-mer analysis is a powerful tool in genomics with several key applications. It plays a fundamental role in genome assembly by enabling the construction of De Bruijn graphs, which help reconstruct sequences from short reads through overlap identification (although this assembly method is less in-favour these days due to modern sequencing technologies whose reads assemble better using the other main approach to genome assembly, the Overlap, Layout, Consensus (OLC) method).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_13.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Additionally, <em>k</em>-mer frequency analysis can detect bacterial contamination in eukaryotic genome assemblies by revealing discrepancies in CG content.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_14.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here you can see two clusters with different GC contents. If your sequencing reads are of high enough quality you can perform a contamination analysis from just the read data alone.</p>

<p class="blog-post_paragraph"><em>k</em>-mer analysis is also useful for correcting errors in next-generation sequencing (NGS) data by identifying abnormal <em>k</em>-mer distributions. Moreover, <em>k</em>-mer compositions across different genomes can indicate horizontal gene transfer events by highlighting unusual patterns. This method is also useful for aiding in locating CpG islands, which are regions rich in cytosine and guanine dinucleotides often associated with gene regulation. All these features and more make <em>k</em>-mer analysis a versatile and essential tool in genomic research. More information about <em>k</em>-mer counting can be found <a href="https://bioinfologics.github.io/post/2018/09/17/k-mer-counting-part-i-introduction/" target="_blank">in this blogpost</a> by Bernardo J. Clavijo, one of the authors of the <em>k</em>-mer counting program <a href="https://doi.org/10.1093/bioinformatics/btw663" target="_blank">KAT</a>, or in this recently submitted <a href="https://arxiv.org/abs/2404.01519" target="_blank">guide to performing <em>k</em>-mer analyses</a>.</p>

<p class="blog-post_paragraph">For our purposes here, we'll explore two other major roles of <em>k</em>-mer analysis that are vital to genome assembly — especially <em>de novo</em> genome assembly — the ability to <strong>estimate genome size and heterozygosity</strong>, which are a major factor in estimating the completeness of a genome assembly.</p>

<p class="blog-post_paragraph">To start investigating this in my data, I created a directory for my <em>Urtica urens</em> inside my home directory</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd ~
mkdir drUrtUren1
cd drUrtUren1
mkdir kmers assembly</code></pre>

<p class="blog-post_paragraph">And I activated my conda environment. This environment contains many bioinformatics programs that will be used in the assembly. The binaries and dependencies were already pulled from Anaconda repositories so this environment did not need to be created, simply activated.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">conda activate eukaryotic_genome_assembly</code></pre>

<p class="blog-post_paragraph">For this workshop, I used the program <a href="https://doi.org/10.1093/bioinformatics/btr011" target="_blank">Jellyfish</a> to count <em>k</em>-mers. There are a few other options like <a href="https://doi.org/10.1093/bioinformatics/btw663" target="_blank">KAT</a> and <a href="https://github.com/KamilSJaron/smudgeplot" target="_blank">Smudgeplot</a>, but these were not investigated.</p>

<p class="blog-post_paragraph">Counting <em>k</em>-mers with Jellyfish works in a couple of steps. First you want to count the <em>k</em>-mers with <code>jellyfish count</code>. I checked the help page to see what the important flags are to be used.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">jellyfish count --help</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Usage: jellyfish count [options] file:path+

Count k-mers in fasta or fastq files

Options (default value in (), *required):
 -m, --mer-len=uint32                    *aLength of mer
 -s, --size=uint64                       *Initial hash size
 -t, --threads=uint32                     Number of threads (1)
     --sam=PATH                           SAM/BAM/CRAM formatted input file
 -F, --Files=uint32                       Number files open simultaneously (1)
 -g, --generator=path                     File of commands generating fast[aq]
 -G, --Generators=uint32                  Number of generators run simultaneously (1)
 -S, --shell=string                       Shell used to run generator commands ($SHELL or /bin/sh)
 -o, --output=string                      Output file (mer_counts.jf)
 -c, --counter-len=Length in bits         Length bits of counting field (7)
     --out-counter-len=Length in bytes    Length in bytes of counter field in output (4)
 -C, --canonical                          Count both strand, canonical representation (false)
     --bc=peath                           Bloom counter to filter out singleton mers
     --bf-size=uint64                     Use bloom filter to count high-frequency mers
     --bf-fp=double                       False positive rate of bloom filter (0.01)
     --if=path                            Count only k-mers in this files
 -Q, --min-qual-char=string               Any base with quality below this character is changed to N
     --quality-start=int32                ASCII for quality values (64)
     --min-quality=int32                  Minimum quality. A base with lesser quality becomes an N
 -p, --reprobes=uint32                    Maximum number of reprobes (126)
     --text                               Dump in text format (false)
     --disk                               Disk operation. Do not do size doubling (false)
 -L, --lower-count=uint64                 Don't output k-mer with count < lower-count
 -U, --upper-count=uint64                 Don't output k-mer with count > upper-count
     --timing=Timing file                 Print timing information
     --usage                              Usage
 -h, --help                               This message
     --full-help                          Detailed help
 -V, --version                            Version</code></pre>

<p class="blog-post_paragraph">Now it's time to determine what size to pick for <code>k</code>. <strong>Choosing <em>k</em> is a tradeoff between specificity and sensitivity</strong>. Using a k that is too small will result in many unrelated sequences being composed of the same <em>k</em>-mers, in a textbook case of specificity loss because there being very few possible <em>k</em>-mers of that size. In the extreme of the small <em>k</em>, <em>k</em>=1 only distinguishes two canonical <em>k</em>-mers: A and C. 1-mer analysis is incredibly popular in biology, but it is best known by the name of GC content analysis. Using extremely large k values would sacrifice many of the benefits and sensitivity of <em>k</em>-mer analyses in the first place. For the <em>k</em>-mer size, I picked <code>31</code>. This is a common <em>k</em>-mer size in genomic analysis because it provides a good balance between specificity and sensitivit. It is long enough to provide specific sequences that help reduce ambiguity (like distinguishing between similar regions, and between mammalian and bacterial DNA) while still being short enough to maintain a high frequency in the data. It is also computationally efficient. While longer <em>k</em>-mers can increase the computational burden due to the larger number of possible combinations, 31 is often a manageable size that balances memory resources and processing effectiveness when analyizing large amounds of DNA bases.</p>

<p class="blog-post_paragraph">I also had to make sure that <code>--canonical</code> was chosen.</p>

<p class="blog-post_paragraph">DNA is typically double-stranded. When sequencing, we can read from either strand. However, we would like to consider every location of the genome once, no matter on which strand we happened to have landed.</p>

<p class="blog-post_paragraph">For example, if we read the sequence <code>ATCGAC</code>, it indicates that both this sequence and its reverse complement, <code>GTCGAT</code>, exist in the genome. One is observed when reading in one direction, while the other is observed in the opposite direction. To ensure completeness in our analyses and to prevent over-counting of <em>k</em>-mers across the genome, we should take both of these sequences into account as a single <em>k</em>-mer, <code>ATCGAC</code>/<code>GTCGAT</code>.</p>

<p class="blog-post_paragraph">To standardize how we refer to <em>k</em>-mers, we can use the concept of <strong>canonical <em>k</em>-mers</strong>. This means selecting the lexicographically smaller of the two reverse complementary sequences, which helps us consistently represent these sequences. Essentially, if we have both <code>ATCGAC</code> and <code>GTCGAT</code>, we only refer to <code>ATCGAC</code>, as it comes first in alphabetical order.</p>

<p class="blog-post_paragraph">Most <em>k</em>-mer counting tools either default to counting canonical <em>k</em>-mers or offer an option to switch between canonical and non-canonical counting. <u>Typically, canonical counting is preferred</u>, meaning that if <code>GTCGAT</code> appears, it will be counted as <code>ATCGAC</code>. However jellyfish by default does not use canonical <em>k</em>-mers, and so this option has to be specified.</p>

<p class="blog-post_paragraph"><code>size</code> and <code>threads</code> were also specified, but this has more to do with computational efficiency and computer memory management than any biologically relevant parameters.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">jellyfish count --mer-len=31 --size=1000 --threads=1 --canonical --output=drUrtUren1.ccs.total.jf drUrtUren1.ccs.total.fasta.gz</code></pre>

<p class="blog-post_paragraph">This generated the file <code>drUrtUren1.jf</code>. This file is what is used in the next step to generate a <em>k</em>-mer histogram using <code>jellyfish histo</code>.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">jellyfish histo drUrtUren1.ccs.total.jf > drUrtUren1.ccs.total.histo</code></pre>

<p class="blog-post_paragraph">Now that a histogram has been generated of these <em>k</em>-mer counts, the next step is to visualize it. To do this, <a href="https://www.nature.com/articles/s41467-020-14998-3" target="_blank">GenomeScope 2.0</a> was used. This program enables you to estimate the size, heterozygosity, and repeat content of the genome based on the frequency of <em>k</em>-mers found in the sequencing data. It comes with the web application accessible at <a href="http://genomescope.org/genomescope2.0/" target="_blank">http://genomescope.org/genomescope2.0/</a> and so the data was pulled off the workshop server first and then uploaded to this website.</p>

<p class="blog-post_paragraph">GenomeScope analyzes the <em>k</em>-mer profile (also called <em>k</em>-mer spectrum) of unassembled, unaligned (i.e. raw) reads to compute the genome and read characteristics. The <em>k</em>-mer profile counts how many <em>k</em>-mers, substrings of length k, occur in the sequence data. A key advantage of the <em>k</em>-mer based approach is that <strong>the <em>k</em>-mer profile can be computed without a reference genome</strong> by directly scanning the sequencing reads, so that it works well for novel genomes. The frequency of unique <em>k</em>-mers are plotted against the <em>k</em>-mer coverage, the formula of which is shown here:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_15.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">So plotted out, a homozygous genome — free of repeats and free of sequencing errors — should be distributed around this location on a <em>k</em>-mer spectrum plot:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_16.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If instead the genome is heterozygous, then the <em>k</em>-mer profile will show a characteristic two-peak profile. The two peaks will be centered at X and 2X (e.g. 50x and 100x) representing heterozygous <em>k</em>-mers that have been sequenced to half the coverage (50x) as the homozygous <em>k</em>-mers sequenced equally from both alleles (100x).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_17.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The relative heights of the peaks will be determined by the overall rate of heterozygosity: low heterozygosity causes the first peak to be relatively short and higher rates of heterozygosity causes the first peak to grow taller and taller. Since each heterozygous base can create <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mn>2</mn>
  <mo>&times;</mo>
  <mi>k</mi>
</math> heterozygous <em>k</em>-mers, the height of the heterozygous peak grows very quickly and matches the height of the homozygous peak at around only 1.2% heterozygosity for <em>k</em>=21.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_18.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Repeats have the opposite effect and cause <em>k</em>-mers to be present more than 2X times, proportional to the number of copies of that repeat in the genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_19.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Sequencing errors and read duplications distort the distribution by creating false <em>k</em>-mers that only occur a few numbers of times or artificially creating duplicated <em>k</em>-mers, which reduces the effective average coverage and increases the variance of the distributions. As with heterozygous variation, since each erroneous base can create <math xmlns="http://www.w3.org/1998/Math/MathML">
  <mn>2</mn>
  <mo>&times;</mo>
  <mi>k</mi>
</math> erroneous <em>k</em>-mers, the height of the erroneousus peak grows very quickly. However because these errors will be scattered randomly across the reads unlike heterozygous variation, these <em>k</em>-mers' peak will be differentiable from the heterozygous peaks because off the much smaller amount of genome coverage.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_20.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">All of these points can be summarized in this figure from <a href="https://doi.org/10.48550/arXiv.2404.01519" target="_blank">this review</a>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_21.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">So for example, this is what a typical <em>k</em>-mer plot would look like for sequencing reads from a diploid organism:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_22.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If the diploid organism had a <strong>high</strong> amount of heterozygosity (greater than 1%), then the plot might look something like this:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_23.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">And if sequencing reads were generated from a diploid sample with <strong>low</strong> heterozygosity, the <em>k</em>-mer spectra might look something like this:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_24.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Notice the little bump around 25-30X coverage. That is the heterozygous peak in this <em>k</em>-mer spectrum.</p>

<p class="blog-post_paragraph">So back to our example, taking the <code>drUrtUren1.ccs.total.histo</code>, I put it into the <a href="http://genomescope.org/genomescope2.0/" target="_blank">GenomeScope 2.0 Web App</a> and I generated the model with <code>k=31</code> and with the rest of the parameters left as default. The resulting plot looked like this:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_25.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">On a logarithmic scale the plot looked like this. You can see the red dotted line (<code>cov-threshold</code>) at the end of the graph. This represents the maximum <em>k</em>-mer coverage used in generating the model. Because this paramter was set to <code>-1</code> in this plot, this means that all the reads are being taken into account to generate the GenomeScope model of the genome. Note: High copy-number DNA such as mitochondria and chloroplasts can confuse the model. Setting a maximum <em>k</em>-mer coverage can avoid this. In GenomScope 1.0, the default <code>Max kmer coverage</code> parameter is set to <code>1000</code>, and so this may be a sensible option in some cases. In fact in this plot, you can see a spike in the sequencing coverage at over 1000X which might correlate to DNA from these organelles:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_26.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The results are summarized here.</p>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">GenomeScope version 2.0
input file = user_uploads/tZG7dlOvvS18tXTlrOsx
output directory = user_data/tZG7dlOvvS18tXTlrOsx
p = 2
k = 31

property                      min               max
Homozygous (aa)               99.8015%          99.8299%
Heterozygous (ab)             0.170073%         0.198476%
Genome Haploid Length         330,562,430 bp    331,325,497 bp
Genome Repeat Length          103,635,705 bp    103,874,937 bp
Genome Unique Length          226,926,725 bp    227,450,560 bp
Model Fit                     71.5733%          97.7745%
Read Error Rate               0.177266%         0.177266%</code></pre>

<p class="blog-post_paragraph">The model parameters are also provided (this isn't that important for my purposes these are just extra statistics provided by the software).</p>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Formula: y_transform ~ x^transform_exp * length * predict2_0(r1, k, d,
    kmercov, bias, x)

Parameters:
         Estimate Std. Error t value Pr(>|t|)
d       3.562e-02  7.920e-04   44.97   <2e-16 ***
r1      1.843e-03  7.101e-05   25.95   <2e-16 ***
kmercov 2.050e+01  1.182e-02 1734.81   <2e-16 ***
bias    1.194e+00  1.089e-02  109.69   <2e-16 ***
length  2.356e+08  4.198e+05  561.20   <2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2022000 on 1995 degrees of freedom

Number of iterations to convergence: 6
Achieved convergence tolerance: 1.49e-08</code></pre>

<p class="blog-post_paragraph">So since the <em>k</em>-mer distribution follows the model lines very well, the data fits the model very well. The estimated model fit is between 71.6% and 97.8%. This means that you can trust the other model parameters very well. From the <em>k</em>-mers found in the reads, the model estimated that the genome size is around 331Mb (found from the haploid genome size), and has a repeat composition of 31.1% (found from subtracting the unique percentage found in the plot from 100).</p>

<p class="blog-post_paragraph">Here in this figure (sections C and D) you can see a few examples where the sequencing data does not fit the model well (aka poor model convergence). <strong>In these cases the numbers predicted by GenomeScope should not be used to estimate your genome characteristics</strong> because there is too much noise in your data.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_27.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Now that I had my <em>k</em>-mer spectra plotted, I went ahead and investigated my raw sequencing data some more</p>

<p class="blog-post_paragraph">I linked all the reads found <code>drUrtUren1.ccs.total.fasta.gz</code> to my personal directory and checked if this link worked by seeing if I could peek into this file.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd ~/urtica-urens/kmers
ln -s /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.ccs.total.fasta.gz
zcat drUrtUren1.ccs.total.fasta.gz | head -2</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">m64089_210117_172547/0/ccs
TAATGAATTTTTGAGCCTTACTTTGGTAGGATTGTTCCATTCCATCAAATAACAATTAGCGTTGGCACAGACTAATGTCTCTATTCTTGATTACATAAACTACTCATTCAAAGTTAATACTAGAATTGTTGATAATAATTGCAGCGATCAAGCAAAAAGTCCTTACTTTGCAATCTTCTGGTTTTCTTTCCCAAACTTCTCTGCGGCAATGGCTATAGATCTCCGGCCAAAGTCATCGTCCAACAAACTGAAATCCAAGCCATTCTCAGCATCATCATTATTTGTGCTGGTCAAGAGACCTGCTTCGTCAAAATAGCCTCTGCCGTACAAATGCCTCGCCAGCAACTTCGCATCATTCGGAAGCTCCGCCACTCTCATCTCCCCAGACTCGATTCCATTCATGGAATCACTTCGGAAGAGCTTCACTGTTTTAAGATTCCTCAAAACACTCGCCAAATTCCTCGAGCCTCTCCTCCGCTCGCTTCTCTCATCGCCGATCGTCTTCTCGCTGGGCTTGATCCGATTCAAGGACTTCGTCCCGAAGCACTTCACTGTTTTCAAATTCCTCAGAACACTCTTCACATTTCTCGAGCCTCTCTCAGCCTCGTTTATCACGCCGCCGTCGATGGTCTTCTCGCCGAACTCCACTATACCAGCGGACTCGGTCTGAGAGCGCTTCTCAGTTCTCAAACTCCTCAAGAAAGTCACGACATGTGTCCAGTCTCCCCCCGAATCGTTCTCTCTTCGTCGATGGACGGAATGCACGGAGTCTCTTTGGGTCTTGGGAGTCTAGAGCCTTTGTTCGCTGGAGTTGCTCTGAAGAGCGAGAAGTCGCAGAAAGGTCTGAGATGGAGCCTCGGGACGAAAGAGCCGTGGTGATTTTGGGTACGGGAAATTAGGGTTTTGAAGACGGACATGAATATGTTTGTTGGGATCTTGCAAAGCAAATTTGGGGAAGTGTCCCTGAAGGAACTTTATATGTAAGGTCCTCTGAAAGTAATAACACAGCAAAAAATTGGCACTAGACTATAATTTCTTATAATATGTGAGAAATGTCCCTGAAATGAAATAAAAGAGTCACCTGACCTGGCTTTCTAACTTAATTTCTGTTTCTTTCAGGTGAGATTTTATTAAAAGAAAGCGCAAAGATTACAATCGCTTTTTAGAAAAGCTTCCACCTTACCCTTGGAGCATTTTTTTACGAGAAAAGGGGCTGAAAGCCCCTAAATTTCGTATTGAATATTAACCGAATCGACTAAAATTGCATCATCAATTAAAAGAGGAAAAACGTTATCAAAATACATAATCTCATCCAAACGGATAACAAATTTAGCAAGAGAATGAGCCGCTACATTAGAAATGCAATGATGAAAAATAATATCAAAATCAACAAAATTATCCGCACCGACGAGAATCTCCGCGATAACAAGGCCTACATCATTAGCCGGATCGACTTCGACCGGACAAACTTGACCACCACCTCGCAATCGGATTGCACAAGTAATCTTGAGTAGCCAAGGGAAGAAGCAGACCTCAAAGCGCTAGAAATAGCCATGCCTTTCGCCACCGTGGGAGTTAAGGGGCGGGCGAGTGGGGAAGCGAAAGCTTTGGAGACCACTCCCTTCACGGTTACGGAGGATGCCCCCAAAGCCGCATTTTTTGCCGTCAGCGGAGAGGACCGCATCAACTCCGAGAGCCATAATCTCGGGGGGAGGAGGGCTCCATCAAAATCGACGAGGGGGGTCACTTGAGCTTTTGGCGTCTTGTGGTTCGCACTCCGAAATTCTTCGAGAAGGCCAATAGACCGATTTACTAGGTCAGCCATGTTGCCCGAGTGGTTGTCGCGAAGCATATTGTTTCGGCGATTCCACAAGCTCCACGAAATCACCAAAAAGAGCTCGAAATCCTGAGAACATAGAAGATCACACATAATATGGGTGAGGAAAAAGATAATGTCTCCATGCATGCGCACTTTAATCATCATCCACCATCCGAGTGATTTCCACACTAGTTTAGACTCCCGACAAGCCCAAAGAGCATGGAAACTCGTCTCCGGGCCTTTACGACACCGAGGGCTTCATTCATCTTTTTGCAAACGTAGCGGTAGCGGGAGCCAACCTCGTGAGAACCTCCACGTGAATGACTGGATTTTAGGGGGATATGGAGCTTCCAAACCTTCTCCCACCATTTAATCGACTCCGAGTTAGAAAAAGTCTCGGCCACCTTCCCCGAGTCACAAAGGACACGGTATCCACTTTTGACCGAGTATTCCCCGTCCTTCGTGAAATGCCAAATGAGGTCATCCGAGATTGTAGTAGACGTCTTCGGGATCTTAAGAATTTTGTGGATGTATTCCTCGCAAAAATGGTTTCGAAGGCGGGCCACATCCCAAGTGCCATCGGGGAGAATAAAATCTGCGACTTTGAAAGATTCTTCTAGTTTCCAAGGTTCCCGCGCTTTGAAGGTTTTGCTTTTAGGAATCCACCCGTCATGCTTGGAAGAGATAGATGCCCCAGAACCCACTCTCCACCTAGTACCTCGACAAATGACTTCTCAGCCCCAATTAAGACTCCTCCAAATAGAAGAAGCTTTATAGGACCAACTAGACGTAAAAATACACGAGTCGGGAAAATAGCAAACGTTGAGGATACAAGCGGATAGGGATCGCGGCGCATTCAACACCCATCACAAATTATTCGCAAGAAGAGCTTGGTTGAAGTCTTCCAAGTTCTTAAAGCCCATGCCTCCGAGCTCTTTTCGGGCGCAAAGCTTCTCCCATCTTCCCTAATGGACCTTTTTCTTGTCGGCACCGGTAACGCACCAAAAGTTTGCCACAACTGCATGGAGTTTGGTCACAAGGCTCTTCGGAAGCATGAATACGCTCATAATATAGCTTGAGATTGCTTGTATGACTGCCTTGATGAGTACTTCTTTCCCAACCCAAGAGAAGTAAGACCCTTTCCAACTTCGGATTCTTCCCAAAACTCGGTCCTTGATGAACTTGAAGAGGAAAGCATCGTTTCGGCCTACTTTCATTGGGAGACCGAGGTATTTGTCATGGTACGCGGTCACCATAACACCGAGGTCTTAATCGAGACGCCCCGCTGACTCCCTCGAAACGTTCTTCCCAAAAGACACCTTTGACTTGTCAAAGTTCACGAGTTGACCGGAAGCCCGAGAGTACAACTCTATAGCGTTTTTGACCGTTTGCACTTCTTTCCGGTTAGCCTTGAAGAAAACCAAGCTATCACCCGCGAAGAAAACGTGAGACACAAAGATCCCCGGTCTCCCAAACTTCACCCCCGACAAACTCCCAACTTCCTCCGCCTTGCGAAACAGCTCCGAGAGCCCTTCCGAACAAAAAAGAAAAAGAAAGGGGGATAGGAGGTCCCCTTGACGCAGCCCCCGAGTCGGAATGATCTCGCCACAAATAGATCCGTTGAGGTTGAACGAGAATCTAACCGAGCGGACACATTGCATAATGAGTTCGACTCATCTATCGTCGTATCCAAGTCTTCTCATAACCTCTTCAAAAGTAGTTCCACTCGACCCGATCGTAAGCTTTACACATATCAAGCTTAAGGGATCCACCATAACCGTTACCAAATCTTCCTTTTCTCATCGCATGAAGGCCCTCGTAGCCGATGATAACCTTAGCGAAGATTTTTCTACCTCCCACAAAGGCACTTTGCGTCTCCGAAATGACATCCCCTATAGACCATTGAAGACGGTTCGTGATGCATTTCACAGTCACTTTGTATAACACGGTGCAAAGGCTAATCGGCCGAAAGTCCTTCACGTGTTTAGGAGTCTTGTTTTTTGGAATGAGAGTGATGACCGTATCGTTCAAGTGGTGAAGATCCCCCTTATTGTTAAGAACCTCGAGGCAAGTATCCGTAACGCTACGACCCACGATATTCCAATACTTTTGAAAAAAAAGACCCGGAGACCGTTATGGCCCGGCGCCTTGAGGGGACCGATATTAAAGATGGCCGTCTTAACCTCCTCTCCAAGAACGAAGGCTGTCAACATAGTGTTGGTGTCATCCATGACCTTTGAGCTCAAGTTGCTAGTCACATCGTATACAACACTCGGACGAGGATTAGTAGAAGAGAACATACGGGCGAAGTAGTTAGAAATTGTACCTTCGATTATCTCCGAAGAGGTTTGCCAAACTCCATAAGAATCAAAGAGACCGAGAATCTTGTTCTTCTTGGATCTTTGATTAGCTTGGCAATGGAAGAATTTTGAGTTGGCATCCCCGCTCTTGAGCCAAAGGAACTTCGTTGTCTCCAATACTTGGATTCTTTCTCCAATAGAGAGTTGAGCGTTTTCTATTCTCGGCACACCTTACTCGCATCGAAAGAGGAAGAGAAAGAGTTTAGAGCCTCCAAGACTCGGCGTTGCTTCTCTATACTCTTCCTCATACCATGGAATTTGTCTCTCCCCCATTCCACTAAAGATTTGGCACATTTGGCCGAGGCTTCATAGTTATCACCCCAAGTCTCCGCCACAATCTCCTCACATCTTTTTTCATCGCACCAAGCTTCCTAAAAAAAGAAGCGCTTCTTAATAGACTGTCCTCGACGCTGCCTTGTCTCTCTACCATCGGCTAGGAGAAGGAGCGGTCGATGATCGGAGGCCCAATGATCAAGATGGACTACTTTGGCTGCCTTGAAATGGTCCGTTCAATCAAGATTGGTAACACATCGGTCTAGGCAAATTCTCGTACACCCCGATTCCGTCTTCCTCTTCCACGTGAAAGGGTCACCTTCGAATCCAATATCTTCCAAATCGCAACCATCAAGGACATTTCGAAAAGCCTCCATAGACCCAATATTTTTGGCGCATCCCCTTTTTTCTCCGACTCACAGAGAATCTCATTAAAGTCGCCCTCGCAAAGCCATGGTCCTCGAAAATTACTATGCACTTGCCGTATGAGAACCCAAAAAAACTTTCTTTGACTTTCTTCCAGGTTGCCGTATATTCCCATAAATCTCCAACAAAACCGGTCAACCTTTACGCATTCCGCCTCAATAAAAAATTCTTGTATAAGACACAATTCGAATATCCTTGTCGCTTCTCCAAATCACAATTCGAGCATTTGCTTTTGGGTTATTTTATAGTTTTTTCTTCGAGCAATTTCGTAATTAAAGATCAACATGAAGATATGCTGATCATAATTCTCGAGTTTCTCATGGCATCTCTTTATTGATCGGTGTACCTTTTACAACAAATATGAATTTTATAAACTGCAATCATAATTCGAAGCAACGTGTCACAATTAGAAGCGATTTGTTGAAGTAGTTTTTCTTATGCGCCAACTATCTTTATTTCAAAGTATCTTTTCTTCGAAAAATCAATTTTTCAATGTTTAGAACGACAAATCATTTTTTTTTGTTTTATGCTATTTGAAGCGATTAGTAGAAGTTATATGTTATTTCTAAATGTCAAATTTTATTGGCTACGCTTCATTTCCACCAAATTTTCAGATTTTTCAAAACATCAAAGTTATTAAACATTTAGAAGCGGTTTTTGAAATAACAAGTTACTTCTAAATGCCAAAAAAATTGAGGGAATAGTTCCACCCTTCTCAACGGTCATCAAAGTTTTCTCATCAATTGTCTTTTAAAAAAACTTTCGTCTTACTTCCAGCATCTAAAGTTTGGCCTTTACAAAAGTAGAGTTCGACCGATATTTTATCCCGCTTTCATGTTCATTATAACAAAAACGGCAAACCATGTCATACTCTATGAATCGAGTATTCCACTCTTATATCAATTAAAACTATGATTAATGAATATAATAATCATTAAAAACATAGAAATTGCTACATAGTCATTGTTTAGTGAGATTTCGTTTATAGCAATAAATAAAAATAAATCGAAAATGGATACCTTGCATTAGAAAGCATAAATAAATATAAAATTGTTGAAAGTAATATGAAACTATTGAAAGTAATATAAAATAAATGAATAAAACACGGTTATCTCGGATTGGTTGCGGGACTAAAGGATGGGAAATAAATGAATTGATTCTATTTTAAAATAAATTTGAGAGTGAAAATAGGTAACCAAACTGTCACACTTCTCTCTCCATTTTTTCTCTACCTAAAATTATCTCTCCACAAAAACAAAAATAAACAAATGGGAACAGTTTCAAGCCTGGAAAAAAACCTTAACGTCAATAAAAGCTTACTCGCGACCACCGAAAGCGTTTGAAGGAAAAGTTAGTATAACTTTTTTCCAAAAATTCAGACCAAAAACGTACGTGCATGTTGTTTTCTTTATATTTAATATTTTTTTACTTCGATTTAGGATTTCTGAAACATTTAGGGTGGATCAATTTTTTATTTTTAAGATCTCCAACATTTTTTTATGAAATGAATCAATAAAAAAAAGAGTTCAATTGGTGTAAAGTATATCAAAATAAATTACCAAAATGAACGTGATTCTTACTATACTTCAAATAAAACAAATACAGGAAAGATGATAAGCATTATTAATAAAAATGTTAATATTGCCGCTTATTTCTTTTAATTGAAAATTAATACATGATCCGTTATTAATGGATTAATATTTCAAATCCCATGAAGAACACACAAAAAAGGCAATTATATTTTAATTAAGGCTTAAAGAAATTGTAACTGTTGTCGCTAAAATTTGGACAACGTTCAGAGACCTAAGCGGTCAGATCCCCGCTCGGGGCGAGTGACCCTCGCTTCGGCCTTAGCGTGTCAGATCCGGTGTCCCGGATCGTCTCTTAGCTGTCGTGTGGTCTTCGAGTCCTTCTGTTATCTGAAAAGAGTAAACAGACGTGAACTAAGAAGGGAAAACAGAGGGGGGGTCCTCTGTTTACCTCTCCGACGCTCACCATTAGTATAATGTGGGTGGAGTTTGTAGGATGACGAGGAAGGATCGACAAGATCGGGCGGAAACGGAGACTTAAGAGTCAACGTCGCTTCGACGCTCGGACCGAGGATCTTAGGGAACAATTACTATGGGAAGTTGTTCCCGACTCAGCTTCCCGAGGAGGGTATGGCCGAGATAGAGAGAGAGTAAGAGAGAGAGAGTATAGTTTTCACTGGGGAAAACTGGTGTACCTTTGTTATGAGGGTTGTGGGGCTATTAATAGGATGAGGAGAGCTCTAGGGCCGGCCAAGAGACCCCTTGGGCGTGTGGAAGGCAAGAGGACGTCCAGGCCACCTTGCAGGACGTCCAGGCCCTCGGTGGACGTCCAGGCCACCTTGCAGGACGTCCAGGCCCCCGGTGGACGTCCAGGCCACCTTGCAGGACGTCCAGGCCCTCGGTGGACGTCCAGGCCACC</code></pre>

<p class="blog-post_paragraph">I first analyzed the read size distrubution, using the <code>plot_fasta_length.py</code> script provided.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cp /home/ubuntu/Share/scripts/plot_fasta_length.py .
python plot_fasta_length.py drUrtUren1.ccs.total.fasta.gz drUrtUren1.ccs.total.fasta.length.png</code></pre>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_28.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">And then I ran <a href="https://github.com/marcelauliano/Teaching/blob/main/asmstats" target="_blank">asmstats</a>, a program provided by Dr. Marcela Uliano-Silva, to view the statistics of the reads</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">export PATH=$PATH:scp /home/ubuntu/Share/scripts/
asmstats drUrtUren1.ccs.total.fasta.gz > drUrtUren1.ccs.total.fasta.stats
cat drUrtUren1.ccs.total.fasta.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1.ccs.total.fasta.gz
A = 5846066817 (31.1%), C = 3568582822 (19.0%), G = 3558705638 (18.9%), T = 5847264174 (31.1%), CpG = 1354752162 (7.2%)
sum = 18820619451, n = 2215315, mean = 8495.68546730375, largest = 31748, smallest = 51
N50 = 9043, L50 = 843287
N60 = 8408, L60 = 1059112
N70 = 7757, L70 = 1291979
N80 = 7012, L80 = 1546627
N90 = 5993, L90 = 1835022
N100 = 51, L100 = 2215315</code></pre>

<p class="blog-post_paragraph">From all this data I can now now estimate how much coverage of the genome I have in this reads file. From the summary statistics I can see that I have 18820619451 total bases in the sequencing reads. From the GenomeScope <em>k</em>-mer spectra, I can see that the estimated genome size is around 331324497 base pairs. Dividing the total number of bases in the reads by the estimated genome size, I can calculate the estimated coverage I have to assemble this genome. 18820619451 / 331324497 ≅ <strong>56.8X coverage</strong>.</p>

<p class="blog-post_paragraph">Now that I've performed my <em>k</em>-mer analysis, I can begin the process of assembling my genome.</p>

<p class="blog-post_paragraph">
    <u><em>Step 2: Performing a Genome Assembly</em></u>
  </p>

<p class="blog-post_paragraph">In order to perform a genome assembly, you have to pick an ideal assembler for your data. This involves understanding the basic mechanisms behind how these assemblers work and also the differences between the popular assemblers used by the research community. This is critical because different assemblers employ different overall strategies and are designed to assemble reads that span the genome at a certain coverage, and depending on which one you choose will effect how you interpret and process your data downstream of the initial assembly step.</p>

<p class="blog-post_paragraph">For diploid organisms, a lot of modern long, high-quality read assemblers will work by first creating a collapsed draft assembly, called the <u>primary assembly</u> (usually a fusion of the paternal and maternal haplotypes). This assembly is as contiguous as possible without scaffolding. Then, an <u>alternate assembly</u> will try to be assembled. This assembly will usually be much more fragmented, because it is usually comprised of all the heterozygous variation not collapsed into the primary assembly. The primary assembly will contain all the homozygous sequences common to both maternal and paternal chromosomes (see the left side of the figure below for reference).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_29.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">For PacBio HiFi data, <u>the current consensus best assembler is the <a href="https://hifiasm.readthedocs.io/en/latest/" target="_blank">hifiasm</a> assembler</u> written by Heng Li (<a href="https://doi.org/10.1038/s41592-020-01056-5" target="_blank">see paper here</a>). This was verified by a <a href="https://doi.org/10.1101/gr.278232.123" target="_blank">benchmarking study</a> published in 2024, however it's always important to stay up to date with the literature and the latest trends and understand the biology of your organism before making your decision of which assembler to choose for your organism.</p>

<p class="blog-post_paragraph">hifiasm works by first correcting errors in the sequencing reads if there are 3 or more reads that overlap at an errorneous location. It then tries to reconstruct both haplotypes into a primary and an alternate assembly using an internal purging tool. It can then phase these assemblies into their correct paternal and maternal chromosomes if complimentary data (either trio binning data or Hi-C data is provided). Trio binning involves taking illumina sequencing reads from the parents of the organism and using that data to resolve the haplotypes of the child's genome assembly. This method is useful but it doesn't work very accurately if the paternal, maternal, and child genomes are too homozygous. Hi-C on the other hand is an assay that uses DNA-DNA proximity ligation to identify long range chromatin contacts in an unbiased, genome-wide fashion. If a high quality Hi-C library is generated, it can be used to phase the genome of the organism (elaborated on in a further section of this post).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_30.png" alt="alt text" />
</div>

<p class="blog-post_paragraph"><a href="http://www.genome.org/cgi/doi/10.1101/gr.263566" target="_blank">HiCanu</a> was one of the first generation of HiFi read assemblers. HiCanu also does read error correction in the approach as hifiasm. It finds the coverage in the reads and if there is enough of a consensus overlap to correctly identify an error it will correct it. However one major difference in it's assembly strategy is it performs homopolymer compression (e.g. <code>CAAAT</code> -> <code>CAT</code>). This is done because when you have a homopolymer in the sequence, there is a higher chance that you will have errors in the sequencing reads. So through this strategy, the OLC graph is created using compressed reads that likely have less errors and then after the assembly graph is created, it is expanded to retain the homopolymer information. Also, because HiFi reads are very long and high quality, the assembler can detect and seperate haplotypes very well. HiCanu tries to find perfect overlaps, so when it finds something like a SNP, it's probably going to thinks it's another haplotype of the same contig. So HiCanu is going to output to you a <strong>pot of contigs with double the expected genome size</strong>. It combines all the haplotype contigs into one output files. This is why you have to run a post processing step on HiCanu assemlbies especially (using tools like <a href="https://github.com/dfguan/purge_dups" target="_blank">purge_dups</a>)</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_31.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Remember that assemblers work at a specific coverage. According to the <a href="https://canu.readthedocs.io/en/latest/quick-start.html" target="_blank">documentation</a>, HiCanu will work best around 30 - 60x coverage. This is important to keep in mind because lets say for example that we were trying to assemble an organellar genome like the mitochondrion. Reads from mitochondria might comprise an extremely high amount of the DNA in the sample, and you could get upwards of 500X coverage for this small genome. If you were to try to use a program like HiCanu to assemble this, HiCanu would not take all 500X reads into account to try to assemble this genome, especially since for HiCanu you need to provide an estimated genome size. 600X coverage for a small genome is overkill and so it would try to <em>downsample</em> the reads to get it to a more optimal coverage range and this downsampling would be <em>random</em>. This means that running the same assembly program with the same input could result in <strong>2 different assemblies</strong> if you are not aware of this behaviour. Once again, it's important to pick the right tool for the job. Organelle genome assemblies should be performed using programs like <a href="https://github.com/marcelauliano/MitoHiFi" target="_blank">MitoHiFi</a> or <a href="https://github.com/c-zhou/oatk" target="_blank">Oatk</a>.</p>

<p class="blog-post_paragraph">In the workshop I ran the sequencing reads through both hifiasm and HiCanu. I started off by creating my directories and setting up my environment.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd ~/drUrtUren1/
mkdir hicanu hifiasm
conda activate eukaryotic_genome_assembly
cd hicanu/
export PATH=scp /home/ubuntu/Share/softwares/canu-2.2/bin:$PATH</code></pre>

<p class="blog-post_paragraph">I performed the first raw genome assembly with HiCanu.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">canu</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">usage:   canu [-version] [-citation] \
              [-haplotype | -correct | -trim | -assemble | -trim-assemble] \
              [-s <assembly-specifications-file>] \
               -p <assembly-prefix> \
               -d <assembly-directory> \
               genomeSize=<number>[g|m|k] \
              [other-options] \
              [-haplotype{NAME} illumina.fastq.gz] \
              [-corrected] \
              [-trimmed] \
              [-pacbio |
               -nanopore |
               -pacbio-hifi] file1 file2 ...

example: canu -d run1 -p godzilla genomeSize=1g -nanopore-raw reads/*.fasta.gz


  To restrict canu to only a specific stage, use:
    -haplotype     - generate haplotype-specific reads
    -correct       - generate corrected reads
    -trim          - generate trimmed reads
    -assemble      - generate an assembly
    -trim-assemble - generate trimmed reads and then assemble them

  The assembly is computed in the -d <assembly-directory>, with output files named
  using the -p <assembly-prefix>.  This directory is created if needed.  It is not
  possible to run multiple assemblies in the same directory.

  The genome size should be your best guess of the haploid genome size of what is being
  assembled.  It is used primarily to estimate coverage in reads, NOT as the desired
  assembly size.  Fractional values are allowed: '4.7m' equals '4700k' equals '4700000'

  Some common options:
    useGrid=string
      - Run under grid control (true), locally (false), or set up for grid control
        but don't submit any jobs (remote)
    rawErrorRate=fraction-error
      - The allowed difference in an overlap between two raw uncorrected reads.  For lower
        quality reads, use a higher number.  The defaults are 0.300 for PacBio reads and
        0.500 for Nanopore reads.
    correctedErrorRate=fraction-error
      - The allowed difference in an overlap between two corrected reads.  Assemblies of
        low coverage or data with biological differences will benefit from a slight increase
        in this.  Defaults are 0.045 for PacBio reads and 0.144 for Nanopore reads.
    gridOptions=string
      - Pass string to the command used to submit jobs to the grid.  Can be used to set
        maximum run time limits.  Should NOT be used to set memory limits; Canu will do
        that for you.
    minReadLength=number
      - Ignore reads shorter than 'number' bases long.  Default: 1000.
    minOverlapLength=number
      - Ignore read-to-read overlaps shorter than 'number' bases long.  Default: 500.
  A full list of options can be printed with '-options'.  All options can be supplied in
  an optional sepc file with the -s option.

  For TrioCanu, haplotypes are specified with the -haplotype{NAME} option, with any
  number of haplotype-specific Illumina read files after.  The {NAME} of each haplotype
  is free text (but only letters and numbers, please).  For example:
    -haplotypeNANNY nanny/*gz
    -haplotypeBILLY billy1.fasta.gz billy2.fasta.gz

  Reads can be either FASTA or FASTQ format, uncompressed, or compressed with gz, bz2 or xz.

  Reads are specified by the technology they were generated with, and any processing performed.

  [processing]
    -corrected
    -trimmed

  [technology]
    -pacbio      <files>
    -nanopore    <files>
    -pacbio-hifi <files>

Complete documentation at http://canu.readthedocs.org/en/latest/</code></pre>

<p class="blog-post_paragraph">I performed the assembly on a subset of 600 sequencing reads in order to save time and computing power. HiCanu requires that you pass in the estimated size of the genome as input. Here I put 16000 since the 600 reads correspond to this organism's mitochondria.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">ln -s scp /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.600.fasta
canu -d run1 -p drUrtUren1.600 genomeSize=16000 -pacbio-hifi drUrtUren1.600.fasta useGrid=false corThreads=2</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">-- canu 2.2
--
-- CITATIONS
--
-- For assemblies of PacBio HiFi reads:
--   Nurk S, Walenz BP, Rhiea A, Vollger MR, Logsdon GA, Grothe R, Miga KH, Eichler EE, Phillippy AM, Koren S.
--   HiCanu: accurate assembly of segmental duplications, satellites, and allelic variants from high-fidelity long reads.
--   biorXiv. 2020.
--   https://doi.org/10.1101/2020.03.14.992248
--
-- Read and contig alignments during correction and consensus use:
--   Šošic M, Šikic M.
--   Edlib: a C/C ++ library for fast, exact sequence alignment using edit distance.
--   Bioinformatics. 2017 May 1;33(9):1394-1395.
--   http://doi.org/10.1093/bioinformatics/btw753
--
-- Overlaps are generated using:
--   Berlin K, et al.
--   Assembling large genomes with single-molecule sequencing and locality-sensitive hashing.
--   Nat Biotechnol. 2015 Jun;33(6):623-30.
--   http://doi.org/10.1038/nbt.3238
--
--   Myers EW, et al.
--   A Whole-Genome Assembly of Drosophila.
--   Science. 2000 Mar 24;287(5461):2196-204.
--   http://doi.org/10.1126/science.287.5461.2196
--
-- Contig consensus sequences are generated using an algorithm derived from pbdagcon:
--   Chin CS, et al.
--   Nonhybrid, finished microbial genome assemblies from long-read SMRT sequencing data.
--   Nat Methods. 2013 Jun;10(6):563-9
--   http://doi.org/10.1038/nmeth.2474
--
-- CONFIGURE CANU
--
-- Detected Java(TM) Runtime Environment '11.0.19' (from 'java') without -d64 support.
--
-- WARNING:
-- WARNING:  Failed to run gnuplot using command 'gnuplot'.
-- WARNING:  Plots will be disabled.
-- WARNING:
--
--
-- Detected 32 CPUs and 124 gigabytes of memory on the local machine.
--
-- Local machine mode enabled; grid support not detected or not allowed.
--
--                                (tag)Concurrency
--                         (tag)Threads          |
--                (tag)Memory         |          |
--        (tag)             |         |          |       total usage      algorithm
--        -------  ----------  --------   --------  --------------------  -----------------------------
-- Local: meryl     12.000 GB    4 CPUs x   8 jobs    96.000 GB  32 CPUs  (k-mer counting)
-- Local: hap        8.000 GB    4 CPUs x   8 jobs    64.000 GB  32 CPUs  (read-to-haplotype assignment)
-- Local: cormhap    6.000 GB   16 CPUs x   2 jobs    12.000 GB  32 CPUs  (overlap detection with mhap)
-- Local: obtovl     4.000 GB    8 CPUs x   4 jobs    16.000 GB  32 CPUs  (overlap detection)
-- Local: utgovl     4.000 GB    8 CPUs x   4 jobs    16.000 GB  32 CPUs  (overlap detection)
-- Local: cor        -.--- GB    2 CPUs x   - jobs     -.--- GB   - CPUs  (read correction)
-- Local: ovb        4.000 GB    1 CPU  x  31 jobs   124.000 GB  31 CPUs  (overlap store bucketizer)
-- Local: ovs        8.000 GB    1 CPU  x  15 jobs   120.000 GB  15 CPUs  (overlap store sorting)
-- Local: red       15.000 GB    4 CPUs x   8 jobs   120.000 GB  32 CPUs  (read error detection)
-- Local: oea        8.000 GB    1 CPU  x  15 jobs   120.000 GB  15 CPUs  (overlap error adjustment)
-- Local: bat       16.000 GB    4 CPUs x   1 job     16.000 GB   4 CPUs  (contig construction with bogart)
-- Local: cns        -.--- GB    4 CPUs x   - jobs     -.--- GB   - CPUs  (consensus)
--
-- Found trimmed raw PacBio HiFi reads in the input files.
--
-- Generating assembly 'drUrtUren1.600' in '/home/user19/drUrtUren1/hicanu/run1':
--   genomeSize:
--     16000
--
--   Overlap Generation Limits:
--     corOvlErrorRate 0.0000 (  0.00%)
--     obtOvlErrorRate 0.0250 (  2.50%)
--     utgOvlErrorRate 0.0100 (  1.00%)
--
--   Overlap Processing Limits:
--     corErrorRate    0.0000 (  0.00%)
--     obtErrorRate    0.0250 (  2.50%)
--     utgErrorRate    0.0003 (  0.03%)
--     cnsErrorRate    0.0500 (  5.00%)
--
--   Stages to run:
--     assemble HiFi reads.
--
--
-- Correction skipped; not enabled.
--
-- Trimming skipped; not enabled.
--
-- BEGIN ASSEMBLY
----------------------------------------
-- Starting command on Wed Nov  6 14:03:30 2024 with 103.005 GB free disk space

    cd .
    ./drUrtUren1.600.seqStore.sh \
    > ./drUrtUren1.600.seqStore.err 2>&1

-- Finished on Wed Nov  6 14:03:30 2024 (furiously fast) with 103.003 GB free disk space
----------------------------------------
--
-- In sequence store './drUrtUren1.600.seqStore':
--   Found 355 reads.
--   Found 3202731 bases (200.17 times coverage).
--    Histogram of corrected reads:
--
--    G=3202731                          sum of  ||               length     num
--    NG         length     index       lengths  ||                range    seqs
--    ----- ------------ --------- ------------  ||  ------------------- -------
--    00010        13553        21       325970  ||       3597-3927            3|--------
--    00020        12053        46       642481  ||       3928-4258            5|-------------
--    00030        10845        74       962643  ||       4259-4589            3|--------
--    00040        10191       105      1289927  ||       4590-4920            5|-------------
--    00050         9683       137      1607664  ||       4921-5251            7|-----------------
--    00060         8945       171      1922234  ||       5252-5582           10|-------------------------
--    00070         8134       209      2244504  ||       5583-5913            5|-------------
--    00080         7603       250      2567202  ||       5914-6244            8|--------------------
--    00090         6561       295      2885096  ||       6245-6575           16|---------------------------------------
--    00100         3597       354      3202731  ||       6576-6906           11|---------------------------
--    001.000x                 355      3202731  ||       6907-7237           17|------------------------------------------
--                                               ||       7238-7568           12|------------------------------
--                                               ||       7569-7899           23|--------------------------------------------------------
--                                               ||       7900-8230           26|---------------------------------------------------------------
--                                               ||       8231-8561           19|-----------------------------------------------
--                                               ||       8562-8892           11|---------------------------
--                                               ||       8893-9223           17|------------------------------------------
--                                               ||       9224-9554           15|-------------------------------------
--                                               ||       9555-9885           15|-------------------------------------
--                                               ||       9886-10216          23|--------------------------------------------------------
--                                               ||      10217-10547          10|-------------------------
--                                               ||      10548-10878          20|-------------------------------------------------
--                                               ||      10879-11209           8|--------------------
--                                               ||      11210-11540           9|----------------------
--                                               ||      11541-11871           5|-------------
--                                               ||      11872-12202           9|----------------------
--                                               ||      12203-12533           5|-------------
--                                               ||      12534-12864           7|-----------------
--                                               ||      12865-13195           7|-----------------
--                                               ||      13196-13526           2|-----
--                                               ||      13527-13857           4|----------
--                                               ||      13858-14188           4|----------
--                                               ||      14189-14519           3|--------
--                                               ||      14520-14850           2|-----
--                                               ||      14851-15181           3|--------
--                                               ||      15182-15512           2|-----
--                                               ||      15513-15843           2|-----
--                                               ||      15844-16174           1|---
--                                               ||      16175-16505           0|
--                                               ||      16506-16836           0|
--                                               ||      16837-17167           0|
--                                               ||      17168-17498           0|
--                                               ||      17499-17829           0|
--                                               ||      17830-18160           0|
--                                               ||      18161-18491           0|
--                                               ||      18492-18822           0|
--                                               ||      18823-19153           0|
--                                               ||      19154-19484           0|
--                                               ||      19485-19815           0|
--                                               ||      19816-20146           1|---
--
--
-- In sequence store './drUrtUren1.600.seqStore':
--   Found 355 reads.
--   Found 3202731 bases (200.17 times coverage).
--    Histogram of corrected-trimmed reads:
--
--    G=3202731                          sum of  ||               length     num
--    NG         length     index       lengths  ||                range    seqs
--    ----- ------------ --------- ------------  ||  ------------------- -------
--    00010        13553        21       325970  ||       3597-3927            3|--------
--    00020        12053        46       642481  ||       3928-4258            5|-------------
--    00030        10845        74       962643  ||       4259-4589            3|--------
--    00040        10191       105      1289927  ||       4590-4920            5|-------------
--    00050         9683       137      1607664  ||       4921-5251            7|-----------------
--    00060         8945       171      1922234  ||       5252-5582           10|-------------------------
--    00070         8134       209      2244504  ||       5583-5913            5|-------------
--    00080         7603       250      2567202  ||       5914-6244            8|--------------------
--    00090         6561       295      2885096  ||       6245-6575           16|---------------------------------------
--    00100         3597       354      3202731  ||       6576-6906           11|---------------------------
--    001.000x                 355      3202731  ||       6907-7237           17|------------------------------------------
--                                               ||       7238-7568           12|------------------------------
--                                               ||       7569-7899           23|--------------------------------------------------------
--                                               ||       7900-8230           26|---------------------------------------------------------------
--                                               ||       8231-8561           19|-----------------------------------------------
--                                               ||       8562-8892           11|---------------------------
--                                               ||       8893-9223           17|------------------------------------------
--                                               ||       9224-9554           15|-------------------------------------
--                                               ||       9555-9885           15|-------------------------------------
--                                               ||       9886-10216          23|--------------------------------------------------------
--                                               ||      10217-10547          10|-------------------------
--                                               ||      10548-10878          20|-------------------------------------------------
--                                               ||      10879-11209           8|--------------------
--                                               ||      11210-11540           9|----------------------
--                                               ||      11541-11871           5|-------------
--                                               ||      11872-12202           9|----------------------
--                                               ||      12203-12533           5|-------------
--                                               ||      12534-12864           7|-----------------
--                                               ||      12865-13195           7|-----------------
--                                               ||      13196-13526           2|-----
--                                               ||      13527-13857           4|----------
--                                               ||      13858-14188           4|----------
--                                               ||      14189-14519           3|--------
--                                               ||      14520-14850           2|-----
--                                               ||      14851-15181           3|--------
--                                               ||      15182-15512           2|-----
--                                               ||      15513-15843           2|-----
--                                               ||      15844-16174           1|---
--                                               ||      16175-16505           0|
--                                               ||      16506-16836           0|
--                                               ||      16837-17167           0|
--                                               ||      17168-17498           0|
--                                               ||      17499-17829           0|
--                                               ||      17830-18160           0|
--                                               ||      18161-18491           0|
--                                               ||      18492-18822           0|
--                                               ||      18823-19153           0|
--                                               ||      19154-19484           0|
--                                               ||      19485-19815           0|
--                                               ||      19816-20146           1|---
--
----------------------------------------
-- Starting command on Wed Nov  6 14:03:30 2024 with 103.003 GB free disk space

    cd unitigging/0-mercounts
    ./meryl-configure.sh \
    > ./meryl-configure.err 2>&1

-- Finished on Wed Nov  6 14:03:30 2024 (like a bat out of hell) with 103.003 GB free disk space
----------------------------------------
--  segments   memory batches
--  -------- -------- -------
--        01  0.01 GB       2
--
--  For 355 reads with 3202731 bases, limit to 1 batch.
--  Will count kmers using 01 jobs, each using 2 GB and 4 threads.
--
-- Finished stage 'merylConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'meryl' concurrent execution on Wed Nov  6 14:03:30 2024 with 103.003 GB free disk space (1 processes; 8 concurrently)

    cd unitigging/0-mercounts
    ./meryl-count.sh 1 > ./meryl-count.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:31 2024 (one second) with 103.001 GB free disk space
----------------------------------------
-- Found 1 Kmer counting (meryl) outputs.
-- Finished stage 'utg-merylCountCheck', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'meryl' concurrent execution on Wed Nov  6 14:03:31 2024 with 103.001 GB free disk space (1 processes; 8 concurrently)

    cd unitigging/0-mercounts
    ./meryl-process.sh 1 > ./meryl-process.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:32 2024 (one second) with 103.001 GB free disk space
----------------------------------------
-- Meryl finished successfully.  Kmer frequency histogram:
--
-- WARNING: gnuplot failed.
--
----------------------------------------
--
--  22-mers                                                                                           Fraction
--    Occurrences   NumMers                                                                         Unique Total
--       1-     1         0                                                                        0.0000 0.0000
--       2-     2     34396 ************************************************                       0.1998 0.0319
--       3-     4     50029 ********************************************************************** 0.3611 0.0706
--       5-     7     48474 *******************************************************************    0.6018 0.1565
--       8-    11     14334 ********************                                                   0.7948 0.2599
--      12-    16      4023 *****                                                                  0.8611 0.3136
--      17-    22      3258 ****                                                                   0.8799 0.3354
--      23-    29      1958 **                                                                     0.8998 0.3676
--      30-    37      2410 ***                                                                    0.9111 0.3918
--      38-    46      2040 **                                                                     0.9246 0.4287
--      47-    56      2809 ***                                                                    0.9365 0.4694
--      57-    67       909 *                                                                      0.9515 0.5307
--      68-    79       725 *                                                                      0.9571 0.5587
--      80-    92       873 *                                                                      0.9609 0.5817
--      93-   106       852 *                                                                      0.9660 0.6167
--     107-   121       884 *                                                                      0.9709 0.6561
--     122-   137      1224 *                                                                      0.9767 0.7092
--     138-   154       732 *                                                                      0.9832 0.7771
--     155-   172       935 *                                                                      0.9874 0.8259
--     173-   191      1264 *                                                                      0.9930 0.8991
--     192-   211         2                                                                        1.0000 0.9999
--
--           0 (max occurrences)
--     2153859 (total mers, non-unique)
--      172131 (distinct mers, non-unique)
--           0 (unique mers)
-- Finished stage 'meryl-process', reset canuIteration.
--
-- Removing meryl database 'unitigging/0-mercounts/drUrtUren1.600.ms22'.
--
-- OVERLAPPER (normal) (assembly) erate=0.01
--
----------------------------------------
-- Starting command on Wed Nov  6 14:03:32 2024 with 103.003 GB free disk space

    cd unitigging/1-overlapper
    /home/ubuntu/Share/softwares/canu-2.2/bin/overlapInCorePartition \
     -S  ../../drUrtUren1.600.seqStore \
     -hl 80000000 \
     -rl 1000000000 \
     -ol 500 \
     -o  ./drUrtUren1.600.partition \
    > ./drUrtUren1.600.partition.err 2>&1

-- Finished on Wed Nov  6 14:03:32 2024 (furiously fast) with 103.003 GB free disk space
----------------------------------------
--
-- Configured 1 overlapInCore jobs.
-- Finished stage 'utg-overlapConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'utgovl' concurrent execution on Wed Nov  6 14:03:32 2024 with 103.003 GB free disk space (1 processes; 4 concurrently)

    cd unitigging/1-overlapper
    ./overlap.sh 1 > ./overlap.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:34 2024 (2 seconds) with 103.003 GB free disk space
----------------------------------------
-- Found 1 overlapInCore output files.
--
-- overlapInCore compute 'unitigging/1-overlapper':
--   kmer hits
--     with no overlap            11902       11902 +- 0
--     with an overlap            18352       18352 +- 0
--
--   overlaps                     18352       18352 +- 0
--     contained                   4941        4941 +- 0
--     dovetail                   13411       13411 +- 0
--
--   overlaps rejected
--     multiple per pair              0           0 +- 0
--     bad short window               0           0 +- 0
--     bad long window                0           0 +- 0
-- Finished stage 'utg-overlapCheck', reset canuIteration.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:34 2024 with 103.003 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/ovStoreConfig \
     -S ../drUrtUren1.600.seqStore \
     -M 4-8 \
     -L ./1-overlapper/ovljob.files \
     -create ./drUrtUren1.600.ovlStore.config \
     > ./drUrtUren1.600.ovlStore.config.txt \
    2> ./drUrtUren1.600.ovlStore.config.err

-- Finished on Wed Nov  6 14:03:34 2024 (in the blink of an eye) with 103.003 GB free disk space
----------------------------------------
--
-- Creating overlap store unitigging/drUrtUren1.600.ovlStore using:
--      1 bucket
--      2 slices
--        using at most 1 GB memory each
-- Finished stage 'utg-overlapStoreConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'ovB' concurrent execution on Wed Nov  6 14:03:34 2024 with 103.003 GB free disk space (1 processes; 31 concurrently)

    cd unitigging/drUrtUren1.600.ovlStore.BUILDING
    ./scripts/1-bucketize.sh 1 > ./logs/1-bucketize.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:34 2024 (lickety-split) with 103.002 GB free disk space
----------------------------------------
-- Overlap store bucketizer finished.
-- Finished stage 'utg-overlapStoreBucketizerCheck', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'ovS' concurrent execution on Wed Nov  6 14:03:34 2024 with 103.002 GB free disk space (2 processes; 15 concurrently)

    cd unitigging/drUrtUren1.600.ovlStore.BUILDING
    ./scripts/2-sort.sh 1 > ./logs/2-sort.000001.out 2>&1
    ./scripts/2-sort.sh 2 > ./logs/2-sort.000002.out 2>&1

-- Finished on Wed Nov  6 14:03:34 2024 (furiously fast) with 103.002 GB free disk space
----------------------------------------
-- Overlap store sorter finished.
-- Finished stage 'utg-overlapStoreSorterCheck', reset canuIteration.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:34 2024 with 103.002 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/ovStoreIndexer \
      -O  ./drUrtUren1.600.ovlStore.BUILDING \
      -S ../drUrtUren1.600.seqStore \
      -C  ./drUrtUren1.600.ovlStore.config \
      -delete \
    > ./drUrtUren1.600.ovlStore.BUILDING.index.err 2>&1

-- Finished on Wed Nov  6 14:03:34 2024 (fast as lightning) with 103.002 GB free disk space
----------------------------------------
-- Overlap store indexer finished.
-- Checking store.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:34 2024 with 103.002 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/ovStoreDump \
     -S ../drUrtUren1.600.seqStore \
     -O  ./drUrtUren1.600.ovlStore \
     -counts \
     > ./drUrtUren1.600.ovlStore/counts.dat 2> ./drUrtUren1.600.ovlStore/counts.err

-- Finished on Wed Nov  6 14:03:34 2024 (fast as lightning) with 103.002 GB free disk space
----------------------------------------
--
-- Overlap store 'unitigging/drUrtUren1.600.ovlStore' successfully constructed.
-- Found 36704 overlaps for 348 reads; 252 reads have no overlaps.
--
--
-- Purged 0 GB in 3 overlap output files.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:34 2024 with 103.002 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/ovStoreStats \
     -C 200.17 \
     -S ../drUrtUren1.600.seqStore \
     -O  ./drUrtUren1.600.ovlStore \
     -o  ./drUrtUren1.600.ovlStore \
     > ./drUrtUren1.600.ovlStore.summary.err 2>&1

-- Finished on Wed Nov  6 14:03:34 2024 (in the blink of an eye) with 103.002 GB free disk space
----------------------------------------
--
-- Overlap store 'unitigging/drUrtUren1.600.ovlStore' contains:
--
--   category            reads     %          read length        feature size or coverage  analysis
--   ----------------  -------  -------  ----------------------  ------------------------  --------------------
--   middle-missing          4    1.13     7935.50 +- 652.49         887.00 +- 847.72     (bad trimming)
--   middle-hump             1    0.28     9506.00 +- 0.00          3234.00 +- 0.00       (bad trimming)
--   no-5-prime             10    2.82     7428.00 +- 1166.44       1751.20 +- 1784.47    (bad trimming)
--   no-3-prime             17    4.79     7357.94 +- 2475.49       1711.82 +- 1465.49    (bad trimming)
--
--   low-coverage          137   38.59     5935.00 +- 1797.66         16.35 +- 17.05      (easy to assemble, potential for lower quality consensus)
--   unique                 54   15.21     4784.57 +- 989.60         143.02 +- 30.12      (easy to assemble, perfect, yay)
--   repeat-cont             0    0.00        0.00 +- 0.00             0.00 +- 0.00       (potential for consensus errors, no impact on assembly)
--   repeat-dove             0    0.00        0.00 +- 0.00             0.00 +- 0.00       (hard to assemble, likely won't assemble correctly or even at all)
--
--   span-repeat             6    1.69    10053.83 +- 781.67        7435.33 +- 98.54      (read spans a large repeat, usually easy to assemble)
--   uniq-repeat-cont      111   31.27     6844.86 +- 1248.48                             (should be uniquely placed, low potential for consensus errors, no impact on assembly)
--   uniq-repeat-dove        8    2.25     9260.50 +- 883.72                              (will end contigs, potential to misassemble)
--   uniq-anchor             0    0.00        0.00 +- 0.00             0.00 +- 0.00       (repeat read, with unique section, probable bad read)
-- Finished stage 'utg-createOverlapStore', reset canuIteration.
--
-- Loading read lengths.
-- Loading number of overlaps per read.
--
-- Configure RED for 15gb memory.
--                   Batches of at most (unlimited) reads.
--                                      500000000 bases.
--                   Expecting evidence of at most 536870912 bases per iteration.
--
--           Total                                               Reads                 Olaps Evidence
--    Job   Memory      Read Range         Reads        Bases   Memory        Olaps   Memory   Memory  (Memory in MB)
--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------
--      1  3463.39         1-600             355      3202731   390.97        36704     0.42  1024.00
--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------
--                                                    3202731                 36704
-- Finished stage 'readErrorDetectionConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'red' concurrent execution on Wed Nov  6 14:03:34 2024 with 103.002 GB free disk space (1 processes; 8 concurrently)

    cd unitigging/3-overlapErrorAdjustment
    ./red.sh 1 > ./red.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:37 2024 (3 seconds) with 103.002 GB free disk space
----------------------------------------
-- Found 1 read error detection output files.
-- Finished stage 'readErrorDetectionCheck', reset canuIteration.
--
-- Loading read lengths.
-- Loading number of overlaps per read.
--
-- Configure OEA for 8gb memory.
--                   Batches of at most (unlimited) reads.
--                                      300000000 bases.
--
--           Total                                               Reads                 Olaps  Adjusts
--    Job   Memory      Read Range         Reads        Bases   Memory        Olaps   Memory   Memory  (Memory in MB)
--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------
--      1  2380.21         1-600             355      3202731    11.07        36704     1.12   320.02
--   ---- -------- ------------------- --------- ------------ -------- ------------ -------- --------
--                                                    3202731                 36704
-- Finished stage 'overlapErrorAdjustmentConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'oea' concurrent execution on Wed Nov  6 14:03:37 2024 with 103.002 GB free disk space (1 processes; 15 concurrently)

    cd unitigging/3-overlapErrorAdjustment
    ./oea.sh 1 > ./oea.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:38 2024 (one second) with 103.002 GB free disk space
----------------------------------------
-- Found 1 overlap error adjustment output files.
-- Finished stage 'overlapErrorAdjustmentCheck', reset canuIteration.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:38 2024 with 103.002 GB free disk space

    cd unitigging/3-overlapErrorAdjustment
    /home/ubuntu/Share/softwares/canu-2.2/bin/loadErates \
      -S ../../drUrtUren1.600.seqStore \
      -O ../drUrtUren1.600.ovlStore \
      -L ./oea.files \
    > ./oea.apply.err 2>&1

-- Finished on Wed Nov  6 14:03:38 2024 (in the blink of an eye) with 103.002 GB free disk space
----------------------------------------
-- No report available.
-- Finished stage 'updateOverlapStore', reset canuIteration.
-- Finished stage 'unitig', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'bat' concurrent execution on Wed Nov  6 14:03:38 2024 with 103.002 GB free disk space (1 processes; 1 concurrently)

    cd unitigging/4-unitigger
    ./unitigger.sh 1 > ./unitigger.000001.out 2>&1

-- Finished on Wed Nov  6 14:03:39 2024 (one second) with 103.002 GB free disk space
----------------------------------------
-- Unitigger finished successfully.
--
--  ERROR RATES
--  -----------
--                                                   --------threshold------
--  397                          fraction error      fraction        percent
--  samples                              (1e-5)         error          error
--                   --------------------------      --------       --------
--  command line (-eg)                           ->     30.00        0.0300%
--  command line (-ef)                           ->  -----.--      ---.----%
--  command line (-eM)                           ->     30.00        0.0300%
--  mean + std.dev       0.23 +-  12 *     2.29  ->     27.77        0.0278%
--  median + mad         0.00 +-  12 *     0.00  ->      0.00        0.0000%
--  90th percentile                              ->      1.00        0.0010%  (enabled)
--
--  BEST EDGE FILTERING
--  -------------------
--  At graph threshold 0.0300%, reads:
--    available to have edges:          108
--    with at least one edge:            91
--
--  At max threshold 0.0300%, reads:  (not computed)
--    available to have edges:            0
--    with at least one edge:             0
--
--  At tight threshold 0.0010%, reads with:
--    both edges below error threshold:        86  (80.00% minReadsBest threshold = 72)
--    one  edge  above error threshold:         5
--    both edges above error threshold:         0
--    at least one edge:                       91
--
--  At loose threshold 0.0278%, reads with:
--    both edges below error threshold:        89  (80.00% minReadsBest threshold = 72)
--    one  edge  above error threshold:         2
--    both edges above error threshold:         0
--    at least one edge:                       91
--
--
--  INITIAL EDGES
--  -------- ----------------------------------------
--       235 reads are contained
--       313 reads have no best edges (singleton)
--        21 reads have only one best edge (spur)
--                 21 are mutual best
--        31 reads have two best edges
--                  1 have one mutual best edge
--                 30 have two mutual best edges
--
--
--  FINAL EDGES
--  -------- ----------------------------------------
--       235 reads are contained
--       313 reads have no best edges (singleton)
--        22 reads have only one best edge (spur)
--                 22 are mutual best
--        30 reads have two best edges
--                  0 have one mutual best edge
--                 30 have two mutual best edges
--
--
--  EDGE FILTERING
--  -------- ------------------------------------------
--         0 reads are ignored
--        56 reads have a gap in overlap coverage
--         0 reads have lopsided best edges
-- Found, in version 1, after unitig construction:
--   contigs:      6 sequences, total length 92486 bp (including 0 repeats of total length 0 bp).
--   bubbles:      1 sequences, total length 11626 bp.
--   unassembled:  87 sequences, total length 580823 bp.
--
-- Contig sizes based on genome size 16kbp:
--
--            NG (bp)  LG (contigs)    sum (bp)
--         ----------  ------------  ----------
--     10       19531             1       19531
--     20       19531             1       19531
--     30       19531             1       19531
--     40       19531             1       19531
--     50       19531             1       19531
--     60       19531             1       19531
--     70       19531             1       19531
--     80       19531             1       19531
--     90       19531             1       19531
--    100       19531             1       19531
--    110       19531             1       19531
--    120       19531             1       19531
--    130       17493             2       37024
--    140       17493             2       37024
--    150       17493             2       37024
--    160       17493             2       37024
--    170       17493             2       37024
--    180       17493             2       37024
--    190       17493             2       37024
--    200       17493             2       37024
--    210       17493             2       37024
--    220       17493             2       37024
--    230       17493             2       37024
--    240       16586             3       53610
--    250       16586             3       53610
--    260       16586             3       53610
--    270       16586             3       53610
--    280       16586             3       53610
--    290       16586             3       53610
--    300       16586             3       53610
--    310       16586             3       53610
--    320       16586             3       53610
--    330       16586             3       53610
--    340       14497             4       68107
--    350       14497             4       68107
--    360       14497             4       68107
--    370       14497             4       68107
--    380       14497             4       68107
--    390       14497             4       68107
--    400       14497             4       68107
--    410       14497             4       68107
--    420       14497             4       68107
--    430       13997             5       82104
--    440       13997             5       82104
--    450       13997             5       82104
--    460       13997             5       82104
--    470       13997             5       82104
--    480       13997             5       82104
--    490       13997             5       82104
--    500       13997             5       82104
--    510       13997             5       82104
--    520       10382             6       92486
--    530       10382             6       92486
--    540       10382             6       92486
--    550       10382             6       92486
--    560       10382             6       92486
--    570       10382             6       92486
--
-- Finished stage 'unitigCheck', reset canuIteration.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:39 2024 with 103.002 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/utgcns \
      -S ../drUrtUren1.600.seqStore \
      -T  ./drUrtUren1.600.ctgStore 1 \
      -partition 0.8 1.5 0.1 \
    > ./drUrtUren1.600.ctgStore/partitioning.log 2>&1

-- Finished on Wed Nov  6 14:03:39 2024 (like a bat out of hell) with 103.001 GB free disk space
----------------------------------------
-- Using slow alignment for consensus (iteration '0').
-- Configured 5 consensus jobs.
-- Finished stage 'consensusConfigure', reset canuIteration.
--
-- Local: cns        0.500 GB    4 CPUs x   8 jobs     4.000 GB  32 CPUs  (consensus)
--
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'cns' concurrent execution on Wed Nov  6 14:03:39 2024 with 103.001 GB free disk space (5 processes; 8 concurrently)

    cd unitigging/5-consensus
    ./consensus.sh 1 > ./consensus.000001.out 2>&1
    ./consensus.sh 2 > ./consensus.000002.out 2>&1
    ./consensus.sh 3 > ./consensus.000003.out 2>&1
    ./consensus.sh 4 > ./consensus.000004.out 2>&1
    ./consensus.sh 5 > ./consensus.000005.out 2>&1

-- Finished on Wed Nov  6 14:03:40 2024 (one second) with 102.999 GB free disk space
----------------------------------------
-- Using slow alignment for consensus (iteration '1').
-- Configured 5 consensus jobs.
-- All 5 consensus jobs finished successfully.
-- Finished stage 'consensusCheck', reset canuIteration.
-- Using slow alignment for consensus (iteration '0').
-- Configured 5 consensus jobs.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:40 2024 with 102.999 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/tgStoreLoad \
      -S ../drUrtUren1.600.seqStore \
      -T  ./drUrtUren1.600.ctgStore 2 \
      -L ./5-consensus/ctgcns.files \
    > ./5-consensus/ctgcns.files.ctgStoreLoad.err 2>&1

-- Finished on Wed Nov  6 14:03:40 2024 (like a bat out of hell) with 102.997 GB free disk space
----------------------------------------
-- Purging consensus output after loading to ctgStore.
-- Purged 5 .cns outputs.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:40 2024 with 102.999 GB free disk space

    cd unitigging
    /home/ubuntu/Share/softwares/canu-2.2/bin/tgStoreDump \
      -S ../drUrtUren1.600.seqStore \
      -T ./drUrtUren1.600.ctgStore 2 \
      -sizes -s 16000 \
    > ./drUrtUren1.600.ctgStore/seqDB.v002.sizes.txt

-- Finished on Wed Nov  6 14:03:40 2024 (lickety-split) with 102.999 GB free disk space
----------------------------------------
-- Found, in version 2, after consensus generation:
--   contigs:      6 sequences, total length 129833 bp (including 0 repeats of total length 0 bp).
--   bubbles:      1 sequences, total length 16734 bp.
--   unassembled:  87 sequences, total length 812116 bp.
--
-- Contig sizes based on genome size 16kbp:
--
--            NG (bp)  LG (contigs)    sum (bp)
--         ----------  ------------  ----------
--     10       26889             1       26889
--     20       26889             1       26889
--     30       26889             1       26889
--     40       26889             1       26889
--     50       26889             1       26889
--     60       26889             1       26889
--     70       26889             1       26889
--     80       26889             1       26889
--     90       26889             1       26889
--    100       26889             1       26889
--    110       26889             1       26889
--    120       26889             1       26889
--    130       26889             1       26889
--    140       26889             1       26889
--    150       26889             1       26889
--    160       26889             1       26889
--    170       24154             2       51043
--    180       24154             2       51043
--    190       24154             2       51043
--    200       24154             2       51043
--    210       24154             2       51043
--    220       24154             2       51043
--    230       24154             2       51043
--    240       24154             2       51043
--    250       24154             2       51043
--    260       24154             2       51043
--    270       24154             2       51043
--    280       24154             2       51043
--    290       24154             2       51043
--    300       24154             2       51043
--    310       24154             2       51043
--    320       23885             3       74928
--    330       23885             3       74928
--    340       23885             3       74928
--    350       23885             3       74928
--    360       23885             3       74928
--    370       23885             3       74928
--    380       23885             3       74928
--    390       23885             3       74928
--    400       23885             3       74928
--    410       23885             3       74928
--    420       23885             3       74928
--    430       23885             3       74928
--    440       23885             3       74928
--    450       23885             3       74928
--    460       23885             3       74928
--    470       21554             4       96482
--    480       21554             4       96482
--    490       21554             4       96482
--    500       21554             4       96482
--    510       21554             4       96482
--    520       21554             4       96482
--    530       21554             4       96482
--    540       21554             4       96482
--    550       21554             4       96482
--    560       21554             4       96482
--    570       21554             4       96482
--    580       21554             4       96482
--    590       21554             4       96482
--    600       21554             4       96482
--    610       19060             5      115542
--    620       19060             5      115542
--    630       19060             5      115542
--    640       19060             5      115542
--    650       19060             5      115542
--    660       19060             5      115542
--    670       19060             5      115542
--    680       19060             5      115542
--    690       19060             5      115542
--    700       19060             5      115542
--    710       19060             5      115542
--    720       19060             5      115542
--    730       14291             6      129833
--    740       14291             6      129833
--    750       14291             6      129833
--    760       14291             6      129833
--    770       14291             6      129833
--    780       14291             6      129833
--    790       14291             6      129833
--    800       14291             6      129833
--    810       14291             6      129833
--
-- Finished stage 'consensusLoad', reset canuIteration.
----------------------------------------
-- Starting command on Wed Nov  6 14:03:40 2024 with 102.999 GB free disk space

    cd .
    /home/ubuntu/Share/softwares/canu-2.2/bin/tgStoreDump \
      -S ./drUrtUren1.600.seqStore \
      -T ./unitigging/drUrtUren1.600.ctgStore 2 \
      -o ./drUrtUren1.600.contigs \
      -layout \
    > ./drUrtUren1.600.contigs.layout.err 2>&1

-- Finished on Wed Nov  6 14:03:40 2024 (lickety-split) with 102.999 GB free disk space
----------------------------------------
----------------------------------------
-- Starting command on Wed Nov  6 14:03:40 2024 with 102.999 GB free disk space

    cd .
    /home/ubuntu/Share/softwares/canu-2.2/bin/tgStoreDump \
      -S ./drUrtUren1.600.seqStore \
      -T ./unitigging/drUrtUren1.600.ctgStore 2 \
      -consensus -fasta \
      -unassembled \
    > ./drUrtUren1.600.unassembled.fasta
    2> ./drUrtUren1.600.unassembled.err

-- Finished on Wed Nov  6 14:03:41 2024 (one second) with 102.998 GB free disk space
----------------------------------------
----------------------------------------
-- Starting command on Wed Nov  6 14:03:41 2024 with 102.998 GB free disk space

    cd .
    /home/ubuntu/Share/softwares/canu-2.2/bin/tgStoreDump \
      -S ./drUrtUren1.600.seqStore \
      -T ./unitigging/drUrtUren1.600.ctgStore 2 \
      -consensus -fasta \
      -contigs \
    > ./drUrtUren1.600.contigs.fasta
    2> ./drUrtUren1.600.contigs.err

-- Finished on Wed Nov  6 14:03:41 2024 (like a bat out of hell) with 102.998 GB free disk space
----------------------------------------
-- Finished stage 'generateOutputs', reset canuIteration.
--
-- Assembly 'drUrtUren1.600' finished in '/home/user19/drUrtUren1/hicanu/run1'.
--
-- Summary saved in 'drUrtUren1.600.report'.
--
-- Sequences saved:
--   Contigs       -> 'drUrtUren1.600.contigs.fasta'
--   Unassembled   -> 'drUrtUren1.600.unassembled.fasta'
--
-- Read layouts saved:
--   Contigs       -> 'drUrtUren1.600.contigs.layout'.
--
-- Bye.</code></pre>

<p class="blog-post_paragraph">Next I performed the assembly with hifiasm.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">hifiasm</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Usage: hifiasm [options] &lt;in_1.fq&gt; &lt;in_2.fq&gt; &lt;...&gt;
Options:
  Input/Output:
    -o STR       prefix of output files [hifiasm.asm]
    -t INT       number of threads [1]
    -h           show help information
    --version    show version number
  Overlap/Error correction:
    -k INT       k-mer length (must be &lt;64) [51]
    -w INT       minimizer window size [51]
    -f INT       number of bits for bloom filter; 0 to disable [37]
    -D FLOAT     drop k-mers occurring &gt;FLOAT*coverage times [5.0]
    -N INT       consider up to max(-D*coverage,-N) overlaps for each oriented read [100]
    -r INT       round of correction [3]
    -z INT       length of adapters that should be removed [0]
    --max-kocc   INT
                 employ k-mers occurring &lt;INT times to rescue repetitive overlaps [2000]
    --hg-size    INT(k, m or g)
                 estimated haploid genome size used for inferring read coverage [auto]
  Assembly:
    -a INT       round of assembly cleaning [4]
    -m INT       pop bubbles of &lt;INT in size in contig graphs [10000000]
    -p INT       pop bubbles of &lt;INT in size in unitig graphs [0]
    -n INT       remove tip unitigs composed of &lt;=INT reads [3]
    -x FLOAT     max overlap drop ratio [0.8]
    -y FLOAT     min overlap drop ratio [0.2]
    -i           ignore saved read correction and overlaps
    -u           post-join step for contigs which may improve N50; 0 to disable; 1 to enable
                 [1] and [1] in default for the UL+HiFi assembly and the HiFi assembly, respectively
    --hom-cov    INT
                 homozygous read coverage [auto]
    --lowQ       INT
                 output contig regions with &gt;=INT% inconsistency in BED format; 0 to disable [70]
    --b-cov      INT
                 break contigs at positions with &lt;INT-fold coverage; work with '--m-rate'; 0 to disable [0]
    --h-cov      INT
                 break contigs at positions with &gt;INT-fold coverage; work with '--m-rate'; -1 to disable [-1]
    --m-rate     FLOAT
                 break contigs at positions with &lt;=FLOAT*coverage exact overlaps;
                 only work with '--b-cov' or '--h-cov'[0.75]
    --primary    output a primary assembly and an alternate assembly
    --ctg-n      INT
                 remove tip contigs composed of &lt;=INT reads [3]
  Trio-partition:
    -1 FILE      hap1/paternal k-mer dump generated by "yak count" []
    -2 FILE      hap2/maternal k-mer dump generated by "yak count" []
    -3 FILE      list of hap1/paternal read names []
    -4 FILE      list of hap2/maternal read names []
    -c INT       lower bound of the binned k-mer's frequency [2]
    -d INT       upper bound of the binned k-mer's frequency [5]
    --t-occ      INT
                 forcedly remove unitigs with &gt;INT unexpected haplotype-specific reads;
                 ignore graph topology; [60]
    --trio-dual  utilize homology information to correct trio phasing errors
  Purge-dups:
    -l INT       purge level. 0: no purging; 1: light; 2/3: aggressive [0 for trio; 3 for unzip]
    -s FLOAT     similarity threshold for duplicate haplotigs in read-level [0.75 for -l1/-l2, 0.55 for -l3]
    -O INT       min number of overlapped reads for duplicate haplotigs [1]
    --purge-max  INT
                 coverage upper bound of Purge-dups [auto]
    --n-hap      INT
                 number of haplotypes [2]
  Hi-C-partition:
    --h1 FILEs   file names of Hi-C R1  [r1_1.fq,r1_2.fq,...]
    --h2 FILEs   file names of Hi-C R2  [r2_1.fq,r2_2.fq,...]
    --seed INT   RNG seed [11]
    --s-base     FLOAT
                 similarity threshold for homology detection in base-level;
                 -1 to disable [0.5]; -s for read-level (see &lt;Purge-dups&gt;)
    --n-weight   INT
                 rounds of reweighting Hi-C links [3]
    --n-perturb  INT
                 rounds of perturbation [10000]
    --f-perturb  FLOAT
                 fraction to flip for perturbation [0.1]
    --l-msjoin   INT
                 detect misjoined unitigs of &gt;=INT in size; 0 to disable [500000]
  Ultra-Long-integration:
    --ul FILEs   file names of Ultra-Long reads [r1.fq,r2.fq,...]
    --ul-rate    FLOAT
                 error rate of Ultra-Long reads [0.2]
    --ul-tip     INT
                 remove tip unitigs composed of &lt;=INT reads for the UL assembly [6]
    --path-max   FLOAT
                 max path drop ratio [0.6]; higher number may make the assembly cleaner
                 but may lead to more misassemblies
    --path-min   FLOAT
                 min path drop ratio [0.2]; higher number may make the assembly cleaner
                 but may lead to more misassemblies
    --ul-cut     INT
                 filter out &lt;INT UL reads during the UL assembly [0]
  Dual-Scaffolding:
    --dual-scaf  output scaffolding
    --scaf-gap   INT
                 max gap size for scaffolding [3000000]
  Telomere-identification:
    --telo-m     STR
                 telomere motif at 5'-end; CCCTAA for human [NULL]
    --telo-p     INT
                 non-telomeric penalty [1]
    --telo-d     INT
                 max drop [2000]
    --telo-s     INT
                 min score for telomere reads [500]
Example: ./hifiasm -o NA12878.asm -t 32 NA12878.fq.gz
See `https://hifiasm.readthedocs.io/en/latest/' or `man ./hifiasm.1' for complete documentation.</code></pre>

<p class="blog-post_paragraph">hifiasm doesn't require you to provide the estimated genome size. All it requires is the sequencing reads for assembly. The <code>--primary</code> flag tells hifiasm to output both a primary and an alternate assembly, with emphasis on generating a more contiguous primary assembly containing homozygous and some heterzygous regions. The <code>-f0</code> flag is used to disable the initial bloom filter which takes 16 GB of memory at the beginning. The bloom filter is used to filter out singleton <em>k</em>-mers when counting all <em>k</em>-mers. It takes <math xmlns="http://www.w3.org/1998/Math/MathML">
    <msup>
      <mn>2</mn>
      <mrow>
        <mo>(</mo>
        <mtext>INT</mtext>
        <mo>-</mo>
        <mn>3</mn>
        <mo>)</mo>
      </mrow>
    </msup>
  </math> bytes of memory. Using a proper value for <code>-f</code> saves memory. <code>-f37</code> is recommended for human assemblies, and for genomes much larger than human, applying <code>-f38</code> or even <code>-f39</code> is preferred to save memory on <em>k</em>-mer counting. However for small genomes it's often recommended to disable this with <code>-f0</code>.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd /home/user19/drUrtUren1/hifiasm
ln -s /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.600.fasta
hifiasm -f0 --primary -t 1 -o drUrtUren1.600.hifiasm drUrtUren1.600.fasta</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">hifiasm -f0 --primary -t 1 -o drUrtUren1.600.hifiasm drUrtUren1.600.fasta
[M::ha_analyze_count] lowest: count[5] = 20459
[M::ha_analyze_count] highest: count[6] = 21871
[M::ha_hist_line]     1: ****************************************************************************************************> 171436
[M::ha_hist_line]     2: ****************************************************************************************************> 36749
[M::ha_hist_line]     3: ************************************************************************************************** 21346
[M::ha_hist_line]     4: ********************************************************************************* 17656
[M::ha_hist_line]     5: ********************************************************************************************** 20459
[M::ha_hist_line]     6: **************************************************************************************************** 21871
[M::ha_hist_line]     7: *************************************************************************************************** 21673
[M::ha_hist_line]     8: ******************************************************** 12224
[M::ha_hist_line]     9: ************************************************************ 13061
[M::ha_hist_line]    10: ************************************************* 10743
[M::ha_hist_line]    11: ***************************** 6268
[M::ha_hist_line]    12: ************************ 5209
[M::ha_hist_line]    13: ******************* 4258
[M::ha_hist_line]    14: ************** 2998
[M::ha_hist_line]    15: *********** 2379
[M::ha_hist_line]    16: *** 657
[M::ha_hist_line]    17: **** 798
[M::ha_hist_line]    18: *** 603
[M::ha_hist_line]    19: * 235
[M::ha_hist_line]    20: * 228
[M::ha_hist_line]    21: * 317
[M::ha_hist_line]    22: ** 533
[M::ha_hist_line]    23: * 225
[M::ha_hist_line]    24: * 135
[M::ha_hist_line]    25: * 227
[M::ha_hist_line]    26: * 227
[M::ha_hist_line]    27: ** 355
[M::ha_hist_line]    28: ** 412
[M::ha_hist_line]    29: * 325
[M::ha_hist_line]  rest: ***************************************************************************************** 19536
[M::ha_analyze_count] left: none
[M::ha_analyze_count] right: count[9] = 13061
[M::ha_ft_gen] peak_hom: 9; peak_het: 6
[M::ha_ct_shrink::0.194*1.01] ==> counted 16411 distinct minimizer k-mers
[M::ha_ft_gen::0.197*1.01@0.056GB] ==> filtered out 16411 k-mers occurring 45 or more times
[M::ha_opt_update_cov] updated max_n_chain to 100
[M::yak_count] collected 208119 minimizers
[M::ha_pt_gen::0.442*1.00] ==> counted 18857 distinct minimizer k-mers
[M::ha_pt_gen] count[4095] = 0 (for sanity check)
[M::ha_analyze_count] lowest: count[5] = 758
[M::ha_analyze_count] highest: count[6] = 854
[M::ha_hist_line]     1: ****************************************************************************************************> 9366
[M::ha_hist_line]     2: ****************************************************************************************************> 1513
[M::ha_hist_line]     3: ************************************************************************************************** 838
[M::ha_hist_line]     4: ********************************************************************************** 700
[M::ha_hist_line]     5: ***************************************************************************************** 758
[M::ha_hist_line]     6: **************************************************************************************************** 854
[M::ha_hist_line]     7: ************************************************************************************************** 836
[M::ha_hist_line]     8: *********************************************************** 501
[M::ha_hist_line]     9: ********************************************************* 489
[M::ha_hist_line]    10: ************************************************* 415
[M::ha_hist_line]    11: *************************** 230
[M::ha_hist_line]    12: ************************ 206
[M::ha_hist_line]    13: ******************* 158
[M::ha_hist_line]    14: ************* 113
[M::ha_hist_line]    15: *********** 91
[M::ha_hist_line]    16: *** 27
[M::ha_hist_line]    17: *** 28
[M::ha_hist_line]    18: ** 20
[M::ha_hist_line]    19: ** 13
[M::ha_hist_line]    20: * 10
[M::ha_hist_line]    21: ** 13
[M::ha_hist_line]    22: **** 30
[M::ha_hist_line]    23: ** 17
[M::ha_hist_line]    24: * 11
[M::ha_hist_line]    25: ** 21
[M::ha_hist_line]    26: **** 33
[M::ha_hist_line]    27: *** 26
[M::ha_hist_line]    28: **** 34
[M::ha_hist_line]    29: *** 26
[M::ha_hist_line]    30: ** 16
[M::ha_hist_line]    31: ** 21
[M::ha_hist_line]    32: * 12
[M::ha_hist_line]    33: ** 21
[M::ha_hist_line]    34: *** 27
[M::ha_hist_line]    35: *** 26
[M::ha_hist_line]    36: * 7
[M::ha_hist_line]    37: *** 29
[M::ha_hist_line]    38: **** 37
[M::ha_hist_line]    39: **** 30
[M::ha_hist_line]    40: ** 19
[M::ha_hist_line]    41: ** 18
[M::ha_hist_line]    42: * 11
[M::ha_hist_line]    43: ** 19
[M::ha_hist_line]    44: **** 32
[M::ha_hist_line]    45: * 8
[M::ha_hist_line]    46: * 5
[M::ha_hist_line]    47: * 12
[M::ha_hist_line]    48: ** 13
[M::ha_hist_line]    49: * 9
[M::ha_hist_line]    50: * 12
[M::ha_hist_line]    51: * 12
[M::ha_hist_line]    52: ** 13
[M::ha_hist_line]    53: ** 15
[M::ha_hist_line]    54: * 10
[M::ha_hist_line]    55: ** 15
[M::ha_hist_line]    56: * 9
[M::ha_hist_line]    57: * 12
[M::ha_hist_line]    58: * 12
[M::ha_hist_line]    59: ** 19
[M::ha_hist_line]    60: ** 14
[M::ha_hist_line]    61: * 7
[M::ha_hist_line]    62: * 8
[M::ha_hist_line]    63: * 6
[M::ha_hist_line]    64: * 11
[M::ha_hist_line]    65: ** 16
[M::ha_hist_line]    66: ** 13
[M::ha_hist_line]    67: ** 20
[M::ha_hist_line]    68: * 10
[M::ha_hist_line]    69: ** 16
[M::ha_hist_line]  rest: **************************************************************************************************** 858
[M::ha_analyze_count] left: none
[M::ha_analyze_count] right: none
[M::ha_pt_gen] peak_hom: 6; peak_het: -1
[M::ha_ct_shrink::0.446*1.00] ==> counted 9491 distinct minimizer k-mers
[M::ha_pt_gen::] counting in normal mode
[M::yak_count] collected 208119 minimizers
[M::ha_pt_gen::0.689*1.00] ==> indexed 198753 positions, counted 9491 distinct minimizer k-mers
[M::pec::16.063] # bases: 5420793; # corrected bases: 8605
[M::ha_assemble::16.773*1.00@0.073GB] ==> corrected reads for round 1
[M::ha_assemble] # bases: 5420793; # corrected bases: 8605
[M::yak_count] collected 208106 minimizers
[M::ha_pt_gen::17.008*1.00] ==> counted 14556 distinct minimizer k-mers
[M::ha_pt_gen] count[4095] = 0 (for sanity check)
[M::ha_analyze_count] lowest: count[5] = 726
[M::ha_analyze_count] highest: count[6] = 873
[M::ha_hist_line]     1: ****************************************************************************************************> 5211
[M::ha_hist_line]     2: ****************************************************************************************************> 1379
[M::ha_hist_line]     3: ********************************************************************************************* 810
[M::ha_hist_line]     4: ************************************************************************** 646
[M::ha_hist_line]     5: *********************************************************************************** 726
[M::ha_hist_line]     6: **************************************************************************************************** 873
[M::ha_hist_line]     7: ************************************************************************************************* 848
[M::ha_hist_line]     8: ********************************************************* 494
[M::ha_hist_line]     9: ********************************************************** 504
[M::ha_hist_line]    10: *********************************************** 408
[M::ha_hist_line]    11: ***************************** 255
[M::ha_hist_line]    12: *********************** 201
[M::ha_hist_line]    13: ******************** 173
[M::ha_hist_line]    14: *************** 132
[M::ha_hist_line]    15: ************ 103
[M::ha_hist_line]    16: *** 24
[M::ha_hist_line]    17: *** 30
[M::ha_hist_line]    18: ** 14
[M::ha_hist_line]    19: * 12
[M::ha_hist_line]    20: ** 14
[M::ha_hist_line]    21: * 11
[M::ha_hist_line]    22: *** 24
[M::ha_hist_line]    23: * 13
[M::ha_hist_line]    24: * 10
[M::ha_hist_line]    25: *** 30
[M::ha_hist_line]    26: *** 23
[M::ha_hist_line]    27: ** 14
[M::ha_hist_line]    28: *** 25
[M::ha_hist_line]    29: **** 34
[M::ha_hist_line]    30: ***** 40
[M::ha_hist_line]    31: ** 21
[M::ha_hist_line]    32: *** 22
[M::ha_hist_line]    33: ** 17
[M::ha_hist_line]    34: **** 33
[M::ha_hist_line]    35: *** 24
[M::ha_hist_line]    36: ** 17
[M::ha_hist_line]    37: * 12
[M::ha_hist_line]    38: ** 19
[M::ha_hist_line]    39: ***** 43
[M::ha_hist_line]    40: ** 18
[M::ha_hist_line]    41: ** 18
[M::ha_hist_line]    42: ** 18
[M::ha_hist_line]    43: ** 14
[M::ha_hist_line]    44: ** 15
[M::ha_hist_line]    45: ** 18
[M::ha_hist_line]    46: ** 21
[M::ha_hist_line]    47: ** 20
[M::ha_hist_line]    48: ** 18
[M::ha_hist_line]    49: * 11
[M::ha_hist_line]    50: ** 14
[M::ha_hist_line]    51: * 8
[M::ha_hist_line]    52: ** 15
[M::ha_hist_line]    53: * 11
[M::ha_hist_line]    54: * 10
[M::ha_hist_line]    55: * 11
[M::ha_hist_line]    56: ** 16
[M::ha_hist_line]    57: ** 14
[M::ha_hist_line]    58: * 7
[M::ha_hist_line]    59: * 12
[M::ha_hist_line]    60: * 8
[M::ha_hist_line]    61: * 5
[M::ha_hist_line]    62: *** 25
[M::ha_hist_line]    63: * 10
[M::ha_hist_line]    64: * 8
[M::ha_hist_line]    65: *** 23
[M::ha_hist_line]    66: * 9
[M::ha_hist_line]    67: * 9
[M::ha_hist_line]    68: * 13
[M::ha_hist_line]    69: ** 16
[M::ha_hist_line]    70: * 8
[M::ha_hist_line]    71: * 8
[M::ha_hist_line]    72: * 11
[M::ha_hist_line]    73: * 8
[M::ha_hist_line]    74: * 10
[M::ha_hist_line]    75: ** 16
[M::ha_hist_line]    76: * 6
[M::ha_hist_line]    77: * 8
[M::ha_hist_line]    78: * 6
[M::ha_hist_line]    79: * 6
[M::ha_hist_line]    80: * 7
[M::ha_hist_line]    81: * 6
[M::ha_hist_line]    82: * 8
[M::ha_hist_line]    83: * 11
[M::ha_hist_line]    84: * 10
[M::ha_hist_line]    85: * 8
[M::ha_hist_line]    86: * 10
[M::ha_hist_line]    87: * 12
[M::ha_hist_line]    88: ** 14
[M::ha_hist_line]    89: * 10
[M::ha_hist_line]    90: * 7
[M::ha_hist_line]  rest: **************************************************************************** 662
[M::ha_analyze_count] left: none
[M::ha_analyze_count] right: none
[M::ha_pt_gen] peak_hom: 6; peak_het: -1
[M::ha_ct_shrink::17.013*1.00] ==> counted 9345 distinct minimizer k-mers
[M::ha_pt_gen::] counting in normal mode
[M::yak_count] collected 208106 minimizers
[M::ha_pt_gen::17.256*1.00] ==> indexed 202895 positions, counted 9345 distinct minimizer k-mers
[M::pec::11.221] # bases: 5419350; # corrected bases: 424
[M::ha_assemble::28.490*1.00@0.081GB] ==> corrected reads for round 2
[M::ha_assemble] # bases: 5419350; # corrected bases: 424
[M::yak_count] collected 207984 minimizers
[M::ha_pt_gen::28.722*1.00] ==> counted 14382 distinct minimizer k-mers
[M::ha_pt_gen] count[4095] = 0 (for sanity check)
[M::ha_analyze_count] lowest: count[5] = 726
[M::ha_analyze_count] highest: count[6] = 872
[M::ha_hist_line]     1: ****************************************************************************************************> 5035
[M::ha_hist_line]     2: ****************************************************************************************************> 1380
[M::ha_hist_line]     3: ********************************************************************************************* 809
[M::ha_hist_line]     4: ************************************************************************** 648
[M::ha_hist_line]     5: *********************************************************************************** 726
[M::ha_hist_line]     6: **************************************************************************************************** 872
[M::ha_hist_line]     7: ************************************************************************************************ 841
[M::ha_hist_line]     8: ********************************************************* 500
[M::ha_hist_line]     9: ********************************************************** 506
[M::ha_hist_line]    10: *********************************************** 408
[M::ha_hist_line]    11: ***************************** 250
[M::ha_hist_line]    12: ************************ 206
[M::ha_hist_line]    13: ******************** 172
[M::ha_hist_line]    14: *************** 133
[M::ha_hist_line]    15: ************ 102
[M::ha_hist_line]    16: *** 24
[M::ha_hist_line]    17: **** 31
[M::ha_hist_line]    18: ** 14
[M::ha_hist_line]    19: * 12
[M::ha_hist_line]    20: ** 14
[M::ha_hist_line]    21: * 12
[M::ha_hist_line]    22: *** 29
[M::ha_hist_line]    23: *** 22
[M::ha_hist_line]    24: * 11
[M::ha_hist_line]    25: ** 15
[M::ha_hist_line]    26: **** 32
[M::ha_hist_line]    27: ***** 40
[M::ha_hist_line]    28: *** 28
[M::ha_hist_line]    29: *** 22
[M::ha_hist_line]    30: ** 21
[M::ha_hist_line]    31: ** 15
[M::ha_hist_line]    32: ** 16
[M::ha_hist_line]    33: * 12
[M::ha_hist_line]    34: ** 18
[M::ha_hist_line]    35: **** 35
[M::ha_hist_line]    36: ** 19
[M::ha_hist_line]    37: *** 22
[M::ha_hist_line]    38: **** 33
[M::ha_hist_line]    39: ***** 47
[M::ha_hist_line]    40: ** 18
[M::ha_hist_line]    41: * 5
[M::ha_hist_line]    42: * 12
[M::ha_hist_line]    43: ** 19
[M::ha_hist_line]    44: * 12
[M::ha_hist_line]    45: * 12
[M::ha_hist_line]    46: * 13
[M::ha_hist_line]    47: * 9
[M::ha_hist_line]    48: *** 22
[M::ha_hist_line]    49: * 11
[M::ha_hist_line]    50: ** 17
[M::ha_hist_line]    51: * 12
[M::ha_hist_line]    52: * 11
[M::ha_hist_line]    53: * 7
[M::ha_hist_line]    54: * 7
[M::ha_hist_line]    55: * 8
[M::ha_hist_line]    56: ** 15
[M::ha_hist_line]    57: ** 16
[M::ha_hist_line]  rest: ****************************************************************************************************> 1024
[M::ha_analyze_count] left: none
[M::ha_analyze_count] right: none
[M::ha_pt_gen] peak_hom: 6; peak_het: -1
[M::ha_ct_shrink::28.725*1.00] ==> counted 9347 distinct minimizer k-mers
[M::ha_pt_gen::] counting in normal mode
[M::yak_count] collected 207984 minimizers
[M::ha_pt_gen::28.968*1.00] ==> indexed 202949 positions, counted 9347 distinct minimizer k-mers
[M::pec::8.111] # bases: 5419340; # corrected bases: 76
[M::ha_assemble::37.186*1.00@0.081GB] ==> corrected reads for round 3
[M::ha_assemble] # bases: 5419340; # corrected bases: 76
[M::yak_count] collected 208001 minimizers
[M::ha_pt_gen::37.423*1.00] ==> counted 14396 distinct minimizer k-mers
[M::ha_pt_gen] count[4095] = 0 (for sanity check)
[M::ha_analyze_count] lowest: count[5] = 726
[M::ha_analyze_count] highest: count[6] = 872
[M::ha_hist_line]     1: ****************************************************************************************************> 5050
[M::ha_hist_line]     2: ****************************************************************************************************> 1379
[M::ha_hist_line]     3: ********************************************************************************************* 809
[M::ha_hist_line]     4: ************************************************************************** 648
[M::ha_hist_line]     5: *********************************************************************************** 726
[M::ha_hist_line]     6: **************************************************************************************************** 872
[M::ha_hist_line]     7: ************************************************************************************************ 841
[M::ha_hist_line]     8: ********************************************************* 499
[M::ha_hist_line]     9: ********************************************************** 504
[M::ha_hist_line]    10: *********************************************** 411
[M::ha_hist_line]    11: ***************************** 250
[M::ha_hist_line]    12: ************************ 206
[M::ha_hist_line]    13: ******************** 172
[M::ha_hist_line]    14: *************** 133
[M::ha_hist_line]    15: ************ 102
[M::ha_hist_line]    16: *** 24
[M::ha_hist_line]    17: **** 31
[M::ha_hist_line]    18: ** 14
[M::ha_hist_line]    19: * 12
[M::ha_hist_line]    20: ** 14
[M::ha_hist_line]    21: * 12
[M::ha_hist_line]    22: *** 29
[M::ha_hist_line]    23: *** 22
[M::ha_hist_line]    24: * 11
[M::ha_hist_line]    25: ** 15
[M::ha_hist_line]    26: **** 32
[M::ha_hist_line]    27: ***** 40
[M::ha_hist_line]    28: *** 28
[M::ha_hist_line]    29: *** 22
[M::ha_hist_line]    30: ** 21
[M::ha_hist_line]    31: ** 15
[M::ha_hist_line]    32: ** 16
[M::ha_hist_line]    33: * 12
[M::ha_hist_line]    34: ** 18
[M::ha_hist_line]    35: **** 35
[M::ha_hist_line]    36: ** 19
[M::ha_hist_line]    37: *** 22
[M::ha_hist_line]    38: **** 33
[M::ha_hist_line]    39: ***** 47
[M::ha_hist_line]    40: ** 18
[M::ha_hist_line]    41: * 5
[M::ha_hist_line]    42: * 12
[M::ha_hist_line]    43: ** 19
[M::ha_hist_line]    44: * 12
[M::ha_hist_line]    45: * 12
[M::ha_hist_line]    46: * 13
[M::ha_hist_line]    47: * 9
[M::ha_hist_line]    48: *** 22
[M::ha_hist_line]    49: * 11
[M::ha_hist_line]    50: ** 17
[M::ha_hist_line]    51: * 12
[M::ha_hist_line]    52: * 11
[M::ha_hist_line]    53: * 7
[M::ha_hist_line]    54: * 7
[M::ha_hist_line]    55: * 8
[M::ha_hist_line]    56: ** 15
[M::ha_hist_line]    57: ** 16
[M::ha_hist_line]  rest: ****************************************************************************************************> 1024
[M::ha_analyze_count] left: none
[M::ha_analyze_count] right: none
[M::ha_pt_gen] peak_hom: 6; peak_het: -1
[M::ha_ct_shrink::37.426*1.00] ==> counted 9346 distinct minimizer k-mers
[M::ha_pt_gen::] counting in normal mode
[M::yak_count] collected 208001 minimizers
[M::ha_pt_gen::37.669*1.00] ==> indexed 202951 positions, counted 9346 distinct minimizer k-mers
[M::ha_print_ovlp_stat_0] # overlaps: 75775
[M::ha_print_ovlp_stat_0] # strong overlaps: 1403
[M::ha_print_ovlp_stat_0] # weak overlaps: 74372
[M::ha_print_ovlp_stat_0] # exact overlaps: 74637
[M::ha_print_ovlp_stat_0] # inexact overlaps: 1138
[M::ha_print_ovlp_stat_0] # overlaps without large indels: 75767
[M::ha_print_ovlp_stat_0] # reverse overlaps: 1196
[M::ha_print_ovlp_stat_0] # running time: 6.624
[M::ha_assemble::44.294*1.00@0.081GB] ==> found overlaps for the final round
[M::ha_opt_update_cov_min] updated max_n_chain to 100
Writing reads to disk...
Reads has been written.
Writing ma_hit_ts to disk...
ma_hit_ts has been written.
Writing ma_hit_ts to disk...
ma_hit_ts has been written.
bin files have been written.
[M::purge_dups] homozygous read coverage threshold: 5
[M::purge_dups] purge duplication coverage threshold: 7
Writing raw unitig GFA to disk...
Writing processed unitig GFA to disk...
[M::purge_dups] homozygous read coverage threshold: 5
[M::purge_dups] purge duplication coverage threshold: 7
[M::mc_solve:: # edges: 0]
[M::mc_solve_core_adv::0.000] ==> Partition
[M::adjust_utg_by_primary] primary contig coverage range: [4, infinity]
Writing primary contig GFA to disk...
Writing alternate contig GFA to disk...
Inconsistency threshold for low-quality regions in BED files: 70%
[M::main] Version: 0.20.0-r639
[M::main] CMD: hifiasm -f0 --primary -t 1 -o drUrtUren1.600.hifiasm drUrtUren1.600.fasta
[M::main] Real time: 44.388 sec; CPU: 44.388 sec; Peak RSS: 0.081 GB</code></pre>

<p class="blog-post_paragraph">Now that hifiasm is done running, we can check the outputs.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd ~/drUrtUren1/hifiasm
ls -ltrh</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">ls -ltrh
total 5.4M
lrwxrwxrwx 1 user19 user19   55 Nov  6 13:51 drUrtUren1.600.fasta -> /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.600.fasta
-rw-rw-r-- 1 user19 user19 1.4M Nov  6 14:01 drUrtUren1.600.hifiasm.ec.bin
-rw-rw-r-- 1 user19 user19 3.1M Nov  6 14:01 drUrtUren1.600.hifiasm.ovlp.source.bin
-rw-rw-r-- 1 user19 user19  53K Nov  6 14:01 drUrtUren1.600.hifiasm.ovlp.reverse.bin
-rw-rw-r-- 1 user19 user19 7.0K Nov  6 14:01 drUrtUren1.600.hifiasm.r_utg.noseq.gfa
-rw-rw-r-- 1 user19 user19 322K Nov  6 14:01 drUrtUren1.600.hifiasm.r_utg.gfa
-rw-rw-r-- 1 user19 user19 1.7K Nov  6 14:01 drUrtUren1.600.hifiasm.r_utg.lowQ.bed
-rw-rw-r-- 1 user19 user19 7.0K Nov  6 14:01 drUrtUren1.600.hifiasm.p_utg.noseq.gfa
-rw-rw-r-- 1 user19 user19 322K Nov  6 14:01 drUrtUren1.600.hifiasm.p_utg.gfa
-rw-rw-r-- 1 user19 user19 1.7K Nov  6 14:01 drUrtUren1.600.hifiasm.p_utg.lowQ.bed
-rw-rw-r-- 1 user19 user19 6.8K Nov  6 14:01 drUrtUren1.600.hifiasm.p_ctg.noseq.gfa
-rw-rw-r-- 1 user19 user19 309K Nov  6 14:01 drUrtUren1.600.hifiasm.p_ctg.gfa
-rw-rw-r-- 1 user19 user19 1.7K Nov  6 14:01 drUrtUren1.600.hifiasm.p_ctg.lowQ.bed
-rw-rw-r-- 1 user19 user19    0 Nov  6 14:01 drUrtUren1.600.hifiasm.a_ctg.noseq.gfa
-rw-rw-r-- 1 user19 user19    0 Nov  6 14:01 drUrtUren1.600.hifiasm.a_ctg.lowQ.bed
-rw-rw-r-- 1 user19 user19    0 Nov  6 14:01 drUrtUren1.600.hifiasm.a_ctg.gfa</code></pre>

<p class="blog-post_paragraph">hifiasm with the <code>--primary</code> flag outputs two assemblies: a primary and an alternate assembly. This is denoted in the output files with the <code>.p_*</code> and <code>.a_*</code> symbols. Also, hifiasm doesn't output any <code>.fasta</code> files. Assemblies need to be converted to <code>.fasta</code> format from the <code>.gfa</code> files that hifiasm outputs (see <a href="https://hifiasm.readthedocs.io/en/latest/faq.html#id1" target="_blank">here</a> for details). After converting the assemblies to <code>.fasta</code> format, I investigate the assembly statistics. I investigated the statistics of the provided whole hifiasm assemblies since the 600 read subset was to save computing resources while practicing using the assembly commands.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">export PATH=/home/ubuntu/Share/scripts/:$PATH
gfa2fa drUrtUren1.600.hifiasm.p_ctg.gfa > drUrtUren1.600.hifiasm.p_ctg.fa
gfa2fa drUrtUren1.600.hifiasm.a_ctg.gfa > drUrtUren1.600.hifiasm.a_ctg.fa
asmstats drUrtUren1.hifiasm.total.p_ctg.fa.gz > drUrtUren1.hifiasm.total.p_ctg.fa.stats
cat drUrtUren1.hifiasm.total.p_ctg.fa.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1.hifiasm.total.p_ctg.fa.gz
A = 141035727 (32.2%), C = 77702296 (17.8%), G = 77615519 (17.7%), T = 141089788 (32.3%), CpG = 29344546 (6.7%)
sum = 437443330, n = 2590, mean = 168897.038610039, largest = 39522977, smallest = 2268
N50 = 17272222, L50 = 9
N60 = 12262420, L60 = 12
N70 = 3097529, L70 = 17
N80 = 232352, L80 = 105
N90 = 41878, L90 = 578
N100 = 2268, L100 = 2590</code></pre>

<p class="blog-post_paragraph">Here you can see that hifiasm assembled 437 Mb across 2590 contigs. The assembly has an N50 of 17272222 bp. <strong>N50</strong> is a genome assembly statistic that measures the contiguity of an assembly. The N50 is defined as <u>the sequence length of the shortest contig at 50% of the total assembly length</u>. It can be thought of as the point of half of the mass of the distribution. The number of bases from all contigs longer than the N50 will be close to the number of bases from all contigs shorter than the N50.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_32.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">I also did the same inspection on the alternate assembly generated by hifiasm.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">asmstats drUrtUren1.hifiasm.total.a_ctg.fa.gz > drUrtUren1.hifiasm.total.a_ctg.fa.stats
cat drUrtUren1.hifiasm.total.a_ctg.fa.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1.hifiasm.total.a_ctg.fa.gz
A = 117948605 (31.8%), C = 67290993 (18.2%), G = 67280909 (18.2%), T = 118000312 (31.8%), CpG = 25346240 (6.8%)
sum = 370520819, n = 11768, mean = 31485.4536879674, largest = 1565515, smallest = 2984
N50 = 34728, L50 = 2551
N60 = 28033, L60 = 3744
N70 = 23401, L70 = 5200
N80 = 20345, L80 = 6894
N90 = 16605, L90 = 8887
N100 = 2984, L100 = 11768</code></pre>

<p class="blog-post_paragraph">Here you can see that the alternate assembly that was generated was 15% smaller (with 370520819 total bases instead of 437443330), and also much more fragmented (containing 11768 contigs instead of 2590). This fragmentation is reflected in the N50 value, which is 34728 bp instead of 17272222 bp. This is expected because the alternate hifiasm assembly primarilily contains the contigs corresponding to the heterozygous regions of the genome.</p>

<p class="blog-post_paragraph">Now I did the same thing for the HiCanu Assembly.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cd ~/drUrtUren1/hicanu/run1
asmstats drUrtUren1.hicanu.total.contigs.fasta.gz > drUrtUren1.hicanu.total.contigs.fasta.stats
cat drUrtUren1.hicanu.total.contigs.fasta.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1.hicanu.total.contigs.fasta.gz
A = 225253791 (32.2%), C = 124732503 (17.8%), G = 124684505 (17.8%), T = 225158632 (32.2%), CpG = 47805276 (6.8%)
sum = 699829431, n = 13242, mean = 52849.2245129135, largest = 17167494, smallest = 4224
N50 = 362274, L50 = 117
N60 = 82996, L60 = 579
N70 = 34337, L70 = 1998
N80 = 23001, L80 = 4566
N90 = 18042, L90 = 7994
N100 = 4224, L100 = 13242</code></pre>

<p class="blog-post_paragraph">Here you can see that the assembly generated was much bigger than that of hifiasm, with almost 700 Mb being found in the assembly. However it is not the case that this assembly is more contiguous. Remember that HiCanu outputs all contigs, even those containing different haplotypes, into the same assembly output file. This means that there will likely be a lot of duplicated sequences assemblies generated by HiCanu. This can be very dangerous if you are not aware of this behaviour. You might think you've found evidence of a gene-family duplication in your organisms, while in reality this is just an artifact of the assembly process (see <a href="https://doi.org/10.1186/s13059-022-02764-1" target="_blank">here</a> for a review detailing this problem).</p>

<p class="blog-post_paragraph">
    <u><em>Step 3: Genome Assembly Quality Control</em></u>
  </p>

<p class="blog-post_paragraph">Now that we have assemblies generated for these organisms, the next question is how do we start verifying whether these assemblies are of high quality. An organism's genomes will often contain hundreds and thousands of A's, C's, G's, and T's seemingly randomly scrambled together in one sequence. Trying to find out whether these sequences are arranged in the correct order or if there are any sequences missing is a challenge that involves clever algorithms and use of existing data in biological databases like the NCBI. A few common tactics for genome assembly quality checking are:</p>

<ol>
  <li>
    <p class="blog-post_paragraph"><strong>Checking for Base Accuracy</strong></p>
  </li>
  <ul>
    <li>This can be done by <strong>realigning reads from the same species' <em>k</em>-mers </strong>. This involves aligning sequencing reads based on <em>k</em>-mers that are specific to the target species. The goal is to minimize errors and ensure that the assembly accurately represents the species' genomic sequence. This step can help identify misassemblies or regions of low accuracy in the genome.</li>
    <li>Another method is by checking for <strong><em>k</em>-mer Completeness with Respect to Illumina or HiFi Data</strong>. This involves comparing the <em>k</em>-mer distribution from the assembled genome against the <em>k</em>-mers obtained from high-quality sequencing data (e.g., Illumina or HiFi reads). This comparison helps determine how well the assembly captures the genetic diversity and completeness of the genome, identifying regions that may be missing or poorly represented.</li>
  </ul>
  <p></p>
  <li>
    <p class="blog-post_paragraph"><strong>Assessing Local Structure Accuracy</strong></p>
  </li>
  <ul>
    <li>This can be done by using <strong>external evidence</strong>, which refers to incorporating additional sequencing data (such as long-read or mate-pair data) to validate the structure of the assembly. External evidence can help correct misalignments and fill gaps, enhancing the overall accuracy of the assembled genome.</li>
    <li>It can also be done by using <strong>Known Adjacencies</strong>, which involves utilizing information from known genomic features, such as PCR primer locations or previously identified structural variants, to improve local structure accuracy. This approach helps verify the order and orientation of genes and other genomic elements, ensuring that the assembly reflects the expected genomic architecture.</li>
  </ul>
  <p></p>
  <li>
    <p class="blog-post_paragraph"><strong>Analyzing Gene Content and Order and Orientation of Genes/Exons of Known Genes</strong></p>
  </li>
  <ul>
    <li>Analyzing the arrangement of genes and exons within the assembled genome is crucial for functional annotation.</li>
    <li>Comparing the gene order and orientation against a reference or known gene set (e.g., housekeeping genes) can reveal structural integrity and evolutionary conservation. This information is vital for understanding gene regulation and potential gene interactions.</li>
  </ul>
  <p></p>
  <li>
    <p class="blog-post_paragraph"><strong>Quantitative Assessment of Genome Assembly and Annotation Completeness</strong></p>
  </li>
  <ul>
    <li>The gold standard for performing a quantitative assessment on genome annotation has for the longest time been <a href="https://doi.org/10.1002/cpz1.323" target="_blank">BUSCO</a> (Benchmarking Universal Single-Copy Orthologs). BUSCO provides a standardized method for evaluating genome assemblies based on the presence and completeness of conserved genes across different species.</li>
    <li>By assessing the assembly against a set of expected single-copy orthologs, BUSCO quantifies the completeness of both the genome assembly and its annotation. This tool helps you determine how well the assembly meets evolutionary expectations and informs decisions on further refinements or analyses.</li>
    <li>Basically, it works by starting from a set of conserved genes from a species lineage, and then finding them in the assembly and assessing whether they are complete (<strong><em>C</em></strong>), single copy (<strong><em>S</em></strong>), duplicated (<strong><em>D</em></strong>), fragmented (<strong><em>F</em></strong>), or missing (<strong><em>M</em></strong>).</li>
    <li>This thus means that the accuracy of the BUSCO measurement is dependent on how accurate the data is in the lineage databases.</li>
    <li>Some newer projects are being built that try to improve on the accuracy of BUSCO, one of these being a faster and more accurate re-implementation of BUSCO called <a href="https://doi.org/10.1093/bioinformatics/btad595" target="_blank">compleasm</a>.</li>
  </ul>
</ol>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_33.png" alt="alt text" />
</div>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_34.png" alt="alt text" />
</div>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_35.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">For the sake of the workshop, the initial genome QC was performed by aligning the sequencing reads back to the read assembly. Mapping the reads was done using <a href="https://github.com/lh3/minimap2" target="_blank">minimap2</a>.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">minimap2 -t 8 --secondary=no -ax map-pb drUrtUren1.hicanu.total.contigs.fasta.gz drUrtUren1.ccs.total.fasta.gz -o drUrtUren1.hicanu.total.contigs.sam</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">[M::mm_idx_gen::6.582*1.58] collected minimizers
[M::mm_idx_gen::7.204*2.13] sorted minimizers
[M::main::7.204*2.13] loaded/built the index for 13242 target sequence(s)
[M::mm_mapopt_update::7.490*2.08] mid_occ = 937
[M::mm_idx_stat] kmer size: 19; skip: 10; is_hpc: 1; #seq: 13242
[M::mm_idx_stat::7.651*2.06] distinct minimizers: 22547367 (38.32% are singletons); average occurrences: 3.853; average spacing: 8.055; total length: 699829431
[M::worker_pipeline::247.530*7.42] mapped 60746 sequences
[M::worker_pipeline::501.683*7.54] mapped 60147 sequences
[M::worker_pipeline::796.664*7.54] mapped 59537 sequences
[M::worker_pipeline::1112.521*7.34] mapped 59125 sequences
[M::worker_pipeline::1689.828*6.13] mapped 59073 sequences
[M::worker_pipeline::2188.936*5.71] mapped 59442 sequences
[M::worker_pipeline::2423.438*5.92] mapped 59153 sequences
[M::worker_pipeline::2694.284*6.10] mapped 59024 sequences
[M::worker_pipeline::2995.926*6.24] mapped 58787 sequences
[M::worker_pipeline::3603.017*5.81] mapped 58793 sequences
[M::worker_pipeline::4219.251*5.51] mapped 58466 sequences
[M::worker_pipeline::4769.299*5.46] mapped 58586 sequences
[M::worker_pipeline::5373.323*5.28] mapped 58420 sequences
[M::worker_pipeline::5896.517*5.26] mapped 58527 sequences
[M::worker_pipeline::6578.290*5.06] mapped 58475 sequences
[M::worker_pipeline::7043.986*5.08] mapped 58287 sequences
[M::worker_pipeline::7788.172*4.91] mapped 58187 sequences
[M::worker_pipeline::8331.583*4.89] mapped 58184 sequences
[M::worker_pipeline::8729.472*4.93] mapped 58572 sequences
[M::worker_pipeline::9386.563*4.83] mapped 58552 sequences
[M::worker_pipeline::9793.511*4.86] mapped 58349 sequences
[M::worker_pipeline::10244.690*4.88] mapped 58409 sequences
[M::worker_pipeline::11132.630*4.73] mapped 58192 sequences
[M::worker_pipeline::11404.993*4.79] mapped 58400 sequences
[M::worker_pipeline::11644.991*4.85] mapped 58386 sequences
[M::worker_pipeline::11890.607*4.92] mapped 58334 sequences
[M::worker_pipeline::12139.916*4.98] mapped 58272 sequences
[M::worker_pipeline::12385.983*5.04] mapped 58448 sequences
[M::worker_pipeline::12633.173*5.09] mapped 58428 sequences
[M::worker_pipeline::12881.003*5.15] mapped 58561 sequences
[M::worker_pipeline::13129.121*5.20] mapped 58920 sequences
[M::worker_pipeline::13377.388*5.25] mapped 59004 sequences
[M::worker_pipeline::13626.954*5.30] mapped 59267 sequences
[M::worker_pipeline::13875.655*5.35] mapped 59141 sequences
[M::worker_pipeline::14126.440*5.39] mapped 59409 sequences
[M::worker_pipeline::14380.264*5.44] mapped 59475 sequences
[M::worker_pipeline::14627.925*5.48] mapped 59862 sequences
[M::worker_pipeline::14788.189*5.50] mapped 38375 sequences
[M::main] Version: 2.28-r1209
[M::main] CMD: minimap2 -t 8 --secondary=no -ax map-pb -o drUrtUren1.hicanu.total.contigs.sam drUrtUren1.hicanu.total.contigs.fasta.gz drUrtUren1.ccs.total.fasta.gz
[M::main] Real time: 14788.295 sec; CPU: 81384.543 sec; Peak RSS: 9.547 GB</code></pre>

<p class="blog-post_paragraph">Now that the mapping has finished, I then used a tool called <a href="https://github.com/samtools/samtools" target="_blank">samtools</a> to manipulate the output file. When the mapping has completed, the information is stored inside a <code>.sam</code> file. This <code>.sam</code> file is human readable but it is also very large and takes up a lot of storage space. When working with reads, it's more common to work with <code>.bam</code> files. These files are not human readable (they are binary files and just contain a series of <code>0</code>'s and <code>1</code>'s) but they store the same information as the <code>.sam</code> files and consume much less memory. It's more common to store alignment data as <code>.bam</code> files and use tools like <code>samtools</code> to manipulate and read these files. Another thing that is often done to these alignment files is to <strong>index them</strong>. Often times you want to view your alignment data in a visual software, or use this data for downstream processing. Often times a prerequisite for these other programs to work with these alignment files is to have them indexed as well, which just means storing certain information from that alignment file in a special data structure that the computer likes. These indexes are usually stored in files with the <code>.bai</code> extension. However, <u>before you can index an alignment file, you have to sort it first</u>. The following commands are what I used to do all the aforementioned steps.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">samtools view -h -Sb drUrtUren1.hicanu.total.contigs.sam -o drUrtUren1.hicanu.total.contigs.bam
samtools sort drUrtUren1.hicanu.total.contigs.bam -o drUrtUren1.hicanu.total.contigs.sorted.bam
samtools index drUrtUren1.hicanu.total.contigs.sorted.bam</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">[bam_sort_core] merging from 3 files and 1 in-memory blocks...</code></pre>

<p class="blog-post_paragraph">Now that these reads are mapped, it's a good idea to visualize how they mapped to this genome assembly. Normally, it's a good idea to use visualization techniques where you can visualize how <em>all</em> of the reads map the the <em>entire</em> genome in a single data visualization plot (which will be elaborated on further in this post). However, it's also good to know how to look at certain mappings in higher resolution. This is what was focused on at this part of the workshop. The goal here was to isolate a single contig that was assembled from the HiCanu assembly and then see how the reads that mapped to that contig looked. In order to get stared on this goal, I used the <code>filterfasta.py</code> that was provided on the workshop's folder. In order to get this script to work, I needed to isolate the ID of the contig I wanted to isolate the sequence for and provide that to the folder. I used the following two lines to isolate and inspect the first contig ID from the HiCanu assembly.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">gunzip -c drUrtUren1.hicanu.total.contigs.fasta.gz | grep ">" | head -n 1 | sed 's/>//g' | cut -d ' ' -f 1 > contig-to-isolate.txt
cat contig-to-isolate.txt</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">tig00000002</code></pre>

<p class="blog-post_paragraph">Then I isolated the sequence from that contig using the provided script.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cp ~/Share/scripts/filterfasta.py .
python filterfasta.py -i contig-to-isolate.txt drUrtUren1.hicanu.total.contigs.fasta > tig00000002.fasta</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Will be writing to stdout.
Using list of IDs to filter.
isNeg? FalseFiltering and writing...
13000 records done
1 out of 13242 sequences remained after filtering.</code></pre>

<p class="blog-post_paragraph">This is what the isolated <code>.fasta</code> file looked like.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cat tig00000002.fasta</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">>tig00000002 len=25969 reads=31 class=contig suggestRepeat=no suggestBubble=yes suggestCircular=no trim=0-25969
ACACGTAATTTTAATTTGTCAAAACATAAATTTAACTTGATAAAAATGTAATTTAACTTT
TCAAAACGTAACTTTAACTTATAAAAAACGTACTTTAACTTGTCAAAATGTAATTTTAAT
GTGTCAAAAAGTAACTTGAAGTTGTTAAAACATAACTTTAACTTGTAAAAAACGTACTTC
AACTTGTAAAAAACGTACTTCAACTTGTAAAAAACGTACTTCATCTTGTAAAATAGTTGC
TTTAACTTGTAAAAAACGCACTTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGT
TACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTACTTTAACTTATCAAAA
TGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTAA
AAAACGTACTTTAACTTGTAAAAAACGTATTTTAACCTGTCAAAACGTAACTTTAACTTG
TAAAAAACCTACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAACATACATTAAC
TTGTAAAAAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTAGTAACAAATGAACTTT
AACTTGTAAAAACGTACTTCAACTTGTAAAAAACATACTTAAACTTGTAAAAAACCTATT
TTAACTTGTCAAAACGTAACTATAATTTGTCAAAACAAAAATTTAACTAGTTAAAAACGT
AACTTAACTTGTCAAAACGTAACTTTAACTTGTAAAAAATTTACTTTAACTTGTCAAAAT
GCAATTTTAATGTGTCAAAACGTAACTTGAAGTTGTTAAAACGTAACTTTAACTTGTAAA
AAACGTACTTCAACTTGTAAAAAACGTACTTCAACTTGTAAAAAACGTACTTCAACTTGT
AAAATAGTTGCTTTAACTTGTAAAAAACGCACTTTAACTTGTAAAAAACGTATTTTAACG
TGTCAAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTACTTTA
ACTTATCAAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTCAAAACGTAACT
TTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGTCAAAACGTA
ACTTTAACTTGTAAAAAACCTACTTTAACTTGTAAAAAAACGTAATTTAACTTATCAAAA
TGTAACTTTAACTTGTAAAAAAAAGTAATATAACTTATCAACATGTAAGTTTAACTTGTC
AAAACGTAACTTTAACTTGTTAAAAACGTACTTTAACTTGTAAAAACATATTTTAACTAG
TAAAAAACGTACTTTAACCTGTCAAAACGTAACTTTAACTTGTAAAAAACCTACTTTAAC
TTGTCAAAACGTAAATTTAACTTGTCAAAAAATAAGTTTAACTTGTTAAAAACATACATT
AACTTGTAAAAAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTAGTAACAAATGAAC
TTTAACTTGTAAAAACGTACTTCAACTTATAAAAAACATACTATAACTTGTAAAAAACCT
ATTTTAACTTGTCAAAACGTAACTATAATTTGTCAAAACGAAAATTTAACTTGTTAAAAA
CGTAACTTAACTTGTCAAAACGTAACTTTAACTTATAAAAAACTTACTTTAACTTGTCAA
AATGTAATTTTAATGTGTCAAAACGTAACTTGAAGTTGTTAAAACGTAACTTTAACTTGT
AAAAAACGTACTTCAACTTGTAAAAAACGTACTTCAACTTGTAAAAAACGTACTTCAACT
TGTAAAATAGTCGCTTTAACTTGTAAAAAACGCACTTTAACTTGTAAAAAACGTATTTTA
ACGTGTCAAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTACT
TTAACTTATCAAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTCAAAACGTA
ACTTTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGTCAAAAC
GTAACTTTAACTTGTAAAAAACCTACTTTAACTTGTAAAAAAACGTAATTTAACTTATCA
ATATGTAACTTTAACTTGTAAAAAAAAGTAATTTAACTTATCAAAATGTAAGTTTAACTT
GTCAAAACGTAACTTTAACTTGTTAAAAACGTACTTTAACTTATCAAAATGTAACTTTAA
CTTGTCAAAACGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTAAAAAACGTACTT
TAACTTGTAAAAAACGTACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTA
CTTTAACTTATCAAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTCAAAACG
TAACTTTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTATTTTAACCTGTCAAA
ACGTAACTTTAACTTGTAAAAAACCTACTTTAACTTGTCAAAACGTAAGTTTAACTTGTC
AAAACGTAAGTTTAACTTGTTAAAAACATACATTAACTTGTAAAAAACGTACTTTAACTT
GTCAAAAATGTGGTTTAACTAGTAACAAATGAACTTTAACTTGTAAAAACGTACTTTAAC
TTGTCAAAACGTAACTTTAACTTGTAAAAAACCTACTTTAACTTGTAAAAAAAACTTACT
TTAACTTGTCAAAATGTAATTTTAATGTGTCAAAACGTAACTTGAAGTTGTTAAAACGTA
ACTTTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTACTTCAACTTGTAAAATA
GTTGCTTTAACTTGTTAAAAACATAAATTTAACTTGTAAAAAACCTACTTTAACTTGTAA
AAAAACGTAATTTATTTTATCAAAATGTAACTTTAACTTGTTAAAAACGTACTTTAACTT
GTAAAAAAATATAATTTAACTTATCTAAATGTAACTTTAACTTGTCAAAACGTAACTTTA
ACTTGTTAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTACT
TTAACCTGTCAAAACGTAACTTTAACTTGTAAAAAACCTACTTTAACTTGTCAAAACGTA
ACTTTAACTTGTCAAAACGTAAGTTTAAATTGTTACAAACATACATTAACTTGTAAAAAA
CATACTTTAACCTGTCAAAACATAACTTTAACTTGTAAAAAACCTACTTTAACTTGTCAA
AACGTAAGTTTAACTTGTTAAAAACATACATTAACTTGTAAAAAAAGTACTTTAACTTGT
CAAAAATGTGGTTTAACTAGTAAAAAATGAACTTTCACTTGTAAAAAACATATTTTAACT
TGTCAAAACGTAACTTTAATTTGTCAAAATGAAAATTTAACTTGTTAAAAACGTAATTTA
TTTTGTCAAAAGTAACTTTAACTTATAAAACACTTACTTTAACTTGTCAAAATGTAATTT
TAATGTGTCAAAACGTAACTTGAAGTTGTTAAAATGTAACTTTAACTTGTAAAAAACGTA
CTTCAACTTGTAAAAAACGTACTTCAACTTGTAAATCAGTTGCTTTAATTTGTAAAAAAC
GTATCTTAACTTGTAAAAAACGTATGTTCACGTGTCAAAATGTAACTTTAACTTGTTAAA
AACAGACTTTAACTTGTAAAAAAACGTAATTTAACTTCTCAAAATGTAACTTTAATTTGT
CAAAACGTAACTTTAATTTGTTAAAAACGTACTTTAACGTGTAAAAAACGTACTTTAACT
TGTAAAAAACGTACATTAACCTGTCAAAACATAACTTTAACTTGTAAAAAACCTACTTGA
ACTTGTCAAAACGTAAGTTTAACTTGTCAAAACGTAAGTTTAACTTGTCAAAACGTAAGT
TTAACTTGTTAAAAACATATATTAACTTGTAAAAAACGTACTTTAACTTGTCAAAAATGT
GGTTTAACTAGTAACAAATGAACTTTAACTTGTAAAAACGTACTTCAACTTGTAAAAAAC
ATACTTTAACTTGTAAAAAACCTATTTTCACTTGTCAAAACGTAACTATAATTTGTCAAA
ACGAAAATTTAACTTGTTAAAAACGTAACTTAACTTGTCAAAACGTAACTTTAACTTATA
AAAAACTTACTTTAACTTGTCAAAATGTAATTTTAATCTGTCAAAACGTAACTTGAAGTT
GTTAAAACGTAACTTTAACTTGTAAAAAAAGTACTTCAACTTGTAAAAAACGTACTTCAA
CTTGTAAAACAGTTGCTTTAACTTGTAAAAAACGTACTTTAACTTGTAAAAAACGTATGT
TAACGTGTCAAAATGTAACTTTAACTTGTTAAAACGTAACTTTAACTTGTTAAAACGTAA
CTTTAACTTGTAATAAACGTACTTTTACCTGTCAAAACGTAACTTTAACTTGTAAAAAAC
GTACTTTAACTTGTAAAAAACGTACTTCAACTTGTAAAACAGTTGCTTTGACTTGTTAAA
AACGTAACTTTAACTTGTAAAAAACCTATTTTAACTTGTAAAAAAAAGTAATTAAACTTA
TCAAAATGTAACTTTAACTTGTTAAAAACGTACTTTAACTTGTTAAAAAACGTAATTTGA
CTTATCTAAATGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTACTT
TAACTTGTAAAAAACGTACTGTTGTCGCCGTATTTCCGGCGACAGCGTTAAATGTATACA
TCCACGATTATATCCGAGGCGCGGAGGGGAGACGCCTTCCCCGAGACGACCGAGACCCAC
CTTCGCGCACCTAACCGACCACCGTCGCGCAGGGGAGGGGCCCGGCCGGCCACCCTCTCC
CATGAGAGGGGCCTGGCCGGCCTTCCCCACCCGAGGGAGGGACCTGGCCGGCCTTCCCCA
CCTAGGGGAGGGGCCTGGCCGGCCTTCCCCACCCGAGGGAGGGACCTGGCCGGCCTTCCC
CACGTAGGGGAGGGGCCTGGCCGGCCTTCCTCACCCGGGGGAGGGACCTGGCCGGCCTTC
CCCACCTAGGGGAAGGGCCTGGCCGGCCTTCCTCACCCGGGGGAGGGACCTGGCCGGCCT
TCCCCACCTAGGGGAGGGGCCTGGCCGGCCTTCCTCGCCCGAGGGAGGGACCTGGCCGGC
CTTCCCCACTTAGGGGAGAGACCTGGCCGGCCTTCCCGACCCACTCGGCCACAAGGCTAC
ACCCGTGGCTATAAGTAGCCTTCGATGCTCTCCACGCAGAGGTAAGCTCTAGAGACAAAA
AATAGACACTTTCTCCCTTACTCCCTATATACATATCATACTCCCAAGTGAGGTATTTCG
GATATTTTCTCCCGACCGGAAGAATACTCCCTAAAGATCTGAGATCCGAGCTCTGAGACG
CCAGTGATCCTAGTATCCCCGTCTCGCTCGATCTCGTCGATCCCTCCTCGTCATCCTGCA
AACTCCGCCCACATTATACTAATGTTGAGCGTCGGAGAGGTAAACAGAGGACCCCCCTCT
GTTTTCCCTTCTTAGTTCACGTCTGTTTACTCTTTTCAGATAACAGAAGGACTCGAAGAC
CACACGACAGCTAAGGGACGATCCGGGACACCGGATCCGACACGCTCAGGCCCAAGCGAG
GGTTACCCGCTCCGAGCGAGGATCTGACCGCTTAGGTCTCGAAACGTTGTCCAAATTTCA
GCGACAACATTTTGGCGCTAGAAGGAGGGGAGATTTTTGACCTTAAAATTCGCCCAACCC
ACATTTCTCTATTTTCATCTATTTCCTTATTTTTCATACGAAATGAGTAGCGTTCCTGCG
ACTCATGCGGATGCTTGATCGAGCAACCCTATCGTCCGAGCCGCCGGCTCGCGGTACGAA
TTCCGGCAATTGCCTCACCGAGAGGCAAGGCCGCAAGCTCCCGCAAGGGGGATCGAAGAC
GTCAATCCGGTCGTTCACGACGGCGGGGACGACCTGCGAGACGAGACTAATCGCCGATTG
GCGGAACTCTTGGCGCGGCCGGGAGTGAACAAGGGTTTGCTTGTTAACCTCCTTGGATAC
GAGGAAGAGGCTAACTCGACCCAAGATGAGGATTACATCGGAAACGATGTAATCCCGGAG
ACGGACACGAGTCAAACTTTCACTCCCTCGCCTTTGGCCCACATGAGCGACGCTCCGTTG
GGGTCCTCGGCTAGCCGAGGACGAATTGGTTCAAAACTTGAGGGGTCCAGTAACCCGCGA
GTGCCGGGAAGCGGTAACGCTTCCCAGGGCAGGCGATGTAGGCGTCCGCGGCGTGCGCAG
TCCCGACGGGGCGACGATGGTGCGGAAAGCCATGAGGTCCACCGCGAAGGTCGCCCGCCG
ACGGCGAGGCCTTCGCACCCTTCGGGGGAGGATCTCAGGGTTCGCCTTAACTCGCGGAGG
GGTCCTCAAGAGGAAGATTTGAGAAGGAAGTTGAACGCCCGTCGGACTGCGGCGGGGGCC
ACAACCCTCGGGAACCCTCCTCCCAACCTCATGGATCAAGGATCTAATGAGCGACAAGGG
CCAACGCCGGTGGGACCCATCGCGGGAACCGCCAACGGTCTCGAGATGATGACAATGATT
CAGGGAGTCCAGAATCAAATCATCGAGCTCCGAGGAGCAGGCTCGGGGGGCTCCAGACCT
TACAACGGTCTGGAGGGTTCATACTCGCCTTTCGCGGCGGTAGTGCTAGCACAACCGCCG
GGAAGGGGATTTAAGCTCCCCACCATCTCGTCTTATGACGGACGCTCCGACCCCAACTTC
CACATCCAACATTACGAGACGTGGATGTTGATGCAAGGATTGTCAGACGCGACTCTCTGC
CAGGCGTTCAGTTTGACATTGACGGGGTCAAGCTTTGAATGGTTCAGGAGCTTGAAACCT
GGCAGCATCGGATCTTTCCAAGCCCTTCGGGAGGCTTTCCTGGCTCGGTTTGCCACACTT
GTGGCACATAAAAAATCTAAATCATACCTATGGACGATCAAACAAAAATCGGACGAGAGC
CTCCGTCAGTATCTGGCCCGATTCACGGAAGAATCCCATAAAGTCGAAGACTTCGATGAT
CGGGATGCCATAGTCGCCATCACCGAGGGTTCCGAGCCTAGCGATTTCTTAAAATCGATT
GTGGGAAAGACCCCGCGAGATATGGCTGAGCTTATGGCATGCGCCAGGACGCACATGGGG
ATAGAGGATTACCTAAGCTGTAGGAAAGGCGATGGGCGCTCAGGGCGCAACCGTCCCGAC
GAAAGGCCTGAAACTCGAGGGAAACGCGGAAGGAATAACCGCGCGGAACCCTCTCCCAGA
TCAAGGAAACCTTTCGTCGACCAGTTTCGGGGGGAAATCCTCACCCCTCTCAATACGTCT
CCAGCCCGGATCTTGGCCCTACATCGAGATAAACTCGGGTCCCCGACGCCTTTGAGGAAG
GGTCCCGCGATGCGGGATAAAAGTAAATTTTGTGAATACCATCGAGATCATGGACACATG
ACCGCCGATTGTTTTCAACTGAAACGACATATAGAAGCCCTGATCCAGGATGGGAAACTC
AAAGAGTTTGTCATGCGGGCCGCAGGGGTAGCTGTCCCGGAGGAACGTCTTCCGAATCAG
AGTAATAAGCTGATAAAGGAAAAGGCCGCCGCCGTCGCCACTATTAGCAGCGAGCCCGAA
AATCCTGAAATGAACATCGTGAGAACAATTACTGGAGGGCCCCAATGTGGCGAATCAAGC
AACCAAAGGAAAAAATACCTAAGGGCGGTGACAAACGAGAGAATCAAGAGTGGCGTTAAC
CTCACTCAGGTGACCCCCCCACCAACTCCGTTTGAGGAAATTCGCTTCACCGAGGATGAG
GCGAAAAGAATTTTGCACCCGCATAGCGTTGCCTTGGTTGTGTCTGCCACGGTGGCTAAC
AACAAGGTCCAGCGCATCCTAATTGACAACGGAAGTGCGGTGGATATAATTTACCAATCG
GCGCTTGATCGAATGAAACTCGACGGAGCGAGGCTCGCCCCAATCCGTACCCCCCTGTAC
GGGTTTTCAGGAGAAAAGGTTTGGACGGAAGGGTCGATCGATCTCCCCGTAACCTTCGGA
ACCCGGGGTGGTCCGGAAGTCACTCGAATTGTGACTTTCTTGATCGTGAGTGAACCGACG
ACGTATAACATGATTATCGGTCGACCGACCCTCGTCGGTTTAAAAGCCATCACCTCAGTC
TACCACCTCATGGTGAAATTCCCAACCAGCAAGGGGATTGCATGCATGATGGGAGATCAA
GCCATGGCGCGAACGTGCCATGTGCGAGCCGTAAAGGAAGAAGTTACCAAACCAACAGGG
AACGCGACAATTTGCACAACCTACCAGCTCTCGGACGAGCTTCCCTGCAAACCTGAGGAA
GGGGGGACCTTCGGAGAATTGGACCAGCGCGACGAGACCGAGCGCCGTCGCGTGGAAACC
CATGACGACCTCGTCACGATAGTCCTGGACATCGCTAAGCCCGACCAAACGGTCAAGATC
GGTGCTCAATTGTCTGCCGAGGAGCAAGAATCGATGGCCCACTTCCTTAGAGAATACAAA
GATGTCTTCGCATGGACACATGCGGATATGCCGGGCATAGACCCGGAGATTATGGTTCAT
CGTTTGAACGATGACTGCAATGTGCCCGAGCGGGTACAGCCTAGAAGGGCATTCGGTGGG
GAGCGGTATCTCGCCATCGGAGAAGAGGTTGACAAACTTCTCGCCGCCGATTTCATACGA
GAAGCGGAACACCCCAAGTGGGTATCCAATGTAGTAATGGTCAAAAAGCCCAACGGGAAG
TGGCGCATTTGCATAGATTATAAAGATCTTAACAAAGCCTGCCCGAAGGATAGCTTTCCA
CTCCCCCGGATTGAGCAACTCGTGGACGCAACAGCGGGCCATGAGCTGCTCAGCTTCATG
GATGCCTACTCAGGCTATAACCAGATCTGGATGCATGAGAGTGACCAGATCAAGACGACC
TTTGTAACCAACAAAGGTCTCTATTGCTACAAGGTAATGCCGTTCGGCCTCAAGAACGCC
GGCGCAACCTACCAGCGATTGGTCAACCGAATGTTCGCCGAGCATATCGGAAAAACCATG
GAAGTCTACGTGGACGACATGTTGGTAAAAAGCCGATTGGCTTCAAACCACATCCGCGAC
CTGAAGACTATGTTCCAAATCCTCCGCCAATACGGGATGAAGCTAAACCCCAAAAAATGC
ACCTTCGGGGTAGAATCGGGAAAATTCCTCGGCTACATGGTTAGCAGGCGAGGAATTGAA
GCCAACCCTGAGAAAATCGCCGCCATACTTGGGATGAAGTCTCCCACGAAACCCAAAGAG
GTCCAAAGCCTGGCGGGTAGGGTCGCGGCTTTAGCTAGATTTGTTTCCAAGGCGACCGAT
AGGTGCCAACCATTTTTTCAGGCGTTGAAGATTAGGGGCGGGTTCGAATGGACATCCGAA
TGCGAGAAGGCATTTTGCGAAATGAAGGCTTACTTCGCAGCACCCCCGTTCTTGTCGAAA
CCCAAGGACCATGAGCCTCTCGTGCTTTACCTCGCTGTGTCCGAAAGGGCCATGAGCGCG
GTCCTTATCCGTGAGGAAGGGCAGGGACAATGGCTGGTGTATTATGTCAGCAAGGCCCTC
CTAGACGTTGAGACCCGGTATCCTGAGATCGAGAAATTGTCGCTCGCACTCGTCATGAGC
GCCAGGAAACTACCCCCGTATTTCCAGGCACACCTGATAGTCGTGCGGACCGGTCACCCG
ATCGAAAAGGCACTACAAAGAGGATCCTCGACGAGGATGGCCATATGGTCACAGGAACTT
AGCGAGTACGGGGTACAGTTCCGACCACGAAACGCCATAAAGGGTCAGGCCCTAGCAGAT
TTTGTGGCAGAATTTACGGCAGGAATCGAGAATGAAGGCGAAGCCCTCACGCCAAAATCT
CCAGAGAAGAGAGGCCCTGGCCCATCCGGGCCAGCGTGGGTGTTATGCGTCGACGGCGCA
TCAAACGCTCAAGGGGCCGGGGCCGGAGTAATTCTAACCAACCCCGACGGGGAAAAGTTA
AAATACGCGCTGCGCTTCGAATTCAAAGCCTCCAACAACGAGGCAGAATACGAGGCCTTA
ATAGTCGGTTTAGAGCTCGCTCACCGACTCGGCGCAGAAAACTTGCACATCTTTTGCGAC
TCGCAATTGATCGTGAATCAAGTCAAAGGCGAATTCGCCGCCAAGGAGCCAAACATGATG
GCTTATTTGGCGAAAACCCGAAAGAACCTAGAAGCGATAGCTTGGTTCGAGATCGTTCAA
ATCCCCAGAGAGGAGAACGCCGAGGCAGATGCTCTGGCGCGACTCGCTTCGGGCATAGAT
TCGAACGAGGATAGGGAAGTCCCGATTGAGATTCTAACCAGGCCTTACATCTCGAGGAGC
GATGTAGGGATGGTTAGTAAATCATCGGATTTGTCTTGGTTGACACCCCTAAAAAACTAC
ATAGCAAAGGGAGTAACTCCAGAGAATAACGAAGAGGCCAGGAAACTGAAACATTTGGCG
TCAAAGTATACGATGCGCGGAGACGTGCTATACAAGAGGGGCTACTCGATGCCTCTCCTT
CGCTGTCTCGACAACAGCGAAGCGAGACGAGCCATGAAAGAAGTCCACGAAGGACTATGT
GGGAACCACGCGGCAGGACAGTCGATGGCCATGAAGCTTCTAAAGCTAGGCTACTTTTGG
CCAACTTTGAAAAAAGACGCCCATGACTGGTCCCGGAAATGCGATAAGTGCCAAAGGTTC
GCCCGTGTTCCCAGACAACCCCCGGCGCCACTGTGCCCGATCATCGTTCCCCGACCATTC
ACCAAGTGGGGAATCGATTTGATCGGTGAATTACCCACGGCTCCAAGGCAGCTCCGATTC
GCCATAGTTGCAATCGACTACTTCACCAAGTGGGTTGAAGCCGAGCCACTGGCAAAGATC
ACGGAGGCAAATACAACGAACTTCATATGGAAGAGCATAATATGCCGCCACGGTATCCCG
GAATCGATCGTGACCGATAATGGCCGACAATTCGACAACGCGCGATTGCGCGAGATATGC
GAACAGCTTGGGATTAGTAAAACATACTCCTCCCCAAGGCACCCCCAGGCTAACGGCCAA
GTGGAGGCGGCAAACCGTACGATAAAAGAAAATTTGAAGAAGAAATTGGAATCCCGAAAG
GGCGCGTGGGCAGACGAATTACCCTTCGTATTATGGGCATATCGTACCACGGCACGAAAC
GCAACGGGGGAGACCCCCTTCGCCTTGACTTTCGGTGTGGAGGCCGTGGTCCCGATAGAG
ACCACGCTCCCCACGTTCCGAATGGAGGTCTTCGACGAAAAAGAAAACGAGGAAGCCATA
AGAGCAAACCTCGACCTAATCGACGAAAAGAGGGCCGACGCACTCACGCGCCTAACAGCC
CAAAAAAGAAAAGTCGAACGGTACTACAACGCAAGGGTTAAAACTCGGAGCTTCCCCGAC
GGCAGCCTGGTTTTGAGGAAAGTATTTCCAAATACCAAGGTACAAGGGGCTGGTGTGCTA
GGCCCAACCTGGGAAGGCCCGTACCGAGTTCGGAGAGCCATTCACAAAGGCACATACGAA
CTAGAAAACATGGACGGGCTCGCGTTCACACACCCTTGGAACGCCGAACATTTGAAAATG
TATCACGCATGAAGGCCAAGCGGAGTGCGAGATTTCCAACTAATGTAAACACAAAGCCAC
GATTGTCGTAATTCAATAAGTGTATCCATTTGAACATCGAATAAAAAATCGTTTTGTGTC
ATCATTTTGGTTATTTTCCCGCTGTAAGTTCCCAGAAATAATAGCTATTATCCCTAGAAA
CTTGGGGGGCTGGAGGTGAACCCAACAAAGCGACGATACGCTAATAGGGACCGAAACCCC
CCACCGATGGGCGAATCGATAACAAAACCATAACGGTTTGCATTCGAGAAAATCAAAGAC
CCACAAAACCGCGGGGTACGAACCCCGTGGCCGGAGAAATCTCCACCGAAAACCAGTAAC
AAGGCAACAACCTTGCATCTGGGCGTCCATGGATTTGAGGAAGGTTGGCAAACCCCGATT
GACACACAACCGGGCCAATGTCAATGCCCAATCCGAAGGCGTAAAACGAATACCCTAGAG
ATGAATGCGCGCGACGCCAACAGGGCAAATCGTAAACCAACGTCACAAGTCGACGAGTCA
TCGATCGAGCCAAATCTCGGGATTCCCCCGTCGTAGAGCCAAGACGTAAGCCGACAGGCC
ATCGTCGAAAGCTCACGAACAAAGAAGCTCAAAGCACGGCTGGAAGTAAAAAATGCCCAA
AACCGAAGAATGCGAGCTCGGATGTCGCCAAGGAAGCAAGGCGTAAGTGAATGAGACATT
GACCGTCCCTCTAAACACAAACGTGCCCAAGAGGAGAGTCGGACGAGTTACCAAATGAGT
CTCCCACCATCGACCACGCAATATACGCGACATGCCGTTAGTGCTAGTTGCCGGCCATAG
TAAGCCGACGACGGCAAAAGAAATTCCCGATCATCGACCGCCCCTCGCCGAAGGGGCCCG
CCGTAAATCGAAGGACCCGAGCGTCTTGGAAACCGACAATCGAAAACTTCACGCAACACA
AGAATCGCAAAAGAAACCAACGCGAACCCGCGAGGTATTTCGTTCGATAAAGAGAAGATA
GCGAGAAGATTGCGAGAAAGTAAATGTCACAAAAATGGAAAGCCTCGAAACATCCAAAAT
GAGTCAGGTAAATAAGAAGAAAGATGCGCTAAACACAGGCGCAGACGAGACAAAAACACT
AATCGGAGCCGAAGTCAAGCTGCTCCTCGCCGGGGCCCGCGCTCGGGTGAGAAGGGGATG
TACCCACCGAGGACTGCGTAGTCTCGGTGGGACGCCGCCCACCGCGATCGCCAAAGCCCG
CGTCGCCCTCGATCGGAGGGAACATGGTATCGAAGAATGCGAAAGACACTTCGCGGGGTT
TGTTCTCTGCGACAATAGCATCCCGGACGCGAACGGCCGCAGCGTGTCGGGCCTCGTATA
CCGCGACTTCCCAATCATCCTTCAGGCGCTCTTTGATGCGCACGGCTTCATCACGGTCGT
CGACGACGGCTTGCATCTCCGCCAAGAGGGCGTCGTTCTCCTCAACGGCTTTCAGGTGGG
CAGCCTCGCTCACAGCTAGCCGACCGCGGACGGAAGCAAGGTCGGCATCACGGGAGGCCA
AGGCCTTCTCCATCGCCTTGCTTTCCGCTAGGAAGGCGCTCGCGGATCGCTCCACTTGCT
CCCTCGAGGCCCTCTCCTCCGCGAATTTCGCCTCGGCCGCGTCCCGATTAGAAATCGCCT
TGGCCAACGCCTCTTTCAACGCTCGTTCACGGCGAGCGGCCTCCTCCATCACGAGCTCCA
TCCGCTCGCAGCGCGCCAAAGCCTACAGTGCAACGACAAATTAGCATGGCATAGAGAGGA
AAAACACGAAATAAGAGGCACGCGAATGACTAACCTTCCGCGCATCGCGAGTCCCTTGTC
GAAAGAGTTCCATATCGCTGACGACGACGAGCTCCTTAAAGCACTTGGGGGATGAGTCCG
CGAGGCCCCGAGCGCTCCGATCCTTCAAGTCTTTCGCTTCGGCGAGATAGAGCGACCCGC
CGCCACCAGTAGAGGTCGCTTCCACCAATCTTCGAGAACCCGACGGCGCGTCACTCGTCG
AGGATACGGCGGGGCGTTTCCGCGGAGGGCCGCTGCTCCCGCCCTCCTGAGGTGCCGACG
AGGGTTGAGATTCCGGCCTCTCCCGTGACGGGCTCGGCCGCCTCGCAGCATGAGCCTCGC
GCGGAGGTGGCGAGGTCGTAACGCCTCCTCGAGGTCGTCGACGAGAACCGCCTAACGAAA
GAGCAACCAAGTTAGTAGGCAAAAATGAACGACACAATAAGGGGTACCTCGTACCGCTAC
ATCGATCAGCCGGAAGGACTTCTTCGTCGGGCTCTGAGTCCGAATCGTTCCGCCCTCGGC
CGCGTGCTCGGGGGAACTGAGGAGTAGTCGAAACGAACCACCCTGTGATTGCCATAGTAT
CGGGCGACGCGAAGATCGAAGCCTGACGACGGACAGACGGGAGACGATGGACGACTCTGC
TAAGCTCCGTGAATTCTGGGTCGGCGGCCGAATCTGGGATCCTCGCGGGCCCAAACTCAA
GAGGGGCCGAGCCGAACGCGGACTCCGAAACCCACACCCAAGGGTATCGGTACTTGGTGG
GCACATTGGTAACCACGGCAAACTGCTTGGCATGGTACGCCCGGTACGACCCGGCCCTAT
TCGACCCTTCCACGATCGATAGGAATTCCAAGAACTCCTCGGCCGTGACCGCCCTCTTCC
ATTCTCGAAAGAACATCATATGGCAGGCAGCCACCATCCGATAGAACGAGGGTAGCATTT
GGGCTGCGTGCACCTCCAGTCGACGCAGAAGCTCCTGTACCACGGCAGGGAAAGGCAACC
GAAGACCGCTCCGGAGACTGGCCACATGAACCGCGGCGAACCCGAGCCGAGGACGCGAGG
GCCGATCCTCCGGGAGAGGACGCAAGATGTCGATACTCGGCGGGATCCCGAACTCCTCGC
GAAGCCCACGGAACTGGGCGAGAAACACGAGGGACGGGGAGTCGACCCCCTCGCAGGGGA
AGGTCCACCGAATCCGGGACGCGGTGATCGGGAAGGCAAGGGTCCGCTCGTTCTCTTCCG
ACACGCGACTACGCCCCAGCTGGGCCAAAGTCGGAAGGGTACCCGCAGCCGACGTCTGCC
TGGGAGGTCTCGGGTCCTCCGTTGGAATCTCGGGCGATTCCCGCCCCGTAGGAGACCCCC
GGCCGCTCGGCCGATGCTCGAAGGTCCTCCGGTCGGCCACCGAACCCCCGTGAGGGGACT
GGCCGGCCATCTCAATCAGGGGAGGCCCCTGGCCGGCCATGGGATCCCTGGGAGGGGGCC
GACCGGCCATCCTATCCAGGGGAGACCCCTGGCCGGCCATCGGGTCCCCGAGAGGGGGCC
GGCCGGCCATCACATCCGGGGGAGACCCCTGGTCGGCCACCACCCCCCTGGGAGAGGGCC
GGCCGGCCATTTGACTTGGGGGAGACCCCTGGTCGGCCACCGCCCCCCTGGGAGAAGGCC
GGCTGGCCATCTGGCTCGGGGGAGGCCCCTGGTCGGCCGCCAGAGCTCCGGGAGGGGGCT
GGCCGACCGCCTGGCCCTGATGGGGCGTGAAGTCGGCGGTGGACGCCTCGGGTGGAGCCC
ACGCCACCGAGTGAACTCCGGCGAAGACAAGCCCGGAGCCCACGACCTGCGACGCGCCTC
CCACGACTCCACACACCAAGGGAGGCTCGCCCCCGTCGCCCAGGGCGCTGTCGCGCTCCT
CCTCGGCCGACAAACCTCCGCCCGCGCCTTCGGGGCTCCCGATCCAGCCAGATCCGTGCG
CCTCCGCCCGGGGCGTGTCCCCCGAGGCACCACGCGCGCGTCTGTGGACGTCTCCGCCTC
GCCCTGCCGCGCCGCCATCTCCGTGCCCGAGGGGCCTCGCCCAAGGATCCGCCGCCACGC
CACGAGAAGATCCGGCGCCGCCCTCACAGCCAAGGGACGCAGACCCATCGTCGGTCCCCG
AGGCCACCTCCCACTCGCCCGCGCCCCCAAGGACAGACCTGGGAGCGCTCCGCCGGCGAT
CTGTCGCCGTCGCGCCCTGCGCGAATCCGGGTTCGCCCTCGCCCCTGGTCGGCGCCGAGC
CCCCCTCGCCGGTAATCTCCAAAGTCCCTCCGACCTCACGCCCCTCGCCGAGCAAGGCCC
CTTTGCCGGGCTGTTCACCGGCCAGGGCCTCCGCCCTCGCCTCCGGCGTCGCGATTCCGT
CGGCGGCCATCCACGCGTCAAGGTCGTCGTCGTCAAAGAAGAGAGAAAGAGAGGATTCGG
GATTTGGGGAAAAGGGTAAATTGGGGATTGAAAAAGTATTTCTCCCAGGCGAGGATAAAA
AGGAACCAACGAAGTCGTCAACGGACCAGTCATCCACGTCAAACGGCACGGGGTCCATGA
CGCAACGCAAAGAAAGAAAAAAGGAAAAATATAATAAATTAAAGGTTAGGGTGCTTATAA
AGCTACCTGATCTCCTCCACCCCCGGGGAGGTTCCCTGGAGCGCTGCCCTCCGGCCCGGC
CCGTCGGTGCACGCCCAATCTTAGGTGCGGAACGCCCGTCGGAAGGCGCCACTTCTGAGT
TCGCTCTCGATCTCAAGTTCGACATCAAACTCGGGGGGCTAATGTTGTCGCCGTATTTCC
GGCGACAGCGTTAAATGTATACATCCACGATTATATCCGAGGCGCGGAGGGGAGACGCCT
TCCCCGAGACGACCGAGACCCACCTTCGCACACCTGACCGACCACCGTCGCGCAGGGGAG
GGGCCCGGCCGGCCACCTTCTCCCATGAGAGGGGCCTGGCCGGCCTTCCCAACCCGAGGG
AGGGACTTGGCCGGCCTTCCCCACCTGTAAAAAACCTACTTTAATTTGTCAAAACGTACT
TCAACTTCTAAAAAACGTACCTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGTA
ACTTTAACTTGTCAAAACGTAAATTTAACTTGTTAAAAACGATCTTTAACTTGTAAAAAA
CTTACTTTAACTTGTCCAAACGTAACTTTTACTAGTCAAAACTTAACTTTAACTTGTTAA
AACGTAATTTAACTTGTAAAAAACGTACTTTAACTTGTGAAAAACGTACTTTAACTTGTC
AAAACATTAATTTAACTTGTAAATAACCTACATTAACTTATCAAAACGTAACTTTAACTT
GTCAAAACGTAAGTTTAACTTGTTAAAAACATTTATTAACTTGTAAAAAACGTACTTTAA
GTTGTCAAAAATGTGGTTTAACTTGTAATAAATGTACTTTAACTTGTAAAAAACGTACTT
CAACTTGTAAAAAACGTACTTTAACTAGTACAAAACCTACTTTAAATAGTCAAAACGTAA
CTTTAACTTGTCAAGACGTAAGTTTAACTTTTTAAAAACGTAATTTAAGTTGAAAAAACG
TAACTTTAACTTGTAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGCCAAAA
CGTAACTTTAACTTGTAAAAAACCTACTTTAACTTGTCAAAACGTAACTTTAATTTGTAA
AAACGTAAGTTTAACTTGTTAAAGTAAATTAACTTGTCAAAACGTACTTCAATTTGTAAA
AAACGTACCTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGTAACTTTAACTTGT
CAAAACGTAAATTTAACTTGTTAAAAACGTAACTTTAACTGGTAAAAAAAACCTACTTTA
ACTTGTCAAAACGTAACTTTAATTTGTAAAAACGTAAGATTAACTTGTTAAAAATGTAAT
TTAACTTGTCAAAACGTACTTCAACTTCTAAAAAACATACCTTAACTTGTAAAAAACGTA
TTTTAACGTGTCAAAATGTAACTTTAACTTGTCAAAACGTAAATTTAACTTGTTACAAAC
GTACTTTAACTTGTATAAAACTTACTTTAACTTCTCCAAACATAACTTTTACTTGTCAAA
ACTTAACTTTAACTTGTTAAAAACGTAGTTTAACTTGTAAAAAACGGACTTTAACTTGTG
AAAAACGTACTTTAACTTGTCAAAACATTAATTTAACTTGTAAAAAACCTACTTTAACTT
ATCAAAACGTAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAACATTTATTAA
CTTGTAAAAAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTTGTAATAAATGTACTT
TAACTTGTAAAAAACGTTCTTCAACTTGTAAAAAACGTACTTTAACTAGTACAAAACCGA
CTTTAAATTTTCAAAACGTAACTTTAACTTGTGAAAACGTAATTTTAACTTGTTAAAAAT
GTAATTTAACTTGTCAAAACATAACTTTAACTTGTTCAAAACGTACTTTAACTTGTAAAA
AACGTACTTTAACTTGCAAAAACGTAACTTTAACTTGTAAAAAACCTACTTTAACTTTTC
AAAACGTAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAATGTAATTTAACTT
GTCAAAACGTACTTCAATTTGTAAAAAACGTACCTTAACTTGTAAAAAACGTATTTTAAC
GTGTCAAAATGTAACTTTAACTTGTCAAAACGTAAATTTAACTTGTTAAAAATGTAACTT
TAACTTGTAAAAAAACCTACTTTAACTTGTCAAAATGTAACTTTAATTTGTAAAAACGTA
AGATTAACTTGTTAAAAATATAATTTAACTTGTCAAAACGTACTTCAACTTCTAAAAAAC
ATACCTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGTAACTTTAACTTGTCAAA
ACGTAAATTTAACTTGTTACAAACGTACTTTAACTTGTAAAAAACTTACTTTAACTTCTC
CAAACGTAACTTTTACTTGTCAAAACTTAACTTTAACTTGTTAAAAACGTAGTTTAACTT
GTAAAAAACGTACTTTAACTTGTGAAAAACGTACTTTAACTTGTCAAAACATTAATTTAA
CTTGTAAAAAACCTACTTTAACTTATCAAAACGTAACTTTAACTTGTCAAAACGTAAGTT
TAACTTGTTAAAAACATTTATTAACTTGTAAAAAACGTACTTTAACTTGTCAAAAATGTG
GTTTGACTTGTAATAAATGTACTTTAACTTGTAAAAAACGTACTTCAACTTGTAAAAAAC
GTACTTTAACTAGTACAAAACCTACTTTAAATTTTCAAAACGTAACTTTAACTTGTCAAA
ACATAAGTTTAACTTGATAAAAACGTAATTTAACTTGTAAAAAAGTAACTTTAACTTGTA
AAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGCCAAAACGTAACTTTAACTTG
TAAAAAACCTACTTTAATTTGTCAAAACGTAAGTTTAGTTTGTAAAAACGTAAGTTTAAC
TTGTTAAAAATGTAATTTAACTTGTCAAAACGGACTTCAATTTGTAAAAAACGTACCTTA
ACTTGTAAAAAAACGTATTTTAACGTGTCAAAATGTAACTTTAACTTGTCAAAACGTAAA
TTTAACTTCTTAAAAACGTAACTTTAACTTGTAAAAAAACCTACTTTAACTTGTCAAAAC
GTAACTTTAATTTGTAAAAACGTAAGTTTAACTTGTTAAAAATGTAATTTAACTTGTCAA
AACGTACTTCAACTTCTAAAAAATGTACCTTAACTTGTTAAAAACGTATTTTAATGTGTC
AAAATGTAACTTTAACTTGTCAAAACGTAAATTTAACTTGTTACAAACGTACTTTAACTT
GTAAAAAACTTACTTTAACTTCTCCAAACGTAACTTTTACTTGTCAAAACTTAACTTTAA
CTTGTTATAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGTGAAAAACGTACTT
TAACTTGTCAAAACATTAATTTAACTTGTAAAAAACCTGCTTTAACTTATCACAACGTAA
CTTTAAATTGTCAAAACGTAAATTTAACTTGTTAAAAACATTTATTAACTTGTAAAAAAC
GTACTTTAACTTGTCAAAAATGTGGTTTAACTTGTTATAAATGTACTTTAATTTGTAAAA
AACGTACTTCAACTTGTAAAAAACGTACTTTAACTAGTAGAAAAACTACTTTAAATTTTC
AAAACGTAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAACGTAATTTAACTT
GTCCAAACGTAACTTTAACTTGTAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAAC
TTGCCAAAACGTAACTTTAACATGTAAAAAACCTACTTTATTTTGTCAAAACGTAACTTT
AATTTGTAAAAACGTAAGTTTAACTTGTTAAAAATGTAATTTAACTTGTCAAAACGTACT
TCAATTTGTAAAAAACGTAACTTAACTTGTAAAAAACGTACTTTAACTTGTGAAAAACGT
ACTTTAACTTGTCTAAACATTAATTTAACTTGTAAAAAACCTACTTTATATTATCAAAAC
GTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACATTTATTAACTTGTAAA
AAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTTGTAATAAATGTACTTTAACTTGT
AAAAAACGTACTTTAACTTTTAAAAAACGTAGTTTAACTAGTATAAAACCTACTTTAAAT
TGTCAAAACGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACGTAATTTA
ACTTGTCAAAACGTAACATTAACTTGTAAAAAACGTATTTTAACTTGTAAAAAACGTACT
TTAACTTGTAAATAACGTACTTCAACTTGTGAAAACGTAACTTTAACTTGTAAAAAACCT
ACTTTAACTTATCAAAACGTAACTTTAACTTGTCTAAACGTAAGTTTAACTTGTTAAAAG
CATACATTAACTTGTAAAAAACGTGCTTTAACTTGTCAAAAATGTGGTTTAACTTGTAAT
AAATGTACTTTAACTTGTAAAAAACGTTCTTCAACTTGTAAAAAACGTACTTTAACTAGT
ACAAAACCGACTTTAAATTTTCAAAACGTAACTTTAACTTGTGAAAACGTAATTTTAACT
TGTTAAAAATGTAATTTAACTTGTCAAAACATAACTTTAACTTGTTCAAAACGTACTTTA
ACTTGTAAAAAACGTACTTTAACTTGCAAAAACGTAACTTTAACTTGTAAAAAACCTACT
TTAACTTTTCAAAACGTAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAATGT
AATTTAACTTGTCAAAACGTACTTCAATTTGTAAAAAACGTACCTTAACTTGTAAAAAAC
GTATTTTAACGTGTCAAAATGTAACTTTAACTTGTCAAAACGTAAATTTAACTTGTTAAA
AATGTAACTTTAACTTGTAAAAAAACCTACTTTAACTTGTCAAAATGTAACTTTAATTTG
TAAAAACGTAAGATTAACTTGTTAAAAATATAATTTAACTTGTCAAAACGTACTTCAACT
TCTAAAAAACATACCTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGTAACTTTA
ACTTGTCAAAACGTAAATTTAACTTGTTACAAACGTACTTTAACTTGTAAAAAACTTACT
TTAACTTCTCCAAACGTAACTTTTACTTGTCAAAACTTAACTTTAACTTGTTAAAAACGT
AGTTTAACTTGTAAAAAACGTACTTTAACTTGTGAAAAACGTACTTTAACTTGTCAAAAC
ATTAATTTAACTTGTAAAAAACCTACTTTAACTTATCAAAACGTAACTTTAACTTGTCAA
AACGTAAGTTTAACTTGTTAAAAACATTTATTAACTTGTAAAAAACGTACTTTAACTTGT
CAAAAATGTGGTTTGACTTGTAATAAATGTACTTTAACTTGTAAAAAACGTACTTCAACT
TGTAAAAAACGTACTTTAACTAGTACAAAACCTACTTTAAATTTTCAAAACGTAACTTTA
ACTTGTCAAAACATAAGTTTAACTTGATAAAAACGTAATTTAACTTGTAAAAAAGTAACT
TTAACTTGTAAAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGCCAAAACGTAA
CTTTAACTTGTAAAAAACCTACTTTAATTTGTCAAAACGTAAGTTTAGTTTGTAAAAACG
TAAGTTTAACTTGTTAAAAATGTAATTTAACTTGTCAAAACGGACTTCAATTTGTAAAAA
ACGTACCTTAACTTGTAAAAAAACGTATTTTAACGTGTCAAAATGTAACTTTAACTTGTC
AAAACGTAAATTTAACTTCTTAAAAACGTAACTTTAACTTGTAAAAAAACCTACTTTAAC
TTGTCAAAACGTAACTTTAATTTGTAAAAACGTAAGTTTAACTTGTTAAAAATGTAATTT
AACTTGTCAAAACGTACTTCAACTTCTAAAAAATGTACCTTAACTTGTTAAAAACGTATT
TTAATGTGTCAAAATGTAACTTTAACTTGTCAAAACGTAAATTTAACTTGTTACAAACGT
ACTTTAACTTGTAAAAAACTTACTTTAACTTCTCCAAACGTAACTTTTACTTGTCAAAAC
TTAACTTTAACTTGTTATAAACGTACTTTAACTTGTAAAAAACGTACTTTAACTTGTGAA
AAACGTACTTTAACTTGTCAAAACATTAATTTAACTTGTAAAAAACCTGCTTTAACTTAT
CACAACGTAACTTTAAATTGTCAAAACGTAAATTTAACTTGTTAAAAACATTTATTAACT
TGTAAAAAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTTGTTATAAATGTACTTTA
ATTTGTAAAAAACGTACTTCAACTTGTAAAAAACGTACTTTAACTAGTAGAAAAACTACT
TTAAATTTTCAAAACGTAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTTAAAAACGT
AATTTAACTTGTCCAAACGTAACTTTAACTTGTAAAAACGTACTTTAACTTGTAAAAAAC
GTACTTTAACTTGCCAAAACGTAACTTTAACATGTAAAAAACCTACTTTATTTTGTCAAA
ACGTAACTTTAATTTGTAAAAACGTAAGTTTAACTTGTTAAAAATGTAATTTAACTTGTC
AAAACGTACTTCAATTTGTAAAAAACGTAACTTAACTTGTAAAAAACGTACTTTAACTTG
TGAAAAACGTACTTTAACTTGTCTAAACATTAATATAACTTGTAAAAAACCTACTTTATA
TTATCAAAACGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAAACATTTATT
AACTTGTAAAAAACGTACTTTAACTTGTCAAAAATGTGGTTTAACTTGTAATAAATGTAC
TTTAACTTGTAAAAAACGTACTTTAACTTTTAAAAAACGTAGTTTAACTAGTATAAAACC
TACTTTAAATTGTCAAAACGTAACTTTAACTTGTCAAAACGTAACTTTAACTTGTTAAAA
ACGTAATTTAACTTGTCAAAACGTAACATTAACTTGTAAAAAACGTATTTTAACTTGTAA
AAAACGTACTTTAACTTGTAAATAACGTACTTCAACTTGTGAAAACGTAACTTTAACTTG
TAAAAAACCTACTTTAACTTATCAAAACGTAACTTTAACTTGTCTAAACGTAAGTTTAAC
TTGTTAAAAGCATACATTAACTTGTAAAAAACGTGCTTTAACTTGTCAAAAATGTGGTTT
AACTTGAAATAAATGTACTTTAACTTCTAAAAAACGTACTTAGACTTGTAAAAAACGTAC
TTTAACTTGTTAAAAACCTAGTTTAACTTGTCAAAACGTAACTTTAACTTGTAAAAACGT
ACTTTAAATTGTAAAAAAAGTATTTTAACTTGTCAAAACGTAACTTTAACTTGTGAAAAA
CCTACTTTAACTTGTCAAAACGTAACTTTAATTTGTAAAAATGTAAGTTTAACTTGTTAA
AAACGTAATTTAACTTGTCAAAACGTACTTAAATTTGTAAAAAACGTACCTTAACTTGTA
AAAAACGTATTTTAACGTGTGAAAATGTAACTTTAACTTGTCAAACCGTAAATTTAACTT
GTTAAAAACGTAACTTTAAATTGTAAAAAACATACTTTAACTTGTCAAAACGTAACTTTA
ATTTGTAAAAACGTAAGTTTAACTTGTTAAAATGTAATTTAACTTGTCAAAACATACTTC
AACTTGTGAAAAACGTACCTTAACTTGTAAAAAACGTATTTTAACGTGTCAAAATGTAAC
TTTAACTTGTCAAAATGTAAATTTAACTTGTTAAAAACGTAGTTTAACTTGTAAAAAACT
TACTTTAACTTGTCAAAACGTAACTTTAACACTTCAAAACGTAACTTTAACTTGTTAAAA
ACGTACTTTAACTTGTAAAAATCGTACTTTAACTTGTGAAAAACCTACTTTAACTTATCA
AAATGTAACTTTAACTTGTCAAAACGTAAGTTTTACTTGTTAAAAACATTTATTGACTTG
TAAAAAACGTACTTTAACTTGTCAAAAATGTTGTTTATCTTTTAATAAATGTACTTTAAC
TTCTGAAAAACGTACTTCAACTTATAAAAAACATACTTTAACTAGTACAAAACCTACTTT
AACTTGTGAAAACATAACTTTAACTTGTCAAAACGTAAGTTTAACTTCTTAAAAATGTAA
TTTAACTTGTCAAAACGTACTTCAACTTGTAAAAAACGTACCTTAGTTTGTAAAAAACGT
ATTTTAACTTGTCAAAATGTAACTTTAACTTGTCAAAACGAAAATTTAACTTATTAAAAA
CGTAATTTAACATGTTAAAAAAACTTACTTTATCTCGTCAAAACGTAAGGTTAACTTGTA
AAAAACTTTCTTTAACTTGTCAAAACGTAACTTTAATTTGTCAAAACGTACTTCAACTTG
TAAAAAAACGTACCTTAACTTGTAAAAAACCTACTTTAACTTGTCAAAAATGTGATTTAA
CTTGTTCAAAACGTAGTTTAACTTGTTAAAAACCTAGTTTAACTTGTCAAAACGTAACTT
TAACTTGTAAAAACGTACTTTAAATTGTAAAAAAAGTACTTTAACTTGTCAAAACGTAAC
TTTAACTTGTGAAAAACCTACTTTAACTTGTCAAAACGTAACTTTAATTTGTAAAAATGT
AAGTTTAACTTGTTAAAAACGTAATTTAACTTGTCAAACCGTACTTAAATTTGTAAAAAA
CGTACCTTAACTTGTAAAAAACGTATTTTAAACGTGTGAAAATGTAACTTTAACTTGTCA
AACCGTAAATTTAACTTGTTAAAAAGGTAACTTTAAATTGTAAAAAACATACTTTAACTT
GTCAAAACGTAACTTTAATTTGTAAAAACGTAAGTTTAACTTGTTAAAATGTAATTTAAC
TTGTCAAAACATACTTCAACTTGTGAAAAACGTACCTTAACTTGTAAAAAAACGTATTTT
AACGTGTCAAAATGTAACTTTAACTTGTCAAAATGTAAATTTAACTTGTTAAAAACGTAC
TTTAACTTGTAAAAAACTTACTTTAACTTGTCAAAACGTAACTTTAACACTTCAAAACGT
AACTTTAACTTGTTAAAAACGTACTTTAACTTGTAAAAATCGTACTTTAACTTGTGAAAA
ACCTACTTTAACTTAAAAAAAATGGAACTTTAACTTGTCAAAACGTAAGTTTAACTTGTT
AAAAACATTTATTGACTTGTAAAAAACGTACTTTAACTTGTCAAAAATGTTGTTTATCTT
TTAATAAATGTACTTTAACTTCTGAAAAACGTACTTCAACTTATAAAAAACATACTTTAA
CTAGTACAAAACCTACTTTAACTTGTGAAAACATAACTTTAACTTGTCAAAACGTAAGTT
TAACTTCTTAAAAATGTAATTTAACTTGTCAAAACGTACTTCAACTTGTAAAAAACGTAC
CTTAGCTTGTAAAAAACGTATTTTAACTTGTCAAAATGTAAATTTAACTTGTCAAAACGA
AAATTTAACTTATTAAAAACGTAATTTATGATGTTAAAAAAACTTACTTTATCTCGTCAA
AACGTAAGTTTAACTTGTAAAAAACTTTCTTTAACTTGTCAAAACGTAACTTTAATTTGT
CAAAACGTACTTCAACTTGTAAAAAACGTACCTTAACTTGTAAAAAACCTACTTCAACTT
GTCAAAAATGTGATTTAACTTGTAACAAATGTACTTTAACTTGTAAAAACGTACTTCAAC
TTGTAAAAAAACGTACTTTAACTTGTAAAAAACCTACTTTAACTTGTCA</code></pre>

<p class="blog-post_paragraph">Now I wanted to visualize how the sequencing reads mapped to the contig I just isolated, using a software like the <a href="https://igv.org/" target="_blank">Integrative Genomics Viewer (IGV)</a>. In order to do this, I needed to filter out the reads that mapped to this file (and that alignment file needs to be indexed). Also, just like the <code>.bam</code> file, <u>the <code>.fasta</code> file needs to be indexed</u> (needs a corresponding <code>.fai</code> file) <u>in order to be opened by IGV</u>. The following commands were used to perform all these tasks.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">samtools faidx tig00000002.fasta
samtools view -b -h drUrtUren1.hicanu.total.contigs.sorted.bam tig00000002 > tig00000002.bam
samtools index tig00000002.bam</code></pre>

<p class="blog-post_paragraph">Now inside this folder, we have these four files</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">ls -1 tig00000002*</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">tig00000002.bam
tig00000002.bam.bai
tig00000002.fasta
tig00000002.fasta.fai</code></pre>

<p class="blog-post_paragraph">Loading the <code>tig00000002.fasta</code> and <code>tig00000002.bam</code> files into IGV, you can see the following alignment.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_36.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">You can see the distribution of coverage on the top channel. You can see that the coverage is distributed in a way where most of the center of the contig is covered, but the coverage of reads tapers off at the ends. This is expected because with lower coverage at the ends, the alignment algorithm starts to lose confidence about what sequences are adjacent to the current contig, and so they will decide to end the building of the contig and that's one of the main determining factor of the contig size. IGV also highlights certain details in the reads, specifically how they differ to the reference sequence (although some of these may be hard to see from this zoomed out view of the contig). Purple I-beam-like shapes indicate single base pair insertions. You can see a lot of them scattered around in these reads. It's known that the current biggest issue with the PacBio technology is the insertions or deletions of bases (indel errors). This is because the rate of synthesis of the polymerase is sporatic and not perfectly matched with the sampling rate of the fluorescence, and so certain incorporations may get lost or mistakenly added. This was more common with PacBio CSS reads but with current HiFi read technology, this is less of an issue. Another issue you may encounter is base mismatches. These are viewed on the reads in IGV as coloured bars. Green bars on the reads indicate a mismatch at a given location between the base on the read and the base in the reference, where the base in the read is an <code>A</code> and the base in the reference sequence is something else. Blue bars indicate the mismatch is a <code>C</code>, orange bars indicate the mismatch is a <code>G</code>, and red bars indicate the mismatch is a <code>T</code>. The last thing you can see is deletions (represented by thin, black lines found between the reads). These occur for much of the same reasons as insertions but are much less common in this dataset than insertions for some reason.</p>

<p class="blog-post_paragraph">In addition to investigating the data contig-by-contig, you can also get the full summary of the alignment statistics of the whole dataset using the <code>samtools flagstat</code> program. <code>flagstat</code> counts and reports the number of reads with different flags set in the file. These flags represent various attributes of each read (such as whether it’s mapped, unmapped, paired, and so on). It generates counts for different categories of reads, such as:</p>
<ul>
  <li>
    Mapped vs. unmapped reads
  </li>
  <li>
    Properly paired vs. improperly paired reads
  </li>
  <li>
    Reads with low mapping quality
  </li>
  <li>
    Reads with specific flags (e.g., duplicate reads, secondary alignments, etc.)
  </li>
</ul>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">samtools flagstat drUrtUren1.hicanu.total.contigs.sorted.bam</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">2224799 + 0 in total (QC-passed reads + QC-failed reads)
2215315 + 0 primary
0 + 0 secondary
9484 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
2203971 + 0 mapped (99.06% : N/A)
2194487 + 0 primary mapped (99.06% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)</code></pre>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">samtools flagstat tig00000002.bam</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">31 + 0 in total (QC-passed reads + QC-failed reads)
31 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
31 + 0 mapped (100.00% : N/A)
31 + 0 primary mapped (100.00% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)</code></pre>

<p class="blog-post_paragraph">However there are much better methods nowadays for assessing read mapping to the assembled genomes and these will be touched on in the next sections.</p>

<p class="blog-post_paragraph">
    <u><em>Step 4: Purging Duplications and Merqury evaluation</em></u>
  </p> 

<p class="blog-post_paragraph">Now that the genomes are assembled, the next required step will be to remove any sort of duplications from the genome. Genome assemblies that don't perform this step are often inaccurate and can lead to a lot of incorrect biological insights. An interesting example that demonstrates this well came from <a href="https://doi.org/10.1186/s13059-022-02764-1" target="_blank">a study</a> that analyzed genomes generated by the Vertabrate Genomes Project and other researchers. The Vertebrate Genomes Project is an initiative aimed at producing high-quality reference genomes for a wide range of vertebrate species. The project focuses on generating genome assemblies that are both comprehensive and accurate utilizing the most modern and advanced sequencing technologies. While most of their genome assemblies are performed <em>de novo</em>, recently this project created assemblies of three species that already contained reference assemblies on NCBI, those being the platypus, zebra finch, and Anna’s hummingbird. Some researcher decided to compare the older genome assemblies with the new ones generated by the Vertebrate Genomes Project, which used methods like haplotype phasing and purging to reduce false duplications. By conducting whole genome alignments, they identified the extent of false duplications, finding that 4 to 16% of sequences in the previous assemblies were inaccurately duplicated, leading to the overestimation of gene family expansions.</p>

<p class="blog-post_paragraph">The main source of these false duplications is what's known as <strong><em>heterotype duplications</em></strong>, where certain haplotype sequences were relatively more divergent than other parts of the genome leading the assembly algorithms to classify them as separate genes or genomic regions (false paralogs). However, there are two other types of errors that are quite important. <strong></em>Switch errors</em></strong> occur in which variants from each haplotype are assembled into the same pseudo-haplotype. You will basically have a hybrid between the maternal and the paternal haplotype. You don't even have necessarily the heterozygosity belonging to only one of them. Also <strong></em>nucleotide consensus errors</em></strong> occur due to the collapse between the two haplotypes, causing mixing of SNPs and of different variants from both haplotypes. To fix this, you need to do some post-processing of the initial assemblies.</p>

<p class="blog-post_paragraph">What happens routinely during genome assembly without phasing, is that two haplotypes that belong to the same region but are somehow divergent, both end up in one of the assemblies.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_37.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">They end up being introduced into one contig with a gap in the middle (kind of like they are scaffolded together, as some assembly algorithms try to perform scaffolding after contigging and this is an artifact of that).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_38.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">You end up with sequences that should not be in the primary assembly, but yet they are. Therefore, often times, your primary assemblies before <em>purging</em> will be larger than what you expect in your final asssembly. This is a very common problem when assembling using poor quality reads like PacBio CLR, but still happens even in HiFi reads.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_39.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">So how can you identify whether this has happened in your assemblies at this stage?</p>

<p class="blog-post_paragraph">One thing you can do is map your reads back to your genome. This will produce a coverage plot. If you have this duplication in, for example a 2<em>n</em> genome, you will expect to see a drop in coverage in the reads to about half this level at the location where the read occurs. Also, you'll likely see a spot with 0 coverage in the middle. This coverage signature is a good indication that something like this has occured. Assemblers will often perform this kind of test periodically during assembly to try to find and remove these errors, but they aren't perfect and so often times you'll find these errors in the primary assembly.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_40.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If you see a coverage signature similar to the one below, then there is a good chance that your assembly contains an erroneous insertion. However, be aware that performing sequence alignment with improper alignment parameters can also cause reads from 2 similar sequences to pileup in a single region and produce a similar pattern. Often times if this step is done as part of a post-processing program, the alignment parameters chosen by the program should be sufficient to prevent this kind of thing, but it is good to be aware about and these regions might need closer inspection.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_41.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">But then after you have found these duplication patterns from the read mapping back to the assembly, how can you ensure that regions that drop in coverage are <em>actually</em> false duplications. You can perform a <strong><em>self-alignment</em></strong>. If these regions are actually similar, you can conclude that they are probably these kind of false duplication assembly artifacts and can safetly remove them. This is essentially what the program <a href="https://github.com/dfguan/purge_dups" target="_blank">purge_dups</a> does.</p>

<p class="blog-post_paragraph">In more detail, <code>purge_dups</code> works as follows: after you run your assembler and you have the contigs of the primary assembly, if you believe the size of this assembly is larger than your true genome size and you suspect haplotype retention, you run <code>purge_dups</code>. <code>purge_dups</code> will map all your reads back to your assembly, and it's going to find an <em>even coverage</em> (aka <em>average coverage</em>), for all the reads in your assembly, in this example that number is 8x.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_42.png" alt="alt text" />
</div>

<p class="blog-post_paragraph"><code>purge_dups</code> will identify retained haplotigs by looking for parts of the genome where there is a drop in coverage. The cutoffs for this drop in coverage can vary (it can be a 2x drop in coverage or it could be more). Unlike its predacessor software <code>purge_haplotigs</code>, <code>purge_dups</code> can identify whole contigs and also <strong>parts of contigs</strong> where this is the case.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_43.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">After <code>purge_dups</code> finds a drop in coverage, it goes ahead and performs a self-BLAST to see if it has some similiarity to some other contig.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_44.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If it finds that that portion of genome has similarity to some other contig, then it's going to delete that portion from the original assembly because it knows that this is a retained haplotype.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_45.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">So by running <code>purge_dups</code>, you are purging out the contigs that are retained haplotigs and you are also purging out parts of your contigs (like tips of your contigs) that there might be haplotypes retained.</p>

<p class="blog-post_paragraph">The Hifiasm assembler has an internal purging, but it is not perfect for highly heterozygous species, so even for Hifiasm which doesn't require you to have a genome size estimate, you still need to have an idea of your estimated genome size for quality control purposes to have an idea if you're dealing with retained haplotypes. <u>The Darwin Tree of Life project runs <code>purge_dups</code> for every single assembly they run with Hifiasm</u> (their main assembler). Then they perform QC using <strong><em>Merqury plots</em></strong> to see if <code>purge_dups</code> has done a good job.</p>

<p class="blog-post_paragraph">For HiCanu, <code>purge_dups</code> will help you parse your primary and your alternate assembly. Then, if you continue the steps in the Darwin Tree of Life Project Pipeline, you will continue the scaffolding steps and curation steps using the primary assembly.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_46.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here is an example of a <em>k</em>-mer spectra plot of a primary hifiasm assembly back to the reads before and after purging.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_47.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">In the beginning the assembly was 3.5 gigabase, which had 3753 contigs. After purging, 300 million bases and a lot of contigs were removed. So purging can automatically remove a large amount of the erroneously assembled genome. If you're plotting only the primary assembly, it means you should have all your <em>k</em>-mers only at 1x coverage. Yet, before purging, you can see a layer of something on the top which are <em>k</em>-mers happening twice (you can mainly see a small purple peak in the plot on the left). This is wrong. <u>These indicate retained haplotypes</u>. Then, when running this <em>k</em>-mer analysis again after purging, you can see that that peak has been nicely removed. This is just the first look into what <em>k</em>-mers can do for you to be able to evaluate your assembly.</p>

<p class="blog-post_paragraph">How do you perform a QC to see if your assembly is well purged? How do you know if you've over-purged and thrown valuable pieces of your genome away? One way to do that is by performing another <em>k</em>-mer analysis using <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02134-9" target="_blank"><strong>Merqury</strong></a>.</p>

<p class="blog-post_paragraph">What Merqury basically does is that it takes the <em>k</em>-mers of you assembly and counts them. It will also count the <em>k</em>-mers of your initial reads (instead of counting <em>k</em>-mers with Jellyfish, <em>k</em>-mers are counted with the </code>meryl</code> program which is a tool inside the merqury program. These <em>k</em>-mers are stored in something called a <em>meryl database</em>). It then compares them looking for the intersection of <em>k</em>-mers and plots them (think of this like a Venn Diagrom. The program looks for <em>k</em>-mers that are in both the assembly and the reads).</p>

<p class="blog-post_paragraph">So baiscally the question you are trying to answer with Merqury is "<u>are all my <em>k</em>-mers in my assembly also in my reads?</u>". If you have a lot of <em>k</em>-mers in your reads that are not present in your assembly, your assembly might be incomplete because they were not all assembled. If on the other hand, you have a lot of <em>k</em>-mers, extra <em>k</em>-mers, or duplicated <em>k</em>-mers in the assembly, when in reality there was supposed to be only one copy in the assembly because there is only one copy of the region (like for example a homozygous region) then you can quickly see that your assembly might contain retained haplotypes.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_48.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here on the right side, you can see a distribution of <em>k</em>-mers in a concatenated primary and alternate assembly. <em>k</em>-mer's either appear 0 times in the assembly (only in the reads), 1 time (<em>k</em>-mers represeting heterozygous regions), 2 times (representing homozygous regions), or 3 or more times (representing repeats). <em>k</em>-mers that appear 1 or more times can also represent <em>k</em>-mers found only in the assembly, although these will be plotted at extremly low coverages, so they will map to the bottom left of the plot and are an easy way to see if there are any errors in the assembly.</p>

<p class="blog-post_paragraph">Below is an example of what a Merqury plot looks like for a diploid organism's <strong>primary assembly only</strong>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_49.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">You can notice that despite being a diploid organism, this Merqury plot only has 1 times coverage of the <em>k</em>-mers. This is because when plotting only a primary assembly, every region of the genome is only plotted once. The haplotypes are absent from this plot. You can see that there is a lot of coverage of <em>k</em>-mers in where the heterozygous peak would be and those are classified as "read only". These <em>k</em>-mers are not actually "read only", but rather the missing <em>k</em>-mers that are found in the alternate assembly.</p>

<p class="blog-post_paragraph">You can also kind of see a <em>little bit</em> of green above the homozygous peak at ~70 times coverage. This is indicative of retained haplotypes in the primary assembly. However, <u>this amount of retention is fine</u>. You're hardly ever going to get perfect genome assembly and this is enough to qualify as a high-quality primary assembly.</p>

<p class="blog-post_paragraph">It's important to understand what the final pool of contigs are in your <code>.fasta</code> file(s) in order to interpret Merqury plots properly. For example, some genome assembly pipelines or consortia will classify their primary and alternate assemblies in different ways. The Darwin Tree of Life project for example will <em>combine</em> the sequences from their alternate assembly from the original assembly run and the sequences that were purged from the primary assembly in a single file (before purging this pool of sequences again). The final pool of sequences is what's plotted in the Merqury plots from their assembly pipelines. Here you can see what the Merqury plot looks like when the primary and alternate assemblies are passed into Merqury as 2 separate files.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_50.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Below are the Merqury output of the same files when they are passed in individually to the Merqury program.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_51.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Back to these graphs, it's important to understand that when performing Genome Quality QC, <u>it's important to perform Merqury QC on every stage of the assembly process to ensure you are going in the correct direction</u>. Here you see Merqury perfromed on the primary assembly before purging and the primary assembly after purging.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_52.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here is an example of what would happen if you <strong>over-purge</strong> your assemblies.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_53.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here you have an example of a Merqury plot for a diploid organism of both haplotypes (the primary and the alternate assembly) together. Because the heterozygous <em>k</em>-mers from both assemblies should be present, you expect to only see one peak at the heterozygous peak location (at 1X coverage). However, you can see a second, overlapping peak at this location corresponding to read-only <em>k</em>-mers. This indicates that you have over-purged as a lot of non-erroneous, <em>k</em>-mers derived from real reads of the organism's genomes are all of a sudden not incorporated into the assembly anymore and are only found in the sequencing reads.</p>

<p class="blog-post_paragraph">So in conclusion, when it comes to purging, in your Merqury plots you can have an extra layer above your peaks with higher coverage (indicating <strong>retained-haplotypes</strong>) or a peak of read-only <em>k</em>-mers (indicating that you <strong>over-purged</strong>). All in all it shows the necessity of performing multiple Merqury QC steps at every step of the assembly. More information on how to evaluate <em>k</em>-mer plots can be found <a href="https://github.com/marbl/merqury/wiki/2.-Overall-k-mer-evaluation" target="_blank">here</a>.</p>

<p class="blog-post_paragraph">The last thing to note about Merqury is that in addition to <em>k</em>-mer spectra plots, Merqury will also output alignment stats to give you <strong>QV</strong> (consensus quality value) estimates in files titled <code>output.qv</code>. QV is an assembly quality metric, based on the number of <em>k</em>-mers that are found only in the assembly but not in the reads (which would imply consensus base pair errors). This is also known as the <strong><em>k</em>-mer survival rate</strong> (i.e., how many <em>k</em>-mers are unaffected by sequencing error). To calculate this, the error rate is first determined. The proportion of unique <em>k</em>-mers in the assembly over the total number of <em>k</em>-mers in the assembly represent the error rate, as they are assumed to have base errors. Then, it is recorded as a log-scaled probability of error for the consensus base calls (Phred Quality Score). The formula for the error rate, with <em>E</em> being error rate is:</p>

<div style="text-align: center;">
  <math xmlns="http://www.w3.org/1998/Math/MathML">
    <mi>QV</mi>
    <mo>=</mo>
    <mo>-</mo>
    <mn>10</mn>
    <mo>&InvisibleTimes;</mo>
    <mi>log</mi>
    <mi>E</mi>
  </math>
</div>
<p class="blog-post_paragraph"></p>
<p class="blog-post_paragraph">This means that <code>Q30</code> corresponds to 99.90% accuracy, <code>Q40</code> to 99.99%, etc.</p>

<p class="blog-post_paragraph">The following table shows column descriptors for an example <em>A. thaliana</em> dataset.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_54.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The above output makes the claim that the assembly is 99.90 - 99.99% accurate.</p>

<p class="blog-post_paragraph">In addition to the <code>output.qv</code> file, Merqury will also output a <code>output.genome.qv</code> file. This file contains the consensus quality of each individual contig. There isn’t too much value with many contigs, and serves more as a diagnostic tool to see any potential contig outliers. It follows a similar format to the whole genome QV. Below are some sample results from the same <em>A. thaliana</em> dataset:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_55.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">In practical sense, when checking the <code>output.qv</code> file, you might see an output like this:</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">head */*.qv</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">==> triocanu_clr/triocanu_.qv <==
col	682359	123511626	35.1183	0.000307729
cvi	506306	122349407	36.3761	0.00023035
Both	1188665	245861033	35.6991	0.00026921</code></pre>

<p class="blog-post_paragraph">Here you can see multiple assemblies in the same output file. This can be the case when you measure different versions of assemblies or primary and alternate versions with Merqury. Merqury will also give you QV estimates of both assemblies combined. Here you can see that these results suggest the <code>cvi</code> assembly has slightly higher QV (36.4) than the <code>col</code> assembly (35.1), but still are very similar. As of writing, <strong>the current estimated QV for assemblies from PacBio HiFi reads is even higher than this (around a QV 40)</strong>. This has been benchmarked to show that SNP calling studies can be accurate.</p>

<p class="blog-post_paragraph">Assembly QV used to be estimated based on mapping based techniques. The problem with these however was that mapping biases occured because you have heterozygosity compounded with errors. You expect each read to map back to their specific haplotypes, especially if they are long reads, however this isn't always the case. Complicated regions, like repetative regions and ribosomal RNA arrays can have multiple mappings and supplemental mappings, and these skew the QV estimates. Some studies exclude these regions and QV estimates can be biased or cherry-picked for better results. <strong><em>k</em>-mer-based QV estimation removes a lot of these priors and offers a more accurate estimate quality of the assembled genomes</strong>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_56.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">However it's important to rememver that genomics is a proactive science. If you are ever insecure of your results, it's often a good idea to use multiple approaches to check the sanity of your assumptions. Be aware of the benefits and short-comings of every method and use every tool at your disposal.</p>

<p class="blog-post_paragraph">After running Merqury QC the next tool you should use is a tool like BUSCO to perform gene completeness estimate QC's. <a href="https://tolqc.cog.sanger.ac.uk/darwin/insects/Trypoxylon_attenuatum/" target="_blank">Click here</a> to see an example of what full QC reports should look like for eukaryotic genome assembly. Formatting all your QC data into a report this way is a convenient way to assemble your genomes as you can perform analyses at every stage of your assembly as shown in the example below:</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_57.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">One more thing to note on performing genome assembly QC is that you will often see <a href="https://github.com/ablab/quast" target="_blank">QUAST</a> used to perform QC on multiple assemblies. However, <strong>QUAST is reference-based</strong>. We live in a time where we are getting quite high resolution with the genomes we produce, which means we are going to have variations. We are going to be able to pick up the total population of variations between the genomes. So if you have a very close sample or the same organism you have sequenced and you want to use QUAST to evaluate because you are just changing the sequencing technology, you can use QUAST and use the reference as a ground truth. However if you have another specimen sequenced, then the QUAST evaluation will be affected by the biology as well. For example, if you take version 38 of the human genome and compare it to the telomere-to-telomere human genome assembly, QUAST is going to give you an output with 20,000 errors on the telomere-to-telomere assembly. However a big part of this is because the assemblies were generated on different human beings with different variation and our technology is able to pick these up very well. <strong>Therefore it's always important to keep reference bias aware in your mind and know which tools are reference-based and which are reference-free and also keep in mind that the reference doesn't mean that it presents the ground-truth of the biology of the organisms</strong>. Genome assemblies are often mosiacs and heuristics to model certain realities in collected from bulk samples in nature.</p>

<p class="blog-post_paragraph">With that background out of the way, these is how I performed Merqury QC on the <em>Urtica urens</em> assembly.</p>

<p class="blog-post_paragraph">I first created an a directory for my QC and linked my primary and alternate assemblies there.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">mkdir ~/drUrtUren1/merqury_hifiasm_eval
cd ~/drUrtUren1/merqury_hifiasm_eval
ln -s /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.hifiasm.total.p_ctg.fa.gz
ln -s /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.hifiasm.total.a_ctg.fa.gz</code></pre>

<p class="blog-post_paragraph">In order to run merqury, you need a meryl datase of the reads you want to compare <em>k</em>-mers with the <em>k</em>-mers in your assembly. In this workshop, the meryl database for <em>Urtica urens</em> PacBio HiFi reads was provided (with <em>k</em>=21). I simply symlinked the meryl database folder to my execution folder.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">ln -s /home/ubuntu/Share/drUrtUren1_data/drUrtUren1.pacbio.k21.meryl/</code></pre>

<p class="blog-post_paragraph">Next I activated my conda environment and checked the Merqury usage.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">conda activate merqury_env
merqury.sh</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Usage: merqury.sh <read-db.meryl> [<mat.meryl> <pat.meryl>] <asm1.fasta> [asm2.fasta] <out>
<read-db.meryl> : k-mer counts of the read set
<mat.meryl> : k-mer counts of the maternal haplotype (ex. mat.hapmer.meryl)
<pat.meryl> : k-mer counts of the paternal haplotype (ex. pat.hapmer.meryl)
<asm1.fasta> : Assembly fasta file (ex. pri.fasta, hap1.fasta or maternal.fasta)
[asm2.fasta] : Additional fasta file (ex. alt.fasta, hap2.fasta or paternal.fasta)
\*asm1.meryl and asm2.meryl will be generated. Avoid using the same names as the hap-mer dbs
<out> : Output prefix
Arang Rhie, 2020-01-29. arrhie@gmail.com
</code></pre>

<p class="blog-post_paragraph">I also investigated what's inside the <code>merqury.sh</code> script</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cat $(which merqury.sh)</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">#! /bin/bash

if [[ "$#" -lt 3 ]]; then
	echo "Usage: merqury.sh <read-db.meryl> [<mat.meryl> <pat.meryl>] <asm1.fasta> [asm2.fasta] <out>"
	echo -e "\t<read-db.meryl>\t: k-mer counts of the read set"
	echo -e "\t<mat.meryl>\t\t: k-mer counts of the maternal haplotype (ex. mat.hapmer.meryl)"
	echo -e "\t<pat.meryl>\t\t: k-mer counts of the paternal haplotype (ex. pat.hapmer.meryl)"
	echo -e "\t<asm1.fasta>\t: Assembly fasta file (ex. pri.fasta, hap1.fasta or maternal.fasta)"
	echo -e "\t[asm2.fasta]\t: Additional fasta file (ex. alt.fasta, hap2.fasta or paternal.fasta)"
	echo -e "\t*asm1.meryl and asm2.meryl will be generated. Avoid using the same names as the hap-mer dbs"
	echo -e "\t<out>\t\t: Output prefix"
	echo -e "Arang Rhie, 2020-01-29. arrhie@gmail.com"
	exit 0
fi

source $MERQURY/util/util.sh

readdb=`link $1`
echo "read: $readdb"
echo

if [[ "$#" -gt 4 ]]; then
	echo "Haplotype dbs provided."
	echo "Running Merqury in trio mode..."
	echo

	hap1=`link $2`
	hap2=`link $3`
	asm1=`link $4`
	echo "hap1: $hap1"
	echo "hap2: $hap2"
	echo "asm1: $asm1"
	if [[ "$#" -eq 5 ]]; then
		out=$5
	else
		asm2=`link $5`
		out=$6
		echo "asm2: $asm2"
	fi

elif [[ "$#" -gt 2 ]]; then
	echo "No haplotype dbs provided."
	echo "Running Merqury in non-trio mode..."
	echo

	asm1=`link $2`
	echo "asm1: $asm1"
	if [[ "$#" -eq 3 ]]; then
		out=$3
	else
		asm2=`link $3`
		out=$4
		echo "asm2: $asm2"
	fi
fi

echo "out : $out"
echo

if [ -e $out ]; then
        echo "$out already exists. Provide a different name. (Are we missing the <out>?)"
        exit -1
fi

mkdir -p logs

echo "
Get spectra-cn plots and QV stats"
name=$out.spectra-cn
log=logs/$name.log
$MERQURY/eval/spectra-cn.sh $readdb $asm1 $asm2 $out > $log 2> $log

if [ -z $hap1 ]; then
	exit 0
fi

echo "
Get blob plots"
name=$out.blob
log=logs/$name.log
$MERQURY/trio/hap_blob.sh $hap1 $hap2 $asm1 $asm2 $out > $log 2> $log

echo "
Get haplotype specfic spectra-cn plots"
name=$out.spectra-hap
log=logs/$name.log
$MERQURY/trio/spectra-hap.sh $readdb $hap1 $hap2 $asm1 $asm2 $out > $log 2> $log

echo "
Get phase blocks"
name=$out.phase-block1
log=logs/$name.log

$MERQURY/trio/phase_block.sh $asm1 $hap1 $hap2 $out.${asm1/.fasta/}  > $log 2> $log
echo

if [ -z $asm2 ] ; then
	echo "Get block N plots"
	name=$out.block_N
	log=logs/$name.log

	$MERQURY/trio/block_n_stats.sh $asm1 $out.${asm1/.fasta/}.*.phased_block.bed $out

	exit 0
fi

name=$out.phase-block2
log=logs/$name.log
$MERQURY/trio/phase_block.sh $asm2 $hap1 $hap2 $out.${asm2/.fasta/}  > $log 2> $log

echo "
Get block N plots"
name=$out.block_N
log=logs/$name.log

$MERQURY/trio/block_n_stats.sh $asm1 $out.${asm1/.fasta/}.*.phased_block.bed $asm2 $out.${asm2/.fasta/}.*.phased_block.bed $out</code></pre>

<p class="blog-post_paragraph">This all shows that in order to run merqury, you need:</p>


<ol>
  <li>
    The meryl database
  </li>
  <li>
    Your assembly (one or two, if you have primary and haplotigs. In our case it's both)
  </li>
  <li>
    An output name.
  </li>
</ol>


<p class="blog-post_paragraph">Therefore I ran:</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">merqury.sh drUrtUren1.pacbio.k21.meryl drUrtUren1.hifiasm.total.p_ctg.fa.gz drUrtUren1.hifiasm.total.a_ctg.fa.gz hifiasm</code></pre>

<p class="blog-post_paragraph">Regarding the Merqury output files, for the plot files, you will have 3 types: <code>st</code> (stacked), <code>ln</code> (line) and <code>fl</code> (filled). These three graphs will represent the same data. They are just a different way of plotting the (same) data. Apart from having <code>st</code>, <code>fn</code> and <code>ln</code>, Merqury is going to output 2 types of files, <code>asm</code> and <code>cn</code>. If the file contains <code>cn</code> it’s going to show you all <em>k</em>-mer counts of primary + haplotigs in the plot, without descriminating them. If it contains <code>asm</code> it is going to show you <em>k</em>-mers belonging to each assembly in different colours. For further information I recommend having a look at the manual <a href="https://github.com/marbl/merqury/wiki/2.-Overall-k-mer-evaluation" target="_blank">here</a>.</p> 

<p class="blog-post_paragraph">
    <u><em>Step 5: Proximity Ligation and Chromosome Conformation Capture</em></u>
  </p> 

<p class="blog-post_paragraph">It's important to remember that in nature, genomes are organized into a highly organized condensed structure made up of DNA, RNA, and proteins that forms the chromosomes in the nucleus of a cell. This structure is known as <strong>chromatin</strong>. Chromatin is formed when long strands of DNA molecules wrap around spool-like complexes of proteins called histones to form structures that resemble beads on a string called nucleosomes (at least in most organisms).</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_58.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This 3D organization is important for gene expression dynamics, but it is also very useful for aiding in assembling genomes. Even though the 3D organization of chromosomes in a nucleus is fluid and constantly changing, there are certain patterns that emerge in DNA organizaiton that can be use to help resolve which regions should be next with other regions. Sometimes distant linear regions of the same chromosome or of other chromosomes tend to stay in close proximity with other distant regions of the genome and this is often done to regulate gene expression. In fact on average, intrachromosomal contact probability is much higher than interchromosomal contact. In other words, <strong>Chromosomes are more likely to make contact with themselves than with other chromosomes</strong>. Another important pattern is that <strong>interaction probability rapidly decays with increasing genomic distance</strong>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_59.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">And additionally, <strong>smaller chromosomes tend to have more interchromosomal contacts than larger chromosomes</strong>. Below is a contact matrix of the human chromosomes mapped to themselves. Without going into too much detail, perfect mapping would be right across the grey diagonal region of the heatmap, with no background signal. However, the background signal indicates reads that are also mapping to other parts of the genome in addition to themsevles. You can see that the smaller chromosomes have a higher intensity of off-target mappings.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_60.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Capturing this information is very useful for resolving gaps and breaks in genome assembly's as contigs can be <strong>scaffolded</strong>. Long distance interactions can help orient contigs and join them together in assemblies.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_61.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This idea is also very useful when performing metagenome assemblies, where the pool of fragmented DNA is even messier. Long distance DNA interactions captured from the same organisms really help to serve as scaffolds for assembling the pool of DNA from a mixture of organisms.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_62.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">In recent history, there have been many Chromatin Capture techniques that have been developped, each with their benefits and limitations. The biggest revolution came from development of the <strong>Hi-C</strong> technique, which allowed for interactions from all regions of the genomes to be screened against all regions from the genome for the first time. A great review about the development of protein capture techniques and their limitations can be found <a href="https://wires.onlinelibrary.wiley.com/doi/10.1002/wdev.395" target="_blank">here</a>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_63.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Hi-C works by the following steps:</p>

<ol>
  <li>Cells are first treated with a chemical crosslinker (<strong>usually formaldehyde</strong>) to freeze interactions between DNA regions that are physically close to each other in the 3D space of the nucleus. This preserves the spatial relationships between different parts of the genome.</li>
  <li>The crosslinked DNA is then cut into smaller fragments using restriction enzymes. These enzymes recognize specific DNA sequences and cleave the DNA at those sites, creating smaller, more manageable pieces.</li>
  <li>The ends of the DNA fragments are repaired and labeled with a small detectable molecule called <strong>biotin</strong>. This step ensures that the ends of the DNA fragments are compatible for ligation and can be later captured for sequencing.</li>
  <li>The DNA fragments are then ligated under dilute conditions, which favors the joining of DNA fragments that are physically close to each other in the nucleus. This step creates chimeric DNA molecules that represent interactions between different regions of the genome.</li>
  <li>The crosslinks are reversed, freeing the DNA from proteins and other molecules. This step also breaks the bonds formed by the crosslinker, which allows the DNA to be processed further.</li>
  <li>The DNA is purified and then sheared into smaller fragments, which are more suitable for sequencing.</li>
  <li>The biotinylated DNA fragments are captured using streptavidin beads, which bind tightly to biotin. This step enriches for the chimeric DNA molecules that represent interactions between different genomic regions.</li>
  <li>The captured DNA fragments are used to prepare a sequencing library (usually as short read-paired ends using Illumina sequencing).
    <ul>
      <li>Paired ends can be thousands of bases apart on a linear DNA sequence.</li>
    </ul>
  </li>
  <li>The sequencing data is analyzed to map the interactions between different regions of the genome.</li>
</ol>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_64.png" alt="alt text" />
</div>
<p class="blog-post_paragraph">The result of Hi-C Chromatin capture data is is a contact matrix (or a contact heatmap), where each entry represents the frequency of interactions between two genomic loci. Different micro-architectures in the 3D conformation of the genome (e.g. chromatin loops, topologically associating domains (TADs), and sub-TADs) will have different contact signatures on the contact matrix.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_65.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Although Hi-C has been used as the "Gold-standard" for chromosome conformation capture technology, researchers began reaching the limits of its ability to capture fine-resolution (~1 kb) architectures. Newer technologies such as Micro-C have been able to increase the resolution of chromatin conformation capture. You can see in the plot less gaps in contact (white crosses) appearing when the sample is analyzed using Micro-C instead of Hi-C. This is incredibly useful for epigenetic studies, however for the purposes of genome assembly, these higher resolution standards might be overkill as collecting this data may be very expensive and the Hi-C data is often sufficient for scaffolding a genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_66.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Nanopore sequencing based chromatin capture techniques are also being developped, such as Pore-C and HiPore-C. These techinques are very promising because they offer similar chromatin capture accuracy as Hi-C, however they claim to offer cheaper and streamlined data collection and analysis.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_67.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Additionally, there is another promising technique called <a href="https://omni-c.readthedocs.io/en/latest/" target="_blank">Omni-C</a>. Omni-C improves upon the Hi-C protocol because <strong>it does not depend on restriction enzymes to digest the crosslinked DNA</strong>. Instead it uses a sequence-independent endonuclease. This reduces the sequence bias of the overall library because restriction biases can only digest in certain areas of the genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_68.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">To perform this scaffolding, you generally need a scaffolding algorithm. One of the most common Hi-C scaffolding algorithms is called <a href="https://academic.oup.com/bioinformatics/article/39/1/btac808/6917071?login=true" target="_blank">YaHS</a>. Reduced to it's most basic form, <strong>this algorithm basically just takes all the mappings from the paired-end reads and tries to find the regions that have the most mappings</strong>. Remember that when mapping softwares do the mapping, there is often a primary mapping as well as many supplemental mappings with lower confidence for each portion of mapped DNA.</p>

<p class="blog-post_paragraph">Below is an example of how to interpret a contact matrix of a draft assembly.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_69.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">You can see two main things. Hi-C contact maps are a very easy way to identify potential sex chromsomes in diploid or even polyploid samples, as the background mapping coverage in those DNA scaffolds or contigs will tend to be around half of what is seen across all the other background samples. Of course these contigs/scaffolds need to be analyzed further to check if they have sex-determining genes present, but this is a good way to get candidate sequences to look at. Another things is that you can see very little off-diagonal mapping in most of the scaffolds/contigs, except at the smaller fragments. This indicates that the draft assembly was very successful and very little manual curation needs to be done for this sample.</p>

<p class="blog-post_paragraph">Here is another good example of sex chromosome identification using Hi-C contact maps.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_70.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If you look at annotation <strong>A</strong> in this diagram, you can see that this contigs/scaffold has 2 unique features.</p>

<ol>
  <li>It has less background mapping than the other contigs/scaffold</li>
  <li>It has a large amount of off-diagonal signal with another, smaller contig/scaffold in particular</li>
</ol>

<p class="blog-post_paragraph">Which chromosomes in the genome share a lot of similarity with each other but are also expected contain half the mapping coverage of other chromosomes? <strong>The sex chromosomes</strong>. Because there are 2 contigs/scaffolds of <strong>different sizes</strong> that share a lot of similarity with each other, these are likely X and Y chromosomes. <strong>Using just a Hi-C map on a non-curated genome, we could quickly deduce that this sample was collected from a male organism</strong>.</p>

<p class="blog-post_paragraph">Annotation <strong>B</strong> is showing some chromosome abberation in the assembly. These aberations can be fixed through manual curation (<a href="https://gitlab.com/wtsi-grit/rapid-curation/-/blob/main/Interpreting_HiC_Maps_guide.pdf" target="_blank">see here for more information</a>).</p>

<p class="blog-post_paragraph">Here is another example of a contact map for another organism's draft assembly.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_71.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here you can see more strong off-target mapping that likely needs manual creation, but in additon you can notice contigs that don't have any Hi-C mapping data mapped to them at all. As the figure explains, this could be due to contamination or incorrect preparation of your Hi-C library.</p>

<p class="blog-post_paragraph">Contamination in your assembly can also manifest as sequences with higher sequencing coverage but also less off-diagonal chromatin-capture mapping to the rest of the genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_72.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Hi-C maps are great for finding misassemblies in certain contigs or scaffolds. These sequences should have strange distribution of Hi-C mappings within their sequences and would likely need manual fixing.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_73.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Another genome assembly phenomenon that you can encounter are <strong>imposter sequences</strong>. These are sequences that somehow got incorporated to your contigs or scaffolds but will share no mappings with any other sequences.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_74.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Another thing you might encounter is <strong>retained haplotigs</strong> in your assemblies. Even after using <code>purge_dups</code> to try and remove these, some of these assembly artificats can bypass detection by the software and still persist in your genome. The Hi-C maps can quickly identify them and allow you to fix them manually.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_75.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Hi-C maps can also give you a good indication of the completeness of your genomes because (in probably most eukaryotic organisms), you should see a nice, repetative, off-diagonal mapping pattern accross the middle and ends of all of your final chromosomes, indicating the shared similarity of centromeric and telomeric sequences across chromosomes.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_76.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">There are many more ways Hi-C can help understand your genome and it is too much to all list here. More information on how to interpret Hi-C maps can be found <a href="https://gitlab.com/wtsi-grit/rapid-curation/-/blob/main/Interpreting_HiC_Maps_guide.pdf" target="_blank">in this document</a>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_77.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">One good way to think of Hi-C maps is as <strong>modern day Karyotypes</strong>. They can quickly show you the number of- and size of the chromosomes in an organisms genome. But in addition, they can quickly reveal a lot of other genomic patterns and unique characteristics of your organism's biology.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_78.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">We tend to take certain things in biology for granted: "Centromeres are very repetative", "Telomeres are organized in specific ways", etc. Now with these chromatin-capture technologies we are able to do this high-resolution biology that we were never able to do before. We can now easily evaluate how our previous biological notions hold up across the tree of life.</p>

<p class="blog-post_paragraph"><strong>Scaffolding using some sort of chromatin-capture technology is a very important ingredient of creating a high-quality genome assembly</strong>. Not only for the myriads of benefits of having better data for identifying contamination, misassemblies, and genome features like sex chromosomes, but also just on a basic level whereby proximity-ligation-based scaffolding can significantly improve the contiguity of your genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_79.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Here you can see an example where just by performing Hi-C scaffolding on an assembly using short-reads, the researchers were able to go from a contig N50 of 64,321 to a scaffold N50 of 140,950,122. <strong>This is a over 2000-fold increase in genome contiguity</strong>.</p>

<p class="blog-post_paragraph">After running a scaffolding software, there genome will still likely not be complete and will need some human intervention for refinement. This is called manual curation. <u>Manual curation should be done using <a href="https://github.com/wtsi-hpag/PretextView" target="_blank">PretextView</a></u>. This is an app that allows inspection of Hi-C reads mapped to an assembly.</p>

<p class="blog-post_paragraph">PretextView can be used to make real-time changes to assemblies:</p>

<ol>
  <li>It can be used to break and join scaffolds.</li>
  <li>It can be used to re-orient scaffolds or parts thereof.</li>
  <li>Additional tracks (bedgraphs) can be injected into Pretext maps to view alongside the Hi-C data, e.g., Coverage, Telomeres, Sequence gaps, Repeat density.</li>
</ol>

<p class="blog-post_paragraph">PretextView can also be used to dump an <code>agp</code> (describes the assembly of a larger sequence object from smaller objects) of changes to an assembly. This <code>agp</code> can then be passed to our in-house script <code>rapid_pretext2tpf.py</code>, which will allow the edited map to be turned back into a corrected fasta file.</p>

<p class="blog-post_paragraph">Some caveats to be aware of are:</p>

<ul>
  <li>PretextView is limited by resolution.</li>
  <li>Mis-assemblies may be hidden in large genomes.</li>
  <li>Very small scaffolds may not have sufficient resolution to correctly determine placement/orientation.</li>
</ul>

<p class="blog-post_paragraph">Here is an example of a real manual curation step, of a very complicated marine organism.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_80.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">As you can see, even after manual curation, there is still a lot of small, off-target mappings in the fragments. Genome assembly is hardly ever perfect, and for a lot of organisms, you aren't striving to spend weeks upon weeks to create the perfect assembly and refine out every little innaccuracy in all of your smallest fragments. At some point you will just have to say "<em>khallas</em> this is good enough for what I need".</p>

<p class="blog-post_paragraph">One of the softwares that is used to scaffold genomes with Hi-C that is called <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-3879-z" target="_blank">SALSA2</a>. In the workshop, this is what I used.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">mkdir ~/drUrtUren1/scaff
cd ~/drUrtUren1/scaff
conda activate salsa
run_pipeline.py -h</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">usage: run_pipeline.py [-h] -a ASSEMBLY -l LENGTH -b BED [-o OUTPUT]
                       [-c CUTOFF] [-g GFA] [-e ENZYME] [-i ITER] [-x DUP]
                       [-s EXP] [-m CLEAN] [-f FILTER] [-p PRNT]

SALSA Iterative Pipeline

optional arguments:
  -h, --help            show this help message and exit
  -a ASSEMBLY, --assembly ASSEMBLY
                        Path to initial assembly
  -l LENGTH, --length LENGTH
                        Length of contigs at start
  -b BED, --bed BED     Bed file of alignments sorted by read names
  -o OUTPUT, --output OUTPUT
                        Output directory to put results
  -c CUTOFF, --cutoff CUTOFF
                        Minimum contig length to scaffold, default=1000
  -g GFA, --gfa GFA     GFA file for assembly
  -e ENZYME, --enzyme ENZYME
                        Restriction Enzyme used for experiment
  -i ITER, --iter ITER  Number of iterations to run, default = 3
  -x DUP, --dup DUP     File containing duplicated contig information
  -s EXP, --exp EXP     Expected Genome size of the assembled genome
  -m CLEAN, --clean CLEAN
                        Set this option to "yes" if you want to find
                        misassemblies in input assembly
  -f FILTER, --filter FILTER
                        Filter bed file for contigs present in the assembly
  -p PRNT, --prnt PRNT  Set this option to "yes" if you want to output the
                        scaffolds sequence and agp file for each iteration</code></pre>

<p class="blog-post_paragraph">To run SALSA2 you need 3 files. The files are a <code>*purged.polish.fa</code> file, a <code>*purged.polish.fa.fai</code> (fasta index), and <code>merge.mkdup.bed</code> (A BED file that contains information about the alignment of sequencing reads to the reference genome). I symlinked all these files to my execution directory and ran:</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">ln -s /home/ubuntu/Share/drUrtUren1_data/HiC/drUrtUren1_hifiasm.purged.polish.fa
ln -s /home/ubuntu/Share/drUrtUren1_data/HiC/drUrtUren1_hifiasm.purged.polish.fa.fai
ln -s /home/ubuntu/Share/drUrtUren1_data/HiC/merge.mkdup.bed
run_pipeline.py -a drUrtUren1_hifiasm.purged.polish.fa -l drUrtUren1_hifiasm.purged.polish.fa.fai -b merge.mkdup.bed</code></pre>

<p class="blog-post_paragraph">And now I ran the following scripts to compare the assembly statistics between the pre-scaffolded assembly and the scaffolded assembly.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">export PATH=$PATH:/home/ubuntu/Share/scripts
conda activate eukaryotic_genome_assembly
asmstats drUrtUren1_hifiasm.purged.polish.fa > drUrtUren1_hifiasm.purged.polish.fa.stats
asmstats scaffolds_FINAL.fasta > scaffolds_FINAL.fasta.stats</code></pre>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cat drUrtUren1_hifiasm.purged.polish.fa.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">drUrtUren1_hifiasm.purged.polish.fa
A = 108642479 (31.8%), C = 62019190 (18.2%), G = 61995971 (18.2%), T = 108760884 (31.9%), CpG = 22963896 (6.7%)
sum = 341418524, n = 413, mean = 826679.234866828, largest = 39522602, smallest = 2262
N50 = 20314943, L50 = 7
N60 = 17272083, L60 = 9
N70 = 15791032, L70 = 11
N80 = 11606222, L80 = 13
N90 = 2266649, L90 = 18
N100 = 2262, L100 = 413</code></pre>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cat scaffolds_FINAL.fasta.stats</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">scaffolds_FINAL.fasta
COMPOSITION	A = 108660503 (31.8%), C = 62042070 (18.2%), G = 61973091 (18.2%), T = 108742860 (31.8%), N = 16000 (0.0%), CpG = 22963896 (6.7%)
SCAFFOLD	sum = 341434524, n = 384, mean = 889152.40625, largest = 51573973, smallest = 2262
SCAFFOLD	N50 = 24132303, L50 = 5
SCAFFOLD	N60 = 20314943, L60 = 7
SCAFFOLD	N70 = 19854874, L70 = 8
SCAFFOLD	N80 = 17272083, L80 = 10
SCAFFOLD	N90 = 2266649, L90 = 15
SCAFFOLD	N100 = 2262, L100 = 384
CONTIG	sum = 341418524, n = 416, mean = 820717.605769231, largest = 39522602, smallest = 2262
CONTIG	N50 = 19854874, L50 = 7
CONTIG	N60 = 17272083, L60 = 9
CONTIG	N70 = 15791032, L70 = 11
CONTIG	N80 = 11133947, L80 = 14
CONTIG	N90 = 2266649, L90 = 19
CONTIG	N100 = 2262, L100 = 416
GAP	sum = 16000, n = 32, mean = 500, largest = 500, smallest = 500</code></pre>

<p class="blog-post_paragraph">These files, as well as a curated one, were visualized the <a href="https://www.aidenlab.org/juicebox/" target="_blank">Juicebox</a> software webserver. On the site, I clicked <code>Load Map</code> > <code>Local File</code> and uploaded my <code>.hic</code> files.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_81.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">In the workshop we also performed a misassembly investigation using Hi-C maps and IGV, for this purpose, a <em>Eristalis pertinax</em> scaffolded assembly was provided and visualized in Juicebox.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_82.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">As you can see, there is a strange distribution of mappings in scaffold 1. This region (scaffold_1:56,972,207-56,977,386) was isolated and viewed in IGV.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_83.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">As you can see, there is a gap in the Pacbio coverage at this region. Therefore, since we have a weird looking Hi-C Heatmap and no reads crossing that region, <strong>we can very safetly assume this is a misassembly</strong>.</p>

<p class="blog-post_paragraph">So overall, <u>this is basically all you need to create a high-quality, chromosome-level nuclear genome assembly. High quality reads and chromatin proximity ligation data.</u> Telomere-to-telomere genome assemblies still need the added ingredient of Ultra Long reads (Have to be Oxford Nanopore Reads as of right now), however this may change soon with how quickly this field evolves. RNA-Sequencing needs to be done, especially in eukaryotic organisms, to produce good gene models for the organism</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_84.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">
    <u><em>Step 6: Organelle Genome Assembly</em></u>
  </p> 



<p class="blog-post_paragraph">The final thing to take note of is organelle genome assemblies. Every time you isolate DNA from an organism, a large percentage of the reads will come from the organelle genomes. Each organelle (mitochondria and chlroplasts (in photosynthetic eukaryotes)) typically contains multiple copies of their own genome. The number of copies can vary depending on the cell type, organism, and metabolic demands, but as a general rule of thumb for well studied organisms, <strong>a single mitochondrion may contain 2 to 10 copies of mitochondrial DNA (mtDNA) and a single chloroplast can contain 10 to 100 copies of chloroplast DNA (cpDNA)</strong>. In addition, a cell can have tenss, to hundreds, to thousands of organelles per cell, <strong>the overall number of organellular DNA copies per cell can range from hundreds to tens of thousands</strong>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_85.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">A lot of assembly programs will output files that end in <code>.gfa</code>. These are <strong>Graphical Fragment Assembly files</strong>. These files contain information from the <em>graph space</em> of the genome assembly (which is like the intermediate version of the assembly, before the assembly is linearized). A <code>.gfa</code> file consists of a set of sequences and tab-delimited pairs of sequences with their overlaps. You can read more about this format from <a href="http://gfa-spec.github.io/GFA-spec/GFA1.html" target="_blank">here</a>. These graphical fragment assembly files can be visualized in programs like <a href="https://rrwick.github.io/Bandage/" target="_blank">Bandage</a>, which help you visualize assembly graphs of your genomes. An assembly graph is used to represent the final assembly of a genome (or metagenomes). In simple terms, <strong>the assembler builds this assembly graph based on reads and their overlap information. Then the assembler resolves paths across the assembly graph and outputs non-branching paths as contigs</strong>.</p>

<p class="blog-post_paragraph">Genome assembly graphs in Bandage can be very confusing to interpret at first. I recommend <a href="https://github.com/rrwick/Bandage/wiki" target="_blank">reading the Bandage wiki</a> and <a href="https://www.youtube.com/playlist?list=PLJU3mQyt10C3rv3SMCF44mVy64GuZpqPT" target="_blank">watching these videos</a> before trying to work with it.</p>

<p class="blog-post_paragraph">Here you can see what a typical graph structure of a chloroplast genome assembly looks like.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_86.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The way to basically interpret this is as follows</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_87.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This is an example of a resolved DNA structure. The way to think about this is to imagine tracing a single path through the graph and then linearizing that trace. This results in a single, circular sequence of DNA.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_88.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">If you think about it, because there are 3 nodes and 4 edges in the above graph, with one of the nodes coming from collapsed repeated sequences, there are 2 ways to traverse the graph above: two separate circular sequences that share a region in common, or a single larger circular sequence with an element that occurs twice.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_89.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">It's important to always take into account the biology of your organism when interpretting this graph. If you combine this information with additional knowledge about chloroplast structures and how they contain long, inverted repeats, then you can rule out the first scenario and deduce that the real structure of the bacterial genome would be the second option. In this way, Bandage has assisted in turning a fragmented assembly of three contigs into a completed genome of one sequence.</p>

<p class="blog-post_paragraph"><strong>Assembly of the chloroplast genome, especially in plants, is usually a very streamlined process</strong>. There are usually no real issues with it and softwares like <a href="https://github.com/c-zhou/oatk" target="_blank">Oatk</a> are really good at assembling them. However, <strong>mitochondrial genome assembly can still be a complicated challenge</strong>. Mitochondrial biology can be more diverse than chloroplast biology generally. Here is just a quick rundown of many differences between mitochondria found in plants and animals.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_90.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This variety extends past animals and plants. Across the Eukaryotic tree of life there is a wide variety of genetic architecture. You will often find many differences in intergenic regions, number and length of introns, the presence and number of pseudogenes, etc., between organisms.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_91.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">However despite the differences in structure between mitochondria of different organisms, their genetic function is well conserved across many Eukaryotic organisms (except for in anaerobic eukaryotes whose mitochondria have evolved into <a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1550-7408.2009.00407.x" target="_blank">hydrogenosomes or mitosomes</a>). There is also no correlation between genome size and gene content</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_92.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Below is an example of how these genes are layed out on a mitochondrial genome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_93.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">It's also important to remember that mitochondria exhibit <strong>heteroplasmy</strong>, which is the presence of more than one type of mtDNA in an individual. This was thought to be a rare phenomenon that was associated with aging and disease, but with modern sequencing technology we are realising that it is quite common amongst many taxa. Heterplasmy can exist at the level of different mtDNA between cells of an organism but also different mtDNA within the same cell.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_94.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Another important concept for mitochondrial genome assembly is <strong>NUMTs</strong>, which stands for <strong>Nuclear Mitochondrial DNA segments</strong>. These are created by the transfer of portions of the mitochondrial genome to the nuclear genome. Because this is usually an evolutionary process that takes a lot of time, the original mitochondrial genome that's inserted into the nuclear genome tend to go through mutations over the generations and becomes diversified away from the original sequence. However, there is still some similarity which makes some software and analytical tools that try to find mitochondrial sequences confused and they predict that NUMTs are really the mitochondrial sequences.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_95.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">This trend is not limited to evolutionary DNA transfer from the mitochondria to the nucleus. In eukaryotic biology, the general trend is summarized below.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_96.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Because of all of these peculiarities, specialized softwares to assemble mitochondrial genomes have been developped that take all of these factors into consideration. One of these is <a href="https://github.com/marcelauliano/MitoHiFi" target="_blank">MitoHiFi</a>.</p>

<p class="blog-post_paragraph">As a broad outline, MitoHiFi works as follows:</p>

<ol>
  <li>
    Assemble the genome
  </li>
  <li>
    Filter out NUMTs / incomplete assemblies
  </li>
  <li>
    Remove redundancy
  </li>
  <li>
    Annotate the genome
  </li>
</ol>

<p class="blog-post_paragraph">More specifically, this is how the MitoHiFi workflow looks like. To run MitoHiFi, you will always need to have a reference genome. This will probably need to be a mitogenome of a closely related organism. The MitoHiFi program comes with a script that helps you find a closely-related reference mitogenome called <code>findMitoReference.py</code>. MitoHiFi can also be run in two different modes: The first one is the <strong>reads mode</strong> (<code>-r</code>). Here you can input you PacBio HiFi reads or your Nanopore high-quality reads together with the reference mitogenome. Then the reads are mapped to the reference genome and only the reads with high-quality mappings are retained because these reads that likely came from the mitochondrial DNA. These reads are then fed into <code>hifiasm</code> to assemble contigs. The contigs are then aligned to the reference mitogenome, and then only the best aligning contigs are retained. The selected contigs are circularized and then annotated with either <code>MitoFinder</code> or <code>MITOS</code>. You can also run this in <strong>contig mode</strong> (<code>-c</code>), so if you already have an initial set of contigs then this you just skip the <code>hifiasm</code> assembly and go straight to the BLAST stage.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_97.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">To run <code>findMitoReference.py</code>, you provide as input the name of your species, your email, an output directory and a minimum mitogenome reference length to ensure that the mitogenome you get back is not an incomplete or partial mitogenome (<strong>suggested to set this to around 15000 for an animal mitogenome</strong>).</p>

<p class="blog-post_paragraph">To run the pipeline (using the main <code>mitohifi.py</code> script), provide your reads or contigs using the <code>-r</code> or <code>-c</code> flags, respectively (for the reads you can input multiple <code>.fasta</code> files). You also need to specify the reference mitogenome with the <code>-f</code> flag and also a GenBank file of the reference mitogenome with the <code>-g</code> flag. <code>-t</code> is used to specificy the number of threads to run this program with. You also need to specify the organism's genetic code following the <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi" target="_blank">NCBI table</a> using the <code>-o</code> flag (for the mitogenome annotation step). The inputs for this parameter are:</p>

<ul>
      <li><code>1</code> - The Standard Code</li>
      <li><code>2</code> - The Vertebrate Mitochondrial Code</li>
      <li><code>3</code> - The Yeast Mitochondrial Code</li>
      <li><code>4</code> - The Mold, Protozoan, and Coelenterate Mitochondrial Code and the Mycoplasma/Spiroplasma Code</li>
      <li><code>5</code> - The Invertebrate Mitochondrial Code</li>
      <li><code>6</code> - The Ciliate, Dasycladacean and Hexamita Nuclear Code</li>
      <li><code>9</code> - The Echinoderm and Flatworm Mitochondrial Code</li>
      <li><code>10</code> - The Euplotid Nuclear Code</li>
      <li><code>11</code> - The Bacterial, Archaeal and Plant Plastid Code</li>
      <li><code>12</code> - The Alternative Yeast Nuclear Code</li>
      <li><code>13</code> - The Ascidian Mitochondrial Code</li>
      <li><code>14</code> - The Alternative Flatworm Mitochondrial Code</li>
      <li><code>16</code> - Chlorophycean Mitochondrial Code</li>
      <li><code>21</code> - Trematode Mitochondrial Code</li>
      <li><code>22</code> - Scenedesmus obliquus Mitochondrial Code</li>
      <li><code>23</code> - Thraustochytrium Mitochondrial Code</li>
      <li><code>24</code> - Pterobranchia Mitochondrial Code</li>
      <li><code>25</code> - Candidate Division SR1 and Gracilibacteria Code</li>
</ul>

<p class="blog-post_paragraph">If you ever need to view the help page or view more options, simply run <code>python mitohifi.py -h</code>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_98.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">More information on how MitoHiFi works under the hood is shown below. First, if you are using read mode, reads are mapped to the reference and mapped reads are assembled with hifiasm to produce contigs. These contigs (or the ones you provided in contig mode) are then BLASTed to the reference mitogenome, and then the results of the alignment are parsed and filtered according to three different criterea:</p>


<ol>
  <li>If the length of the query (contig) is <code>5</code> times higher than the size of the subject (reference mitogenome), then it is filtered out because this contig likely represents a NUMT, not a mitogenome</li>
  <li>If the coverage of the query (contig) is <code>80%</code> smaller than the length of the subject (reference mitogenome), then it is filtered out because it suggests that your contig is an incomplete or partial mitogenome.</li>
  <li>If the percentage of the query (contig) in the BLAST match is smaller than <code>50%</code>, the contig is filtered out because you don't have strong evidence that this contig really is a mitogenome.</li>
</ol>

<p class="blog-post_paragraph">You can tweak all of these parameters from their default values. The most important parameter out of all of these to be aware of is the last one, which you can change with the <code>-p</code> flag. If you are dealing with mitogenomes from taxa that are usually well conserved, like mammals or vertebrates, then you can likely increase this default <code>50%</code> filter to something higher like <code>70%</code> or <code>80%</code>. The default value of <code>50%</code> was chosen because the Darwin Tree of Life team noticed that for some invertebrate mitogenomes, a lot of their mitogenomes are really dissimilar to each other, especially in the repeated regions, and that sometimes picking a very stringent value would cause the researcher to <em>overfilter</em> real mitogenomes that were just divergent sequence-wise.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_99.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Once the pool of contigs are filtered, the program then proceeds to the circularization step. Because of the circular nature of the mitochondrial genome, you will have sequencing reads that support overlaps all across the sequence. So for example, if you look at the diagram below, you will have many reads that support connections from regions <code>A <-> B</code>, <code>B <-> C</code>, and <code>C <-> A</code>. Sometimes, during the assembly process, the assemblers can get confused and you might end up having a linear contig that contains the same sequence at the start and end positions. So sometimes with circular DNA you will get an assembly artifact where you get a contig with redundant information.</p>

<p class="blog-post_paragraph">To solve this issue, MitoHiFi performs a <strong>self-blast</strong> on the contigs. If a contig has a redundant region at the beginning and the end, the region at the end is deleted. This is performed using the script inside <code>circularizationCheck.py</code>.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_100.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The next thing that MitoHiFi does is <strong>rotate</strong> the contig. This is to ensure that all the potential mitogenomes have the same start location. The creators of MitoHiFi decided to standardize this so at the mitogenome sequence starts that the <strong>tRNA-Phe</strong> coding gene. However, in order rotate the sequence so that it starts at this gene, the contig has to be annotated first, The mitogenome is annotated using the GenBank information of the reference mitogenome.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_101.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">The last step in the MitoHiFi pipeline is to take all the potential mitogenomes that have remained passed the previous filters, and then choose the <strong>final mitogenome</strong>. The final mitogenome is that one that satisfies these three criteria:</p>

<ol>
  <li>The genome is circularized</li>
  <li>The genome is similar to the reference mitogenome (in overall size)</li>
  <li>There are no frameshifts in the annotated genes.</li>
</ol> 

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_102.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">Once MitoHifi finishes running, you expect to have these <strong>3 main files as outputs</strong>: <code>final_mitogenome.fasta</code>, <code>final_mitogenome.gb</code>, and <code>final_mitogenome.annotation.png</code>. If you are running MitoHifi in the reads mode (<code>-r</code>), you will also have another important file called <code>final_mitogenome.coverage.png</code>. This is a graph that shows the sequencing coverage of the reads. This is a sanity check to see if there could have been any drops in coverage across the genome that could have caused a misassembly of the final mitogenome. You will likely see a drop in coverage at the beginning and the end of the mitogenome but this is okay. This is a common artifact of a mitogenome assembly</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_103.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">However there are also more files that allow you to perform a more in-depth analysis MitoHiFi. These files let you view details about the <strong>other contigs that made it to the last step of the pipeline but were not selected as the final mitogenome</strong> (other potential mitogenomes). <code>shared_genes.tsv</code> is also a useful output to check the similarity between your final mitogenome and the mitogenome used as the reference.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_104.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">And also there are these files that in most cases are probably too much information but are present regardless if you need to investigate further.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_105.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">An alternative to MitoHiFi is <a href="https://github.com/c-zhou/oatk" target="_blank">Oatk</a>. This program is especially useful for chloropast genomes, but can also be use for assembling mitogenomes as well.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_106.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">A final note on mitogenome assembly is that organelle genome assembly is a thing that is kept in mind during the entire genome assembly process. For example, if you want to run MitoHiFi in contig mode, the best time do so would be right after you perform the initial assembly step. This is because other programs in the assembly pipeline like <code>purge_dups</code> might purge out the mitogenome contigs because the mitochondria has a different coverage to the nuclear genome. For the Darwin Tree of Life Genome Assembly pipeline specifically, there is another step that was not really touched on here which is contamination filtering before manual curation. They have <a href="https://pipelines.tol.sanger.ac.uk/blobtoolkit" target="_blank">their own pipeline that uses <code>blobtoolkit</code></a> and they use that to find mitochondrial contigs and other organelle contigs that they then remove. The organelle genomes are then assembled seperated from the nuclear genomes at the end. The final genome <code>.fasta</code> files that are deposited in databases like the NCBI will have the organelle genomes concatenated to the end of the file but it's important to be aware that these are all assembled seperately.</p>

<p class="blog-post_paragraph">Back to the workshop, this are the commands performed to perform Mitogenome assembly with MitoHiFi. We performed this assembly on <em>Phalera bucephala</em>. First I prepared my working environemnt and looked for a reference mitogenome.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">mkdir mitohifi_test_01/
cd mitohifi_test_01/
conda activate mitohifi_env
export PATH=$PATH:/home/ubuntu/Share/softwares/MitoHiFi/src/
findMitoReference.py --species "Phalera bucephala" --email mark.pampuch@kaust.edu.sa --outfolder refData --min_length 15000</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">Looking for mitochondrion for Phalera bucephala
Looking for an appropriate organelle among Phalera bucephala
Looking for an appropriate organelle among Phalera
output is written to refData/OQ830676.1.[gb,fasta]</code></pre>

<p class="blog-post_paragraph">Next I got the <code>mitohifi.py</code> ready for execution in my execution directory.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">export PATH=$PATH:/home/ubuntu/Share/softwares/MitoFinder/
ln -s /home/ubuntu/Share/softwares/MitoHiFi/src/mitohifi.py
cp /home/ubuntu/Share/MitoHiFi_data/test.fa .
python mitohifi.py -h</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">usage: MitoHiFi [-h] (-r &lt;reads&gt;.fasta | -c &lt;contigs&gt;.fasta) -f
                &lt;relatedMito&gt;.fasta -g &lt;relatedMito&gt;.gbk -t &lt;THREADS&gt; [-d]
                [-a {animal,plant,fungi}] [-p &lt;PERC&gt;] [-m &lt;BLOOM FILTER&gt;]
                [--max-read-len MAX_READ_LEN] [--mitos]
                [--circular-size CIRCULAR_SIZE]
                [--circular-offset CIRCULAR_OFFSET] [-winSize WINSIZE]
                [-covMap COVMAP] [-v] [-o &lt;GENETIC CODE&gt;]

required arguments:
  -r &lt;reads&gt;.fasta      -r: Pacbio Hifi Reads from your species
  -c &lt;contigs&gt;.fasta    -c: Assembled fasta contigs/scaffolds to be searched
                        to find mitogenome
  -f &lt;relatedMito&gt;.fasta
                        -f: Close-related Mitogenome is fasta format
  -g &lt;relatedMito&gt;.gbk  -k: Close-related species Mitogenome in genebank
                        format
  -t &lt;THREADS&gt;          -t: Number of threads for (i) hifiasm and (ii) the
                        blast search

optional arguments:
  -d                    -d: debug mode to output additional info on log
  -a {animal,plant,fungi}
                        -a: Choose between animal (default) or plant
  -p &lt;PERC&gt;             -p: Percentage of query in the blast match with close-
                        related mito
  -m &lt;BLOOM FILTER&gt;     -m: Number of bits for HiFiasm bloom filter [it maps
                        to -f in HiFiasm] (default = 0)
  --max-read-len MAX_READ_LEN
                        Maximum lenght of read relative to related mito
                        (default = 1.0x related mito length)
  --mitos               Use MITOS2 for annotation (opposed to default
                        MitoFinder
  --circular-size CIRCULAR_SIZE
                        Size to consider when checking for circularization
  --circular-offset CIRCULAR_OFFSET
                        Offset from start and finish to consider when looking
                        for circularization
  -winSize WINSIZE      Size of windows to calculate coverage over the
                        final_mitogenom
  -covMap COVMAP        Minimum mapping quality to filter reads when building
                        final coverage plot
  -v, --version         show program's version number and exit
  -o &lt;GENETIC CODE&gt;     -o: Organism genetic code following NCBI table (for
                        mitogenome annotation): 1. The Standard Code 2. The
                        Vertebrate MitochondrialCode 3. The Yeast
                        Mitochondrial Code 4. The Mold,Protozoan, and
                        Coelenterate Mitochondrial Code and the
                        Mycoplasma/Spiroplasma Code 5. The Invertebrate
                        Mitochondrial Code 6. The Ciliate, Dasycladacean and
                        Hexamita Nuclear Code 9. The Echinoderm and Flatworm
                        Mitochondrial Code 10. The Euplotid Nuclear Code 11.
                        The Bacterial, Archaeal and Plant Plastid Code 12. The
                        Alternative Yeast Nuclear Code 13. The Ascidian
                        Mitochondrial Code 14. The Alternative Flatworm
                        Mitochondrial Code 16. Chlorophycean Mitochondrial
                        Code 21. Trematode Mitochondrial Code 22. Scenedesmus
                        obliquus Mitochondrial Code 23. Thraustochytrium
                        Mitochondrial Code 24. Pterobranchia Mitochondrial
                        Code 25. Candidate Division SR1 and Gracilibacteria
                        Code</code></pre>

<p class="blog-post_paragraph">I ran <code>mitohifi.py</code> in contigs mode and using the reference mitogenome that I found, as well as using the Invertebrate Mitochondrial genetic code (<code>-o 5</code>).</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">python mitohifi.py -c test.fa -f refData/OQ830676.1.fasta -g refData/OQ830676.1.gb -t 1 -o 5</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">2024-11-08 16:42:12 [INFO] Welcome to MitoHifi v2. Starting pipeline...
2024-11-08 16:42:12 [INFO] Length of related mitogenome is: 15397 bp
2024-11-08 16:42:12 [INFO] Number of genes on related mitogenome: 37
2024-11-08 16:42:12 [INFO] Running MitoHifi pipeline in contigs mode...
2024-11-08 16:42:12 [INFO] 1. Fixing potentially conflicting FASTA headers
2024-11-08 16:42:12 [INFO] 2. Let's run the blast of the contigs versus the close-related mitogenome
2024-11-08 16:42:12 [INFO] 2.1. Creating BLAST database:
2024-11-08 16:42:12 [INFO] makeblastdb -in refData/OQ830676.1.fasta -dbtype nucl
2024-11-08 16:42:12 [INFO] Makeblastdb done.
2024-11-08 16:42:12 [INFO] 2.2. Running blast of contigs against close-related mitogenome:
2024-11-08 16:42:12 [INFO] blastn -query fixed_header_contigs.fasta -db refData/OQ830676.1.fasta -num_threads 1 -out contigs.blastn -outfmt 6 std qlen slen
2024-11-08 16:42:12 [INFO] Blast done.
2024-11-08 16:42:12 [INFO] 3. Filtering BLAST output to select target sequences
2024-11-08 16:42:12 [INFO] Filtering thresholds applied:
2024-11-08 16:42:12 [INFO] Minimum query percentage = 50
2024-11-08 16:42:12 [INFO] Minimum query length = 80% subject length
2024-11-08 16:42:12 [INFO] Maximum query length = 5 times subject length
2024-11-08 16:42:12 [INFO] Filtering BLAST finished. A list of the filtered contigs was saved on ./contigs_filtering/contigs_ids.txt file
2024-11-08 16:42:12 [INFO] 4. Now we are going to circularize, annotate and rotate each filtered contig. Those are potential mitogenome(s).
2024-11-08 16:42:12 [INFO] Annotation will be done using MitoFinder (default)
2024-11-08 16:42:12 [INFO] Working with contig tig00007550_1
2024-11-08 16:42:12 [INFO] Started tig00007550_1 circularization
2024-11-08 16:42:13 [INFO] tig00007550_1 circularization done. Circularization info saved on ./potential_contigs/tig00007550_1/tig00007550_1.circularisationCheck.txt
2024-11-08 16:42:13 [INFO] Started tig00007550_1 (MitoFinder) annotation
2024-11-08 16:45:14 [INFO] tig00007550_1 annotation done. Annotation log saved on ./potential_contigs/tig00007550_1/tig00007550_1.annotation_MitoFinder.log
2024-11-08 16:45:14 [INFO] Started tig00007550_1 rotation.
2024-11-08 16:45:14 [INFO] Rotation of tig00007550_1 done. Rotated is at tig00007550_1.mitogenome.rotated.fa
2024-11-08 16:45:14 [INFO] Started calculating mitocontig stats... tig00007550_1
2024-11-08 16:45:14 [INFO] 5. Now the rotated contigs will be aligned
2024-11-08 16:45:14 [INFO] List of contigs that will be aligned: ['tig00007550_1.mitogenome.rotated.fa']

2024-11-08 16:45:14 [INFO] MAFFT alignment will be called with:
mafft --quiet --clustalout --thread 1 all_mitogenomes.rotated.fa > all_mitogenomes.rotated.aligned.aln

2024-11-08 16:45:14 [INFO] Alignment done and saved at ./final_mitogenome_choice/all_mitogenomes.rotated.aligned.aln

2024-11-08 16:45:14 [INFO] 6. Now we will choose the most representative contig

2024-11-08 16:45:14 [INFO] Representative contig is tig00007550_1 that belongs to Cluster 0. This contig will be our final mitogenome. See all contigs and clusters in cdhit.out.clstr
2024-11-08 16:45:14 [INFO] 7. Calculating final stats for final mitogenome and other potential contigs.
    Stats will be saved on contigs_stats.tsv file.
tig00007550_1 list of genes: ['tRNA-Phe', 'tRNA-Glu', 'tRNA-Ser2', 'tRNA-Asn', 'tRNA-Arg', 'tRNA-Ala', 'ND3', 'tRNA-Gly', 'COX3', 'ATP6', 'tRNA-Asp', 'tRNA-Lys', 'COX2', 'tRNA-Leu2', 'COX1', 'tRNA-Tyr', 'tRNA-Cys', 'tRNA-Trp', 'ND2', 'tRNA-Gln', 'tRNA-Ile', 'tRNA-Met', 'tRNA-Val', 'tRNA-Leu', 'ND1', 'tRNA-Ser', 'CYTB', 'ND6', 'tRNA-Pro', 'tRNA-Thr', 'ND4L', 'ND4', 'tRNA-His', 'ND5']
2024-11-08 16:45:14 [INFO] 8. Building annotation plots for all contigs
2024-11-08 16:45:14 [INFO] Pipeline finished!
2024-11-08 16:45:14 [INFO] Run time: 182.57 seconds</code></pre>

<p class="blog-post_paragraph">Taking a look at the final annotated assembly, I get this.</p>

<pre class="command-line" data-user="" data-host=""><code class="language-commandline">cat final_mitogenome.gb</code></pre>

<pre class="output no-lang-display language-output" tabindex="0"><code class="language-output">LOCUS       tig00007550_1_rotated  15507 bp    DNA              UNK 01-JAN-1980
DEFINITION  tig00007550_1.
ACCESSION   tig00007550_1_rotated
VERSION     tig00007550_1_rotated
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     gene            1..68
                     /product="tRNA-Phe"
     tRNA            1..68
                     /product="tRNA-Phe"
     gene            complement(85..152)
                     /product="tRNA-Glu"
     tRNA            complement(85..152)
                     /product="tRNA-Glu"
     gene            complement(168..234)
                     /product="tRNA-Ser2"
     tRNA            complement(168..234)
                     /product="tRNA-Ser2"
     gene            complement(244..309)
                     /product="tRNA-Asn"
     tRNA            complement(244..309)
                     /product="tRNA-Asn"
     gene            complement(333..397)
                     /product="tRNA-Arg"
     tRNA            complement(333..397)
                     /product="tRNA-Arg"
     gene            complement(398..463)
                     /product="tRNA-Ala"
     tRNA            complement(398..463)
                     /product="tRNA-Ala"
     gene            complement(485..838)
                     /gene="ND3"
     CDS             complement(485..838)
                     /translation="IILLNNILIMIFFISNIMMFLSIILSKKSLLDREKCSPFECGFDP
                     KSCARIPFSLHFFLITMIFLIFDVEIALIFPIIPLFKMVNFLMWTKISIFFIIILLIGL
                     YHEWNQNMLNWIN*"
                     /gene="ND3"
     gene            complement(839..904)
                     /product="tRNA-Gly"
     tRNA            complement(839..904)
                     /product="tRNA-Gly"
     gene            complement(907..1695)
                     /gene="COX3"
     CDS             complement(907..1695)
                     /translation="MKTTFNHPFHLVDYSPWPLTGAIGVMTLVTGMVKWFHNFNMNLLI
                     LGYFIVILTMYQWWRDICREGTLQGKHTILVTKGLRWGMILFIVSEIFFFLSFFWAFFH
                     SSLSPNIEIGAMWPPLNILPFNPFQIPLLNTIILISSGVSVTWAHHALMKNNNTQMTQG
                     LFFTIILGIYFTILQAYEYIEAPFTIADSIYGSTFFMATGFHGLHVMIGTIFLFTCLIR
                     HLNNHFSKNHHFGFEAAAWYWHFVDVVWLFLYISIYWWGN*"
                     /gene="COX3"
     gene            complement(1695..2372)
                     /gene="ATP6"
     CDS             complement(1695..2372)
                     /translation="MMSNLFSIFDPSTNIFNLSLNWLSTILGILFIPYSFWLTPNRHHL
                     FWNFILIKLHNEFKTLLKNNYFQGSTIIFISMFSFVLFNNFLGLFPYIFTSTSHLNLSL
                     SISLPLWLSFMIYGWINNTQNMFIHMIPQGTPTILMPFMVLIETISNIIRPGTLAVRLT
                     ANMIAGHLLMTLLSGTGPSMPSYLIMMLILIQILLMILESAVAVIQSYVIAILSTLYSS
                     EAN*"
                     /gene="ATP6"
     gene            complement(2525..2595)
                     /product="tRNA-Asp"
     tRNA            complement(2525..2595)
                     /product="tRNA-Asp"
     gene            complement(2631..2701)
                     /product="tRNA-Lys"
     tRNA            complement(2631..2701)
                     /product="tRNA-Lys"
     gene            complement(2703..3383)
                     /gene="COX2"
     CDS             complement(2703..3383)
                     /translation="MATWSNFNLQNGASPLMEQIIFFHDHTLIILIMITMLVGYLMLSL
                     FFNKYINRFLLEGQMIELIWTILPAITLIFIALPSLRLLYLLDELNNPLITLKSIGHQW
                     YWSYEYSDFNNIEFDSYMIPTNELNSNNFRLLDVDNRIILPMNNQIRIMVTATDVIHSW
                     TIPSLGVKVDANPGRLNQTNFFMNRPGIFYGQCSEICGANHSFMPIVIESISIKNFINW
                     INNYS"
                     /gene="COX2"
     gene            complement(3384..3450)
                     /product="tRNA-Leu2"
     tRNA            complement(3384..3450)
                     /product="tRNA-Leu2"
     gene            complement(3446..4981)
                     /gene="COX1"
     CDS             complement(3446..4981)
                     /translation="RKWLYSTNHKDIGTLYFIFGIWAGMVGTSLSLLIRAELGNPGSLI
                     GDDQIYNTIVTAHAFIMIFFMVMPIMIGGFGNWLIPLMLGAPDMAFPRMNNMSFWLLPP
                     SLTLLISSSIVENGAGTGWTVYPPLSSNIAHGGSSVDLAIFSLHLAGISSILGAINFIT
                     TIINMRLNNMSFDQMPLFIWAVGITAFLLLLSLPVLAGAITMLLTDRNLNTSFFDPAGG
                     GDPILYQHLFWFFGHPEVYILILPGFGMISHIISQESGKKETFGCLGMIYAMMAIGLLG
                     FIVWAHHMFTVGMDIDTRAYFTSATMIIAVPTGIKIFSWLATLHGTQINYSPSMLWSLG
                     FVFLFTVGGLTGVVLANSSIDITLHDTYYVVAHFHYVLSMGAVFTILGGFIHWYPLFTG
                     LNMNPFLLKIQFFIMFIGVNLTFFPQHFLGLAGMPRRYSDYPDSYITWNIISSLGSYIS
                     LLAVMFILIIIWESMINQRIALFTLNLSSSIEWYQNLPPAEHSYNELPILSNF*"
                     /gene="COX1"
     gene            4986..5052
                     /product="tRNA-Tyr"
     tRNA            4986..5052
                     /product="tRNA-Tyr"
     gene            5053..5121
                     /product="tRNA-Cys"
     tRNA            5053..5121
                     /product="tRNA-Cys"
     gene            complement(5114..5185)
                     /product="tRNA-Trp"
     tRNA            complement(5114..5185)
                     /product="tRNA-Trp"
     gene            complement(5198..6211)
                     /gene="ND2"
     CDS             complement(5198..6211)
                     /translation="IYFTYNSPKMFFFFFLIFSTLISISSNSWLGCWIGLEINLLSFIP
                     IISNSFNLFASEASLKYFLTQSIASINFLFSTMMNMILMKNFETNLIISIMMNSSLLMK
                     MGSVPFHFWFPIIIEGLSWFNSFILMTWQKISPMILLSYYFNSNFIIIIVIFNIMVGAI
                     GGLNQTSLRKLLTFSSINNLGWMIMAILISENLWMFYFCFYSFLISIMCLLFYNLNMFY
                     INQLFINNMNFMMKINLLINFLSLGGLPPFIGFFPKWVIINFLINNNFYFMAFIFMMMS
                     LITLFFYIRIIYSSFMFNYFKMKWFKIFIKNKSFLIINFFSFISISGMIFSTFFFM*"
                     /gene="ND2"
     gene            6268..6336
                     /product="tRNA-Gln"
     tRNA            6268..6336
                     /product="tRNA-Gln"
     gene            complement(6334..6399)
                     /product="tRNA-Ile"
     tRNA            complement(6334..6399)
                     /product="tRNA-Ile"
     gene            complement(6401..6468)
                     /product="tRNA-Met"
     tRNA            complement(6401..6468)
                     /product="tRNA-Met"
     gene            7642..7713
                     /product="tRNA-Val"
     tRNA            7642..7713
                     /product="tRNA-Val"
     gene            9097..9164
                     /product="tRNA-Leu"
     tRNA            9097..9164
                     /product="tRNA-Leu"
     gene            9166..10101
                     /gene="ND1"
     CDS             9166..10101
                     /translation="MLLKDMYMIFLGLLILILGVLIGVAYLTLLERKVLGYIQIRKGPN
                     KVGLMGILQPFSDAIKLFTKEQTYPNFSNYLSYYFSPVISFLLSLMIWVLVPYYYNMVS
                     FNLGMLFFLCCTSLGVYTIMIAGWSSNSNYALLGGLRAVAQTISYEVSLALILMSNIIM
                     IMDFNMMNFYIYQVNIWFIFLMFPLSLCWMSSSLAETNRTPFDFAEGESELVSGFNVEY
                     SSGGFALIFLAEYSSILFMSLLFVLLYLGGYVLSIFFYLKLFFISFLFIWVRGTLPRYR
                     YDKLMYLAWKSYLPISLNFLLFFLGVKIFF*"
                     /gene="ND1"
     gene            complement(10133..10197)
                     /product="tRNA-Ser"
     tRNA            complement(10133..10197)
                     /product="tRNA-Ser"
     gene            complement(10197..11348)
                     /gene="CYTB"
     CDS             complement(10197..11348)
                     /translation="MTNKYLPIRKTHPIIKMINNSLIDLPSPSNISSWWNFGSLLALCL
                     IIQILTGLFLTMYYTANIELAFYSVNYICRNVNYGWLIRTLHANGASFFFICIYLHIGR
                     GIYYESFNLMYTWMIGVMILFLLMATAFMGYVLPWGQMSFWGATVITNLLSAIPYLGPM
                     LVNWIWGGFAVDNATLTRFYTFHFLLPFMILMMTMIHLLFLHQTGSNNPLGLNSNLDKI
                     PFHPFFTYKDLIGFIIILFLLVMLTLINPYLLGDPDNFIPANPLVTPVHIQPEWYFLFA
                     YAILRSIPNKLGGVIALIMSILILIILPLTFNKKIQGIQFYPINQILFWLLMMMIILLT
                     WIGARPVEEPYIITGQLLTIFYFSYFIINPIMNKMWDNLLFNK*"
                     /gene="CYTB"
     gene            complement(11358..11888)
                     /gene="ND6"
     CDS             complement(11358..11888)
                     /translation="MTKMFLSLMIIMLSIMMFFFNNPLSMGLMILLQTLLTCLLSGMMI
                     KTYWFSYILFLTFLGGLLVLFIYVSSIASNELFNMPIIMKSFFFICILIILMIQIFYMN
                     NLMWMNFSKNSDMENFMNMYINMNIENKINLNKLYNNQTFLMMMMLVIYLFITLIAVVK
                     ITNIFYGPLRSSF*"
                     /gene="ND6"
     gene            11894..11958
                     /product="tRNA-Pro"
     tRNA            11894..11958
                     /product="tRNA-Pro"
     gene            complement(11959..12024)
                     /product="tRNA-Thr"
     tRNA            complement(11959..12024)
                     /product="tRNA-Thr"
     gene            12041..12331
                     /gene="ND4L"
     CDS             12041..12331
                     /translation="MMILNIWVVFIMMYLIGNLIFVSKRKHLLIVLLSLEFLVLSIFFF
                     MLIFLMLIEYDMYMLMVFLVFSVCEGALGLSILVSMIRTHGNDYFQSFNLL*"
                     /gene="ND4L"
     gene            12365..13702
                     /gene="ND4"
     CDS             12365..13702
                     /translation="MMKFLMMMIFMMPLCFMKNMFWMVQMMLMLMMFLFMNLNISIMNF
                     CNISYMMSIDVMSYGLILLSIWIVILMIMSSENLFKENFYINFFLFNIIFLLIMLYLTF
                     SVMNLFMFYLFFEGSLIPTLLLIIGWGYQPERIQAGMYLLFYTLFVSLPLLIGIFYIFN
                     NMNCLMIYFLKFLSLNYYILYFSMVMAFLVKMPMYFVHLWLPKAHVEAPVSGSMILAGI
                     MLKLGGYGLLRIMIFLQEINMKLNYLWIIISLLGGFYISLKCFCQVDIKSLIAYSSVAH
                     MSIVISGIMVMNYWGYMGSYILMIGHGLCSSGMFCLANILYERLHSRSLYINKGMMNFM
                     PSMSLWWFLLLSSNMAAPPSLNLMGEISLINSLLSWSFFSMIMLMLISFFSAGYSLYLY
                     SYIQHGDYYQGVYSYYTGVSREYLMLMLHWFPLNILILKVDYFMIWF"
                     /gene="ND4"
     gene            13704..13769
                     /product="tRNA-His"
     tRNA            13704..13769
                     /product="tRNA-His"
     gene            13770..15485
                     /gene="ND5"
     CDS             13770..15485
                     /translation="INMKYCIWVSSSFFLFFFSMMSFFFGLYFIMNNMFIFFEWEIISL
                     NSLSIVMSILLDWMSLLFMMFVLLISSVVIYYSKSYMSSDLNLFRFIILVLLFVFSMLL
                     LIISPNIISILLGWDGLGLVSYCLVIYYQNLKSYNAGMLTALSNRIGDVLILMVISWMM
                     NYGSWNYIFYLDLMKIDFEMKMISSMIILAAMTKSAQIPFSSWLPAAMAAPTPVSALVH
                     SSTLVTAGVYLLIRFNNLLIEMFFIKLLLLLSGLTMFMAGIAANYEFDLKKIIALSTLS
                     QLGLMMSILSMGMPILAFFHLLTHAMFKALLFMCAGMIIHLMNDIQDIRYMGGISYFIP
                     LTSLCMNISNMALCGIPFLAGFYSKDMILEMVSFSNLNLMVFLLYYISTGLTMFYSFRL
                     MMYLMINDYNLLSIYNLYDEDFIMLKSMFILLFMSVVSGSMLSWLIFSYPYMIYLPFNL
                     KMMVIYISLLGGILGYFVSNMNTYSLNKFLMTYNLSNFFCLMWFMPNLSTYGVSYNFLN
                     LGGNLLKNIDLGWSEIYSGMGLMMIMKKYSMFINVLQMNNFKIYLFSFIMWIIIC"
                     /gene="ND5"
ORIGIN
        1 atttaaatag cttataaatt aagagtataa tattgaagat attaaggaga taaatttatc
       61 tttaaatatt tataataata ttttttttta taaaaaaaat ttttttattt ttacaatgaa
      121 aatgtaatgt tttaatatta aactatataa aaggtaaatt aaagaaaaga aatattaatg
      181 gaatttaacc actaaaaatt aagttagcag cttaatttta taactttata tttcttaata
      241 aattttaatt ggaattttta ttcaatttta tcattaacag tgatatacct ctattttggt
      301 ttcaattaat ataattattt aattttatta attaaataag gagtagataa aactctatct
      361 tttaagtcga aattaaatgc aattttgctt catatttaag ataaattgat ataaattcaa
      421 taatttttga atgcaaatca aatgtttttt aaactataat ccttataata aaaatttatt
      481 gaatttaatt aattcaattt aatatatttt gatttcattc atggtataaa ccaattaata
      541 agataataat aaaaaaaatt ctaattttag ttcatattaa aaaatttact attttaaata
      601 aaggaataat ggggaaaatt aatgcaattt ctacatcaaa aattaaaaaa attatagtaa
      661 ttaaaaaaaa atgtaaagaa aaaggaattc gagcacatga ttttgggtca aatccacatt
      721 caaatgggga gcatttttct cgatctaata aagatttttt agataaaata atagataaga
      781 atattataat attagaaata aaaaaaatta taattaaaat attatttaat aaaataatta
      841 tttatattaa aaattattaa actttttgat tggaaatcaa atatactaaa tatattatat
      901 aaataattaa tttccccatc aatagataga aatataaagg aataatcata ctacgtcaac
      961 gaaatgtcaa tatcaagctg cagcttcaaa tccaaagtgg tgattttttg agaagtgatt
     1021 gtttaagtga cgaattaggc aagtaaataa aaaaattgtt ccaattataa catgaagtcc
     1081 atgaaatcct gttgccataa aaaatgttga tccgtaaatt ctatctgcaa ttgtaaaagg
     1141 tgcttcaata tattcatatg cttgtaaaat agtaaaataa atacctaaaa taattgtaaa
     1201 gaaaagacct tgagttattt gggtattatt attttttatt aaagcatgat gagctcatgt
     1261 cacagatact cctgatctaa ttaaaataat agtattaagt aatggaattt gaaatggatt
     1321 aaatggtaga atatttaaag ggggtcatat tgctccaatt tcaatattag gggataaact
     1381 tctgtgaaaa aaagctcaga agaatgataa aaaaaaaaaa atttctgaca caataaataa
     1441 aattatacct caacgaagtc cttttgttac gagaatagta tgtttacctt gaagtgttcc
     1501 ttctcggcag atatctcgtc atcattgata tatagttaaa attacaataa aatatcctaa
     1561 aataagtaaa tttatgttaa aattatgaaa tcattttact attcctgtta ctaatgttat
     1621 aactccaatt gccccagtta agggtcatgg gctataatct actaaatgaa atggatggtt
     1681 aaaagttgtt ttcattaatt tgcttcacta gaatatagag ttcttaaaat tgcaattaca
     1741 taagattgaa ttactgcaac tgctgattct aaaattatca gaagaatttg aattaaaatg
     1801 agtattataa ttaaataaga gggtatagaa ggtccagtgc ctcttaaaag agttattaaa
     1861 agatgtccag caattatatt tgctgttaaa cgaacagcta aagttcctgg tcgaatgata
     1921 tttctaattg tttcaattaa gactataaat ggtattaaaa tagtaggtgt tccttgagga
     1981 attatatgga taaatatatt ttgtgtatta ttaattcacc cataaattat gaatcttaat
     2041 cataatggta aagaaataga taatgacaaa tttaaatgac tagttcttgt aaaaatgtat
     2101 gggaataatc ctagaaaatt attaaataat acaaaggaaa atattgaaat aaaaataatt
     2161 gttgatcctt gaaaataatt attttttaat aaagttttaa attcattatg aagtttaatt
     2221 aaaataaaat ttcaaaataa gtgatgtcga ttaggggtta atcaaaatga ataaggaata
     2281 aataagattc ctaaaatagt tcttaatcaa ttaagagata aattaaaaat attagttgaa
     2341 ggatcaaaaa ttgaaaataa atttcttatc attttcaatt tatttttttt tttttaaaat
     2401 tattaattaa attattatta aatttataat taaaaataaa ataatttatt atattaaaaa
     2461 ttataaaaat aataataaat aaaaatagat atataattca attaataggt attatttgtg
     2521 ggataaaaag atattacttt ttaagtaata atttaaattt gacatattta tgttatataa
     2581 attaactaat cttttataca tacttcctat atatatatat atatatatta tcattagaag
     2641 taattgctaa tttactatga aatggtttaa gagaccagta cttgctttca gtcatctaat
     2701 gaagaataat tattaattca attaataaaa tttttaattg aaattctttc aattacaatt
     2761 ggtataaaac tatggttagc cccacagatt tcagaacatt gaccataaaa aattcctggt
     2821 cgatttataa aaaaattagt ttggtttagg cggcctggat tagcatctac ttttacccct
     2881 aatgatggaa ttgttcaaga gtggattaca tcagttgcag tcactataat acgaatttga
     2941 ttatttattg gtaaaataat tcgattatca acatccaata aacgaaaatt attagaattt
     3001 aattcattag tgggaattat ataagagtca aattcaatat tattaaaatc agagtattca
     3061 tatcttcaat atcattgatg tccaatagat ttcaaagtaa ttaaaggatt atttaattca
     3121 tcaagtaagt aaagtaatcg taaagatgga agggcaataa aaattaaagt aattgctggt
     3181 aaaatggttc aaattaattc aattatttga ccttctaata aaaatcgatt aatatattta
     3241 ttaaaaaata atcttaatat taaataacct actaatatag taattataat taaaataatt
     3301 aaagtatgat catgaaaaaa aataatttgt tctattaaag gggaagctcc attttgtaaa
     3361 ttgaaattag atcatgttgc catttctaaa aaaaggataa acctttataa atggggttta
     3421 aatccattac atataatctg ccatattaga aattacttaa aattggaagt tcattataag
     3481 aatgttcagc tggtggtaga ttttggtatc attcaataga agaagaaaga tttaaagtaa
     3541 ataaagcaat tcgttgatta attatagatt ctcaaataat aattaaaata aatattactg
     3601 ctaataaaga aatataagaa cctaaagaag aaataatgtt tcaagtaata taagaatctg
     3661 ggtaatcaga ataacgacga ggtattcctg caagtcctaa aaaatgttga gggaaaaagg
     3721 ttaaatttac cccaataaat ataataaaaa attgaatttt taataaaaaa gggtttatat
     3781 ttaatcctgt aaataaagga tatcaatgaa taaatcctcc taaaattgta aatacagccc
     3841 ctatagaaag aacataatga aaatgagcca ctacataata tgtatcatga agggtaatat
     3901 caatagaaga attagctaaa acaactcctg ttaatcctcc tacagtaaat aaaaaaacaa
     3961 atcctaatct tcataatatt gaaggtctat aattaatttg tgttccatga agagttgcta
     4021 atcagctaaa aattttaatt cctgtaggta cagcaatgat tatagtagct gaagtaaaat
     4081 atgctcgagt atcaatatct ataccaacag taaatatatg atgagctcaa acaataaaac
     4141 ctaataaacc aatagctatt attgcgtaaa ttatacctaa acatccaaag gtttcttttt
     4201 ttcctctttc ttgagaaata atatgagaaa ttatcccaaa tcctggtaaa attaaaatat
     4261 aaacttctgg gtgaccaaaa aatcaaaata aatgttgata taaaattggg tcacctcctc
     4321 ctgcaggatc aaaaaaagaa gtattcaaat ttcggtcagt taataatata gtaatagctc
     4381 ctgctaaaac tggtaaagat aaaagtaata aaaaagctgt aattccaaca gctcaaataa
     4441 ataaaggtat ttgatcaaaa gatatattat taagtcgtat attaataatt gttgtaataa
     4501 aattaattgc tcctaaaata gaagaaattc cagctaaatg tagagagaaa atagctaaat
     4561 caactgatct tcctccatga gcaatattag atgaaagtgg gggataaact gttcatccag
     4621 ttcctgctcc attttctaca attcttcttg aaattagaag ggttaaggag gggggtaaaa
     4681 gtcagaaact tatattattt atacgaggga atgctatatc tggagctcct aatattaagg
     4741 ggattaatca atttccaaat cctccaatta taataggtat tactataaaa aaaattataa
     4801 tgaaagcatg agctgtgaca atagtattat aaatttgatc atctccaatt aaagatccag
     4861 ggttacctaa ttctgctcga attaataatc ttaatgaagt tcctactatt cctgcccaaa
     4921 ttccaaaaat gaaatataat gttccaatat ctttatgatt tgtagagtaa agtcattttc
     4981 gctataataa aatggctgag attaagcgat aaattgtaaa tttatttacg agatataata
     5041 tctcttttat taagtcttat agtcaaaaat aatgatataa aattgcaaat tttaaggaat
     5101 aataatttat tattaaggct taaagaacaa ttattttctt tataaataat tttgaagatt
     5161 attagtttat ttaacttaaa accttaataa aaattattta tataaaaaaa aaagttctaa
     5221 aaattattcc tctaatagaa ataaaagaaa aaaaattaat aattaaaaat gatttatttt
     5281 taataaaaat tttaaatcat tttattttaa aataattaaa tatgaaagag gaataaataa
     5341 tacgaatata aaaaaataat gtaattaatc ttattattat aaaaataaat gctataaaat
     5401 aaaaattatt attaattaaa aaattaataa taactcattt agggaaaaat ccaataaaag
     5461 gtggtagtcc ccctaaagat aaaaaattaa ttaataaatt aatttttatt ataaaattta
     5521 tattattaat aaaaagttga ttgatataaa atatattaag attataaaat aataaacata
     5581 taattctaat aagaaaagaa taaaaacaaa aataaaatat tcataaattt tctctaatta
     5641 aaattgctat aattattcat cctaaattat taatagaaga aaatgttaaa agcttacgta
     5701 aggaagtttg gtttaatcct ccaattgccc caactataat attaaaaata acaataataa
     5761 taataaaatt tctattaaaa taatatgata ataaaattat gggggaaatt ttttgtcatg
     5821 ttattaaaat aaaactatta aatcaagata aaccttcaat aataatagga aatcaaaaat
     5881 gaaaaggtac agatcctatt tttataagta atgaagaatt tattataatt gaaataatta
     5941 aatttgtttc gaaatttttt attaagatta tatttattat agttgaaaat aaaaagttaa
     6001 ttgaggcaat agattgagtt aaaaaatatt ttaaagaggc ttctgaggca aataaattaa
     6061 atgaattaga aataataggg ataaaagaaa gtaaattaat ttctaaaccg attcaacatc
     6121 ctaatcaaga attagaagaa attgaaatta aggttctaaa aattaaaaaa aaaaaaaaaa
     6181 atattttggg ggaattatat gtaaaataaa tttaaaataa agaatatttc tttttattga
     6241 atttaattaa ttcattaatt tattatttat attttagtgt aagatgcaca ataatttttg
     6301 atattataag atatagtttg attctataaa atataataaa ggtaaattta tttacttaat
     6361 ttatcctatc agaataatcc tttaatcagg cactttattt tttaaaaaaa aggcatttcc
     6421 tttatatttg gggtatgaac ccaaaagctt aatttagctt atttttaatt tatttattta
     6481 tatgtaaaat tattaaaaac ggtataaatt aatcaataaa ttaattttta tattataatt
     6541 atatatatat atatattaaa tttttaatat attattatat ttattattct ttaacttaat
     6601 tgtaagaaaa taaataataa ttatatatat atatagttaa atataagaat tatcaagttt
     6661 taatttccaa gtttgttttt aatatgaata ttataaagag aaaaaaagag aaattattca
     6721 atgtttaata acttatatta taaattattt aatatttata atatataata aatattttaa
     6781 taaaaaaaaa aaaaaaaaaa aattctattt aagaaaattt acatataagt aaaagtttta
     6841 tagtttaaaa aatttatttt aaatttattt tacatataaa ttttagtgtt gaattttgat
     6901 tatttaaaaa ataattgaaa tttaaatgta gtaattaata ttaaaaattt aagaaattaa
     6961 gatttagtta tattaaaatt agcaacaaat tttgtgccag cagttgcggt tatacaaaag
     7021 ttaaattaaa tttttatagg taattaatta ataaataatg attaaatata ttaaaaattt
     7081 taaattatta agtgaaattt taatttaatt aaaatttata tgaaaattat gatttaataa
     7141 attttataaa aaactaggat tagataccct attattaaaa tttaaattta taatactaaa
     7201 atagtaaata ataatttatt gaaacttaaa taaattggcg gtattttagt tcatttagag
     7261 gaatctgttt aataattgat aatccacgaa taaatttact taatttatta ttttgtatat
     7321 cgtcgttaaa aaaatatttt ttaataaaaa taatatttta aattttttat ttaaatttaa
     7381 atcagatcaa gatgcagatt ataattaaga atataatgga ttacaataaa ttaatttata
     7441 tgaatattaa tttttaaaat taattgaagg tggatttaaa tgtaatttta tttaatttaa
     7501 taaagtgatt aaaaattaaa atatgtacat attgcccgtc gctttcatat gtaaatatat
     7561 gcatatatat atacttaaat atatttaaaa ttgaaataag tcgtaacaaa gtagaggtac
     7621 tggaaagtgt ttctagaaag acaaattaga gcttgaattt taagtatttc atttacattg
     7681 aaaagaatat taataaaaaa ttaattaatt tgagggggca aattaattaa tttataatta
     7741 ataaaagaat ttttataatt aaaatattta atgggggtta aagtatttta atagaaaaaa
     7801 ttgaaaaatt tatagtaaat tagtattgtg aaagaatttt tgaaatagtt atgaaatatg
     7861 gaattattta aaaagtaatt tttatttaat gtatcttggg tatcagagtt tattaaaaat
     7921 tatttttata aaaaaaattt ctcgaattta aaagagataa ttaatttaat aaagttattg
     7981 tagcataaat attttaaata attaattgga aatgatatgt taatcgtttt taaatatatc
     8041 tagttttttt agaaaaaaat ttaattttaa aattttattt ataaaaaaaa atttattaat
     8101 aaattttttt ttataaaatt tttatatttt aagggataag ctttaaatta aaattatata
     8161 ataatttttg ttttaaattg aaaattttta taatttatat tgttaataaa ttttatttta
     8221 ttataaataa tttcaatatt atataaaatt taaaataaat ttatattttt ataaaaaaat
     8281 aatttttaat aaataaaaat tagttataat gataaaatta gtatattaat ttaattaaat
     8341 taaaattata ttttaattta aagttaaata aaaggaattc ggcaaatatt tatattcact
     8401 tgtttatcaa aaacatgtct ttttgtaaat aatttaaagt ctaatctgcc cactgattta
     8461 ttattaaagg gctgcagtat tttgactgta caaaggtagc ataatcaata gtcatttaat
     8521 tggtgacttg tatgaaagat tggatgaaat ataaactgtc tcttatttaa aatttagaat
     8581 ttaatttttc aattaaaaag ttgaaatata ttaaaaagac gagaagaccc tatagagttt
     8641 tatatttatt tttaattaat attatttata aaataaaaat tgaaaattaa taaaaatatt
     8701 ttattggggt gataaaaaaa taaaaataac tttttttatt taataacata aataaatgaa
     8761 ttaatgatcc aatattattg attaaaagaa aaaattacct tagggataac agcgtaattt
     8821 ttttttttag ttcaaataaa aaaaaaagtt tgcgacctcg atgttggatt aagataaaat
     8881 ttaaatgcaa aagtttaaaa ttttgatctg ttcgatcatt aaaatcttac atgatctgag
     8941 ttcaaaccgg tgtgagccag gttggtttct atcttttata aattttttat attttagtac
     9001 gaaaggaata aatattaaaa ttaaatttat ttaattgaga atattattaa ttaaaaatta
     9061 ttattaaaaa tttaaattta aattaaagat tttaaaacta ttttggcaga aaaatgtaat
     9121 ggttttagaa atcataaata tataaattaa ttatatatat agtaaatgtt gttaaaagat
     9181 atatatataa tttttttagg tttattgatt ttaattttag gggttttaat tggggttgca
     9241 tatttaactt tattagaacg taaagtttta ggttatattc aaattcgaaa aggtcctaat
     9301 aaggttggtt taataggaat tttacaacct ttttcagatg caattaagtt atttactaaa
     9361 gaacaaactt atcctaattt ttctaattat ttatcatatt atttttctcc ggtaattaga
     9421 tttttattat ctttaataat ttgagtatta gtcccttatt attataatat ggtgagattt
     9481 aatttaggta tattattttt tctttgttgt acgagattag gggtttatac tattataatt
     9541 gctggttggt cttctaattc taattatgct ttattaggag gtttacgggc tgtcgctcaa
     9601 actatttctt atgaagtaag attagcttta attcttatat ctaatattat tataattata
     9661 gattttaata taataaattt ttatatttat caagtaaata tttgatttat ttttttaata
     9721 tttcctttaa gtttatgttg aatatcttct agattagcgg aaactaatcg tactccattt
     9781 gattttgctg aaggtgaaag agagttagtt tcagggttta atgttgaata tagaagtggg
     9841 ggatttgcat taattttttt agcagaatac tctagaattt tatttatgag attattattt
     9901 gttttattat atttaggggg gtatgtatta agaatttttt tttatttgaa attatttttt
     9961 atttcttttt tatttatttg agttcgtggt actttacctc gttatcgata tgataagctt
    10021 atatatttag cttggaaaag ttatttacca atttcattaa attttttatt gtttttttta
    10081 ggggtaaaaa ttttttttta atatatatat atatatataa tatttttagt ataaattaat
    10141 agaattttta ttctttctta agttttcaaa acaaatgctt ttccaagctc attaatttat
    10201 ttattaaata gtaaattatc tcatatttta tttataattg ggttaataat aaaatatgaa
    10261 aaataaaaaa tagttaataa ttgtcctgta ataatataag gttcttctac tggtcgagct
    10321 ccaattcatg ttaataaaat aattattatt attaataatc aaaataaaat ttggttaatt
    10381 ggataaaatt gaattccttg aattttttta ttaaaggtta gaggtaaaat aattaaaatt
    10441 aaaatagata taattaaagc aataactcct cctaatttat tagggattga tcgaagaata
    10501 gcataagcaa ataaaaaata tcattctggt tgaatatgaa ctggggtaac aagaggatta
    10561 gcaggaataa aattatcagg atctcctaat aagtatggat taattaatgt taatattacc
    10621 aataaaaata aaataataat aaatccaatt aaatctttgt atgtaaaaaa tggatgaaaa
    10681 ggaattttat ctaaatttct atttaatcct aaaggattat tagatcctgt ttggtgtaaa
    10741 aaaagtaagt gaattatagt tattattaaa attataaatg gtaataaaaa atggaaagta
    10801 taaaatcgag ttaaagttgc attgtcaact gcaaatcccc ctcaaattca gttaactaat
    10861 atagggccta agtaagggat tgctgataaa agattagtaa ttactgttgc ccctcaaaat
    10921 gatatttgtc ctcaaggtag aacatatcct ataaatgctg ttgctattaa taaaaataaa
    10981 attattactc cgattattca tgtgtatatt aaattaaagg attcatagta aattcctcgt
    11041 ccaatatgaa gataaataca aataaaaaaa aaagatgctc catttgcatg taaagttcga
    11101 attaatcaac cataatttac atttcggcaa atataattta ctctataaaa agctaattca
    11161 atattagctg tataatatat agttaggaat aatccagtta aaatttgaat aattaaacat
    11221 aaagctaata aagatccaaa atttcatcat gaagagatat tagagggaga tggtaaatcg
    11281 attaaagaat tgttgattat tttaataatt gggtgagttt ttcgaattgg taaatattta
    11341 tttgtcattt atttaattta aaaagaagat cgtaatgggc cataaaaaat gttagtaatt
    11401 tttacaactg caattaaagt aataaataaa taaattacta atattattat tattaaaaat
    11461 gtttgattat tatataattt atttaaatta attttatttt caatatttat attaatatat
    11521 atatttataa aattttctat atctgaattt ttggaaaagt ttattcatat taaattattt
    11581 atataaaaaa tttgaattat aagaataatt aaaatacaaa taaaaaaaaa agattttata
    11641 ataattggta tattaaaaag ttcattagaa gcaattcttg atacataaat aaataatact
    11701 aataatcctc ctaaaaatgt taaaaataaa atataagaaa atcaataagt tttaattatt
    11761 attcctgata ataagcaagt taataaagtt tgtaataaga ttattaatcc tatagataaa
    11821 ggattattaa aaaaaaatat tataattgat aatataataa ttattaatga taaaaatatt
    11881 tttgtcattt attcaaggag tagtttaatt aaaataataa ttttggagat tatagataaa
    11941 gaaattcttt ttctttgaag tttttaaagt aaattactta tttttgattt acaagaccaa
    12001 tatttttttt taaattataa aaacatttaa tgtttattta atgataattt taaatatttg
    12061 agttgttttt attataatat atttaattgg caatttgatt tttgtatcta aacgtaagca
    12121 tttattaatt gttttattaa gattagaatt tttagtatta agaatttttt tttttatatt
    12181 aattttttta atattaattg aatatgatat atacatatta atggtttttt tagttttttc
    12241 tgtttgtgag ggggctttag gtttatctat tttagtttct ataattcgta ctcatggtaa
    12301 tgattatttt caaagattta atttattata aaaagtaatt gttagataaa atatattata
    12361 attaatgata aaatttttaa taataataat ttttataata cctttatgtt ttataaaaaa
    12421 tatattttga atggttcaaa taatattaat attaataatg tttttattta taaatttaaa
    12481 tattagaatt ataaattttt gtaatattag atatataata tctattgatg ttatatctta
    12541 tggtttaatt ttattaagta tttgaattgt tattttgata attatatcta gagaaaattt
    12601 atttaaagaa aatttttata ttaatttttt tttattcaat attatttttt tattaattat
    12661 gttatattta acttttagag ttataaattt atttatattt tatttatttt ttgaaggaag
    12721 tttaattcct acattattat taattattgg gtgggggtat caacctgagc ggattcaggc
    12781 ggggatatat cttttatttt atactttatt tgtttcctta cctttattaa ttggtatttt
    12841 ttatattttt aataatataa attgtttaat aatttatttt ttaaaatttt taagtttaaa
    12901 ttattatatt ttatattttt ctatagtaat agctttttta gtaaaaatac ctatatattt
    12961 tgttcattta tgattaccta aagctcatgt tgaagctcct gtttcaggtt ctataatttt
    13021 agcgggaatt atattaaaat tggggggtta tggactatta cgaattataa tttttttaca
    13081 ggaaattaat ataaaattaa attatttatg aattattatt agattattag ggggatttta
    13141 tattagatta aaatgttttt gtcaagttga tattaaatct ttaattgcat attcttctgt
    13201 tgcccatatg agaattgtta ttagagggat tatagtaata aattattgag gttatatggg
    13261 ctcatatatt ttaataattg gccatggatt atgttcatca ggaatatttt gtttagctaa
    13321 tattttatat gaacgtttac atagacgaag attatatatt aataaaggta taataaattt
    13381 tataccttct ataagattat gatgattttt attattatct tctaatatag cagcccctcc
    13441 aagtttaaat ttaatagggg aaattagttt aattaataga ttgttgagat ggtctttttt
    13501 ttcaatgatt atattaatat taatttcttt ttttagagct ggctatagat tgtatttata
    13561 ttcttatatt caacatggtg attattatca aggagtttat agatattata ctggtgtttc
    13621 tcgtgaatat ttaatgttaa tattacattg atttccctta aatattttaa ttttaaaagt
    13681 tgattatttt ataatttgat tttatttaaa taatttaaga gaaaatattg atttgtggta
    13741 tcaaaaatat gaaattatca ttttaaatta ttaatataaa atattgtatt tgggtttctt
    13801 cttcattttt tttatttttt tttagaataa taagtttttt ttttggttta tattttatta
    13861 taaataatat atttattttt tttgaatggg aaattatttc tttaaattca ttgagaattg
    13921 taatatctat tttattagat tgaatatctt tattatttat aatatttgtt ttattaattt
    13981 cttctgtggt tatttattat agtaaaagtt atataagatc tgatttaaat ttatttcgtt
    14041 ttattatttt agttttatta tttgtatttt ctatattatt attaattatt agtccaaata
    14101 ttattagaat tttattaggt tgagatggat taggtttagt ttcttattgt ttagtaattt
    14161 attatcaaaa tttaaaatct tataatgctg ggatattaac tgctttatca aatcgtattg
    14221 gggatgtttt aattttaata gttatttctt gaataataaa ttatggtaga tgaaattata
    14281 ttttttattt agatttgata aaaattgatt ttgaaataaa aataattaga agtatgatta
    14341 ttttagcagc tataactaaa agagctcaaa ttccttttag atcttgatta cctgcagcta
    14401 tagctgctcc aactcctgtt tcagctttag tacattcttc aactttagta actgctggag
    14461 tatatttatt aattcgattt aataatttat taattgaaat attttttatt aaattattat
    14521 tattattatc tggtttaaca atatttatgg cgggaattgc tgccaattat gaatttgatt
    14581 taaaaaaaat tattgcttta tctactttaa gacaattagg gttaataatg agaatcttaa
    14641 gaataggaat acctatttta gctttttttc atttattaac acatgctata tttaaagctt
    14701 tattatttat atgtgctggg ataattattc atttgataaa tgatattcaa gatattcgtt
    14761 atataggggg tattagatat tttatccctc taacttcttt atgtataaat atttcaaata
    14821 tagctttatg tggaattcct tttttagcag gattttattc aaaggatata attttagaaa
    14881 tagtaagatt tagaaattta aatttgatag tttttttatt atattatatt tctacgggat
    14941 taactatatt ttatagtttt cgtttaataa tatatttaat aattaatgat tataatttat
    15001 taagaattta taatttatat gatgaagatt ttattatatt aaaaagaata tttattttat
    15061 tatttataag tgtggtaagt gggagaatat taagttgatt aattttttct tatccttata
    15121 taatttattt accgtttaat ttgaaaataa tagtaattta tattagatta ttagggggaa
    15181 ttttgggtta ttttgtgaga aatataaata cttattcttt aaataagttt ttaatgactt
    15241 ataatttaag aaattttttt tgtttaatat gatttatacc taatttatcg acttatgggg
    15301 ttagatataa ttttcttaat ttagggggaa atttgttaaa aaatattgat ttaggttgaa
    15361 gtgaaattta tagaggtata gggttaataa taattataaa aaaatattca atatttatta
    15421 atgttttaca aataaataat tttaaaattt atttatttag atttattata tgaattatta
    15481 tttgtttatt aatattatta gtatttt
//</code></pre>

<p class="blog-post_paragraph">Visually, this is what the final annotated mitogenome looks like.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_107.png" alt="alt text" />
</div>

<p class="blog-post_paragraph">
    <u><em>Conclusion</em></u>
  </p> 

<p class="blog-post_paragraph">Overall, I learned a great deal from this workshop. It has been an invaluable experience that has significantly enhanced my technical skills and deepened my understanding of genomic data analysis. The hands-on training and insights gained will undoubtedly accelerate my progress in effectively working towards my PhD research, which focuses on sequencing challenges. Beyond my immediate project, this workshop has equipped me with tools and knowledge that will make me a more versatile and effective scientist, as sequencing technologies underpin a lot of the modern biological analyses performed today. I feel much more capable of thinking critical about these techniques problems and effectively collaborating with other research groups. There is still a lot to learn on this topic, but I feel lucky to learn the best practices from members of some of the best genome sequencing initiatives in the world and to have access to their public resources to streamline my future sequencing projects.</p>

<div class="workshowcase-img-container">
  <img class="img-fluid" src="assets/img/blog_posts/genome-assembly/image_108.png" alt="alt text" />
</div>
</div>